<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.alignment.Alignment &mdash; scikit-bio 0.2.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.2.2 documentation" href="../index.html" />
    <link rel="up" title="Sequence collections and alignments (skbio.alignment)" href="../alignment.html" />
    <link rel="next" title="skbio.alignment.Alignment.__contains__" href="skbio.alignment.Alignment.__contains__.html" />
    <link rel="prev" title="skbio.alignment.SequenceCollection.write" href="skbio.alignment.SequenceCollection.write.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.alignment.Alignment</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-alignment-alignment">
<h1>skbio.alignment.Alignment<a class="headerlink" href="#skbio-alignment-alignment" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.alignment.Alignment">
<em class="property">class </em><tt class="descclassname">skbio.alignment.</tt><tt class="descname">Alignment</tt><big>(</big><em>seqs</em>, <em>validate=False</em>, <em>score=None</em>, <em>start_end_positions=None</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.2.2/skbio/alignment/_alignment.py#L936"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.alignment.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for storing alignments of biological sequences.</p>
<p>The <tt class="docutils literal"><span class="pre">Alignment</span></tt> class adds convenience methods to the
<tt class="docutils literal"><span class="pre">SequenceCollection</span></tt> class to make it easy to work with alignments of
biological sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seqs</strong> : list of <cite>skbio.sequence.BiologicalSequence</cite> objects</p>
<blockquote>
<div><p>The <cite>skbio.sequence.BiologicalSequence</cite> objects to load into
a new <cite>Alignment</cite> object.</p>
</div></blockquote>
<p><strong>validate</strong> : bool, optional</p>
<blockquote>
<div><p>If True, runs the <cite>is_valid</cite> method after construction and raises
<cite>SequenceCollectionError</cite> if <tt class="docutils literal"><span class="pre">is_valid</span> <span class="pre">==</span> <span class="pre">False</span></tt>.</p>
</div></blockquote>
<p><strong>score</strong> : float, optional</p>
<blockquote>
<div><p>The score of the alignment, if applicable (usually only if the
alignment was just constructed).</p>
</div></blockquote>
<p><strong>start_end_positions</strong> : iterable of two-item tuples, optional</p>
<blockquote>
<div><p>The start and end positions of each input sequence in the alignment,
if applicable (usually only if the alignment was just constructed using
a local alignment algorithm). Note that these should be indexes into
the unaligned sequences, though the <cite>Alignment</cite> object itself doesn&#8217;t
know about these.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>skbio.alignment.SequenceCollectionError</strong></p>
<blockquote>
<div><p>If <tt class="docutils literal"><span class="pre">validate</span> <span class="pre">==</span> <span class="pre">True</span></tt> and <tt class="docutils literal"><span class="pre">is_valid</span> <span class="pre">==</span> <span class="pre">False</span></tt>.</p>
</div></blockquote>
<p><strong>skbio.alignment.AlignmentError</strong></p>
<blockquote class="last">
<div><p>If not all the sequences have the same length.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.sequence.BiologicalSequence.html#skbio.sequence.BiologicalSequence" title="skbio.sequence.BiologicalSequence"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.sequence.BiologicalSequence</span></tt></a>, <a class="reference internal" href="skbio.sequence.NucleotideSequence.html#skbio.sequence.NucleotideSequence" title="skbio.sequence.NucleotideSequence"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.sequence.NucleotideSequence</span></tt></a>, <a class="reference internal" href="skbio.sequence.DNASequence.html#skbio.sequence.DNASequence" title="skbio.sequence.DNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.sequence.DNASequence</span></tt></a>, <a class="reference internal" href="skbio.sequence.RNASequence.html#skbio.sequence.RNASequence" title="skbio.sequence.RNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.sequence.RNASequence</span></tt></a>, <a class="reference internal" href="skbio.alignment.SequenceCollection.html#skbio.alignment.SequenceCollection" title="skbio.alignment.SequenceCollection"><tt class="xref py py-obj docutils literal"><span class="pre">SequenceCollection</span></tt></a>, <a class="reference internal" href="../parse.sequences.html#module-skbio.parse.sequences" title="skbio.parse.sequences"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.parse.sequences</span></tt></a>, <a class="reference internal" href="skbio.parse.sequences.parse_fasta.html#skbio.parse.sequences.parse_fasta" title="skbio.parse.sequences.parse_fasta"><tt class="xref py py-obj docutils literal"><span class="pre">skbio.parse.sequences.parse_fasta</span></tt></a></p>
</div>
<p class="rubric">Notes</p>
<p>By definition, all of the sequences in an alignment must be of the same
length. For this reason, an alignment can be thought of as a matrix of
sequences (rows) by positions (columns).</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.sequence</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="p">(</span><span class="s">&#39;A--CCGT&#39;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;seq1&quot;</span><span class="p">),</span>
<span class="gp">... </span>             <span class="n">DNA</span><span class="p">(</span><span class="s">&#39;AACCGGT&#39;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;seq2&quot;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a1</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a1</span>
<span class="go">&lt;Alignment: n=2; mean +/- std length=7.00 +/- 0.00&gt;</span>
</pre></div>
</div>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.__contains__.html#skbio.alignment.Alignment.__contains__" title="skbio.alignment.Alignment.__contains__"><tt class="xref py py-obj docutils literal"><span class="pre">__contains__</span></tt></a>(id)</td>
<td>The in operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.__eq__.html#skbio.alignment.Alignment.__eq__" title="skbio.alignment.Alignment.__eq__"><tt class="xref py py-obj docutils literal"><span class="pre">__eq__</span></tt></a>(other)</td>
<td>The equality operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.__getitem__.html#skbio.alignment.Alignment.__getitem__" title="skbio.alignment.Alignment.__getitem__"><tt class="xref py py-obj docutils literal"><span class="pre">__getitem__</span></tt></a>(index)</td>
<td>The indexing operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.__iter__.html#skbio.alignment.Alignment.__iter__" title="skbio.alignment.Alignment.__iter__"><tt class="xref py py-obj docutils literal"><span class="pre">__iter__</span></tt></a>()</td>
<td>The iter operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.__len__.html#skbio.alignment.Alignment.__len__" title="skbio.alignment.Alignment.__len__"><tt class="xref py py-obj docutils literal"><span class="pre">__len__</span></tt></a>()</td>
<td>The len operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.__ne__.html#skbio.alignment.Alignment.__ne__" title="skbio.alignment.Alignment.__ne__"><tt class="xref py py-obj docutils literal"><span class="pre">__ne__</span></tt></a>(other)</td>
<td>The inequality operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.__repr__.html#skbio.alignment.Alignment.__repr__" title="skbio.alignment.Alignment.__repr__"><tt class="xref py py-obj docutils literal"><span class="pre">__repr__</span></tt></a>()</td>
<td>The repr method.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.__reversed__.html#skbio.alignment.Alignment.__reversed__" title="skbio.alignment.Alignment.__reversed__"><tt class="xref py py-obj docutils literal"><span class="pre">__reversed__</span></tt></a>()</td>
<td>The reversed method.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.__str__.html#skbio.alignment.Alignment.__str__" title="skbio.alignment.Alignment.__str__"><tt class="xref py py-obj docutils literal"><span class="pre">__str__</span></tt></a>()</td>
<td>The str method.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.degap.html#skbio.alignment.Alignment.degap" title="skbio.alignment.Alignment.degap"><tt class="xref py py-obj docutils literal"><span class="pre">degap</span></tt></a>()</td>
<td>Return a new <cite>SequenceCollection</cite> with all gap characters removed.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.distances.html#skbio.alignment.Alignment.distances" title="skbio.alignment.Alignment.distances"><tt class="xref py py-obj docutils literal"><span class="pre">distances</span></tt></a>([distance_fn])</td>
<td>Compute distances between all pairs of sequences</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.distribution_stats.html#skbio.alignment.Alignment.distribution_stats" title="skbio.alignment.Alignment.distribution_stats"><tt class="xref py py-obj docutils literal"><span class="pre">distribution_stats</span></tt></a>([center_f,&nbsp;spread_f])</td>
<td>Return sequence count, and center and spread of sequence lengths</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.from_fasta_records.html#skbio.alignment.Alignment.from_fasta_records" title="skbio.alignment.Alignment.from_fasta_records"><tt class="xref py py-obj docutils literal"><span class="pre">from_fasta_records</span></tt></a>(fasta_records,&nbsp;...[,&nbsp;...])</td>
<td>Initialize a <cite>SequenceCollection</cite> object</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.get_seq.html#skbio.alignment.Alignment.get_seq" title="skbio.alignment.Alignment.get_seq"><tt class="xref py py-obj docutils literal"><span class="pre">get_seq</span></tt></a>(id)</td>
<td>Return a sequence from the <cite>SequenceCollection</cite> by its id.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.ids.html#skbio.alignment.Alignment.ids" title="skbio.alignment.Alignment.ids"><tt class="xref py py-obj docutils literal"><span class="pre">ids</span></tt></a>()</td>
<td>Returns the <cite>BiologicalSequence</cite> ids</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.int_map.html#skbio.alignment.Alignment.int_map" title="skbio.alignment.Alignment.int_map"><tt class="xref py py-obj docutils literal"><span class="pre">int_map</span></tt></a>([prefix])</td>
<td>Create an integer-based mapping of sequence ids</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.is_empty.html#skbio.alignment.Alignment.is_empty" title="skbio.alignment.Alignment.is_empty"><tt class="xref py py-obj docutils literal"><span class="pre">is_empty</span></tt></a>()</td>
<td>Return True if the SequenceCollection is empty</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.is_valid.html#skbio.alignment.Alignment.is_valid" title="skbio.alignment.Alignment.is_valid"><tt class="xref py py-obj docutils literal"><span class="pre">is_valid</span></tt></a>()</td>
<td>Return True if the SequenceCollection is valid</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.iter_positions.html#skbio.alignment.Alignment.iter_positions" title="skbio.alignment.Alignment.iter_positions"><tt class="xref py py-obj docutils literal"><span class="pre">iter_positions</span></tt></a>([constructor])</td>
<td>Generator of Alignment positions (i.e., columns)</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.iteritems.html#skbio.alignment.Alignment.iteritems" title="skbio.alignment.Alignment.iteritems"><tt class="xref py py-obj docutils literal"><span class="pre">iteritems</span></tt></a>()</td>
<td>Generator of id, sequence tuples</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.k_word_frequencies.html#skbio.alignment.Alignment.k_word_frequencies" title="skbio.alignment.Alignment.k_word_frequencies"><tt class="xref py py-obj docutils literal"><span class="pre">k_word_frequencies</span></tt></a>(k[,&nbsp;overlapping])</td>
<td>Return frequencies of length k words for sequences in Alignment</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.lower.html#skbio.alignment.Alignment.lower" title="skbio.alignment.Alignment.lower"><tt class="xref py py-obj docutils literal"><span class="pre">lower</span></tt></a>()</td>
<td>Converts all sequences to lowercase</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.majority_consensus.html#skbio.alignment.Alignment.majority_consensus" title="skbio.alignment.Alignment.majority_consensus"><tt class="xref py py-obj docutils literal"><span class="pre">majority_consensus</span></tt></a>([constructor])</td>
<td>Return the majority consensus sequence for the <cite>Alignment</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.omit_gap_positions.html#skbio.alignment.Alignment.omit_gap_positions" title="skbio.alignment.Alignment.omit_gap_positions"><tt class="xref py py-obj docutils literal"><span class="pre">omit_gap_positions</span></tt></a>(maximum_gap_frequency)</td>
<td>Returns Alignment with positions filtered based on gap frequency</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.omit_gap_sequences.html#skbio.alignment.Alignment.omit_gap_sequences" title="skbio.alignment.Alignment.omit_gap_sequences"><tt class="xref py py-obj docutils literal"><span class="pre">omit_gap_sequences</span></tt></a>(maximum_gap_frequency)</td>
<td>Returns Alignment with sequences filtered based on gap frequency</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.position_counters.html#skbio.alignment.Alignment.position_counters" title="skbio.alignment.Alignment.position_counters"><tt class="xref py py-obj docutils literal"><span class="pre">position_counters</span></tt></a>()</td>
<td>Return collections.Counter object for positions in Alignment</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.position_entropies.html#skbio.alignment.Alignment.position_entropies" title="skbio.alignment.Alignment.position_entropies"><tt class="xref py py-obj docutils literal"><span class="pre">position_entropies</span></tt></a>([base,&nbsp;...])</td>
<td>Return Shannon entropy of positions in Alignment</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.position_frequencies.html#skbio.alignment.Alignment.position_frequencies" title="skbio.alignment.Alignment.position_frequencies"><tt class="xref py py-obj docutils literal"><span class="pre">position_frequencies</span></tt></a>()</td>
<td>Return frequencies of characters for positions in Alignment</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.read.html#skbio.alignment.Alignment.read" title="skbio.alignment.Alignment.read"><tt class="xref py py-obj docutils literal"><span class="pre">read</span></tt></a>(fp[,&nbsp;format])</td>
<td>Create a new <tt class="docutils literal"><span class="pre">Alignment</span></tt> instance from a file.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.score.html#skbio.alignment.Alignment.score" title="skbio.alignment.Alignment.score"><tt class="xref py py-obj docutils literal"><span class="pre">score</span></tt></a>()</td>
<td>Returns the score of the alignment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.sequence_count.html#skbio.alignment.Alignment.sequence_count" title="skbio.alignment.Alignment.sequence_count"><tt class="xref py py-obj docutils literal"><span class="pre">sequence_count</span></tt></a>()</td>
<td>Return the count of sequences in the <cite>SequenceCollection</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.sequence_length.html#skbio.alignment.Alignment.sequence_length" title="skbio.alignment.Alignment.sequence_length"><tt class="xref py py-obj docutils literal"><span class="pre">sequence_length</span></tt></a>()</td>
<td>Return the number of positions in Alignment</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.sequence_lengths.html#skbio.alignment.Alignment.sequence_lengths" title="skbio.alignment.Alignment.sequence_lengths"><tt class="xref py py-obj docutils literal"><span class="pre">sequence_lengths</span></tt></a>()</td>
<td>Return lengths of the sequences in the <cite>SequenceCollection</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.start_end_positions.html#skbio.alignment.Alignment.start_end_positions" title="skbio.alignment.Alignment.start_end_positions"><tt class="xref py py-obj docutils literal"><span class="pre">start_end_positions</span></tt></a>()</td>
<td>Returns the (start, end) positions for each aligned sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.subalignment.html#skbio.alignment.Alignment.subalignment" title="skbio.alignment.Alignment.subalignment"><tt class="xref py py-obj docutils literal"><span class="pre">subalignment</span></tt></a>([seqs_to_keep,&nbsp;...])</td>
<td>Returns new <cite>Alignment</cite> that is a subset of the current <cite>Alignment</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.toFasta.html#skbio.alignment.Alignment.toFasta" title="skbio.alignment.Alignment.toFasta"><tt class="xref py py-obj docutils literal"><span class="pre">toFasta</span></tt></a>()</td>
<td>Return fasta-formatted string representing the <cite>SequenceCollection</cite></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.to_fasta.html#skbio.alignment.Alignment.to_fasta" title="skbio.alignment.Alignment.to_fasta"><tt class="xref py py-obj docutils literal"><span class="pre">to_fasta</span></tt></a>()</td>
<td>Return fasta-formatted string representing the <cite>SequenceCollection</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.to_phylip.html#skbio.alignment.Alignment.to_phylip" title="skbio.alignment.Alignment.to_phylip"><tt class="xref py py-obj docutils literal"><span class="pre">to_phylip</span></tt></a>([map_labels,&nbsp;label_prefix])</td>
<td>Return phylip-formatted string representing the <cite>SequenceCollection</cite></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.update_ids.html#skbio.alignment.Alignment.update_ids" title="skbio.alignment.Alignment.update_ids"><tt class="xref py py-obj docutils literal"><span class="pre">update_ids</span></tt></a>([ids,&nbsp;fn,&nbsp;prefix])</td>
<td>Update sequence IDs on the sequence collection.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.alignment.Alignment.upper.html#skbio.alignment.Alignment.upper" title="skbio.alignment.Alignment.upper"><tt class="xref py py-obj docutils literal"><span class="pre">upper</span></tt></a>()</td>
<td>Converts all sequences to uppercase</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.alignment.Alignment.write.html#skbio.alignment.Alignment.write" title="skbio.alignment.Alignment.write"><tt class="xref py py-obj docutils literal"><span class="pre">write</span></tt></a>(fp[,&nbsp;format])</td>
<td>Write an instance of <tt class="docutils literal"><span class="pre">Alignment</span></tt> to a file.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>