
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Store a multiple sequence alignment in tabular (row/column) form." />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.alignment.TabularMSA &#8212; scikit-bio 0.7.1.post1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=603e9a7d"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.alignment.TabularMSA';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = '0.7.1.post1';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.alignment.TabularMSA.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.alignment.TabularMSA.append" href="skbio.alignment.TabularMSA.append.html" />
    <link rel="prev" title="Sequence Alignments (skbio.alignment)" href="../alignment.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.7.1.post1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.7.1.post1 documentation"/>
    <img src="../_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.7.1.post1 documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">skbio.alignment.TabularMSA</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.append.html">skbio.alignment.TabularMSA.append</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.consensus.html">skbio.alignment.TabularMSA.consensus</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.conservation.html">skbio.alignment.TabularMSA.conservation</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.extend.html">skbio.alignment.TabularMSA.extend</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.from_dict.html">skbio.alignment.TabularMSA.from_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.from_path_seqs.html">skbio.alignment.TabularMSA.from_path_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.gap_frequencies.html">skbio.alignment.TabularMSA.gap_frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.iter_positions.html">skbio.alignment.TabularMSA.iter_positions</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.join.html">skbio.alignment.TabularMSA.join</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.read.html">skbio.alignment.TabularMSA.read</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.reassign_index.html">skbio.alignment.TabularMSA.reassign_index</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.sort.html">skbio.alignment.TabularMSA.sort</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.to_dict.html">skbio.alignment.TabularMSA.to_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.TabularMSA.write.html">skbio.alignment.TabularMSA.write</a></li>
</ul>
</details></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.AlignPath.html">skbio.alignment.AlignPath</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.PairAlignPath.html">skbio.alignment.PairAlignPath</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align.html">skbio.alignment.pair_align</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_nucl.html">skbio.alignment.pair_align_nucl</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_prot.html">skbio.alignment.pair_align_prot</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.align_score.html">skbio.alignment.align_score</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_nucleotide.html">skbio.alignment.global_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_protein.html">skbio.alignment.global_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align.html">skbio.alignment.global_pairwise_align</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_nucleotide.html">skbio.alignment.local_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_protein.html">skbio.alignment.local_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align.html">skbio.alignment.local_pairwise_align</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../alignment.html" class="nav-link">Sequence Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">skbio.alignment.TabularMSA</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-alignment-tabularmsa">
<h1>skbio.alignment.TabularMSA<a class="headerlink" href="#skbio-alignment-tabularmsa" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">skbio.alignment.</span></span><span class="sig-name descname"><span class="pre">TabularMSA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positional_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L31"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA" title="Link to this definition">#</a></dt>
<dd><p>Store a multiple sequence alignment in tabular (row/column) form.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">iterable of GrammaredSequence, TabularMSA</span></dt><dd><p>Aligned sequences in the MSA. Sequences must all be the same type and
length. For example, <cite>sequences</cite> could be an iterable of <code class="docutils literal notranslate"><span class="pre">DNA</span></code>,
<code class="docutils literal notranslate"><span class="pre">RNA</span></code>, or <code class="docutils literal notranslate"><span class="pre">Protein</span></code> sequences. If <cite>sequences</cite> is a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>,
its <cite>metadata</cite>, <cite>positional_metadata</cite>, and <cite>index</cite> will be used unless
overridden by parameters <cite>metadata</cite>, <cite>positional_metadata</cite>, and
<cite>minter</cite>/<cite>index</cite>, respectively.</p>
</dd>
<dt><strong>metadata</strong><span class="classifier">dict, optional</span></dt><dd><p>Arbitrary metadata which applies to the entire MSA. A shallow copy of
the <code class="docutils literal notranslate"><span class="pre">dict</span></code> will be made.</p>
</dd>
<dt><strong>positional_metadata</strong><span class="classifier">pd.DataFrame consumable, optional</span></dt><dd><p>Arbitrary metadata which applies to each position in the MSA. Must be
able to be passed directly to <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> constructor. Each column
of metadata must be the same length as the number of positions in the
MSA. A shallow copy of the positional metadata will be made.</p>
</dd>
<dt><strong>minter</strong><span class="classifier">callable or metadata key, optional</span></dt><dd><p>If provided, defines an index label for each sequence in <cite>sequences</cite>.
Can either be a callable accepting a single argument (each sequence) or
a key into each sequence’s <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute. Note that <cite>minter</cite>
cannot be combined with <cite>index</cite>.</p>
</dd>
<dt><strong>index</strong><span class="classifier">pd.Index consumable, optional</span></dt><dd><p>Index containing labels for <cite>sequences</cite>. Must be the same length as
<cite>sequences</cite>. Must be able to be passed directly to <code class="docutils literal notranslate"><span class="pre">pd.Index</span></code>
constructor. Note that <cite>index</cite> cannot be combined with <cite>minter</cite> and the
contents of <cite>index</cite> must be hashable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If <cite>minter</cite> and <cite>index</cite> are both provided.</p>
</dd>
<dt>ValueError</dt><dd><p>If <cite>index</cite> is not the same length as <cite>sequences</cite>.</p>
</dd>
<dt>TypeError</dt><dd><p>If <cite>sequences</cite> contains an object that isn’t a <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code>.</p>
</dd>
<dt>TypeError</dt><dd><p>If <cite>sequences</cite> does not contain exactly the same type of
<code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code> objects.</p>
</dd>
<dt>ValueError</dt><dd><p>If <cite>sequences</cite> does not contain <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code> objects of the
same length.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.sequence.DNA</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.sequence.RNA</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.sequence.Protein</span></code></a></dt><dd></dd>
<dt><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></dt><dd></dd>
<dt><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.html#pandas.Index" title="(in pandas v2.3.3)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.Index</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.alignment.TabularMSA.reassign_index.html#skbio.alignment.TabularMSA.reassign_index" title="skbio.alignment.TabularMSA.reassign_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reassign_index</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>If neither <cite>minter</cite> nor <cite>index</cite> are provided, default index labels will be
used: <code class="docutils literal notranslate"><span class="pre">pd.RangeIndex(start=0,</span> <span class="pre">stop=len(sequences),</span> <span class="pre">step=1)</span></code>.</p>
<p class="rubric">Examples</p>
<p>Create a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object with three DNA sequences and four positions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AG-T&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;-C-T&#39;</span><span class="p">)</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">AG-T</span>
<span class="go">-C-T</span>
</pre></div>
</div>
<p>Since <cite>minter</cite> or <cite>index</cite> wasn’t provided, the MSA has default index
labels:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">RangeIndex(start=0, stop=3, step=1)</span>
</pre></div>
</div>
<p>Create an MSA with metadata, positional metadata, and non-default index
labels:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;seq3&#39;</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;msa-id&#39;</span><span class="p">},</span>
<span class="gp">... </span>                 <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;prob&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">--------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;msa-id&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;prob&#39;: &lt;dtype: int64&gt;</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 4</span>
<span class="go">--------------------------</span>
<span class="go">ACGT</span>
<span class="go">AG-T</span>
<span class="go">-C-T</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;seq1&#39;, &#39;seq2&#39;, &#39;seq3&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.default_write_format" title="skbio.alignment.TabularMSA.default_write_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_write_format</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.dtype" title="skbio.alignment.TabularMSA.dtype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dtype</span></code></a></p></td>
<td><p>Data type of the stored sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.iloc" title="skbio.alignment.TabularMSA.iloc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iloc</span></code></a></p></td>
<td><p>Slice the MSA on either axis by index position.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.index" title="skbio.alignment.TabularMSA.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a></p></td>
<td><p>Index containing labels along the sequence axis.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.loc" title="skbio.alignment.TabularMSA.loc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">loc</span></code></a></p></td>
<td><p>Slice the MSA on first axis by index label, second axis by position.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.shape" title="skbio.alignment.TabularMSA.shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shape</span></code></a></p></td>
<td><p>Number of sequences (rows) and positions (columns).</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Attributes (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.MetadataMixin.html#skbio.metadata.MetadataMixin.metadata" title="skbio.metadata.MetadataMixin.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">dict</span></code> containing metadata which applies to the entire object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.metadata.PositionalMetadataMixin.html#skbio.metadata.PositionalMetadataMixin.positional_metadata" title="skbio.metadata.PositionalMetadataMixin.positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_metadata</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> containing metadata along an axis.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.append.html#skbio.alignment.TabularMSA.append" title="skbio.alignment.TabularMSA.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a>(sequence[, minter, index, reset_index])</p></td>
<td><p>Append a sequence to the MSA without recomputing alignment.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.consensus.html#skbio.alignment.TabularMSA.consensus" title="skbio.alignment.TabularMSA.consensus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">consensus</span></code></a>()</p></td>
<td><p>Compute the majority consensus sequence for this MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.conservation.html#skbio.alignment.TabularMSA.conservation" title="skbio.alignment.TabularMSA.conservation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">conservation</span></code></a>([metric, degenerate_mode, gap_mode])</p></td>
<td><p>Apply metric to compute conservation for all alignment positions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.extend.html#skbio.alignment.TabularMSA.extend" title="skbio.alignment.TabularMSA.extend"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend</span></code></a>(sequences[, minter, index, reset_index])</p></td>
<td><p>Extend this MSA with sequences without recomputing alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.from_dict.html#skbio.alignment.TabularMSA.from_dict" title="skbio.alignment.TabularMSA.from_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_dict</span></code></a>(dictionary)</p></td>
<td><p>Create a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.from_path_seqs.html#skbio.alignment.TabularMSA.from_path_seqs" title="skbio.alignment.TabularMSA.from_path_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_path_seqs</span></code></a>(path, seqs)</p></td>
<td><p>Create a tabular MSA from an alignment path and the original sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.gap_frequencies.html#skbio.alignment.TabularMSA.gap_frequencies" title="skbio.alignment.TabularMSA.gap_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_frequencies</span></code></a>([axis, relative])</p></td>
<td><p>Compute frequency of gap characters across an axis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.iter_positions.html#skbio.alignment.TabularMSA.iter_positions" title="skbio.alignment.TabularMSA.iter_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_positions</span></code></a>([reverse, ignore_metadata])</p></td>
<td><p>Iterate over positions (columns) in the MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.join.html#skbio.alignment.TabularMSA.join" title="skbio.alignment.TabularMSA.join"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join</span></code></a>(other[, how])</p></td>
<td><p>Join this MSA with another by sequence (horizontally).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.read.html#skbio.alignment.TabularMSA.read" title="skbio.alignment.TabularMSA.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a>([format])</p></td>
<td><p>Create a new <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> instance from a file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.reassign_index.html#skbio.alignment.TabularMSA.reassign_index" title="skbio.alignment.TabularMSA.reassign_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reassign_index</span></code></a>([mapping, minter])</p></td>
<td><p>Reassign index labels to sequences in this MSA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.sort.html#skbio.alignment.TabularMSA.sort" title="skbio.alignment.TabularMSA.sort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code></a>([level, ascending])</p></td>
<td><p>Sort sequences by index label in-place.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.to_dict.html#skbio.alignment.TabularMSA.to_dict" title="skbio.alignment.TabularMSA.to_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dict</span></code></a>()</p></td>
<td><p>Create a <code class="docutils literal notranslate"><span class="pre">dict</span></code> from this <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.TabularMSA.write.html#skbio.alignment.TabularMSA.write" title="skbio.alignment.TabularMSA.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>(file[, format])</p></td>
<td><p>Write an instance of <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> to a file.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.MetadataMixin.has_metadata.html#skbio.metadata.MetadataMixin.has_metadata" title="skbio.metadata.MetadataMixin.has_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.metadata.PositionalMetadataMixin.has_positional_metadata.html#skbio.metadata.PositionalMetadataMixin.has_positional_metadata" title="skbio.metadata.PositionalMetadataMixin.has_positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_positional_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has positional metadata.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__bool__" title="skbio.alignment.TabularMSA.__bool__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__bool__</span></code></a>()</p></td>
<td><p>Boolean indicating whether the MSA is empty or not.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__contains__" title="skbio.alignment.TabularMSA.__contains__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__contains__</span></code></a>(label)</p></td>
<td><p>Determine if an index label is in this MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__copy__" title="skbio.alignment.TabularMSA.__copy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__copy__</span></code></a>()</p></td>
<td><p>Return a shallow copy of this MSA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__deepcopy__" title="skbio.alignment.TabularMSA.__deepcopy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__deepcopy__</span></code></a>(memo)</p></td>
<td><p>Return a deep copy of this MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__eq__" title="skbio.alignment.TabularMSA.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code></a>(other)</p></td>
<td><p>Determine if this MSA is equal to another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__getitem__" title="skbio.alignment.TabularMSA.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a>(indexable)</p></td>
<td><p>Slice the MSA on either axis.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__iter__" title="skbio.alignment.TabularMSA.__iter__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__iter__</span></code></a>()</p></td>
<td><p>Iterate over sequences in the MSA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__len__" title="skbio.alignment.TabularMSA.__len__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__len__</span></code></a>()</p></td>
<td><p>Return number of sequences in the MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__ne__" title="skbio.alignment.TabularMSA.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ne__</span></code></a>(other)</p></td>
<td><p>Determine if this MSA is not equal to another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__reversed__" title="skbio.alignment.TabularMSA.__reversed__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__reversed__</span></code></a>()</p></td>
<td><p>Iterate in reverse order over sequences in the MSA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.TabularMSA.__str__" title="skbio.alignment.TabularMSA.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__str__</span></code></a>()</p></td>
<td><p>Return string summary of this MSA.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ge__</span></code>(value, /)</p></td>
<td><p>Return self&gt;=value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getstate__</span></code>(/)</p></td>
<td><p>Helper for pickle.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__gt__</span></code>(value, /)</p></td>
<td><p>Return self&gt;value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__le__</span></code>(value, /)</p></td>
<td><p>Return self&lt;=value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__lt__</span></code>(value, /)</p></td>
<td><p>Return self&lt;value.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Details</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.default_write_format">
<span class="sig-name descname"><span class="pre">default_write_format</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'fasta'</span></em><a class="headerlink" href="#skbio.alignment.TabularMSA.default_write_format" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.dtype">
<span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#skbio.alignment.TabularMSA.dtype" title="Link to this definition">#</a></dt>
<dd><p>Data type of the stored sequences.</p>
<p class="rubric">Notes</p>
<p>This property is not writeable.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">dtype</span>
<span class="go">&lt;class &#39;skbio.sequence._dna.DNA&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">dtype</span> <span class="ow">is</span> <span class="n">DNA</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.iloc">
<span class="sig-name descname"><span class="pre">iloc</span></span><a class="headerlink" href="#skbio.alignment.TabularMSA.iloc" title="Link to this definition">#</a></dt>
<dd><p>Slice the MSA on either axis by index position.</p>
<p>This will return an object with the following interface:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">seq_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">seq_idx</span><span class="p">,</span> <span class="n">pos_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">)[</span><span class="n">seq_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;position&#39;</span><span class="p">)[</span><span class="n">pos_idx</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_idx</strong><span class="classifier">int, slice, iterable (int and slice), 1D array_like (bool)</span></dt><dd><p>Slice the first axis of the MSA. When this value is a scalar, a
sequence of <code class="docutils literal notranslate"><span class="pre">msa.dtype</span></code> will be returned. This may be further
sliced by <cite>pos_idx</cite>.</p>
</dd>
<dt><strong>pos_idx</strong><span class="classifier">(same as seq_idx), optional</span></dt><dd><p>Slice the second axis of the MSA. When this value is a scalar, a
sequence of type <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> will be returned.
This represents a column of the MSA and may have been additionally
sliced by <cite>seq_idx</cite>.</p>
</dd>
<dt><strong>axis</strong><span class="classifier">{‘sequence’, ‘position’, 0, 1, None}, optional</span></dt><dd><p>Limit the axis to slice on. When set, a tuple as the argument will
no longer be split into <cite>seq_idx</cite> and <cite>pos_idx</cite>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>TabularMSA, GrammaredSequence, Sequence</dt><dd><p>A <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> is returned when <cite>seq_idx</cite> and <cite>pos_idx</cite> are
non-scalars. A <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code> of type <code class="docutils literal notranslate"><span class="pre">msa.dtype</span></code> is
returned when <cite>seq_idx</cite> is a scalar (this object will match the
dtype of the MSA). A <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> is returned when <cite>seq_idx</cite> is
non-scalar and <cite>pos_idx</cite> is scalar.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.__getitem__" title="skbio.alignment.TabularMSA.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.loc" title="skbio.alignment.TabularMSA.loc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">loc</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>If the slice operation results in a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> without any
sequences, the MSA’s <code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code> will be unset.</p>
<p class="rubric">Examples</p>
<p>First we need to set up an MSA to slice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;A-GT&quot;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;AC-T&quot;</span><span class="p">),</span>
<span class="gp">... </span>                  <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGA&quot;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 4</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">A-GT</span>
<span class="go">AC-T</span>
<span class="go">ACGA</span>
</pre></div>
</div>
<p>When we slice by a scalar we get the original sequence back out of the
MSA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: True</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 A-GT</span>
</pre></div>
</div>
<p>Similarly when we slice the second axis by a scalar we get a column of
the MSA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">-------------</span>
<span class="go">0 C-CC</span>
</pre></div>
</div>
<p>Note: we return an <code class="docutils literal notranslate"><span class="pre">skbio.Sequence</span></code> object because the column of an
alignment has no biological meaning and many operations defined for the
MSA’s sequence <cite>dtype</cite> would be meaningless.</p>
<p>When we slice both axes by a scalar, operations are applied left to
right:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 0.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 A</span>
</pre></div>
</div>
<p>In other words, it exactly matches slicing the resulting sequence
object directly:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 0.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 A</span>
</pre></div>
</div>
<p>When our slice is non-scalar we get back an MSA of the same <cite>dtype</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">AC-T</span>
</pre></div>
</div>
<p>We can similarly slice out a column of that:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 2</span>
<span class="go">-------------</span>
<span class="go">0 G-</span>
</pre></div>
</div>
<p>Slice syntax works as well:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">A-GT</span>
<span class="go">AC-T</span>
</pre></div>
</div>
<p>We can also use boolean vectors:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">],</span> <span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 1</span>
<span class="go">---------------------</span>
<span class="go">G</span>
<span class="go">G</span>
</pre></div>
</div>
<p>Here we sliced the first axis by a boolean vector, but then restricted
the columns to a single column. Because the second axis was given a
nonscalar we still receive an MSA even though only one column is
present.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.index">
<span class="sig-name descname"><span class="pre">index</span></span><a class="headerlink" href="#skbio.alignment.TabularMSA.index" title="Link to this definition">#</a></dt>
<dd><p>Index containing labels along the sequence axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>pd.Index</dt><dd><p>Index containing sequence labels.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.alignment.TabularMSA.reassign_index.html#skbio.alignment.TabularMSA.reassign_index" title="skbio.alignment.TabularMSA.reassign_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reassign_index</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>This property can be set and deleted. Deleting the index will reset the
index to the <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> constructor’s default.</p>
<p class="rubric">Examples</p>
<p>Create a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object with sequences labeled by sequence
identifier:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;a&#39;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;b&#39;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;c&#39;</span><span class="p">})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">minter</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Retrieve index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;a&#39;, &#39;b&#39;, &#39;c&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<p>Set index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;seq3&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;seq1&#39;, &#39;seq2&#39;, &#39;seq3&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<p>Deleting the index resets it to the <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> constructor’s
default:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">RangeIndex(start=0, stop=3, step=1)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.loc">
<span class="sig-name descname"><span class="pre">loc</span></span><a class="headerlink" href="#skbio.alignment.TabularMSA.loc" title="Link to this definition">#</a></dt>
<dd><p>Slice the MSA on first axis by index label, second axis by position.</p>
<p>This will return an object with the following interface:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">seq_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">seq_idx</span><span class="p">,</span> <span class="n">pos_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">)[</span><span class="n">seq_idx</span><span class="p">]</span>
<span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;position&#39;</span><span class="p">)[</span><span class="n">pos_idx</span><span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_idx</strong><span class="classifier">label, slice, 1D array_like (bool or label)</span></dt><dd><p>Slice the first axis of the MSA. When this value is a scalar, a
sequence of <code class="docutils literal notranslate"><span class="pre">msa.dtype</span></code> will be returned. This may be further
sliced by <cite>pos_idx</cite>.</p>
</dd>
<dt><strong>pos_idx</strong><span class="classifier">(same as seq_idx), optional</span></dt><dd><p>Slice the second axis of the MSA. When this value is a scalar, a
sequence of type <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> will be returned.
This represents a column of the MSA and may have been additionally
sliced by <cite>seq_idx</cite>.</p>
</dd>
<dt><strong>axis</strong><span class="classifier">{‘sequence’, ‘position’, 0, 1, None}, optional</span></dt><dd><p>Limit the axis to slice on. When set, a tuple as the argument will
no longer be split into <cite>seq_idx</cite> and <cite>pos_idx</cite>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>TabularMSA, GrammaredSequence, Sequence</dt><dd><p>A <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> is returned when <cite>seq_idx</cite> and <cite>pos_idx</cite> are
non-scalars. A <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code> of type <code class="docutils literal notranslate"><span class="pre">msa.dtype</span></code> is
returned when <cite>seq_idx</cite> is a scalar (this object will match the
dtype of the MSA). A <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> is returned when <cite>seq_idx</cite> is
non-scalar and <cite>pos_idx</cite> is scalar.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.iloc" title="skbio.alignment.TabularMSA.iloc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iloc</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.__getitem__" title="skbio.alignment.TabularMSA.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>If the slice operation results in a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> without any
sequences, the MSA’s <code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code> will be unset.</p>
<p>When the MSA’s index is a <code class="docutils literal notranslate"><span class="pre">pd.MultiIndex</span></code> a tuple may be given to
<cite>seq_idx</cite> to indicate the slicing operations to perform on each
component index.</p>
<p class="rubric">Examples</p>
<p>First we need to set up an MSA to slice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;A-GT&quot;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;AC-T&quot;</span><span class="p">),</span>
<span class="gp">... </span>                  <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGA&quot;</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 4</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">A-GT</span>
<span class="go">AC-T</span>
<span class="go">ACGA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;a&#39;, &#39;b&#39;, &#39;c&#39;, &#39;d&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<p>When we slice by a scalar we get the original sequence back out of the
MSA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;b&#39;</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: True</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 A-GT</span>
</pre></div>
</div>
<p>Similarly when we slice the second axis by a scalar we get a column of
the MSA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">-------------</span>
<span class="go">0 C-CC</span>
</pre></div>
</div>
<p>Note: we return an <code class="docutils literal notranslate"><span class="pre">skbio.Sequence</span></code> object because the column of an
alignment has no biological meaning and many operations defined for the
MSA’s sequence <cite>dtype</cite> would be meaningless.</p>
<p>When we slice both axes by a scalar, operations are applied left to
right:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 0.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 A</span>
</pre></div>
</div>
<p>In other words, it exactly matches slicing the resulting sequence
object directly:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 0.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 A</span>
</pre></div>
</div>
<p>When our slice is non-scalar we get back an MSA of the same <cite>dtype</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">]]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">AC-T</span>
</pre></div>
</div>
<p>We can similarly slice out a column of that:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">],</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 2</span>
<span class="go">-------------</span>
<span class="go">0 G-</span>
</pre></div>
</div>
<p>Slice syntax works as well:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[:</span><span class="s1">&#39;c&#39;</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">A-GT</span>
<span class="go">AC-T</span>
</pre></div>
</div>
<p>Notice how the end label is included in the results. This is different
from how positional slices behave:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">],</span> <span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 1</span>
<span class="go">---------------------</span>
<span class="go">G</span>
<span class="go">G</span>
</pre></div>
</div>
<p>Here we sliced the first axis by a boolean vector, but then restricted
the columns to a single column. Because the second axis was given a
nonscalar we still receive an MSA even though only one column is
present.</p>
<p>Duplicate labels can be an unfortunate reality in the real world,
however <cite>loc</cite> is capable of handling this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Notice how the label ‘a’ happens twice. If we were to access ‘a’ we get
back an MSA with both sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">ACGT</span>
<span class="go">A-GT</span>
</pre></div>
</div>
<p>Remember that <cite>iloc</cite> can always be used to differentiate sequences with
duplicate labels.</p>
<p>More advanced slicing patterns are possible with different index types.</p>
<p>Let’s use a <cite>pd.MultiIndex</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)]</span>
</pre></div>
</div>
<p>Here we will explicitly set the axis that we are slicing by to make
things easier to read:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">)[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 50.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 ACGT</span>
</pre></div>
</div>
<p>This selected the first sequence because the complete label was
provided. In other words <cite>(‘a’, 0)</cite> was treated as a scalar for this
index.</p>
<p>We can also slice along the component indices of the multi-index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;sequence&#39;</span><span class="p">)[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">A-GT</span>
<span class="go">ACGA</span>
</pre></div>
</div>
<p>If we were to do that again without the <cite>axis</cite> argument, it would look
like this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="nb">slice</span><span class="p">(</span><span class="kc">None</span><span class="p">),</span> <span class="mi">1</span><span class="p">),</span> <span class="o">...</span><span class="p">]</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 4</span>
<span class="go">---------------------</span>
<span class="go">A-GT</span>
<span class="go">ACGA</span>
</pre></div>
</div>
<p>Notice how we needed to specify the second axis. If we had left that
out we would have simply gotten the 2nd column back instead. We also
lost the syntactic sugar for slice objects. These are a few of the
reasons specifying the <cite>axis</cite> preemptively can be useful.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.shape">
<span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#skbio.alignment.TabularMSA.shape" title="Link to this definition">#</a></dt>
<dd><p>Number of sequences (rows) and positions (columns).</p>
<p class="rubric">Notes</p>
<p>This property is not writeable.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
</pre></div>
</div>
<p>Create a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object with 2 sequences and 3 positions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span>
<span class="go">Shape(sequence=2, position=3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Dimensions can be accessed by index or by name:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">shape</span><span class="o">.</span><span class="n">position</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__bool__">
<span class="sig-name descname"><span class="pre">__bool__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L856"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__bool__" title="Link to this definition">#</a></dt>
<dd><p>Boolean indicating whether the MSA is empty or not.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>bool</dt><dd><p><code class="docutils literal notranslate"><span class="pre">False</span></code> if there are no sequences, OR if there are no positions
(i.e., all sequences are empty). <code class="docutils literal notranslate"><span class="pre">True</span></code> otherwise.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
</pre></div>
</div>
<p>MSA with sequences and positions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>No sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
<p>No positions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__contains__">
<span class="sig-name descname"><span class="pre">__contains__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L892"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__contains__" title="Link to this definition">#</a></dt>
<dd><p>Determine if an index label is in this MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>label</strong><span class="classifier">hashable</span></dt><dd><p>Label to search for in this MSA.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether <cite>label</cite> is in this MSA.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;l1&#39;</span><span class="p">,</span> <span class="s1">&#39;l2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;l1&#39;</span> <span class="ow">in</span> <span class="n">msa</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;l2&#39;</span> <span class="ow">in</span> <span class="n">msa</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;l3&#39;</span> <span class="ow">in</span> <span class="n">msa</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__copy__">
<span class="sig-name descname"><span class="pre">__copy__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L1079"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__copy__" title="Link to this definition">#</a></dt>
<dd><p>Return a shallow copy of this MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>TabularMSA</dt><dd><p>Shallow copy of this MSA. Sequence objects will be shallow-copied.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.__deepcopy__" title="skbio.alignment.TabularMSA.__deepcopy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__deepcopy__</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">copy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="o">==</span> <span class="n">msa</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="ow">is</span> <span class="n">msa</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L1110"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__deepcopy__" title="Link to this definition">#</a></dt>
<dd><p>Return a deep copy of this MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>TabularMSA</dt><dd><p>Deep copy of this MSA. Sequence objects will be deep-copied.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.__copy__" title="skbio.alignment.TabularMSA.__copy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__copy__</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">copy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="o">==</span> <span class="n">msa</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa_copy</span> <span class="ow">is</span> <span class="n">msa</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L988"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__eq__" title="Link to this definition">#</a></dt>
<dd><p>Determine if this MSA is equal to another.</p>
<p><code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> objects are equal if their sequences, index, metadata,
and positional metadata are equal.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong><span class="classifier">TabularMSA</span></dt><dd><p>MSA to test for equality against.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether this MSA is equal to <cite>other</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">RNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">msa</span>
<span class="go">True</span>
</pre></div>
</div>
<p>MSAs with different sequence characters are not equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;--G&#39;</span><span class="p">)])</span>
<span class="go">False</span>
</pre></div>
</div>
<p>MSAs with different types of sequences (different <code class="docutils literal notranslate"><span class="pre">dtype</span></code>) are not
equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="go">False</span>
</pre></div>
</div>
<p>MSAs with different sequence metadata are not equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;a&#39;</span><span class="p">}),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="go">False</span>
</pre></div>
</div>
<p>MSAs with different index labels are not equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)],</span> <span class="n">minter</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
<p>MSAs with different metadata are not equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)],</span>
<span class="gp">... </span>                  <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;msa-id&#39;</span><span class="p">})</span>
<span class="go">False</span>
</pre></div>
</div>
<p>MSAs with different positional metadata are not equal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">==</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)],</span>
<span class="gp">... </span>                  <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;prob&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]})</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indexable</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L1142"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>Slice the MSA on either axis.</p>
<p>This is a pass-through for <a class="reference internal" href="#skbio.alignment.TabularMSA.iloc" title="skbio.alignment.TabularMSA.iloc"><code class="xref py py-func docutils literal notranslate"><span class="pre">skbio.alignment.TabularMSA.iloc</span></code></a>.
Please refer to the associated documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.iloc" title="skbio.alignment.TabularMSA.iloc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iloc</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.loc" title="skbio.alignment.TabularMSA.loc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">loc</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>Axis restriction is not possible for this method.</p>
<p>To slice by labels, use <code class="docutils literal notranslate"><span class="pre">loc</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L944"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__iter__" title="Link to this definition">#</a></dt>
<dd><p>Iterate over sequences in the MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Yields<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>GrammaredSequence</dt><dd><p>Each sequence in the order they are stored in the MSA.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">msa</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&#39;ACG&#39;</span>
<span class="go">&#39;AC-&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L919"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__len__" title="Link to this definition">#</a></dt>
<dd><p>Return number of sequences in the MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>Number of sequences in the MSA (i.e., size of the 1st dimension).</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This is equivalent to <code class="docutils literal notranslate"><span class="pre">msa.shape[0]</span></code>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__ne__">
<span class="sig-name descname"><span class="pre">__ne__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L1056"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__ne__" title="Link to this definition">#</a></dt>
<dd><p>Determine if this MSA is not equal to another.</p>
<p><code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> objects are not equal if their sequences, index,
metadata, or positional metadata are not equal.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong><span class="classifier">TabularMSA</span></dt><dd><p>MSA to test for inequality against.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether this MSA is not equal to <cite>other</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#skbio.alignment.TabularMSA.__eq__" title="skbio.alignment.TabularMSA.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code></a></dt><dd></dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__reversed__">
<span class="sig-name descname"><span class="pre">__reversed__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L964"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__reversed__" title="Link to this definition">#</a></dt>
<dd><p>Iterate in reverse order over sequences in the MSA.</p>
<dl class="field-list simple">
<dt class="field-odd">Yields<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>GrammaredSequence</dt><dd><p>Each sequence in reverse order from how they are stored in the MSA.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACG&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AC-&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">msa</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&#39;AC-&#39;</span>
<span class="go">&#39;ACG&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.TabularMSA.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.1.post1/skbio/alignment/_tabular_msa.py#L984"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.TabularMSA.__str__" title="Link to this definition">#</a></dt>
<dd><p>Return string summary of this MSA.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../alignment.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Sequence Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.alignment.TabularMSA.append.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.alignment.TabularMSA.append</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA"><code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.default_write_format"><code class="docutils literal notranslate"><span class="pre">TabularMSA.default_write_format</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.dtype"><code class="docutils literal notranslate"><span class="pre">TabularMSA.dtype</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.iloc"><code class="docutils literal notranslate"><span class="pre">TabularMSA.iloc</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.index"><code class="docutils literal notranslate"><span class="pre">TabularMSA.index</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.loc"><code class="docutils literal notranslate"><span class="pre">TabularMSA.loc</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.shape"><code class="docutils literal notranslate"><span class="pre">TabularMSA.shape</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__bool__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__bool__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__contains__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__contains__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__copy__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__copy__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__deepcopy__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__deepcopy__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__eq__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__eq__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__getitem__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__getitem__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__iter__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__iter__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__len__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__len__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__ne__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__ne__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__reversed__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__reversed__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.TabularMSA.__str__"><code class="docutils literal notranslate"><span class="pre">TabularMSA.__str__</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>