
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Store an alignment path between sequences." />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.alignment.AlignPath &#8212; scikit-bio 0.6.4-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=dcd4525e"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.alignment.AlignPath';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.alignment.AlignPath.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.alignment.AlignPath.from_aligned" href="skbio.alignment.AlignPath.from_aligned.html" />
    <link rel="prev" title="skbio.alignment.TabularMSA.write" href="skbio.alignment.TabularMSA.write.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.4-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.4-dev documentation"/>
    <img src="../_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.6.4-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-external" href="../../../roadmap.html">
    Roadmap
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../roadmap.html">
    Roadmap
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.TabularMSA.html">skbio.alignment.TabularMSA</a></li>
</ul>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">skbio.alignment.AlignPath</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.from_aligned.html">skbio.alignment.AlignPath.from_aligned</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.from_bits.html">skbio.alignment.AlignPath.from_bits</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.from_coordinates.html">skbio.alignment.AlignPath.from_coordinates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.from_indices.html">skbio.alignment.AlignPath.from_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.from_tabular.html">skbio.alignment.AlignPath.from_tabular</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.to_aligned.html">skbio.alignment.AlignPath.to_aligned</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.to_bits.html">skbio.alignment.AlignPath.to_bits</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.to_coordinates.html">skbio.alignment.AlignPath.to_coordinates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.alignment.AlignPath.to_indices.html">skbio.alignment.AlignPath.to_indices</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.PairAlignPath.html">skbio.alignment.PairAlignPath</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align.html">skbio.alignment.pair_align</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_nucl.html">skbio.alignment.pair_align_nucl</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_prot.html">skbio.alignment.pair_align_prot</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.align_score.html">skbio.alignment.align_score</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_nucleotide.html">skbio.alignment.global_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_protein.html">skbio.alignment.global_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align.html">skbio.alignment.global_pairwise_align</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_nucleotide.html">skbio.alignment.local_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_protein.html">skbio.alignment.local_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align.html">skbio.alignment.local_pairwise_align</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../alignment.html" class="nav-link">Sequence Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">skbio.alignment.AlignPath</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-alignment-alignpath">
<h1>skbio.alignment.AlignPath<a class="headerlink" href="#skbio-alignment-alignpath" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">skbio.alignment.</span></span><span class="sig-name descname"><span class="pre">AlignPath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lengths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">states</span></span></em>, <em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">ranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/main/skbio/alignment/_path.py#L39"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.AlignPath" title="Link to this definition">#</a></dt>
<dd><p>Store an alignment path between sequences.</p>
<p>It is a compact data structure that stores the operations of sequence alignment:
inserting gaps into designated positions between subsequences. It does not store
the sequence data.</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl>
<dt><strong>lengths</strong><span class="classifier">array_like of int of shape (n_segments,)</span></dt><dd><p>Length of each segment in the alignment.</p>
</dd>
<dt><strong>states</strong><span class="classifier">array_like of uint8 of shape (n_segments,) or (n_packs, n_segments)</span></dt><dd><p>Packed bits representing character (0) or gap (1) status per sequence per
segment in the alignment.</p>
</dd>
<dt><strong>ranges</strong><span class="classifier">array_like of int of shape (n_sequences, 2), optional</span></dt><dd><p>Start and stop positions of each sequence in the alignment.</p>
</dd>
<dt><strong>starts</strong><span class="classifier">array_like of int of shape (n_sequences,), optional</span></dt><dd><p>Start position of each sequence in the alignment.</p>
</dd>
<dt><strong>stops</strong><span class="classifier">array_like of int of shape (n_sequences,), optional</span></dt><dd><p>Stop position of each sequence in the alignment.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The ranges can be defined by supplying <code class="docutils literal notranslate"><span class="pre">starts</span></code>, <code class="docutils literal notranslate"><span class="pre">stops</span></code> or <code class="docutils literal notranslate"><span class="pre">ranges</span></code>.
With either of the first two, the program will locate the other side based
on <code class="docutils literal notranslate"><span class="pre">lengths</span></code> and <code class="docutils literal notranslate"><span class="pre">states</span></code>. For the last (<code class="docutils literal notranslate"><span class="pre">ranges</span></code>), the program will
NOT validate the correctness of the values but simply take them.</p>
</div>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.alignment.TabularMSA</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The underlying data structure of the <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> class efficiently represents a
sequence alignment as two equal-length vectors: <code class="docutils literal notranslate"><span class="pre">lengths</span></code> and <code class="docutils literal notranslate"><span class="pre">states</span></code>. The
lengths vector contains the lengths of individual segments of the alignment with
consistent gap status. The states vector contains the packed bits of gap (1) and
and character (0) status for each position in the alignment.</p>
<p>This data structure is calculated by performing
<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Run-length_encoding">run-length encoding</a> (RLE) on the alignment,
considering each segment with consistent gap status to be a unit in the encoding.
This resembles the CIGAR string (see <a class="reference internal" href="skbio.alignment.PairAlignPath.to_cigar.html#skbio.alignment.PairAlignPath.to_cigar" title="skbio.alignment.PairAlignPath.to_cigar"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PairAlignPath.to_cigar</span></code></a>), and is
generalized to an arbitrary number of sequences.</p>
<p>An <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> object is detached from the original or aligned sequences and is
highly memory efficient. The more similar the sequences are (i.e., the fewer gaps),
the more compact this data structure is. In the worst case, this object consumes
1/8 memory space of the aligned sequences.</p>
<p>This class permits fully vectorized operations and enables efficient conversion
between various formats such as aligned sequences, indices (Biotite), and
coordinates (Biopython).</p>
<p>In addition to alignment operations, an <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> object also stores the ranges
(start and stop positions) of the aligned region in the original sequences. This
facilitates extraction of aligned sequences. The positions are 0-based and
half-open, consistent with Python indexing, and compatible with the
<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/BED_(file_format)">BED format</a>.</p>
<p class="rubric">Examples</p>
<p>Create an <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> object from a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object with three DNA
sequences and 20 positions.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">AlignPath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span>
<span class="gp">... </span>   <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGGTCGTAACGCGTA---CA&#39;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CAG--GTAAG-CATACCTCA&#39;</span><span class="p">),</span>
<span class="gp">... </span>   <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGGTCGTCAC-TGTACACTA&#39;</span><span class="p">),</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">AlignPath</span><span class="o">.</span><span class="n">from_tabular</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span>
<span class="go">&lt;AlignPath, sequences: 3, positions: 20, segments: 7&gt;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">lengths</span>
<span class="go">array([3, 2, 5, 1, 4, 3, 2])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">states</span>
<span class="go">array([[0, 2, 0, 6, 0, 1, 0]], dtype=uint8)</span>
</pre></div>
</div>
<p>In the above example, the first three positions of the alignment contain no gaps,
so the first value in the lengths array is 3, and that in the states array is 0.
The fourth segment, which has length 1, would have gap status (0, 1, 1), which
then becomes 6 after bit packing.</p>
<p>An <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> object is rarely created from scratch. But one still could, like:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">AlignPath</span><span class="p">(</span><span class="n">lengths</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="n">states</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="n">starts</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
<p>The parameter <code class="docutils literal notranslate"><span class="pre">starts</span></code> defines the start positions of the aligned region of each
sequence. The program will automatically calculate the stop positions.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">starts</span>
<span class="go">array([5, 1, 0])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">stops</span>
<span class="go">array([22, 18, 19])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">ranges</span>
<span class="go">array([[ 5, 22],</span>
<span class="go">       [ 1, 18],</span>
<span class="go">       [ 0, 19]])</span>
</pre></div>
</div>
<p>With the ranges, one can extract aligned subsequences from the original sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;NNNNNCGGTCGTAACGCGTACANNNNNNN&quot;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;NCAGGTAAGCATACCTCA&quot;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;CGGTCGTCACTGTACACTANN&quot;</span><span class="p">),</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">path</span><span class="o">.</span><span class="n">ranges</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">])</span>
<span class="go">CGGTCGTAACGCGTACA</span>
<span class="go">CAGGTAAGCATACCTCA</span>
<span class="go">CGGTCGTCACTGTACACTA</span>
</pre></div>
</div>
<p>Alternatively, one can extract the aligned sequences with gap characters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">path</span><span class="o">.</span><span class="n">to_aligned</span><span class="p">(</span><span class="n">seqs</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="go">CGGTCGTAACGCGTA---CA</span>
<span class="go">CAG--GTAAG-CATACCTCA</span>
<span class="go">CGGTCGTCAC-TGTACACTA</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.lengths" title="skbio.alignment.AlignPath.lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lengths</span></code></a></p></td>
<td><p>Array of lengths of segments in alignment path.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.ranges" title="skbio.alignment.AlignPath.ranges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ranges</span></code></a></p></td>
<td><p>Array of (start, stop) positions of sequences in the alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.shape" title="skbio.alignment.AlignPath.shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shape</span></code></a></p></td>
<td><p>Number of sequences (rows) and positions (columns).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.starts" title="skbio.alignment.AlignPath.starts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">starts</span></code></a></p></td>
<td><p>Array of start positions of sequences in the alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.states" title="skbio.alignment.AlignPath.states"><code class="xref py py-obj docutils literal notranslate"><span class="pre">states</span></code></a></p></td>
<td><p>Array of gap status of segments in alignment path.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.stops" title="skbio.alignment.AlignPath.stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stops</span></code></a></p></td>
<td><p>Array of stop positions of sequences in the alignment.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.from_aligned.html#skbio.alignment.AlignPath.from_aligned" title="skbio.alignment.AlignPath.from_aligned"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_aligned</span></code></a>(aln[, gap_chars, starts])</p></td>
<td><p>Create an alignment path from aligned sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.from_bits.html#skbio.alignment.AlignPath.from_bits" title="skbio.alignment.AlignPath.from_bits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_bits</span></code></a>(bits[, starts])</p></td>
<td><p>Create an alignment path from a bit array (0 - character, 1 - gap).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.from_coordinates.html#skbio.alignment.AlignPath.from_coordinates" title="skbio.alignment.AlignPath.from_coordinates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_coordinates</span></code></a>(coords)</p></td>
<td><p>Create an alignment path from an array of segment coordinates.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.from_indices.html#skbio.alignment.AlignPath.from_indices" title="skbio.alignment.AlignPath.from_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_indices</span></code></a>(indices[, gap])</p></td>
<td><p>Create an alignment path from character indices in the original sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.from_tabular.html#skbio.alignment.AlignPath.from_tabular" title="skbio.alignment.AlignPath.from_tabular"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_tabular</span></code></a>(msa[, starts])</p></td>
<td><p>Create an alignment path from a <cite>TabularMSA</cite> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.to_aligned.html#skbio.alignment.AlignPath.to_aligned" title="skbio.alignment.AlignPath.to_aligned"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_aligned</span></code></a>(seqs[, gap_char, flanking])</p></td>
<td><p>Extract aligned regions from original sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.to_bits.html#skbio.alignment.AlignPath.to_bits" title="skbio.alignment.AlignPath.to_bits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_bits</span></code></a>([expand])</p></td>
<td><p>Unpack the alignment path into an array of bits.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.to_coordinates.html#skbio.alignment.AlignPath.to_coordinates" title="skbio.alignment.AlignPath.to_coordinates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_coordinates</span></code></a>()</p></td>
<td><p>Generate an array of segment coordinates in the original sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.alignment.AlignPath.to_indices.html#skbio.alignment.AlignPath.to_indices" title="skbio.alignment.AlignPath.to_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_indices</span></code></a>([gap])</p></td>
<td><p>Generate an array of indices of characters in the original sequences.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.alignment.AlignPath.__str__" title="skbio.alignment.AlignPath.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__str__</span></code></a>()</p></td>
<td><p>Return string representation of this alignment path.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code>(value, /)</p></td>
<td><p>Return self==value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ge__</span></code>(value, /)</p></td>
<td><p>Return self&gt;=value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getstate__</span></code>(/)</p></td>
<td><p>Helper for pickle.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__gt__</span></code>(value, /)</p></td>
<td><p>Return self&gt;value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__hash__</span></code>(/)</p></td>
<td><p>Return hash(self).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__le__</span></code>(value, /)</p></td>
<td><p>Return self&lt;=value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__lt__</span></code>(value, /)</p></td>
<td><p>Return self&lt;value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ne__</span></code>(value, /)</p></td>
<td><p>Return self!=value.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Details</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.lengths">
<span class="sig-name descname"><span class="pre">lengths</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.lengths" title="Link to this definition">#</a></dt>
<dd><p>Array of lengths of segments in alignment path.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.ranges">
<span class="sig-name descname"><span class="pre">ranges</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.ranges" title="Link to this definition">#</a></dt>
<dd><p>Array of (start, stop) positions of sequences in the alignment.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.shape">
<span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.shape" title="Link to this definition">#</a></dt>
<dd><p>Number of sequences (rows) and positions (columns).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.starts">
<span class="sig-name descname"><span class="pre">starts</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.starts" title="Link to this definition">#</a></dt>
<dd><p>Array of start positions of sequences in the alignment.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.states">
<span class="sig-name descname"><span class="pre">states</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.states" title="Link to this definition">#</a></dt>
<dd><p>Array of gap status of segments in alignment path.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.stops">
<span class="sig-name descname"><span class="pre">stops</span></span><a class="headerlink" href="#skbio.alignment.AlignPath.stops" title="Link to this definition">#</a></dt>
<dd><p>Array of stop positions of sequences in the alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.alignment.AlignPath.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/main/skbio/alignment/_path.py#L230"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.AlignPath.__str__" title="Link to this definition">#</a></dt>
<dd><p>Return string representation of this alignment path.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.alignment.TabularMSA.write.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.alignment.TabularMSA.write</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.alignment.AlignPath.from_aligned.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.alignment.AlignPath.from_aligned</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath"><code class="docutils literal notranslate"><span class="pre">AlignPath</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.lengths"><code class="docutils literal notranslate"><span class="pre">AlignPath.lengths</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.ranges"><code class="docutils literal notranslate"><span class="pre">AlignPath.ranges</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.shape"><code class="docutils literal notranslate"><span class="pre">AlignPath.shape</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.starts"><code class="docutils literal notranslate"><span class="pre">AlignPath.starts</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.states"><code class="docutils literal notranslate"><span class="pre">AlignPath.states</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.stops"><code class="docutils literal notranslate"><span class="pre">AlignPath.stops</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.AlignPath.__str__"><code class="docutils literal notranslate"><span class="pre">AlignPath.__str__</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>