
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>FASTA/QUAL format (skbio.io.format.fasta) &#8212; scikit-bio 0.6.3-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=4317e44f"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.io.format.fasta';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.4';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.io.format.fasta.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="FASTQ format (skbio.io.format.fastq)" href="skbio.io.format.fastq.html" />
    <link rel="prev" title="Embedding format (skbio.io.format.embed)." href="skbio.io.format.embed.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.3-dev documentation"/>
    <script>document.write(`<img src="../_static/logo_inv.svg" class="logo__image only-dark" alt="scikit-bio 0.6.3-dev documentation"/>`);</script>
  
  
</a></div>
    
      <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.binary_dm.html">Simple binary dissimilarity matrix format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.binary_dm</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.biom.html">BIOM-Format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.biom</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.blast6.html">BLAST+6 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast6</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.blast7.html">BLAST+7 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast7</span></code>)</a></li>


<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.clustal.html">Clustal format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.clustal</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.embl.html">EMBL format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.embl</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.embed.html">Embedding format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.embed</span></code>).</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">FASTA/QUAL format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.fastq.html">FASTQ format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.genbank.html">GenBank format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.genbank</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.gff3.html">GFF3 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.gff3</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.lsmat.html">Labeled square matrix format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.lsmat</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.newick.html">Newick format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.newick</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.ordination.html">Ordination results format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.ordination</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.phylip.html">PHYLIP multiple sequence alignment format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.phylip</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.qseq.html">QSeq format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.qseq</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.stockholm.html">Stockholm format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.stockholm</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.taxdump.html">Taxdump format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.taxdump</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.sample_metadata.html">Format Support</a></li>

</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.write.html">skbio.io.registry.write</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.read.html">skbio.io.registry.read</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.sniff.html">skbio.io.registry.sniff</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.util.html">I/O utilities (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.util</span></code>)</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.html">I/O Registry (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.registry</span></code>)</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../io.html" class="nav-link">Input and Output (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">FASTA/QUAL...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="fasta-qual-format-skbio-io-format-fasta">
<span id="module-skbio.io.format.fasta"></span><h1>FASTA/QUAL format (<a class="reference internal" href="#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>)<a class="headerlink" href="#fasta-qual-format-skbio-io-format-fasta" title="Link to this heading">#</a></h1>
<p>The FASTA file format (<code class="docutils literal notranslate"><span class="pre">fasta</span></code>) stores biological (i.e., nucleotide or
protein) sequences in a simple plain text format that is both human-readable
and easy to parse. The file format was first introduced and used in the FASTA
software package <a class="reference internal" href="#rc6adbf0d8a87-1" id="id1">[1]</a>. Additional descriptions of the file format can be found
in <a class="reference internal" href="#rc6adbf0d8a87-2" id="id2">[2]</a> and <a class="reference internal" href="#rc6adbf0d8a87-3" id="id3">[3]</a>.</p>
<p>An example of a FASTA-formatted file containing two DNA sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">149855</span>
<span class="n">CGATGTCGATCGATCGATCGATCAG</span>
<span class="o">&gt;</span><span class="n">seq2</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">34989</span>
<span class="n">CATCGATCGATCGATGCATGCATGCATG</span>
</pre></div>
</div>
<p>The QUAL file format is an additional format related to FASTA. A FASTA file is
sometimes accompanied by a QUAL file, particuarly when the FASTA file contains
sequences generated on a high-throughput sequencing instrument. QUAL files
store a Phred quality score (nonnegative integer) for each base in a sequence
stored in FASTA format (see <a class="reference internal" href="#rc6adbf0d8a87-4" id="id4">[4]</a> for more details). scikit-bio supports reading
and writing FASTA (and optionally QUAL) file formats.</p>
<section id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Link to this heading">#</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Reader</p></th>
<th class="head"><p>Writer</p></th>
<th class="head"><p>Object Class</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p>generator of <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> objects</p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment.TabularMSA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.DNA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.RNA</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Protein</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All readers and writers support an optional QUAL file via the
<code class="docutils literal notranslate"><span class="pre">qual</span></code> parameter. If one is provided, quality scores will be read/written
in addition to FASTA sequence data.</p>
</div>
</section>
<section id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Link to this heading">#</a></h2>
<p>The following sections define the FASTA and QUAL file formats in detail.</p>
<section id="fasta-format">
<h3>FASTA Format<a class="headerlink" href="#fasta-format" title="Link to this heading">#</a></h3>
<p>A FASTA file contains one or more biological sequences. The sequences are
stored sequentially, with a <em>record</em> for each sequence (also referred to as a
<em>FASTA record</em>). Each <em>record</em> consists of a single-line <em>header</em> (sometimes
referred to as a <em>defline</em>, <em>label</em>, <em>description</em>, or <em>comment</em>) followed by
the sequence data, optionally split over multiple lines.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Blank or whitespace-only lines are only allowed at the beginning of
the file, between FASTA records, or at the end of the file. A blank or
whitespace-only line after the header line, within the sequence (for FASTA
files), or within quality scores (for QUAL files) will raise an error.</p>
<p>scikit-bio will ignore leading and trailing whitespace characters on each
line while reading.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio does not currently support legacy FASTA format (i.e.,
headers/comments denoted with a semicolon). The format supported by
scikit-bio (described below in detail) most closely resembles the
description given in NCBI’s BLAST documentation <a class="reference internal" href="#rc6adbf0d8a87-3" id="id5">[3]</a>. See <a class="reference internal" href="#rc6adbf0d8a87-2" id="id6">[2]</a> for more
details on legacy FASTA format. If you would like legacy FASTA format
support added to scikit-bio, please consider submitting a feature request on
the scikit-bio
<a class="reference external" href="https://github.com/scikit-bio/scikit-bio/issues">issue tracker</a>
(pull requests are also welcome!).</p>
</div>
<section id="sequence-header">
<h4>Sequence Header<a class="headerlink" href="#sequence-header" title="Link to this heading">#</a></h4>
<p>Each sequence header consists of a single line beginning with a greater-than
(<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>) symbol. Immediately following this is a sequence identifier (ID) and
description separated by one or more whitespace characters.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When reading a FASTA-formatted file, the sequence ID and description
are stored in the sequence <cite>metadata</cite> attribute, under the <cite>‘id’</cite> and
<cite>‘description’</cite> keys, repectively. Both are optional. Each will be
represented as the empty string (<code class="docutils literal notranslate"><span class="pre">''</span></code>) in <cite>metadata</cite> if it is not present
in the header.</p>
<p>When writing a FASTA-formatted file, sequence <cite>metadata</cite> identified by keys
<cite>‘id’</cite> and <cite>‘description’</cite> will be converted to strings and written as the
sequence identifier and description, respectively. Each will be written as
the empty string if not present in sequence <cite>metadata</cite>.</p>
</div>
<p>A sequence ID consists of a single <em>word</em>: all characters after the greater-
than symbol and before the first whitespace character (if any) are taken as the
sequence ID. Unique sequence IDs are not strictly enforced by the FASTA format
itself. A single standardized ID format is similarly not enforced by the FASTA
format, though it is often common to use a unique library accession number for
a sequence ID (e.g., NCBI’s FASTA defline format <a class="reference internal" href="#rc6adbf0d8a87-5" id="id7">[5]</a>).</p>
<p>If a description is present, it is taken as the remaining characters that
follow the sequence ID and initial whitespace(s). The description is considered
additional information about the sequence (e.g., comments about the source of
the sequence or the molecule that it encodes).</p>
<p>For example, consider the following header:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">149855</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">seq1</span></code> is the sequence ID and <code class="docutils literal notranslate"><span class="pre">db-accession-149855</span></code> is the sequence
description.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio’s readers will remove all leading and trailing whitespace
from the description. If a header line begins with whitespace following the
<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, the ID is assumed to be missing and the remainder of the line is
taken as the description.</p>
</div>
</section>
<section id="sequence-data">
<h4>Sequence Data<a class="headerlink" href="#sequence-data" title="Link to this heading">#</a></h4>
<p>Biological sequence data follows the header, and can be split over multiple
lines. The sequence data (i.e., nucleotides or amino acids) are stored using
the standard IUPAC lexicon (single-letter codes).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio supports both upper and lower case characters.
This functionality depends on the type of object the data is
being read into. For <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects, sciki-bio doesn’t care about the
case. Other sequence objects do, but all provide the <cite>lowercase</cite> parameter
to control case functionality. Refer to each class’s respective constructor
documentation for details.</p>
<p>Both <code class="docutils literal notranslate"><span class="pre">-</span></code> and <code class="docutils literal notranslate"><span class="pre">.</span></code> are supported as gap characters when reading into
<code class="docutils literal notranslate"><span class="pre">DNA</span></code>, <code class="docutils literal notranslate"><span class="pre">RNA</span></code>, and <code class="docutils literal notranslate"><span class="pre">Protein</span></code> sequence objects.</p>
<p>Validation is performed when reading into scikit-bio sequence objects that
enforce an alphabet (e.g., <code class="docutils literal notranslate"><span class="pre">DNA</span></code>, <code class="docutils literal notranslate"><span class="pre">RNA</span></code>, <code class="docutils literal notranslate"><span class="pre">Protein</span></code>). If any invalid
characters are found while reading from the FASTA file, an exception is
raised.</p>
</div>
</section>
</section>
<section id="qual-format">
<h3>QUAL Format<a class="headerlink" href="#qual-format" title="Link to this heading">#</a></h3>
<p>A QUAL file contains quality scores for one or more biological sequences stored
in a corresponding FASTA file. QUAL format is very similar to FASTA format: it
stores records sequentially, with each record beginning with a header line
containing a sequence ID and description. The same rules apply to QUAL headers
as FASTA headers (see the above sections for details). scikit-bio processes
FASTA and QUAL headers in exactly the same way.</p>
<p>Instead of storing biological sequence data in each record, a QUAL file stores
a Phred quality score for each base in the corresponding sequence. Quality
scores are represented as nonnegative integers separated by whitespace
(typically a single space or newline), and can span multiple lines.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When reading a QUAL-formatted file, quality scores are stored in the
sequence’s <cite>positional_metadata</cite> attribute under the <cite>‘quality’</cite> column.</p>
<p>When writing a QUAL-formatted file, a sequence’s <cite>positional_metadata</cite>
<cite>‘quality’</cite> column will be written as the quality scores.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When reading FASTA and QUAL files, scikit-bio requires records to be
in the same order in both files (i.e., each FASTA and QUAL record must have
the same ID and description after being parsed). In addition to having the
same order, the number of FASTA records must match the number of QUAL
records (i.e., missing or additonal records are not allowed). scikit-bio
also requires that the number of quality scores match the number of bases in
the corresponding sequence.</p>
<p>When writing FASTA and QUAL files, scikit-bio will maintain the same
ordering of records in both files (i.e., using the same ID and description
in both records) to support future reading.</p>
</div>
</section>
</section>
<section id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Link to this heading">#</a></h2>
<p>The following parameters are available to change how FASTA/QUAL files are read
or written in scikit-bio.</p>
<section id="qual-file-parameter-readers-and-writers">
<h3>QUAL File Parameter (Readers and Writers)<a class="headerlink" href="#qual-file-parameter-readers-and-writers" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">qual</span></code> parameter is available to all FASTA format readers and writers. It
can be any file-like type supported by scikit-bio’s I/O registry (e.g., file
handle, file path, etc.). If <code class="docutils literal notranslate"><span class="pre">qual</span></code> is provided when reading, quality scores
will be included in each in-memory <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object, in addition
to sequence data stored in the FASTA file. When writing, quality scores will be
written in QUAL format in addition to the sequence data being written in FASTA
format.</p>
</section>
<section id="reader-specific-parameters">
<h3>Reader-specific Parameters<a class="headerlink" href="#reader-specific-parameters" title="Link to this heading">#</a></h3>
<p>The available reader parameters differ depending on which reader is used.</p>
<section id="generator-and-tabularmsa-reader-parameters">
<h4>Generator and TabularMSA Reader Parameters<a class="headerlink" href="#generator-and-tabularmsa-reader-parameters" title="Link to this heading">#</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">constructor</span></code> parameter can be used with the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> generator and
<code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> FASTA readers. <code class="docutils literal notranslate"><span class="pre">constructor</span></code> specifies the type of in-memory
sequence object to read each sequence into. For example, if you know that the
FASTA file you’re reading contains protein sequences, you would pass
<code class="docutils literal notranslate"><span class="pre">constructor=Protein</span></code> to the reader call.</p>
<p>When reading into a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> generator, <code class="docutils literal notranslate"><span class="pre">constructor</span></code> defaults to
<code class="docutils literal notranslate"><span class="pre">Sequence</span></code> and must be a subclass of <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> if supplied.</p>
<p>When reading into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>, <code class="docutils literal notranslate"><span class="pre">constructor</span></code> is a required format
parameter and must be a subclass of <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code> (e.g., <code class="docutils literal notranslate"><span class="pre">DNA</span></code>,
<code class="docutils literal notranslate"><span class="pre">RNA</span></code>, <code class="docutils literal notranslate"><span class="pre">Protein</span></code>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The FASTA sniffer will not attempt to guess the <code class="docutils literal notranslate"><span class="pre">constructor</span></code>
parameter.</p>
</div>
</section>
<section id="sequence-reader-parameters">
<h4>Sequence Reader Parameters<a class="headerlink" href="#sequence-reader-parameters" title="Link to this heading">#</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">seq_num</span></code> parameter can be used with the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code>,
<code class="docutils literal notranslate"><span class="pre">DNA</span></code>, <code class="docutils literal notranslate"><span class="pre">RNA</span></code>, and <code class="docutils literal notranslate"><span class="pre">Protein</span></code> FASTA readers. <code class="docutils literal notranslate"><span class="pre">seq_num</span></code> specifies which
sequence to read from the FASTA file (and optional QUAL file), and defaults to
1 (i.e., such that the first sequence is read). For example, to read the 50th
sequence from a FASTA file, you would pass <code class="docutils literal notranslate"><span class="pre">seq_num=50</span></code> to the reader call.</p>
</section>
</section>
<section id="writer-specific-parameters">
<h3>Writer-specific Parameters<a class="headerlink" href="#writer-specific-parameters" title="Link to this heading">#</a></h3>
<p>The following parameters are available to all FASTA format writers:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code>: string to replace <strong>each</strong> whitespace
character in a sequence ID. This parameter is useful for cases where an
in-memory sequence ID contains whitespace, which would result in an on-disk
representation that would not be read back into memory as the same ID (since
IDs in FASTA format cannot contain whitespace). Defaults to <code class="docutils literal notranslate"><span class="pre">_</span></code>. If
<code class="docutils literal notranslate"><span class="pre">None</span></code>, no whitespace replacement is performed and IDs are written as they
are stored in memory (this has the potential to create an invalid
FASTA-formatted file; see note below). This parameter also applies to a QUAL
file if one is provided.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code>: string to replace <strong>each</strong> newline
character in a sequence description. Since a FASTA header must be a single
line, newlines are not allowed in sequence descriptions and must be replaced
in order to write a valid FASTA file. Defaults to a single space. If
<code class="docutils literal notranslate"><span class="pre">None</span></code>, no newline replacement is performed and descriptions are written as
they are stored in memory (this has the potential to create an invalid
FASTA-formatted file; see note below). This parameter also applies to a QUAL
file if one is provided.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">max_width</span></code>: integer specifying the maximum line width (i.e., number of
characters) for sequence data and/or quality scores. If a sequence or its
quality scores are longer than <code class="docutils literal notranslate"><span class="pre">max_width</span></code>, it will be split across
multiple lines, each with a maximum width of <code class="docutils literal notranslate"><span class="pre">max_width</span></code>. Note that there
are some caveats when splitting quality scores. A single quality score will
<em>never</em> be split across multiple lines, otherwise it would become two
different quality scores when read again. Thus, splitting only occurs
<em>between</em> quality scores. This makes it possible to have a single long
quality score written on its own line that exceeds <code class="docutils literal notranslate"><span class="pre">max_width</span></code>. For
example, the quality score <code class="docutils literal notranslate"><span class="pre">12345</span></code> would not be split across multiple lines
even if <code class="docutils literal notranslate"><span class="pre">max_width=3</span></code>. Thus, a 5-character line would be written. Default
behavior is to not split sequence data or quality scores across multiple
lines.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lowercase</span></code>: String or boolean array. If a string, it is treated as a key
into the positional metadata of the object. If a boolean array, it
indicates characters to write in lowercase. Characters in the sequence
corresponding to <cite>True</cite> values will be written in lowercase. The boolean
array must be the same length as the sequence.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The FASTA format writers will have noticeably better runtime
performance if <code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code> are set to <code class="docutils literal notranslate"><span class="pre">None</span></code> so that whitespace
replacement is not performed during writing. However, this can potentially
create invalid FASTA files, especially if there are newline characters in
the IDs or descriptions. For IDs with whitespace, this can also affect how
the IDs are read into memory in a subsequent read operation. For example, if
an in-memory sequence ID is <code class="docutils literal notranslate"><span class="pre">'seq</span> <span class="pre">1'</span></code> and
<code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement=None</span></code>, reading the FASTA file back into memory
would result in an ID of <code class="docutils literal notranslate"><span class="pre">'seq'</span></code>, and <code class="docutils literal notranslate"><span class="pre">'1'</span></code> would be part of the
sequence description.</p>
</div>
</section>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">#</a></h2>
<section id="reading-and-writing-fasta-files">
<h3>Reading and Writing FASTA Files<a class="headerlink" href="#reading-and-writing-fasta-files" title="Link to this heading">#</a></h3>
<p>Suppose we have the following FASTA file with five equal-length sequences
(example modified from <a class="reference internal" href="#rc6adbf0d8a87-6" id="id8">[6]</a>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span> <span class="n">Turkey</span>
<span class="n">AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span>
<span class="o">&gt;</span><span class="n">seq2</span> <span class="n">Salmo</span> <span class="n">gair</span>
<span class="n">AAGCCTTGGCAGTGCAGGGTGAGCCGTGG</span>
<span class="n">CCGGGCACGGTAT</span>
<span class="o">&gt;</span><span class="n">seq3</span> <span class="n">H</span><span class="o">.</span> <span class="n">Sapiens</span>
<span class="n">ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span>
<span class="o">&gt;</span><span class="n">seq4</span> <span class="n">Chimp</span>
<span class="n">AAACCCTTGCCG</span>
<span class="n">TTACGCTTAAAC</span>
<span class="n">CGAGGCCGGGAC</span>
<span class="n">ACTCAT</span>
<span class="o">&gt;</span><span class="n">seq5</span> <span class="n">Gorilla</span>
<span class="n">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Original copyright notice for the above example file:</p>
<p><em>(c) Copyright 1986-2008 by The University of Washington. Written by Joseph
Felsenstein. Permission is granted to copy this document provided that no
fee is charged for it and that this copyright notice is not removed.</em></p>
</div>
<p>Note that the sequences are not required to be of equal length in order for the
file to be a valid FASTA file (this depends on the object that you’re reading
the file into). Also note that some of the sequences occur on a single line,
while others are split across multiple lines.</p>
<p>Let’s define this file in-memory as a <code class="docutils literal notranslate"><span class="pre">StringIO</span></code>, though this could be a real
file path, file handle, or anything that’s supported by scikit-bio’s I/O
registry in practice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fl</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&gt;seq1 Turkey</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;&gt;seq2 Salmo gair</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;AAGCCTTGGCAGTGCAGGGTGAGCCGTGG</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;CCGGGCACGGTAT</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;&gt;seq3 H. Sapiens</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;&gt;seq4 Chimp</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;AAACCCTTGCCG</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;TTACGCTTAAAC</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;CGAGGCCGGGAC</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;ACTCAT</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;&gt;seq5 Gorilla</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Since these sequences are of equal length (presumably because they’ve been
aligned), let’s read the FASTA file into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fl</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">------------------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 5</span>
<span class="go">    position count: 42</span>
<span class="go">------------------------------------------</span>
<span class="go">AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span>
<span class="go">AAGCCTTGGCAGTGCAGGGTGAGCCGTGGCCGGGCACGGTAT</span>
<span class="go">ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span>
<span class="go">AAACCCTTGCCGTTACGCTTAAACCGAGGCCGGGACACTCAT</span>
<span class="go">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>
</pre></div>
</div>
<p>Note that we didn’t specify a file format in the <code class="docutils literal notranslate"><span class="pre">read</span></code> call. The FASTA
sniffer detected the correct file format for us!</p>
<p>To write the <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> in FASTA format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">StringIO</span><span class="p">()</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 Turkey</span>
<span class="go">AAGCTNGGGCATTTCAGGGTGAGCCCGGGCAATACAGGGTAT</span>
<span class="go">&gt;seq2 Salmo gair</span>
<span class="go">AAGCCTTGGCAGTGCAGGGTGAGCCGTGGCCGGGCACGGTAT</span>
<span class="go">&gt;seq3 H. Sapiens</span>
<span class="go">ACCGGTTGGCCGTTCAGGGTACAGGTTGGCCGTTCAGGGTAA</span>
<span class="go">&gt;seq4 Chimp</span>
<span class="go">AAACCCTTGCCGTTACGCTTAAACCGAGGCCGGGACACTCAT</span>
<span class="go">&gt;seq5 Gorilla</span>
<span class="go">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> loads all of the sequences from the FASTA file into memory at
once. If the FASTA file is large (which is often the case), this may be
infeasible if you don’t have enough memory. To work around this issue, you can
stream the sequences using scikit-bio’s generator-based FASTA reader and
writer. The generator-based reader yields <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects (or subclasses
if <code class="docutils literal notranslate"><span class="pre">constructor</span></code> is supplied) one at a time, instead of loading all sequences
into memory. For example, let’s use the generator-based reader to process a
single sequence at a time in a <code class="docutils literal notranslate"><span class="pre">for</span></code> loop:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">skbio.io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fl</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">seq</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Turkey&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq1&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAGCTNGGGC ATTTCAGGGT GAGCCCGGGC AATACAGGGT AT</span>

<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Salmo gair&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq2&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAGCCTTGGC AGTGCAGGGT GAGCCGTGGC CGGGCACGGT AT</span>

<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;H. Sapiens&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq3&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 ACCGGTTGGC CGTTCAGGGT ACAGGTTGGC CGTTCAGGGT AA</span>

<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Chimp&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq4&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAACCCTTGC CGTTACGCTT AAACCGAGGC CGGGACACTC AT</span>

<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Gorilla&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq5&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAACCCTTGC CGGTACGCTT AAACCATTGC CGGTACGCTT AA</span>
</pre></div>
</div>
<p>A single sequence can also be read into a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> (or subclass):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fl</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Turkey&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq1&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAGCTNGGGC ATTTCAGGGT GAGCCCGGGC AATACAGGGT AT</span>
</pre></div>
</div>
<p>By default, the first sequence in the FASTA file is read. This can be
controlled with <code class="docutils literal notranslate"><span class="pre">seq_num</span></code>. For example, to read the fifth sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fl</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Sequence</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Gorilla&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq5&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAACCCTTGC CGGTACGCTT AAACCATTGC CGGTACGCTT AA</span>
</pre></div>
</div>
<p>We can use the same API to read the fifth sequence into a <code class="docutils literal notranslate"><span class="pre">DNA</span></code> sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna_seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fl</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna_seq</span>
<span class="go">DNA</span>
<span class="go">------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;Gorilla&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq5&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 42</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 50.00%</span>
<span class="go">------------------------------------------------</span>
<span class="go">0 AAACCCTTGC CGGTACGCTT AAACCATTGC CGGTACGCTT AA</span>
</pre></div>
</div>
<p>Individual sequence objects can also be written in FASTA format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">StringIO</span><span class="p">()</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">dna_seq</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq5 Gorilla</span>
<span class="go">AAACCCTTGCCGGTACGCTTAAACCATTGCCGGTACGCTTAA</span>
</pre></div>
</div>
</section>
<section id="reading-and-writing-fasta-qual-files">
<h3>Reading and Writing FASTA/QUAL Files<a class="headerlink" href="#reading-and-writing-fasta-qual-files" title="Link to this heading">#</a></h3>
<p>In addition to reading and writing standalone FASTA files, scikit-bio supports
reading and writing FASTA and QUAL files together. Suppose we have the
following FASTA file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">149855</span>
<span class="n">CGATGTC</span>
<span class="o">&gt;</span><span class="n">seq2</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">34989</span>
<span class="n">CATCGTC</span>
</pre></div>
</div>
<p>Also suppose we have the following QUAL file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">149855</span>
<span class="mi">40</span> <span class="mi">39</span> <span class="mi">39</span> <span class="mi">4</span>
<span class="mi">50</span> <span class="mi">1</span> <span class="mi">100</span>
<span class="o">&gt;</span><span class="n">seq2</span> <span class="n">db</span><span class="o">-</span><span class="n">accession</span><span class="o">-</span><span class="mi">34989</span>
<span class="mi">3</span> <span class="mi">3</span> <span class="mi">10</span> <span class="mi">42</span> <span class="mi">80</span> <span class="mi">80</span> <span class="mi">79</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta_fl</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="s2">&quot;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;CGATGTC</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;CATCGTC</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qual_fl</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="s2">&quot;&gt;seq1 db-accession-149855</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;40 39 39 4</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;50 1 100</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;&gt;seq2 db-accession-34989</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;3 3 10 42 80 80 79</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>To read in a single <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> at a time, we can use the
generator-based reader as we did above, providing both FASTA and QUAL files:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">skbio</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_fl</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">qual_fl</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">seq</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">Sequence</span>
<span class="go">----------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;db-accession-149855&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq1&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 7</span>
<span class="go">----------------------------------------</span>
<span class="go">0 CGATGTC</span>

<span class="go">Sequence</span>
<span class="go">---------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;db-accession-34989&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq2&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 7</span>
<span class="go">---------------------------------------</span>
<span class="go">0 CATCGTC</span>
</pre></div>
</div>
<p>Note that the sequence objects have quality scores stored as positional
metadata since we provided a QUAL file. The other FASTA readers operate in a
similar manner.</p>
<p>Now let’s load the sequences and their quality scores into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fasta_fl</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">qual_fl</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 7</span>
<span class="go">---------------------</span>
<span class="go">CGATGTC</span>
<span class="go">CATCGTC</span>
</pre></div>
</div>
<p>To write the sequence data and quality scores in the <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> to FASTA
and QUAL files, respectively:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fasta_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_qual_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fasta_fh</span><span class="p">,</span> <span class="n">qual</span><span class="o">=</span><span class="n">new_qual_fh</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">new_fasta_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 db-accession-149855</span>
<span class="go">CGATGTC</span>
<span class="go">&gt;seq2 db-accession-34989</span>
<span class="go">CATCGTC</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">new_qual_fh</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;seq1 db-accession-149855</span>
<span class="go">40 39 39 4 50 1 100</span>
<span class="go">&gt;seq2 db-accession-34989</span>
<span class="go">3 3 10 42 80 80 79</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fasta_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_qual_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div role="list" class="citation-list">
<div class="citation" id="rc6adbf0d8a87-1" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Lipman, DJ; Pearson, WR (1985). “Rapid and sensitive protein similarity
searches”. Science 227 (4693): 1435-41.</p>
</div>
<div class="citation" id="rc6adbf0d8a87-2" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id6">2</a>)</span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">http://en.wikipedia.org/wiki/FASTA_format</a></p>
</div>
<div class="citation" id="rc6adbf0d8a87-3" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id3">1</a>,<a role="doc-backlink" href="#id5">2</a>)</span>
<p><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml">http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml</a></p>
</div>
<div class="citation" id="rc6adbf0d8a87-4" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">4</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="https://www.broadinstitute.org/crd/wiki/index.php/Qual">https://www.broadinstitute.org/crd/wiki/index.php/Qual</a></p>
</div>
<div class="citation" id="rc6adbf0d8a87-5" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">5</a><span class="fn-bracket">]</span></span>
<p>Madden T. The BLAST Sequence Analysis Tool. 2002 Oct 9
[Updated 2003 Aug 13]. In: McEntyre J, Ostell J, editors. The NCBI Handbook
[Internet]. Bethesda (MD): National Center for Biotechnology Information
(US); 2002-. Chapter 16. Available from:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK21097/">http://www.ncbi.nlm.nih.gov/books/NBK21097/</a></p>
</div>
<div class="citation" id="rc6adbf0d8a87-6" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">6</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a></p>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.io.format.embed.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Embedding format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.embed</span></code>).</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.io.format.fastq.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">FASTQ format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code>)</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-support">Format Support</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-specification">Format Specification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fasta-format">FASTA Format</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-header">Sequence Header</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-data">Sequence Data</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#qual-format">QUAL Format</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-parameters">Format Parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#qual-file-parameter-readers-and-writers">QUAL File Parameter (Readers and Writers)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reader-specific-parameters">Reader-specific Parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#generator-and-tabularmsa-reader-parameters">Generator and TabularMSA Reader Parameters</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-reader-parameters">Sequence Reader Parameters</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#writer-specific-parameters">Writer-specific Parameters</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#examples">Examples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-fasta-files">Reading and Writing FASTA Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-fasta-qual-files">Reading and Writing FASTA/QUAL Files</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.0.2.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>