
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.sequence.Protein &#8212; scikit-bio 0.6.0-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=ad750166" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=9ff6a3c4"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.sequence.Protein';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.2';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.sequence.Protein.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.sequence.Protein.concat" href="skbio.sequence.Protein.concat.html" />
    <link rel="prev" title="skbio.sequence.RNA.__str__" href="skbio.sequence.RNA.__str__.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>


  <div class="bd-header-announcement container-fluid bd-header-announcement">
    <div class="bd-header-announcement__content">scikit-bio is back in active development! Check out our <a href='https://github.com/scikit-bio/scikit-bio/discussions/1935'>announcement of revitalization</a>.</div>
  </div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

   

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.0-dev documentation"/>
    <script>document.write(`<img src="../_static/logo_inv.svg" class="logo__image only-dark" alt="scikit-bio 0.6.0-dev documentation"/>`);</script>
  
  
</a></div>
    
      <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm navbar-btn dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../install.html">
                        Install
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../learn.html">
                        Learn
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../contribute.html">
                        Contribute
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/master/CHANGELOG.md">
                        Releases
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../about.html">
                        About
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../io.html">
                        File I/O (skbio.io)
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
                        Sequences (skbio.sequence)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
                        Alignments (skbio.alignment)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../tree.html">
                        Tree representations (skbio.tree)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
                        Constructing workflows (skbio.workflow)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
                        Diversity calculations (skbio.diversity)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../stats.html">
                        Statistics (skbio.stats)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
                        Metadata (skbio.metadata)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../util.html">
                        Utility functionality (skbio.util)
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../install.html">
                        Install
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../learn.html">
                        Learn
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../contribute.html">
                        Contribute
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/master/CHANGELOG.md">
                        Releases
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-external" href="../../../about.html">
                        About
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../io.html">
                        File I/O (skbio.io)
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
                        Sequences (skbio.sequence)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
                        Alignments (skbio.alignment)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../tree.html">
                        Tree representations (skbio.tree)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
                        Constructing workflows (skbio.workflow)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
                        Diversity calculations (skbio.diversity)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../stats.html">
                        Statistics (skbio.stats)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
                        Metadata (skbio.metadata)
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../util.html">
                        Utility functionality (skbio.util)
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.Sequence.html">skbio.sequence.Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GrammaredSequence.html">skbio.sequence.GrammaredSequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.DNA.html">skbio.sequence.DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.RNA.html">skbio.sequence.RNA</a></li>
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">skbio.sequence.Protein</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.concat.html">skbio.sequence.Protein.concat</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.count.html">skbio.sequence.Protein.count</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.definites.html">skbio.sequence.Protein.definites</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.degap.html">skbio.sequence.Protein.degap</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.degenerates.html">skbio.sequence.Protein.degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.distance.html">skbio.sequence.Protein.distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.expand_degenerates.html">skbio.sequence.Protein.expand_degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.find_motifs.html">skbio.sequence.Protein.find_motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.find_with_regex.html">skbio.sequence.Protein.find_with_regex</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.frequencies.html">skbio.sequence.Protein.frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.gaps.html">skbio.sequence.Protein.gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_definites.html">skbio.sequence.Protein.has_definites</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_degenerates.html">skbio.sequence.Protein.has_degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_gaps.html">skbio.sequence.Protein.has_gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_interval_metadata.html">skbio.sequence.Protein.has_interval_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_metadata.html">skbio.sequence.Protein.has_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_nondegenerates.html">skbio.sequence.Protein.has_nondegenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_positional_metadata.html">skbio.sequence.Protein.has_positional_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_stops.html">skbio.sequence.Protein.has_stops</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.index.html">skbio.sequence.Protein.index</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.iter_contiguous.html">skbio.sequence.Protein.iter_contiguous</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.iter_kmers.html">skbio.sequence.Protein.iter_kmers</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.kmer_frequencies.html">skbio.sequence.Protein.kmer_frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.lowercase.html">skbio.sequence.Protein.lowercase</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.match_frequency.html">skbio.sequence.Protein.match_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.matches.html">skbio.sequence.Protein.matches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.mismatch_frequency.html">skbio.sequence.Protein.mismatch_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.mismatches.html">skbio.sequence.Protein.mismatches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.nondegenerates.html">skbio.sequence.Protein.nondegenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.read.html">skbio.sequence.Protein.read</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.replace.html">skbio.sequence.Protein.replace</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.stops.html">skbio.sequence.Protein.stops</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.to_indices.html">skbio.sequence.Protein.to_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.to_regex.html">skbio.sequence.Protein.to_regex</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.write.html">skbio.sequence.Protein.write</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__bool__.html">skbio.sequence.Protein.__bool__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__contains__.html">skbio.sequence.Protein.__contains__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__copy__.html">skbio.sequence.Protein.__copy__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__deepcopy__.html">skbio.sequence.Protein.__deepcopy__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__eq__.html">skbio.sequence.Protein.__eq__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__ge__.html">skbio.sequence.Protein.__ge__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__getitem__.html">skbio.sequence.Protein.__getitem__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__getstate__.html">skbio.sequence.Protein.__getstate__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__gt__.html">skbio.sequence.Protein.__gt__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__iter__.html">skbio.sequence.Protein.__iter__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__le__.html">skbio.sequence.Protein.__le__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__len__.html">skbio.sequence.Protein.__len__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__lt__.html">skbio.sequence.Protein.__lt__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__ne__.html">skbio.sequence.Protein.__ne__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__reversed__.html">skbio.sequence.Protein.__reversed__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.__str__.html">skbio.sequence.Protein.__str__</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.concat.html">skbio.sequence.Protein.concat</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.count.html">skbio.sequence.Protein.count</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.definites.html">skbio.sequence.Protein.definites</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.degap.html">skbio.sequence.Protein.degap</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.degenerates.html">skbio.sequence.Protein.degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.distance.html">skbio.sequence.Protein.distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.expand_degenerates.html">skbio.sequence.Protein.expand_degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.find_motifs.html">skbio.sequence.Protein.find_motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.find_with_regex.html">skbio.sequence.Protein.find_with_regex</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.frequencies.html">skbio.sequence.Protein.frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.gaps.html">skbio.sequence.Protein.gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_definites.html">skbio.sequence.Protein.has_definites</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_degenerates.html">skbio.sequence.Protein.has_degenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_gaps.html">skbio.sequence.Protein.has_gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_interval_metadata.html">skbio.sequence.Protein.has_interval_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_metadata.html">skbio.sequence.Protein.has_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_nondegenerates.html">skbio.sequence.Protein.has_nondegenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_positional_metadata.html">skbio.sequence.Protein.has_positional_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.has_stops.html">skbio.sequence.Protein.has_stops</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.index.html">skbio.sequence.Protein.index</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.iter_contiguous.html">skbio.sequence.Protein.iter_contiguous</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.iter_kmers.html">skbio.sequence.Protein.iter_kmers</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.kmer_frequencies.html">skbio.sequence.Protein.kmer_frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.lowercase.html">skbio.sequence.Protein.lowercase</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.match_frequency.html">skbio.sequence.Protein.match_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.matches.html">skbio.sequence.Protein.matches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.mismatch_frequency.html">skbio.sequence.Protein.mismatch_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.mismatches.html">skbio.sequence.Protein.mismatches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.nondegenerates.html">skbio.sequence.Protein.nondegenerates</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.read.html">skbio.sequence.Protein.read</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.replace.html">skbio.sequence.Protein.replace</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.stops.html">skbio.sequence.Protein.stops</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.to_indices.html">skbio.sequence.Protein.to_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.to_regex.html">skbio.sequence.Protein.to_regex</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Protein.write.html">skbio.sequence.Protein.write</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GeneticCode.html">skbio.sequence.GeneticCode</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.html">skbio.sequence.SubstitutionMatrix</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.distance.html">Sequence distance metrics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.distance</span></code>)</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../sequence.html" class="nav-link">Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">skbio.sequen...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-sequence-protein">
<h1>skbio.sequence.Protein<a class="headerlink" href="#skbio-sequence-protein" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="skbio.sequence.Protein">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">skbio.sequence.</span></span><span class="sig-name descname"><span class="pre">Protein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positional_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/master/skbio/sequence/_protein.py#L16"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Protein" title="Link to this definition">#</a></dt>
<dd><p>Store protein sequence data and optional associated metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, Sequence, or 1D np.ndarray (np.uint8 or ‘|S1’)</span></dt><dd><p>Characters representing the protein sequence itself.</p>
</dd>
<dt><strong>metadata</strong><span class="classifier">dict, optional</span></dt><dd><p>Arbitrary metadata which applies to the entire sequence.</p>
</dd>
<dt><strong>positional_metadata</strong><span class="classifier">Pandas DataFrame consumable, optional</span></dt><dd><p>Arbitrary per-character metadata. For example, quality data from
sequencing reads. Must be able to be passed directly to the Pandas
DataFrame constructor.</p>
</dd>
<dt><strong>interval_metadata</strong><span class="classifier">IntervalMetadata</span></dt><dd><p>Arbitrary interval metadata which applies to intervals within
a sequence to store interval features (such as protein domains).</p>
</dd>
<dt><strong>lowercase</strong><span class="classifier">bool or str, optional</span></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, lowercase sequence characters will be converted to
uppercase characters in order to be valid IUPAC Protein characters. If
<code class="docutils literal notranslate"><span class="pre">False</span></code>, no characters will be converted. If a str, it will be
treated as a key into the positional metadata of the object. All
lowercase characters will be converted to uppercase, and a <code class="docutils literal notranslate"><span class="pre">True</span></code>
value will be stored in a boolean array in the positional metadata
under the key.</p>
</dd>
<dt><strong>validate</strong><span class="classifier">bool, optional</span></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, validation will be performed to ensure that all sequence
characters are in the IUPAC protein character set. If <code class="docutils literal notranslate"><span class="pre">False</span></code>,
validation will not be performed. Turning off validation will improve
runtime performance. If invalid characters are present, however, there
is <strong>no guarantee that operations performed on the resulting object
will work or behave as expected.</strong> Only turn off validation if you are
certain that the sequence characters are valid. To store sequence data
that is not IUPAC-compliant, use <code class="docutils literal notranslate"><span class="pre">Sequence</span></code>.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>According to the IUPAC notation <a class="reference internal" href="#r079cf422404d-1" id="id1">[1]</a> , a protein sequence may contain the
following 20 definite characters (canonical amino acids):</p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Code</p></th>
<th class="head"><p>3-letter</p></th>
<th class="head"><p>Amino acid</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><cite>A</cite></p></td>
<td><p>Ala</p></td>
<td><p>Alanine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>C</cite></p></td>
<td><p>Cys</p></td>
<td><p>Cysteine</p></td>
</tr>
<tr class="row-even"><td><p><cite>D</cite></p></td>
<td><p>Asp</p></td>
<td><p>Aspartic acid</p></td>
</tr>
<tr class="row-odd"><td><p><cite>E</cite></p></td>
<td><p>Glu</p></td>
<td><p>Glutamic acid</p></td>
</tr>
<tr class="row-even"><td><p><cite>F</cite></p></td>
<td><p>Phe</p></td>
<td><p>Phenylalanine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>G</cite></p></td>
<td><p>Gly</p></td>
<td><p>Glycine</p></td>
</tr>
<tr class="row-even"><td><p><cite>H</cite></p></td>
<td><p>His</p></td>
<td><p>Histidine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>I</cite></p></td>
<td><p>Ile</p></td>
<td><p>Isoleucine</p></td>
</tr>
<tr class="row-even"><td><p><cite>K</cite></p></td>
<td><p>Lys</p></td>
<td><p>Lysine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>L</cite></p></td>
<td><p>Leu</p></td>
<td><p>Leucine</p></td>
</tr>
<tr class="row-even"><td><p><cite>M</cite></p></td>
<td><p>Met</p></td>
<td><p>Methionine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>N</cite></p></td>
<td><p>Asn</p></td>
<td><p>Asparagine</p></td>
</tr>
<tr class="row-even"><td><p><cite>P</cite></p></td>
<td><p>Pro</p></td>
<td><p>Proline</p></td>
</tr>
<tr class="row-odd"><td><p><cite>Q</cite></p></td>
<td><p>Gln</p></td>
<td><p>Glutamine</p></td>
</tr>
<tr class="row-even"><td><p><cite>R</cite></p></td>
<td><p>Arg</p></td>
<td><p>Arginine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>S</cite></p></td>
<td><p>Ser</p></td>
<td><p>Serine</p></td>
</tr>
<tr class="row-even"><td><p><cite>T</cite></p></td>
<td><p>Thr</p></td>
<td><p>Threonine</p></td>
</tr>
<tr class="row-odd"><td><p><cite>V</cite></p></td>
<td><p>Val</p></td>
<td><p>Valine</p></td>
</tr>
<tr class="row-even"><td><p><cite>W</cite></p></td>
<td><p>Trp</p></td>
<td><p>Tryptophan</p></td>
</tr>
<tr class="row-odd"><td><p><cite>Y</cite></p></td>
<td><p>Tyr</p></td>
<td><p>Tyrosine</p></td>
</tr>
</tbody>
</table>
<p>And the following four degenerate characters, each of which representing
two or more amino acids:</p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Code</p></th>
<th class="head"><p>3-letter</p></th>
<th class="head"><p>Amino acids</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><cite>B</cite></p></td>
<td><p>Asx</p></td>
<td><p>D or N</p></td>
</tr>
<tr class="row-odd"><td><p><cite>Z</cite></p></td>
<td><p>Glx</p></td>
<td><p>E or Q</p></td>
</tr>
<tr class="row-even"><td><p><cite>J</cite></p></td>
<td><p>Xle</p></td>
<td><p>I or L</p></td>
</tr>
<tr class="row-odd"><td><p><cite>X</cite></p></td>
<td><p>Xaa</p></td>
<td><p>All 20</p></td>
</tr>
</tbody>
</table>
<p>Plus one stop character: <cite>*</cite> (Ter), and two gap characters: <cite>-</cite> and <cite>.</cite>.</p>
<p>Characters other than the above 27 are not allowed. If you intend to use
additional characters to represent non-canonical amino acids, such as <cite>U</cite>
(Sec, Selenocysteine) and <cite>O</cite> (Pyl, Pyrrolysine), you may create a custom
alphabet using <code class="docutils literal notranslate"><span class="pre">GrammaredSequence</span></code>. Directly modifying the alphabet of
<code class="docutils literal notranslate"><span class="pre">Protein</span></code> may break functions that rely on the IUPAC alphabet.</p>
<p>It should be noted that some functions do not support certain characters.
For example, the BLOSUM and PAM substitution matrices do not support <cite>J</cite>
(Xle). In such circumstances, unsupported characters will be replaced with
<cite>X</cite> to represent any of the canonical amino acids.</p>
<p class="rubric">References</p>
<div role="list" class="citation-list">
<div class="citation" id="r079cf422404d-1" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Nomenclature for incompletely specified bases in nucleic acid
sequences: recommendations 1984.
Nucleic Acids Res. May 10, 1985; 13(9): 3021-3030.
A Cornish-Bowden</p>
</div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Protein</span><span class="p">(</span><span class="s1">&#39;PAW&#39;</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: False</span>
<span class="go">--------------------------</span>
<span class="go">0 PAW</span>
</pre></div>
</div>
<p>Convert lowercase characters to uppercase:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Protein</span><span class="p">(</span><span class="s1">&#39;paW&#39;</span><span class="p">,</span> <span class="n">lowercase</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Protein</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: False</span>
<span class="go">--------------------------</span>
<span class="go">0 PAW</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">alphabet</span></code></dt><dd><p>Return valid characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_gap_char</span></code></dt><dd><p>Gap character to use when constructing a new gapped sequence.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">definite_chars</span></code></dt><dd><p>Return definite characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerate_chars</span></code></dt><dd><p>Return degenerate characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerate_map</span></code></dt><dd><p>Return mapping of degenerate to definite characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_chars</span></code></dt><dd><p>Return characters defined as gaps.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval_metadata</span></code></dt><dd><p><code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code> object containing info about interval features.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></dt><dd><p><code class="docutils literal notranslate"><span class="pre">dict</span></code> containing metadata which applies to the entire object.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">nondegenerate_chars</span></code></dt><dd><p>Return non-degenerate characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">observed_chars</span></code></dt><dd><p>Set of observed characters in the sequence.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_metadata</span></code></dt><dd><p><code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> containing metadata along an axis.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">stop_chars</span></code></dt><dd><p>Return characters representing translation stop codons.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code></dt><dd><p>Array containing underlying sequence characters.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">wildcard_char</span></code></dt><dd><p>Return wildcard character.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.concat.html#skbio.sequence.Protein.concat" title="skbio.sequence.Protein.concat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code></a>(sequences[, how])</p></td>
<td><p>Concatenate an iterable of <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.count.html#skbio.sequence.Protein.count" title="skbio.sequence.Protein.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(subsequence[, start, end])</p></td>
<td><p>Count occurrences of a subsequence in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.definites.html#skbio.sequence.Protein.definites" title="skbio.sequence.Protein.definites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">definites</span></code></a>()</p></td>
<td><p>Find positions containing definite characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.degap.html#skbio.sequence.Protein.degap" title="skbio.sequence.Protein.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>()</p></td>
<td><p>Return a new sequence with gap characters removed.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.degenerates.html#skbio.sequence.Protein.degenerates" title="skbio.sequence.Protein.degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerates</span></code></a>()</p></td>
<td><p>Find positions containing degenerate characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.distance.html#skbio.sequence.Protein.distance" title="skbio.sequence.Protein.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other[, metric])</p></td>
<td><p>Compute the distance to another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.expand_degenerates.html#skbio.sequence.Protein.expand_degenerates" title="skbio.sequence.Protein.expand_degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">expand_degenerates</span></code></a>()</p></td>
<td><p>Yield all possible definite versions of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.find_motifs.html#skbio.sequence.Protein.find_motifs" title="skbio.sequence.Protein.find_motifs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_motifs</span></code></a>(motif_type[, min_length, ignore])</p></td>
<td><p>Search the biological sequence for motifs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.find_with_regex.html#skbio.sequence.Protein.find_with_regex" title="skbio.sequence.Protein.find_with_regex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_with_regex</span></code></a>(regex[, ignore])</p></td>
<td><p>Generate slices for patterns matched by a regular expression.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.frequencies.html#skbio.sequence.Protein.frequencies" title="skbio.sequence.Protein.frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frequencies</span></code></a>([chars, relative])</p></td>
<td><p>Compute frequencies of characters in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.gaps.html#skbio.sequence.Protein.gaps" title="skbio.sequence.Protein.gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gaps</span></code></a>()</p></td>
<td><p>Find positions containing gaps in the biological sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_definites.html#skbio.sequence.Protein.has_definites" title="skbio.sequence.Protein.has_definites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_definites</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more definite characters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_degenerates.html#skbio.sequence.Protein.has_degenerates" title="skbio.sequence.Protein.has_degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_degenerates</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_gaps.html#skbio.sequence.Protein.has_gaps" title="skbio.sequence.Protein.has_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_gaps</span></code></a>()</p></td>
<td><p>Determine if the sequence contains one or more gap characters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_interval_metadata.html#skbio.sequence.Protein.has_interval_metadata" title="skbio.sequence.Protein.has_interval_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_interval_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has interval metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_metadata.html#skbio.sequence.Protein.has_metadata" title="skbio.sequence.Protein.has_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_nondegenerates.html#skbio.sequence.Protein.has_nondegenerates" title="skbio.sequence.Protein.has_nondegenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_nondegenerates</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more non-degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_positional_metadata.html#skbio.sequence.Protein.has_positional_metadata" title="skbio.sequence.Protein.has_positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_positional_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has positional metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_stops.html#skbio.sequence.Protein.has_stops" title="skbio.sequence.Protein.has_stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_stops</span></code></a>()</p></td>
<td><p>Determine if the sequence contains one or more stop characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.index.html#skbio.sequence.Protein.index" title="skbio.sequence.Protein.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a>(subsequence[, start, end])</p></td>
<td><p>Find position where subsequence first occurs in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.iter_contiguous.html#skbio.sequence.Protein.iter_contiguous" title="skbio.sequence.Protein.iter_contiguous"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_contiguous</span></code></a>(included[, min_length, invert])</p></td>
<td><p>Yield contiguous subsequences based on <cite>included</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.iter_kmers.html#skbio.sequence.Protein.iter_kmers" title="skbio.sequence.Protein.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a>(k[, overlap])</p></td>
<td><p>Generate kmers of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.kmer_frequencies.html#skbio.sequence.Protein.kmer_frequencies" title="skbio.sequence.Protein.kmer_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kmer_frequencies</span></code></a>(k[, overlap, relative])</p></td>
<td><p>Return counts of words of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.lowercase.html#skbio.sequence.Protein.lowercase" title="skbio.sequence.Protein.lowercase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lowercase</span></code></a>(lowercase)</p></td>
<td><p>Return a case-sensitive string representation of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.match_frequency.html#skbio.sequence.Protein.match_frequency" title="skbio.sequence.Protein.match_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">match_frequency</span></code></a>(other[, relative])</p></td>
<td><p>Return count of positions that are the same between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.matches.html#skbio.sequence.Protein.matches" title="skbio.sequence.Protein.matches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matches</span></code></a>(other)</p></td>
<td><p>Find positions that match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.mismatch_frequency.html#skbio.sequence.Protein.mismatch_frequency" title="skbio.sequence.Protein.mismatch_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatch_frequency</span></code></a>(other[, relative])</p></td>
<td><p>Return count of positions that differ between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.mismatches.html#skbio.sequence.Protein.mismatches" title="skbio.sequence.Protein.mismatches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatches</span></code></a>(other)</p></td>
<td><p>Find positions that do not match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.nondegenerates.html#skbio.sequence.Protein.nondegenerates" title="skbio.sequence.Protein.nondegenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nondegenerates</span></code></a>()</p></td>
<td><p>Find positions containing non-degenerate characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.read.html#skbio.sequence.Protein.read" title="skbio.sequence.Protein.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a>(file[, format])</p></td>
<td><p>Create a new <code class="docutils literal notranslate"><span class="pre">Protein</span></code> instance from a file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.replace.html#skbio.sequence.Protein.replace" title="skbio.sequence.Protein.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a>(where, character)</p></td>
<td><p>Replace values in this sequence with a different character.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.stops.html#skbio.sequence.Protein.stops" title="skbio.sequence.Protein.stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stops</span></code></a>()</p></td>
<td><p>Find positions containing stop characters in the protein sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.to_indices.html#skbio.sequence.Protein.to_indices" title="skbio.sequence.Protein.to_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_indices</span></code></a>([alphabet, mask_gaps, wildcard, ...])</p></td>
<td><p>Convert the sequence into indices of characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.to_regex.html#skbio.sequence.Protein.to_regex" title="skbio.sequence.Protein.to_regex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_regex</span></code></a>([within_capture])</p></td>
<td><p>Return regular expression object that accounts for degenerate chars.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.write.html#skbio.sequence.Protein.write" title="skbio.sequence.Protein.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>(file[, format])</p></td>
<td><p>Write an instance of <code class="docutils literal notranslate"><span class="pre">Protein</span></code> to a file.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">alphabet</span></code></p></td>
<td><p>Return valid characters.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_gap_char</span></code></p></td>
<td><p>Gap character to use when constructing a new gapped sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_write_format</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">definite_chars</span></code></p></td>
<td><p>Return definite characters.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerate_chars</span></code></p></td>
<td><p>Return degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerate_map</span></code></p></td>
<td><p>Return mapping of degenerate to definite characters.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_chars</span></code></p></td>
<td><p>Return characters defined as gaps.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval_metadata</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code> object containing info about interval features.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">dict</span></code> containing metadata which applies to the entire object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">nondegenerate_chars</span></code></p></td>
<td><p>Return non-degenerate characters.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">observed_chars</span></code></p></td>
<td><p>Set of observed characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_metadata</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> containing metadata along an axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">stop_chars</span></code></p></td>
<td><p>Return characters representing translation stop codons.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code></p></td>
<td><p>Array containing underlying sequence characters.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">wildcard_char</span></code></p></td>
<td><p>Return wildcard character.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Built-ins</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__bool__.html#skbio.sequence.Protein.__bool__" title="skbio.sequence.Protein.__bool__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__bool__</span></code></a>()</p></td>
<td><p>Return truth value (truthiness) of sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__contains__.html#skbio.sequence.Protein.__contains__" title="skbio.sequence.Protein.__contains__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__contains__</span></code></a>(subsequence)</p></td>
<td><p>Determine if a subsequence is contained in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__copy__.html#skbio.sequence.Protein.__copy__" title="skbio.sequence.Protein.__copy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__copy__</span></code></a>()</p></td>
<td><p>Return a shallow copy of this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__deepcopy__.html#skbio.sequence.Protein.__deepcopy__" title="skbio.sequence.Protein.__deepcopy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__deepcopy__</span></code></a>(memo)</p></td>
<td><p>Return a deep copy of this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__eq__.html#skbio.sequence.Protein.__eq__" title="skbio.sequence.Protein.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code></a>(other)</p></td>
<td><p>Determine if this sequence is equal to another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__ge__.html#skbio.sequence.Protein.__ge__" title="skbio.sequence.Protein.__ge__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ge__</span></code></a>(value, /)</p></td>
<td><p>Return self&gt;=value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__getitem__.html#skbio.sequence.Protein.__getitem__" title="skbio.sequence.Protein.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a>(indexable)</p></td>
<td><p>Slice this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__getstate__.html#skbio.sequence.Protein.__getstate__" title="skbio.sequence.Protein.__getstate__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getstate__</span></code></a>(/)</p></td>
<td><p>Helper for pickle.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__gt__.html#skbio.sequence.Protein.__gt__" title="skbio.sequence.Protein.__gt__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__gt__</span></code></a>(value, /)</p></td>
<td><p>Return self&gt;value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__iter__.html#skbio.sequence.Protein.__iter__" title="skbio.sequence.Protein.__iter__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__iter__</span></code></a>()</p></td>
<td><p>Iterate over positions in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__le__.html#skbio.sequence.Protein.__le__" title="skbio.sequence.Protein.__le__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__le__</span></code></a>(value, /)</p></td>
<td><p>Return self&lt;=value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__len__.html#skbio.sequence.Protein.__len__" title="skbio.sequence.Protein.__len__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__len__</span></code></a>()</p></td>
<td><p>Return the number of characters in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__lt__.html#skbio.sequence.Protein.__lt__" title="skbio.sequence.Protein.__lt__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__lt__</span></code></a>(value, /)</p></td>
<td><p>Return self&lt;value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__ne__.html#skbio.sequence.Protein.__ne__" title="skbio.sequence.Protein.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ne__</span></code></a>(other)</p></td>
<td><p>Determine if this sequence is not equal to another.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.__reversed__.html#skbio.sequence.Protein.__reversed__" title="skbio.sequence.Protein.__reversed__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__reversed__</span></code></a>()</p></td>
<td><p>Iterate over positions in this sequence in reverse order.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.__str__.html#skbio.sequence.Protein.__str__" title="skbio.sequence.Protein.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__str__</span></code></a>()</p></td>
<td><p>Return sequence characters as a string.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.concat.html#skbio.sequence.Protein.concat" title="skbio.sequence.Protein.concat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code></a>(sequences[, how])</p></td>
<td><p>Concatenate an iterable of <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.count.html#skbio.sequence.Protein.count" title="skbio.sequence.Protein.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(subsequence[, start, end])</p></td>
<td><p>Count occurrences of a subsequence in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.definites.html#skbio.sequence.Protein.definites" title="skbio.sequence.Protein.definites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">definites</span></code></a>()</p></td>
<td><p>Find positions containing definite characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.degap.html#skbio.sequence.Protein.degap" title="skbio.sequence.Protein.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>()</p></td>
<td><p>Return a new sequence with gap characters removed.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.degenerates.html#skbio.sequence.Protein.degenerates" title="skbio.sequence.Protein.degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degenerates</span></code></a>()</p></td>
<td><p>Find positions containing degenerate characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.distance.html#skbio.sequence.Protein.distance" title="skbio.sequence.Protein.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other[, metric])</p></td>
<td><p>Compute the distance to another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.expand_degenerates.html#skbio.sequence.Protein.expand_degenerates" title="skbio.sequence.Protein.expand_degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">expand_degenerates</span></code></a>()</p></td>
<td><p>Yield all possible definite versions of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.find_motifs.html#skbio.sequence.Protein.find_motifs" title="skbio.sequence.Protein.find_motifs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_motifs</span></code></a>(motif_type[, min_length, ignore])</p></td>
<td><p>Search the biological sequence for motifs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.find_with_regex.html#skbio.sequence.Protein.find_with_regex" title="skbio.sequence.Protein.find_with_regex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_with_regex</span></code></a>(regex[, ignore])</p></td>
<td><p>Generate slices for patterns matched by a regular expression.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.frequencies.html#skbio.sequence.Protein.frequencies" title="skbio.sequence.Protein.frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frequencies</span></code></a>([chars, relative])</p></td>
<td><p>Compute frequencies of characters in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.gaps.html#skbio.sequence.Protein.gaps" title="skbio.sequence.Protein.gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gaps</span></code></a>()</p></td>
<td><p>Find positions containing gaps in the biological sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_definites.html#skbio.sequence.Protein.has_definites" title="skbio.sequence.Protein.has_definites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_definites</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more definite characters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_degenerates.html#skbio.sequence.Protein.has_degenerates" title="skbio.sequence.Protein.has_degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_degenerates</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_gaps.html#skbio.sequence.Protein.has_gaps" title="skbio.sequence.Protein.has_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_gaps</span></code></a>()</p></td>
<td><p>Determine if the sequence contains one or more gap characters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_interval_metadata.html#skbio.sequence.Protein.has_interval_metadata" title="skbio.sequence.Protein.has_interval_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_interval_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has interval metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_metadata.html#skbio.sequence.Protein.has_metadata" title="skbio.sequence.Protein.has_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_nondegenerates.html#skbio.sequence.Protein.has_nondegenerates" title="skbio.sequence.Protein.has_nondegenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_nondegenerates</span></code></a>()</p></td>
<td><p>Determine if sequence contains one or more non-degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_positional_metadata.html#skbio.sequence.Protein.has_positional_metadata" title="skbio.sequence.Protein.has_positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_positional_metadata</span></code></a>()</p></td>
<td><p>Determine if the object has positional metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.has_stops.html#skbio.sequence.Protein.has_stops" title="skbio.sequence.Protein.has_stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_stops</span></code></a>()</p></td>
<td><p>Determine if the sequence contains one or more stop characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.index.html#skbio.sequence.Protein.index" title="skbio.sequence.Protein.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a>(subsequence[, start, end])</p></td>
<td><p>Find position where subsequence first occurs in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.iter_contiguous.html#skbio.sequence.Protein.iter_contiguous" title="skbio.sequence.Protein.iter_contiguous"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_contiguous</span></code></a>(included[, min_length, invert])</p></td>
<td><p>Yield contiguous subsequences based on <cite>included</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.iter_kmers.html#skbio.sequence.Protein.iter_kmers" title="skbio.sequence.Protein.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a>(k[, overlap])</p></td>
<td><p>Generate kmers of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.kmer_frequencies.html#skbio.sequence.Protein.kmer_frequencies" title="skbio.sequence.Protein.kmer_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kmer_frequencies</span></code></a>(k[, overlap, relative])</p></td>
<td><p>Return counts of words of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.lowercase.html#skbio.sequence.Protein.lowercase" title="skbio.sequence.Protein.lowercase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lowercase</span></code></a>(lowercase)</p></td>
<td><p>Return a case-sensitive string representation of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.match_frequency.html#skbio.sequence.Protein.match_frequency" title="skbio.sequence.Protein.match_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">match_frequency</span></code></a>(other[, relative])</p></td>
<td><p>Return count of positions that are the same between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.matches.html#skbio.sequence.Protein.matches" title="skbio.sequence.Protein.matches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matches</span></code></a>(other)</p></td>
<td><p>Find positions that match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.mismatch_frequency.html#skbio.sequence.Protein.mismatch_frequency" title="skbio.sequence.Protein.mismatch_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatch_frequency</span></code></a>(other[, relative])</p></td>
<td><p>Return count of positions that differ between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.mismatches.html#skbio.sequence.Protein.mismatches" title="skbio.sequence.Protein.mismatches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatches</span></code></a>(other)</p></td>
<td><p>Find positions that do not match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.nondegenerates.html#skbio.sequence.Protein.nondegenerates" title="skbio.sequence.Protein.nondegenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nondegenerates</span></code></a>()</p></td>
<td><p>Find positions containing non-degenerate characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.read.html#skbio.sequence.Protein.read" title="skbio.sequence.Protein.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a>(file[, format])</p></td>
<td><p>Create a new <code class="docutils literal notranslate"><span class="pre">Protein</span></code> instance from a file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.replace.html#skbio.sequence.Protein.replace" title="skbio.sequence.Protein.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a>(where, character)</p></td>
<td><p>Replace values in this sequence with a different character.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.stops.html#skbio.sequence.Protein.stops" title="skbio.sequence.Protein.stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">stops</span></code></a>()</p></td>
<td><p>Find positions containing stop characters in the protein sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.to_indices.html#skbio.sequence.Protein.to_indices" title="skbio.sequence.Protein.to_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_indices</span></code></a>([alphabet, mask_gaps, wildcard, ...])</p></td>
<td><p>Convert the sequence into indices of characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Protein.to_regex.html#skbio.sequence.Protein.to_regex" title="skbio.sequence.Protein.to_regex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_regex</span></code></a>([within_capture])</p></td>
<td><p>Return regular expression object that accounts for degenerate chars.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Protein.write.html#skbio.sequence.Protein.write" title="skbio.sequence.Protein.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>(file[, format])</p></td>
<td><p>Write an instance of <code class="docutils literal notranslate"><span class="pre">Protein</span></code> to a file.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.sequence.RNA.__str__.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.sequence.RNA.__str__</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.sequence.Protein.concat.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.sequence.Protein.concat</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Protein"><code class="docutils literal notranslate"><span class="pre">Protein</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.2.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>