
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Scoring matrix between characters in biological sequences." />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.sequence.SubstitutionMatrix &#8212; scikit-bio 0.6.4-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=dcd4525e"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.sequence.SubstitutionMatrix';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.sequence.SubstitutionMatrix.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.sequence.SubstitutionMatrix.by_name" href="skbio.sequence.SubstitutionMatrix.by_name.html" />
    <link rel="prev" title="skbio.sequence.GeneticCode.translate_six_frames" href="skbio.sequence.GeneticCode.translate_six_frames.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.4-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.4-dev documentation"/>
    <img src="../_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.6.4-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.Sequence.html">skbio.sequence.Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GrammaredSequence.html">skbio.sequence.GrammaredSequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.DNA.html">skbio.sequence.DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.RNA.html">skbio.sequence.RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.Protein.html">skbio.sequence.Protein</a></li>
</ul>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GeneticCode.html">skbio.sequence.GeneticCode</a></li>
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">skbio.sequence.SubstitutionMatrix</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.by_name.html">skbio.sequence.SubstitutionMatrix.by_name</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.copy.html">skbio.sequence.SubstitutionMatrix.copy</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.from_dict.html">skbio.sequence.SubstitutionMatrix.from_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.get_names.html">skbio.sequence.SubstitutionMatrix.get_names</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.identity.html">skbio.sequence.SubstitutionMatrix.identity</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.to_dict.html">skbio.sequence.SubstitutionMatrix.to_dict</a></li>
</ul>
</details></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.distance.html">Sequence distance metrics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.distance</span></code>)</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.NucleotideMixin.html">skbio.sequence.NucleotideMixin</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../sequence.html" class="nav-link">Biological Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">skbio.sequence.SubstitutionMatrix</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-sequence-substitutionmatrix">
<h1>skbio.sequence.SubstitutionMatrix<a class="headerlink" href="#skbio-sequence-substitutionmatrix" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="skbio.sequence.SubstitutionMatrix">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">skbio.sequence.</span></span><span class="sig-name descname"><span class="pre">SubstitutionMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scores</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/main/skbio/sequence/_substitution.py#L18"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.SubstitutionMatrix" title="Link to this definition">#</a></dt>
<dd><p>Scoring matrix between characters in biological sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>alphabet</strong><span class="classifier">iterable</span></dt><dd><p>Characters that constitute the alphabet.</p>
</dd>
<dt><strong>scores</strong><span class="classifier">2D array-like</span></dt><dd><p>Scores of substitutions from one character (row, or axis=0) to another
character (column, or axis=1).</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments for the <code class="docutils literal notranslate"><span class="pre">DissimilarityMatrix</span></code> constructor.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix" title="skbio.stats.distance.DissimilarityMatrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.stats.distance.DissimilarityMatrix</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>A substitution matrix (a.k.a. replacement matrix) scores the substitution
of each character by each other character and itself in an alphabet. The
score usually represents the rate of substitution over evolutionary time in
a biological sequence. A higher score usually indicates higher similarity
in chemical properties or functional roles of two molecules, therefore a
mutation from one to the other is easier. In sequence alignment, the score
can measure the likelihood that a pair of aligned characters are homologous
rather than by chance.</p>
<p>This class provides a generalized interface for substitution matrices. The
alphabet usually consists of individual characters, such as nucleotides or
amino acids, but it can be generalized to any iterable of scalars (numbers,
strings, etc.). Therefore, you may use this class to construct substitution
matrices of complicated biological units (such as codons or non-canonical
amino acids). The score matrix may be symmetric, as many existing matrices
are, or asymmetric, where the score of one character substituted by another
is unequal to the other way around. Only square matrices (i.e., numbers of
rows and columns are equal) are supported.</p>
<p>Multiple commonly used nucleotide and amino acid substitution matrices are
pre-defined and can be referred to by name. Examples include NUC.4.4 for
nucleotides, and variants of BLOSUM and PAM matrices for amino acids.</p>
<p><code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code> is a subclass of <code class="docutils literal notranslate"><span class="pre">DissimilarityMatrix</span></code>. Therefore,
all attributes and methods of the latter also apply to the former.</p>
<p>The default floating-point data type of <code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code> is float32. If you
create with an integer array or nested lists, it will be automatically casted into
float32. However, if the input array is already in float64 type, the type will be
preserved.</p>
<p>The underlying array of an <code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code> object is guaranteed to be
C-contiguous, which facilitates row-major operations.</p>
<p class="rubric">Examples</p>
<p>Create a simple substitution matrix between the four nucleotide bases “A”, “C”, “G”
and “T”. The value 2 in the main diagonal represents a match between two identical
bases. The value -1 in the upper-right and lower-left triangles represents a
mismatch between two different bases.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SubstitutionMatrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">SubstitutionMatrix</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">(&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">scores</span>
<span class="go">array([[ 2., -1., -1., -1.],</span>
<span class="go">       [-1.,  2., -1., -1.],</span>
<span class="go">       [-1., -1.,  2., -1.],</span>
<span class="go">       [-1., -1., -1.,  2.]], dtype=float32)</span>
</pre></div>
</div>
<p>Look up the substitution score between a pair of bases.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="go">-1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">]</span>
<span class="go">2.0</span>
</pre></div>
</div>
<p>This matrix can also be created using the <code class="docutils literal notranslate"><span class="pre">identity</span></code> method.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">SubstitutionMatrix</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>Load the pre-defined substitution matrix “NUC.4.4”, which covers the four bases plus
degenerate codes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">SubstitutionMatrix</span><span class="o">.</span><span class="n">by_name</span><span class="p">(</span><span class="s1">&#39;NUC.4.4&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">(&#39;A&#39;, &#39;T&#39;, &#39;G&#39;, &#39;C&#39;, &#39;S&#39;, &#39;W&#39;, &#39;R&#39;, &#39;Y&#39;, &#39;K&#39;, &#39;M&#39;, &#39;B&#39;, &#39;V&#39;, &#39;H&#39;, &#39;D&#39;, &#39;N&#39;)</span>
</pre></div>
</div>
<p>With a <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a> object, one can efficiently map all
characters to their indices in the substitution matrix.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GGATCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">to_indices</span><span class="p">(</span><span class="n">sm</span><span class="p">)</span>
<span class="go">array([2, 2, 0, 1, 3, 3])</span>
</pre></div>
</div>
<p>This approach enables various subsequent operations. For example, one can
efficiently create a position-to-position scoring matrix between two sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GGATCC&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AGATCT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">idx1</span><span class="p">,</span> <span class="n">idx2</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">to_indices</span><span class="p">(</span><span class="n">sm</span><span class="p">),</span> <span class="n">seq2</span><span class="o">.</span><span class="n">to_indices</span><span class="p">(</span><span class="n">sm</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">idx1</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">],</span> <span class="n">idx2</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:]]</span>
<span class="go">array([[-4.,  5., -4., -4., -4., -4.],</span>
<span class="go">       [-4.,  5., -4., -4., -4., -4.],</span>
<span class="go">       [ 5., -4.,  5., -4., -4., -4.],</span>
<span class="go">       [-4., -4., -4.,  5., -4.,  5.],</span>
<span class="go">       [-4., -4., -4., -4.,  5., -4.],</span>
<span class="go">       [-4., -4., -4., -4.,  5., -4.]], dtype=float32)</span>
</pre></div>
</div>
<p>Finding indices of sequence characters is most efficient when the alphabet consists
of only ASCII codes (0 to 127). This can be determined by the <code class="docutils literal notranslate"><span class="pre">is_ascii</span></code> flag of
a substitution matrix. Most common nucleotide and amino acid substitution matrices
only contain ASCII codes.</p>
<p>However, one is not limited to ASCII characters. Using Unicode characters beyond
code point 127 is valid.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">SubstitutionMatrix</span><span class="p">(</span><span class="s1">&#39;äëïöü&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">(&#39;ä&#39;, &#39;ë&#39;, &#39;ï&#39;, &#39;ö&#39;, &#39;ü&#39;)</span>
</pre></div>
</div>
<p>Any iterables of scalars are valid alphabets, granting flexibility in working with
non-character data types. For example, one can include words or tokens in a
substitution matrix.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tokens</span> <span class="o">=</span> <span class="s1">&#39;lorem ipsum dolor sit amet&#39;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">SubstitutionMatrix</span><span class="p">(</span><span class="n">tokens</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">(&#39;lorem&#39;, &#39;ipsum&#39;, &#39;dolor&#39;, &#39;sit&#39;, &#39;amet&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span><span class="p">[</span><span class="s1">&#39;lorem&#39;</span><span class="p">,</span> <span class="s1">&#39;ipsum&#39;</span><span class="p">]</span>
<span class="go">1.0</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.SubstitutionMatrix.alphabet" title="skbio.sequence.SubstitutionMatrix.alphabet"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alphabet</span></code></a></p></td>
<td><p>Alphabet of the substitution matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.SubstitutionMatrix.is_ascii" title="skbio.sequence.SubstitutionMatrix.is_ascii"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_ascii</span></code></a></p></td>
<td><p>Whether alphabet consists of single ASCII characters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.SubstitutionMatrix.scores" title="skbio.sequence.SubstitutionMatrix.scores"><code class="xref py py-obj docutils literal notranslate"><span class="pre">scores</span></code></a></p></td>
<td><p>Matrix of substitution scores.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Attributes (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.T" title="skbio.stats.distance.DissimilarityMatrix.T"><code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code></a></p></td>
<td><p>Transpose of the dissimilarity matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.data" title="skbio.stats.distance.DissimilarityMatrix.data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">data</span></code></a></p></td>
<td><p>Array of dissimilarities.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.default_write_format" title="skbio.stats.distance.DissimilarityMatrix.default_write_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_write_format</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.dtype" title="skbio.stats.distance.DissimilarityMatrix.dtype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dtype</span></code></a></p></td>
<td><p>Data type of the dissimilarities.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.ids" title="skbio.stats.distance.DissimilarityMatrix.ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ids</span></code></a></p></td>
<td><p>Tuple of object IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.shape" title="skbio.stats.distance.DissimilarityMatrix.shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shape</span></code></a></p></td>
<td><p>Two-element tuple containing the dissimilarity matrix dimensions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.size" title="skbio.stats.distance.DissimilarityMatrix.size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">size</span></code></a></p></td>
<td><p>Total number of elements in the dissimilarity matrix.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.by_name.html#skbio.sequence.SubstitutionMatrix.by_name" title="skbio.sequence.SubstitutionMatrix.by_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">by_name</span></code></a>(name)</p></td>
<td><p>Load a pre-defined substitution matrix by its name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.copy.html#skbio.sequence.SubstitutionMatrix.copy" title="skbio.sequence.SubstitutionMatrix.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Return a deep copy of the substitution matrix.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.from_dict.html#skbio.sequence.SubstitutionMatrix.from_dict" title="skbio.sequence.SubstitutionMatrix.from_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_dict</span></code></a>(dictionary[, dtype])</p></td>
<td><p>Create a substitution matrix from a 2D dictionary.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.get_names.html#skbio.sequence.SubstitutionMatrix.get_names" title="skbio.sequence.SubstitutionMatrix.get_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_names</span></code></a>()</p></td>
<td><p>List names of pre-defined substitution matrices.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.identity.html#skbio.sequence.SubstitutionMatrix.identity" title="skbio.sequence.SubstitutionMatrix.identity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">identity</span></code></a>(alphabet, match, mismatch[, dtype])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.to_dict.html#skbio.sequence.SubstitutionMatrix.to_dict" title="skbio.sequence.SubstitutionMatrix.to_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dict</span></code></a>()</p></td>
<td><p>Create a 2D dictionary from the substitution matrix.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.between.html#skbio.stats.distance.DissimilarityMatrix.between" title="skbio.stats.distance.DissimilarityMatrix.between"><code class="xref py py-obj docutils literal notranslate"><span class="pre">between</span></code></a>(from_, to_[, allow_overlap])</p></td>
<td><p>Obtain the distances between the two groups of IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.filter.html#skbio.stats.distance.DissimilarityMatrix.filter" title="skbio.stats.distance.DissimilarityMatrix.filter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter</span></code></a>(ids[, strict])</p></td>
<td><p>Filter the dissimilarity matrix by IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.from_iterable.html#skbio.stats.distance.DissimilarityMatrix.from_iterable" title="skbio.stats.distance.DissimilarityMatrix.from_iterable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_iterable</span></code></a>(iterable, metric[, key, keys])</p></td>
<td><p>Create DissimilarityMatrix from an iterable given a metric.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.index.html#skbio.stats.distance.DissimilarityMatrix.index" title="skbio.stats.distance.DissimilarityMatrix.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a>(lookup_id)</p></td>
<td><p>Return the index of the specified ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.plot.html#skbio.stats.distance.DissimilarityMatrix.plot" title="skbio.stats.distance.DissimilarityMatrix.plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot</span></code></a>([cmap, title])</p></td>
<td><p>Create a heatmap of the dissimilarity matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.read.html#skbio.stats.distance.DissimilarityMatrix.read" title="skbio.stats.distance.DissimilarityMatrix.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a>([format])</p></td>
<td><p>Create a new <code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code> instance from a file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.redundant_form.html#skbio.stats.distance.DissimilarityMatrix.redundant_form" title="skbio.stats.distance.DissimilarityMatrix.redundant_form"><code class="xref py py-obj docutils literal notranslate"><span class="pre">redundant_form</span></code></a>()</p></td>
<td><p>Return an array of dissimilarities in redundant format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.rename.html#skbio.stats.distance.DissimilarityMatrix.rename" title="skbio.stats.distance.DissimilarityMatrix.rename"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rename</span></code></a>(mapper[, strict])</p></td>
<td><p>Rename IDs in the dissimilarity matrix.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.to_data_frame.html#skbio.stats.distance.DissimilarityMatrix.to_data_frame" title="skbio.stats.distance.DissimilarityMatrix.to_data_frame"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_data_frame</span></code></a>()</p></td>
<td><p>Create a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> from this <code class="docutils literal notranslate"><span class="pre">DissimilarityMatrix</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.transpose.html#skbio.stats.distance.DissimilarityMatrix.transpose" title="skbio.stats.distance.DissimilarityMatrix.transpose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transpose</span></code></a>()</p></td>
<td><p>Return the transpose of the dissimilarity matrix.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.within.html#skbio.stats.distance.DissimilarityMatrix.within" title="skbio.stats.distance.DissimilarityMatrix.within"><code class="xref py py-obj docutils literal notranslate"><span class="pre">within</span></code></a>(ids)</p></td>
<td><p>Obtain all the distances among the set of IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.write.html#skbio.stats.distance.DissimilarityMatrix.write" title="skbio.stats.distance.DissimilarityMatrix.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>(file[, format])</p></td>
<td><p>Write an instance of <code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code> to a file.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.__contains__" title="skbio.stats.distance.DissimilarityMatrix.__contains__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__contains__</span></code></a>(lookup_id)</p></td>
<td><p>Check if the specified ID is in the dissimilarity matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.__eq__" title="skbio.stats.distance.DissimilarityMatrix.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code></a>(other)</p></td>
<td><p>Compare this dissimilarity matrix to another for equality.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ge__</span></code>(value, /)</p></td>
<td><p>Return self&gt;=value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.__getitem__" title="skbio.stats.distance.DissimilarityMatrix.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a>(index)</p></td>
<td><p>Slice into dissimilarity data by object ID or numpy indexing.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getstate__</span></code>(/)</p></td>
<td><p>Helper for pickle.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__gt__</span></code>(value, /)</p></td>
<td><p>Return self&gt;value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__le__</span></code>(value, /)</p></td>
<td><p>Return self&lt;=value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__lt__</span></code>(value, /)</p></td>
<td><p>Return self&lt;value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.__ne__" title="skbio.stats.distance.DissimilarityMatrix.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ne__</span></code></a>(other)</p></td>
<td><p>Determine whether two dissimilarity matrices are not equal.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.stats.distance.DissimilarityMatrix.html#skbio.stats.distance.DissimilarityMatrix.__str__" title="skbio.stats.distance.DissimilarityMatrix.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__str__</span></code></a>()</p></td>
<td><p>Return a string representation of the dissimilarity matrix.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Details</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.SubstitutionMatrix.alphabet">
<span class="sig-name descname"><span class="pre">alphabet</span></span><a class="headerlink" href="#skbio.sequence.SubstitutionMatrix.alphabet" title="Link to this definition">#</a></dt>
<dd><p>Alphabet of the substitution matrix.</p>
<p>Each element (character) corresponds to one row/column in the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>tuple</dt><dd><p>Alphabet of the substitution matrix.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This is an alias of <code class="docutils literal notranslate"><span class="pre">ids</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.SubstitutionMatrix.is_ascii">
<span class="sig-name descname"><span class="pre">is_ascii</span></span><a class="headerlink" href="#skbio.sequence.SubstitutionMatrix.is_ascii" title="Link to this definition">#</a></dt>
<dd><p>Whether alphabet consists of single ASCII characters.</p>
<p><cite>True</cite> if every character in the alphabet can be represented by a
single ASCII character within code point range 0 to 127.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>bool</dt><dd><p>Whether alphabet consists of single ASCII characters.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.SubstitutionMatrix.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><a class="headerlink" href="#skbio.sequence.SubstitutionMatrix.scores" title="Link to this definition">#</a></dt>
<dd><p>Matrix of substitution scores.</p>
<p>Each value corresponds to the score of substituting the row character
with the column character.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>2D np.ndarray</dt><dd><p>Matrix of substitution scores.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This is an alias of <code class="docutils literal notranslate"><span class="pre">data</span></code>.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.sequence.GeneticCode.translate_six_frames.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.sequence.GeneticCode.translate_six_frames</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.sequence.SubstitutionMatrix.by_name.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.sequence.SubstitutionMatrix.by_name</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.SubstitutionMatrix"><code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.SubstitutionMatrix.alphabet"><code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix.alphabet</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.SubstitutionMatrix.is_ascii"><code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix.is_ascii</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.SubstitutionMatrix.scores"><code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix.scores</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>