
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Perform pairwise alignment of two sequences." />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.alignment.pair_align &#8212; scikit-bio 0.6.4-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=dcd4525e"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.alignment.pair_align';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.alignment.pair_align.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.alignment.pair_align_nucl" href="skbio.alignment.pair_align_nucl.html" />
    <link rel="prev" title="skbio.alignment.PairAlignPath.to_cigar" href="skbio.alignment.PairAlignPath.to_cigar.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.4-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.4-dev documentation"/>
    <img src="../_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.6.4-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.TabularMSA.html">skbio.alignment.TabularMSA</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.AlignPath.html">skbio.alignment.AlignPath</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.PairAlignPath.html">skbio.alignment.PairAlignPath</a></li>
</ul>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">skbio.alignment.pair_align</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_nucl.html">skbio.alignment.pair_align_nucl</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.pair_align_prot.html">skbio.alignment.pair_align_prot</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.align_score.html">skbio.alignment.align_score</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_nucleotide.html">skbio.alignment.global_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align_protein.html">skbio.alignment.global_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.global_pairwise_align.html">skbio.alignment.global_pairwise_align</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_nucleotide.html">skbio.alignment.local_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align_protein.html">skbio.alignment.local_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.alignment.local_pairwise_align.html">skbio.alignment.local_pairwise_align</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../alignment.html" class="nav-link">Sequence Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">skbio.alignment.pair_align</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-alignment-pair-align">
<h1>skbio.alignment.pair_align<a class="headerlink" href="#skbio-alignment-pair-align" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="skbio.alignment.pair_align">
<span class="sig-prename descclassname"><span class="pre">skbio.alignment.</span></span><span class="sig-name descname"><span class="pre">pair_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="positional-only-separator o"><abbr title="Positional-only parameter separator (PEP 570)"><span class="pre">/</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'global'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sub_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.0,</span> <span class="pre">-1.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_cost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">free_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_paths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_matrices</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/main/skbio/alignment/_pair.py#L33"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.alignment.pair_align" title="Link to this definition">#</a></dt>
<dd><p>Perform pairwise alignment of two sequences.</p>
<p>A versatile, efficient and generalizable function for pairwise alignment using
dynamic programming. It supports:</p>
<ul class="simple">
<li><p>Global, local and semi-global alignments (with all four ends customizable).</p></li>
<li><p>Nucleotide, protein, and un-grammared sequences, plain strings (ASCII and
Unicode), words/tokens, and numbers.</p></li>
<li><p>Match/mismatch scores or substitution matrix.</p></li>
<li><p>Linear and affine gap penalties.</p></li>
<li><p>Integer, decimal and infinite scores.</p></li>
<li><p>Returning one, multiple or all optimal alignment paths.</p></li>
</ul>
<div class="versionadded">
<p><span class="versionmodified added">Added in version 0.6.4.</span></p>
</div>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl>
<dt><strong>seq1</strong><span class="classifier"><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, str, or sequence of scalar</span></dt><dd><p>The first sequence to be aligned.</p>
</dd>
<dt><strong>seq2</strong><span class="classifier"><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, str, or sequence of scalar</span></dt><dd><p>The second sequence to be aligned.</p>
</dd>
<dt><strong>mode</strong><span class="classifier">{‘global’, ‘local’}, optional</span></dt><dd><p>Mode of alignment. Options are:</p>
<ul class="simple">
<li><p>“global” (default): Global alignment, as in the Needleman-Wunsch algorithm.
The two sequences are aligned end-to-end. See also <code class="docutils literal notranslate"><span class="pre">free_ends</span></code> below.</p></li>
<li><p>“local”: Local alignment, as in the Smith-Waterman algorithm. It identifies
similar regions between two sequences.</p></li>
</ul>
</dd>
<dt><strong>sub_score</strong><span class="classifier">tuple of (float, float), SubstitutionMatrix, or str, optional</span></dt><dd><p>Score of a substitution. May be one of the following:</p>
<ul class="simple">
<li><p>Tuple of two numbers: Match score (same symbol) and mismatch score (different
symbols).</p></li>
<li><p><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix" title="skbio.sequence.SubstitutionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a>: A matrix of substitution scores
between all symbols in the alphabet.</p></li>
<li><p>String: Name of the substitution matrix that can be recognized by
<code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix.by_name</span></code>, such as “NUC.4.4” or “BLOSUM62”.</p></li>
</ul>
<p>Default is (1.0, -1.0) (match, mismatch).</p>
</dd>
<dt><strong>gap_cost</strong><span class="classifier">float or tuple of (float, float), optional</span></dt><dd><p>Penalty of a gap. Values are usually positive, representing subtractions from
the alignment score. May be one of the following:</p>
<ul class="simple">
<li><p>One number: Linear gap penalty. Each gap position is penalized by this value
(<em>g</em>). A contiguous gap of length <em>k</em> has a total penalty of <em>g</em> * <em>k</em>.</p></li>
<li><p>Tuple of two numbers: Affine gap penalty. The two numbers (<em>o</em>, <em>e</em>)
represent gap opening penalty and gap extension penalty, respectively. A
contiguous gap of length <em>k</em> has a total penalty of <em>o</em> + <em>e</em> * <em>k</em>.
See also notes below.</p></li>
</ul>
<p>Default is 2.0 (linear gap penalty).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Infinites are valid values for <code class="docutils literal notranslate"><span class="pre">sub_score</span></code> and <code class="docutils literal notranslate"><span class="pre">gap_cost</span></code>. For example,
<code class="docutils literal notranslate"><span class="pre">gap_cost=np.inf</span></code> disables gaps. <code class="docutils literal notranslate"><span class="pre">sub_score=(...,</span> <span class="pre">-np.inf)</span></code> disables
mismatches.</p>
</div>
</dd>
<dt><strong>free_ends</strong><span class="classifier">bool, 2-tuple of bool, or 4-tuple of bool, optional</span></dt><dd><p>Whether gaps at the sequence terminals are free from penalization. Relevant
when <code class="docutils literal notranslate"><span class="pre">mode</span></code> is “global”. Alignment with free ends is known as semi-global
(or “glocal”) alignment. Values can be:</p>
<ul class="simple">
<li><p>False: Penalize terminal gaps using the same method defined by <code class="docutils literal notranslate"><span class="pre">gap_cost</span></code>.
This behavior is the true global alignment.</p></li>
<li><p>True: Do not penalize any terminal gap. This behavior is known as “overlap”,
as it identifies the maximum overlap between two sequences.</p></li>
<li><p>Tuple of two Booleans: Whether terminal gaps of seq1 and seq2 are free,
respectively. For example, (True, False) is the typical semi-global
alignment, useful for aligning a short seq1 within a long seq2.</p></li>
<li><p>Tuple of four Booleans: Whether leading gap of seq1, trailing gap of seq1,
leading gap of seq2, and trailing gap of seq2 are free, respectively. For
example, (False, True, True, False) joins the tail of seq1 with the head
of seq2.</p></li>
</ul>
<p>Default is True (overlap).</p>
</dd>
<dt><strong>trim_ends</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, penalty-free terminal gaps defined by <code class="docutils literal notranslate"><span class="pre">free_ends</span></code> will be excluded
from returned paths. Useful for locating a short query in a long reference.
Default is False.</p>
</dd>
<dt><strong>max_paths</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of alignment paths to return. Default is 1, which is generated
through a performance-oriented traceback algorithm. A value larger than 1 will
trigger a less efficient traceback algorithm to enumerate up to this number of
paths. Setting it as None will return all paths. However, be cautious that the
total number of paths may be extremely large. Setting it as 0 will disable
traceback and return no path (None).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When <code class="docutils literal notranslate"><span class="pre">mode=&quot;local&quot;</span></code>, it is possible that there is no path with a score &gt;
0. In this case, an empty list will be returned. <code class="docutils literal notranslate"><span class="pre">mode=&quot;global&quot;</span></code>
guarantees to return at least one path, even if it’s completely misaligned.</p>
</div>
</dd>
<dt><strong>atol</strong><span class="classifier">float, optional</span></dt><dd><p>Absolute tolerance in comparing scores of alternative alignment paths. This is
to ensure floating-point arithmetic safety when <code class="docutils literal notranslate"><span class="pre">sub_score</span></code> or <code class="docutils literal notranslate"><span class="pre">gap_cost</span></code>
involve decimal numbers. Default is 1e-5. Setting it to 0 or None will slightly
increase performance, and is usually safe when there are only integers (e.g.,
2.0) or half integers (e.g., 2.5) involved.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Relative tolerance is not involved in the calculation.</p>
</div>
</dd>
<dt><strong>keep_matrices</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to include the alignment matrix(ces) in the returned value. They are
typically for diagnostic or educational purposes. Default is False, which lets
the memory space free up after the function finishes.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl>
<dt><strong>score</strong><span class="classifier">float</span></dt><dd><p>Optimal alignment score.</p>
</dd>
<dt><strong>paths</strong><span class="classifier">list of <a class="reference internal" href="skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-class docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a>, optional</span></dt><dd><p>Alignment paths. Up to <code class="docutils literal notranslate"><span class="pre">max_paths</span></code> paths will be returned. Note that all
paths are optimal and share the same alignment score.</p>
</dd>
<dt><strong>matrices</strong><span class="classifier">tuple of ndarray of shape (m + 1, n + 1), optional</span></dt><dd><p>Alignment matrices generated during the computation (if <code class="docutils literal notranslate"><span class="pre">keep_matrices</span></code> is
True). <em>m</em> and <em>n</em> are the lengths of seq1 and seq2, respectively.</p>
<ul class="simple">
<li><p>For linear gap penalty, one main matrix will be returned.</p></li>
<li><p>For affine gap penalty, the main matrix, plus an insertion matrix (gap in
seq1) and a deletion matrix (gap in seq2) will be returned.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.alignment.align_score.html#skbio.alignment.align_score" title="skbio.alignment.align_score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">align_score</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skbio.alignment.PairAlignPath</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>This function implements the classic dynamic programming (DP) method for pairwise
sequence alignment. It is commonly known as the Needleman-Wunsch algorithm <a class="reference internal" href="#rb1a998e68c16-1" id="id1">[1]</a> for
global alignment, or the Smith-Waterman algorithm <a class="reference internal" href="#rb1a998e68c16-2" id="id2">[2]</a> for local alignment. These
two algorithms use linear gap penalty. When affine gap penalty is specified, the
underlying method is the Gotoh algorithm <a class="reference internal" href="#rb1a998e68c16-3" id="id3">[3]</a>, with later modifications <a class="reference internal" href="#rb1a998e68c16-4" id="id4">[4]</a>. The
algorithms are exact and the output alignments are guaranteed to be optimal.</p>
<p>The algorithms are quadratic (<em>O</em>(<em>mn</em>)) in both time and space, where <em>m</em> and
<em>n</em> are the lengths of the two sequences, respectively. Affine gap penalty costs
3x as much memory and 2-3x as much runtime than linear gap penalty.</p>
<p><strong>Parameters</strong></p>
<p>The default parameters <code class="docutils literal notranslate"><span class="pre">sub_score=(1,</span> <span class="pre">-1),</span> <span class="pre">gap_cost=2</span></code> is a simple scoring scheme
that quickly captures sequence similarity. However, in bioinformatics applications,
one usually wants to replace them with more realistic parameters according to the
specific task. For reference, below are the default parameters of some common
bioinformatics programs:</p>
<ul class="simple">
<li><p>NCBI MegaBLAST (nucleotide): <code class="docutils literal notranslate"><span class="pre">sub_score=(1,</span> <span class="pre">-2),</span> <span class="pre">gap_cost=2.5</span></code></p></li>
<li><p>NCBI BLASTN (nucleotide): <code class="docutils literal notranslate"><span class="pre">sub_score=(2,</span> <span class="pre">-3),</span> <span class="pre">gap_cost=(5,</span> <span class="pre">2)</span></code>
(see also <a class="reference internal" href="skbio.alignment.pair_align_nucl.html#skbio.alignment.pair_align_nucl" title="skbio.alignment.pair_align_nucl"><code class="xref py py-func docutils literal notranslate"><span class="pre">pair_align_nucl</span></code></a>)</p></li>
<li><p>NCBI BLASTP (protein): <code class="docutils literal notranslate"><span class="pre">sub_score=&quot;BLOSUM62&quot;,</span> <span class="pre">gap_cost=(11,</span> <span class="pre">1)</span></code>
(see also <a class="reference internal" href="skbio.alignment.pair_align_prot.html#skbio.alignment.pair_align_prot" title="skbio.alignment.pair_align_prot"><code class="xref py py-func docutils literal notranslate"><span class="pre">pair_align_prot</span></code></a>)</p></li>
<li><p>EMBOSS Needle/Water (protein): <code class="docutils literal notranslate"><span class="pre">sub_score=&quot;BLOSUM62&quot;,</span> <span class="pre">gap_cost=(9.5,</span> <span class="pre">0.5)</span></code>
(for nucleotide, replace <code class="docutils literal notranslate"><span class="pre">sub_score</span></code> with <code class="docutils literal notranslate"><span class="pre">&quot;NUC.4.4&quot;</span></code>)</p></li>
<li><p>EBI FASTA (protein): <code class="docutils literal notranslate"><span class="pre">sub_score=&quot;BLOSUM50&quot;,</span> <span class="pre">gap_cost=(8,</span> <span class="pre">2)</span></code></p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These are scoring schemes only. <code class="docutils literal notranslate"><span class="pre">pair_align</span></code> does not reproduce the output
of the programs mentioned.</p>
</div>
<p>The flexibility of these parameters, especially the support for infinity, enables
some common tasks in text analysis:</p>
<ul class="simple">
<li><p>Edit (Levenshtein) distance: <code class="docutils literal notranslate"><span class="pre">sub_score=(0,</span> <span class="pre">-1),</span> <span class="pre">gap_cost=1</span></code> (negate the score
to get the distance)</p></li>
<li><p>Longest common subsequence (LCS):
<code class="docutils literal notranslate"><span class="pre">mode=&quot;local&quot;,</span> <span class="pre">sub_score=(1,</span> <span class="pre">-np.inf),</span> <span class="pre">gap_cost=0</span></code></p></li>
<li><p>Longest common substring:
<code class="docutils literal notranslate"><span class="pre">mode=&quot;local&quot;,</span> <span class="pre">sub_score=(1,</span> <span class="pre">-np.inf),</span> <span class="pre">gap_cost=np.inf</span></code></p></li>
</ul>
<p><strong>Input sequences</strong></p>
<p>The format of input sequences is flexible. The function is most efficient when they
are subclasses of <a class="reference internal" href="skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a>, such as <code class="docutils literal notranslate"><span class="pre">DNA</span></code> and
<code class="docutils literal notranslate"><span class="pre">Protein</span></code>, or any user-customized class. They have a finite alphabet, and permit
auto-replacing degenerate codes with wildcard (such as nucleotides to “N” and amino
acids to “X”), as long as the substitution matrix has the wildcard character.</p>
<p>For non-grammared sequences, such as plain strings, bytes, and integer arrays, the
the function will attempt to encode them into ASCII codes (<em>n</em> = 128), which permit
efficient indexing. If ASCII encoding is not possible (for example lists of words,
large or negative integers, or floats), unique symbols will be extracted from the
sequences and indexed.</p>
<p><strong>Performance</strong></p>
<p>The underlying dynamic programming kernel is a plain dual loop structure, without
further vectorization or parallelization techniques.</p>
<p>This function does not discriminate between seq1 and seq2. Nevertheless, aligning a
shorter seq1 (often referred to as “query”) against a longer seq2 (often referred to
as “target” or “reference”) is usually more efficient than the other way around.</p>
<p>The algorithm defaults to float32, which is sufficient for most use cases. It also
supports float64, with a higher memory cost (2x) and moderately increased runtime.
If float64 is what you need, you may supply a substitution matrix of float64 type,
using e.g., <code class="docutils literal notranslate"><span class="pre">SubstitutionMatrix.identity(&quot;ACGT&quot;,</span> <span class="pre">1,</span> <span class="pre">-2,</span> <span class="pre">dtype=&quot;float64&quot;)</span></code>, which
will be respected by the function without casting throughout calculation.</p>
<p><strong>Affine gap penalty</strong></p>
<p>Under the affine gap penalty mode, the penalty of a contiguous gap of length
<span class="math notranslate nohighlight">\(k\)</span> is calculated as:</p>
<div class="math notranslate nohighlight">
\[G(k) = o + e \times k \tag{1}\]</div>
<p>where <span class="math notranslate nohighlight">\(o\)</span> is the gap opening penalty and <span class="math notranslate nohighlight">\(e\)</span> is the gap extension
penalty.</p>
<p>It should be noted that, discrepancy exists among literature and implementations
regarding whether gap extension penalty should apply to the first position of a
gap. scikit-bio’s equation is consistent with multiple common alignment tools,
such as BLAST <a class="reference internal" href="#rb1a998e68c16-5" id="id5">[5]</a>, Minimap2, SeqAn3, and WFA2-lib.</p>
<p>Meanwhile, multiple other tools, such as EMBOSS, parasail, Biopython and Biotite,
use the following equation instead:</p>
<div class="math notranslate nohighlight">
\[G(k) = o + e \times (k - 1) \tag{2}\]</div>
<p>Therefore, if you intend to reproduce the behavior of a software tool of the
latter category using scikit-bio, you will need to subtract <span class="math notranslate nohighlight">\(e\)</span> from
<span class="math notranslate nohighlight">\(o\)</span> when adopting its parameters. For example, EMBOSS’ default parameters
<code class="docutils literal notranslate"><span class="pre">o=10,</span> <span class="pre">e=0.5</span></code> will become <code class="docutils literal notranslate"><span class="pre">o=9.5,</span> <span class="pre">e=0.5</span></code> in scikit-bio. Vice versa.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.6.4: </span>Previous alignment algorithms in scikit-bio used Eq. 2. These functions were
deprecated in 0.5.x and will be removed in 0.6.x. Future functions will
uniformly use Eq. 1.</p>
</div>
<p>Related: A simple but less common scenario is constant gap penalty, where a
contiguous gap has a constant penalty <span class="math notranslate nohighlight">\(g\)</span> regardless of its length. This can
be specified as <code class="docutils literal notranslate"><span class="pre">gap_cost=(g,</span> <span class="pre">0)</span></code>.</p>
<p class="rubric">References</p>
<div role="list" class="citation-list">
<div class="citation" id="rb1a998e68c16-1" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Needleman, S. B., &amp; Wunsch, C. D. (1970). A general method applicable to the
search for similarities in the amino acid sequence of two proteins. J Mol Biol,
48(3), 443-453.</p>
</div>
<div class="citation" id="rb1a998e68c16-2" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>Smith, T. F., &amp; Waterman, M. S. (1981). Identification of common molecular
subsequences. J Mol Biol, 147(1), 195-197.</p>
</div>
<div class="citation" id="rb1a998e68c16-3" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">3</a><span class="fn-bracket">]</span></span>
<p>Gotoh, O. (1982). An improved algorithm for matching biological sequences.
J Mol Biol, 162(3), 705-708.</p>
</div>
<div class="citation" id="rb1a998e68c16-4" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">4</a><span class="fn-bracket">]</span></span>
<p>Altschul, S. F., &amp; Erickson, B. W. (1986). Optimal sequence alignment using
affine gap costs. Bull Math Biol, 48, 603-616.</p>
</div>
<div class="citation" id="rb1a998e68c16-5" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">5</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK62051/">https://www.ncbi.nlm.nih.gov/books/NBK62051/</a></p>
</div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">pair_align</span>
</pre></div>
</div>
<p>Align two DNA sequences using default parameters (global alignment with free end
gaps, match/mismatch scores, linear gap penalty, returning just one path):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACTACCAGATTACTTACGGATCAGGTACTTGCCAACAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGAAACTACTAGATTACGGATCTTACTTTCCAGCAAGG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
</pre></div>
</div>
<p>The result is a named tuple consisting of score, path(s), and matrices (off by
default). The score represents the “goodness” of the alignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">res</span><span class="o">.</span><span class="n">score</span>
<span class="go">12.0</span>
</pre></div>
</div>
<p>The alignment path can be represented by a
<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Sequence_alignment#CIGAR_Format">CIGAR string</a> . See
<a class="reference internal" href="skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-class docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a> for details.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span>
<span class="go">&lt;PairAlignPath, positions: 44, segments: 7, CIGAR: &#39;4I13M4D6M2D13M2I&#39;&gt;</span>
</pre></div>
</div>
<p>Extract aligned sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">to_aligned</span><span class="p">((</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">[&#39;----ACTACCAGATTACTTACGGATCAGGTACTTGCCAACAA--&#39;,</span>
<span class="go"> &#39;CGAAACTACTAGATTAC----GGATCT--TACTTTCCAGCAAGG&#39;]</span>
</pre></div>
</div>
<p>Alternatively, you can do <code class="docutils literal notranslate"><span class="pre">TabularMSA.from_path_seqs(path,</span> <span class="pre">(seq1,</span> <span class="pre">seq2))</span></code> to
create a dedicated sequence alignment structure. See <a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-class docutils literal notranslate"><span class="pre">TabularMSA</span></code></a> for
details.</p>
<p>Align two protein sequences using local alignment, substitution matrix, and affine
gap penalty (the default parameters of BLASTP):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s1">&#39;MKRTLKGHFVQWC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s1">&#39;MQMLKTHYAQTRN&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span><span class="p">,</span> <span class="n">paths</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;local&#39;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">sub_score</span><span class="o">=</span><span class="s1">&#39;BLOSUM62&#39;</span><span class="p">,</span> <span class="n">gap_cost</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">23.0</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_aligned</span><span class="p">((</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="go">[&#39;LKGHFVQ&#39;, &#39;LKTHYAQ&#39;]</span>
</pre></div>
</div>
<p>Search a sequencing read against a reference genome using the default parameters of
MegaBLAST and return locations of all hits.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;ACCGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;AAACGCTACCGTCCGTAGACCGTGACCGTGCGAAGC&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span><span class="p">,</span> <span class="n">paths</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;global&#39;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">sub_score</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">),</span> <span class="n">gap_cost</span><span class="o">=</span><span class="mf">2.5</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">free_ends</span><span class="o">=</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span> <span class="n">trim_ends</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">max_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">paths</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">ranges</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">[ 7 12]</span>
<span class="go">[18 23]</span>
<span class="go">[24 29]</span>
</pre></div>
</div>
<p>Note: <code class="docutils literal notranslate"><span class="pre">ranges[1]</span></code> stores the start and stop positions of the aligned region in
the target (<code class="docutils literal notranslate"><span class="pre">ranges[0]</span></code> stores those of the query). The numbers are BED-like
(i.e., 0-based, half-open).</p>
<p>Calculate the edit distance between two sentences (lists of words).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">text1</span> <span class="o">=</span> <span class="s1">&#39;The quick brown fox jumps over the lazy dog&#39;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">text2</span> <span class="o">=</span> <span class="s1">&#39;The slow brown wolf jumps over a lazy dog&#39;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">text1</span><span class="p">,</span> <span class="n">text2</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;global&#39;</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="n">sub_score</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">gap_cost</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="n">free_ends</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">max_paths</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.alignment.PairAlignPath.to_cigar.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.alignment.PairAlignPath.to_cigar</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.alignment.pair_align_nucl.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.alignment.pair_align_nucl</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.alignment.pair_align"><code class="docutils literal notranslate"><span class="pre">pair_align</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>