
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>FASTQ format (skbio.io.format.fastq) &#8212; scikit-bio 0.6.3-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=87e54e7c" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=4317e44f"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.io.format.fastq';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.4';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.io.format.fastq.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="GenBank format (skbio.io.format.genbank)" href="skbio.io.format.genbank.html" />
    <link rel="prev" title="FASTA/QUAL format (skbio.io.format.fasta)" href="skbio.io.format.fasta.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.3-dev documentation"/>
    <script>document.write(`<img src="../_static/logo_inv.svg" class="logo__image only-dark" alt="scikit-bio 0.6.3-dev documentation"/>`);</script>
  
  
</a></div>
    
      <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.binary_dm.html">Simple binary dissimilarity matrix format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.binary_dm</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.biom.html">BIOM-Format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.biom</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.blast6.html">BLAST+6 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast6</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.blast7.html">BLAST+7 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.blast7</span></code>)</a></li>


<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.clustal.html">Clustal format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.clustal</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.embl.html">EMBL format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.embl</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.embed.html">Embedding format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.embed</span></code>).</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.fasta.html">FASTA/QUAL format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code>)</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">FASTQ format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.genbank.html">GenBank format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.genbank</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.gff3.html">GFF3 format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.gff3</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.lsmat.html">Labeled square matrix format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.lsmat</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.newick.html">Newick format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.newick</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.ordination.html">Ordination results format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.ordination</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.phylip.html">PHYLIP multiple sequence alignment format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.phylip</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.qseq.html">QSeq format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.qseq</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.stockholm.html">Stockholm format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.stockholm</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.taxdump.html">Taxdump format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.taxdump</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.format.sample_metadata.html">Format Support</a></li>

</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.write.html">skbio.io.registry.write</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.read.html">skbio.io.registry.read</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.sniff.html">skbio.io.registry.sniff</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.util.html">I/O utilities (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.util</span></code>)</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.io.registry.html">I/O Registry (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.registry</span></code>)</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../io.html" class="nav-link">Input and Output (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">FASTQ...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="fastq-format-skbio-io-format-fastq">
<span id="module-skbio.io.format.fastq"></span><h1>FASTQ format (<a class="reference internal" href="#module-skbio.io.format.fastq" title="skbio.io.format.fastq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fastq</span></code></a>)<a class="headerlink" href="#fastq-format-skbio-io-format-fastq" title="Link to this heading">#</a></h1>
<p>The FASTQ file format (<code class="docutils literal notranslate"><span class="pre">fastq</span></code>) stores biological (e.g., nucleotide)
sequences and their quality scores in a simple plain text format that is both
human-readable and easy to parse. The file format was invented by Jim Mullikin
at the Wellcome Trust Sanger Institute but wasn’t given a formal definition,
though it has informally become a standard file format for storing
high-throughput sequence data. More information about the format and its
variants can be found in <a class="reference internal" href="#r951e5a4cbfda-1" id="id1">[1]</a> and <a class="reference internal" href="#r951e5a4cbfda-2" id="id2">[2]</a>.</p>
<p>Conceptually, a FASTQ file is similar to paired FASTA and QUAL files in that it
stores both biological sequences and their quality scores. FASTQ differs from
FASTA/QUAL because the quality scores are stored in the same file as the
biological sequence data.</p>
<p>An example FASTQ-formatted file containing two DNA sequences and their quality
scores:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>@seq1 description 1
AACACCAAACTTCTCCACCACGTGAGCTACAAAAG
+
````Y^T]`]c^cabcacc`^Lb^ccYT\T\Y\WF
@seq2 description 2
TATGTATATATAACATATACATATATACATACATA
+
]KZ[PY]_[YY^```ac^\\`bT``c`\aT``bbb
</pre></div>
</div>
<section id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Link to this heading">#</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Reader</p></th>
<th class="head"><p>Writer</p></th>
<th class="head"><p>Object Class</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p>generator of <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a> objects</p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment.TabularMSA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Sequence</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.DNA</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.RNA</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p>Yes</p></td>
<td><p>Yes</p></td>
<td><p><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.Protein</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Link to this heading">#</a></h2>
<p>A FASTQ file contains one or more biological sequences and their corresponding
quality scores stored sequentially as <em>records</em>. Each <em>record</em> consists of four
sections:</p>
<ol class="arabic simple">
<li><p>Sequence header line consisting of a sequence identifier (ID) and
description (both optional)</p></li>
<li><p>Biological sequence data (typically stored using the standard IUPAC
lexicon), optionally split over multiple lines</p></li>
<li><p>Quality header line separating sequence data from quality scores (optionally
repeating the ID and description from the sequence header line)</p></li>
<li><p>Quality scores as printable ASCII characters, optionally split over multiple
lines. Decoding of quality scores will depend on the specified FASTQ variant
(see below for more details)</p></li>
</ol>
<p>For the complete FASTQ format specification, see <a class="reference internal" href="#r951e5a4cbfda-1" id="id3">[1]</a>. scikit-bio’s FASTQ
implementation follows the format specification described in this excellent
publication, including validating the implementation against the FASTQ example
files provided in the publication’s supplementary data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>IDs and descriptions will be parsed from sequence header lines in
exactly the same way as FASTA headers (<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>). IDs,
descriptions, and quality scores are also stored on, and written from,
sequence objects in the same way as with FASTA.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Blank or whitespace-only lines are only allowed at the beginning of
the file, between FASTQ records, or at the end of the file. A blank or
whitespace-only line after the header line, within the sequence, or within
quality scores will raise an error.</p>
<p>scikit-bio will ignore leading and trailing whitespace characters on each
line while reading.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Validation may be performed depending on the type of object the data
is being read into. This behavior matches that of FASTA files.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>scikit-bio will write FASTQ files in a normalized format, with each
record section on a single line. Thus, each record will be composed of
<em>exactly</em> four lines. The quality header line won’t have the sequence ID and
description repeated.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><cite>lowercase</cite> functionality is supported the same as with FASTA.</p>
</div>
<section id="quality-score-variants">
<h3>Quality Score Variants<a class="headerlink" href="#quality-score-variants" title="Link to this heading">#</a></h3>
<p>FASTQ associates quality scores with sequence data, with each quality score
encoded as a single printable ASCII character. In scikit-bio, all quality
scores are decoded as Phred quality scores. This is the most common quality
score metric, though there are others (e.g., Solexa quality scores).
Unfortunately, different sequencers have different ways of encoding quality
scores as ASCII characters, notably Sanger and Illumina. Below is a table
highlighting the different encoding variants supported by scikit-bio, as well
as listing the equivalent variant names used in the Open Bioinformatics
Foundation (OBF) <a class="reference internal" href="#r951e5a4cbfda-3" id="id4">[3]</a> projects (e.g., Biopython, BioPerl, etc.).</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Variant</p></th>
<th class="head"><p>ASCII
Range</p></th>
<th class="head"><p>Offset</p></th>
<th class="head"><p>Quality
Range</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sanger</p></td>
<td><p>33 to 126</p></td>
<td><p>33</p></td>
<td><p>0 to 93</p></td>
<td><p>Equivalent to OBF’s fastq-sanger.</p></td>
</tr>
<tr class="row-odd"><td><p>illumina1.3</p></td>
<td><p>64 to 126</p></td>
<td><p>64</p></td>
<td><p>0 to 62</p></td>
<td><p>Equivalent to OBF’s fastq-illumina. Use
this if your data was generated using
Illumina 1.3-1.7 software.</p></td>
</tr>
<tr class="row-even"><td><p>illumina1.8</p></td>
<td><p>33 to 95</p></td>
<td><p>33</p></td>
<td><p>0 to 62</p></td>
<td><p>Equivalent to sanger but with 0 to 62
quality score range check. Use this if
your data was generated using Illumina
1.8 software or later.</p></td>
</tr>
<tr class="row-odd"><td><p>solexa</p></td>
<td><p>59 to 126</p></td>
<td><p>64</p></td>
<td><p>-5 to 62</p></td>
<td><p>Not currently implemented.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When writing, Phred quality scores will be truncated to the maximum
value in the variant’s range and a warning will be issued. This is
consistent with the OBF projects.</p>
<p>When reading, an error will be raised if a decoded quality score is outside
the variant’s range.</p>
</div>
</section>
</section>
<section id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Link to this heading">#</a></h2>
<p>The following parameters are available to all FASTQ format readers and writers:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">variant</span></code>: A string indicating the quality score variant used to
decode/encode Phred quality scores. Must be one of <code class="docutils literal notranslate"><span class="pre">sanger</span></code>,
<code class="docutils literal notranslate"><span class="pre">illumina1.3</span></code>, <code class="docutils literal notranslate"><span class="pre">illumina1.8</span></code>, or <code class="docutils literal notranslate"><span class="pre">solexa</span></code>. This parameter is preferred
over <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> because additional quality score range checks and
conversions can be performed. It is also more explicit.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">phred_offset</span></code>: An integer indicating the ASCII code offset used to
decode/encode Phred quality scores. Must be in the range <code class="docutils literal notranslate"><span class="pre">[33,</span> <span class="pre">126]</span></code>. All
decoded scores will be assumed to be Phred scores (i.e., no additional
conversions are performed). Prefer using <code class="docutils literal notranslate"><span class="pre">variant</span></code> over this parameter
whenever possible.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You must provide <code class="docutils literal notranslate"><span class="pre">variant</span></code> or <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> when reading or
writing a FASTQ file. <code class="docutils literal notranslate"><span class="pre">variant</span></code> and <code class="docutils literal notranslate"><span class="pre">phred_offset</span></code> cannot both be
provided at the same time.</p>
</div>
<p>The following additional parameters are the same as in FASTA format
(<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code></a>):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">constructor</span></code>: see <code class="docutils literal notranslate"><span class="pre">constructor</span></code> parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq_num</span></code>: see <code class="docutils literal notranslate"><span class="pre">seq_num</span></code> parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code>: see <code class="docutils literal notranslate"><span class="pre">id_whitespace_replacement</span></code> parameter in
FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code>: see <code class="docutils literal notranslate"><span class="pre">description_newline_replacement</span></code>
parameter in FASTA format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lowercase</span></code>: see <code class="docutils literal notranslate"><span class="pre">lowercase</span></code> parameter in FASTA format</p></li>
</ul>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">#</a></h2>
<p>Suppose we have the following FASTQ file with two DNA sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@seq1</span> <span class="n">description</span> <span class="mi">1</span>
<span class="n">AACACCAAACTTCTCCACC</span>
<span class="n">ACGTGAGCTACAAAAG</span>
<span class="o">+</span><span class="n">seq1</span> <span class="n">description</span> <span class="mi">1</span>
<span class="sd">&#39;&#39;&#39;&#39;Y^T]&#39;]C^CABCACC</span>
<span class="sd">`^LB^CCYT\T\Y\WF</span>
<span class="sd">@seq2 description 2</span>
<span class="sd">TATGTATATATAACATATACATATATACATACATA</span>
<span class="sd">+</span>
<span class="sd">]KZ[PY]_[YY^&#39;&#39;&#39;</span><span class="n">AC</span><span class="o">^</span>\\<span class="s1">&#39;BT&#39;&#39;C&#39;</span>\<span class="n">AT</span><span class="s1">&#39;&#39;</span><span class="n">BBB</span>
</pre></div>
</div>
<p>Note that the first sequence and its quality scores are split across multiple
lines, while the second sequence and its quality scores are each on a single
line. Also note that the first sequence has a duplicate ID and description on
the quality header line, while the second sequence does not.</p>
<p>Let’s define this file in-memory as a <code class="docutils literal notranslate"><span class="pre">StringIO</span></code>, though this could be a real
file path, file handle, or anything that’s supported by scikit-bio’s I/O
registry in practice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;@seq1 description 1&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;AACACCAAACTTCTCCACC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;ACGTGAGCTACAAAAG&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;+seq1 description 1&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;&#39;&#39;&#39;&#39;Y^T]&#39;]C^CABCACC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;&#39;^LB^CCYT\T\Y\WF&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;@seq2 description 2&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;TATGTATATATAACATATACATATATACATACATA&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="sa">r</span><span class="s2">&quot;]KZ[PY]_[YY^&#39;&#39;&#39;AC^</span><span class="se">\\</span><span class="s2">&#39;BT&#39;&#39;C&#39;\AT&#39;&#39;BBB&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>To load the sequences into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>, we run:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;sanger&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">-----------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 35</span>
<span class="go">-----------------------------------</span>
<span class="go">AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span>
<span class="go">TATGTATATATAACATATACATATATACATACATA</span>
</pre></div>
</div>
<p>Note that quality scores are decoded from Sanger. To load the second sequence
as <code class="docutils literal notranslate"><span class="pre">DNA</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span> <span class="c1"># reload the StringIO to read from the beginning again</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;sanger&#39;</span><span class="p">,</span> <span class="n">seq_num</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">----------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;description 2&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq2&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 35</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 14.29%</span>
<span class="go">----------------------------------------</span>
<span class="go">0 TATGTATATA TAACATATAC ATATATACAT ACATA</span>
</pre></div>
</div>
<p>To write our <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> to a FASTQ file with quality scores encoded using
the <code class="docutils literal notranslate"><span class="pre">illumina1.3</span></code> variant:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_fh</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fastq&#39;</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s1">&#39;illumina1.3&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">@seq1 description 1</span>
<span class="go">AACACCAAACTTCTCCACCACGTGAGCTACAAAAG</span>
<span class="go">+</span>
<span class="go">FFFFx}s|F|b}b`ab`bbF}ka}bbxs{s{x{ve</span>
<span class="go">@seq2 description 2</span>
<span class="go">TATGTATATATAACATATACATATATACATACATA</span>
<span class="go">+</span>
<span class="go">|jyzox|~zxx}FFF`b}{{FasFFbF{`sFFaaa</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">new_fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Note that the file has been written in normalized format: sequence and quality
scores each only occur on a single line and the sequence header line is
not repeated in the quality header line. Note also that the quality scores are
different because they have been encoded using a different variant.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div role="list" class="citation-list">
<div class="citation" id="r951e5a4cbfda-1" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id1">1</a>,<a role="doc-backlink" href="#id3">2</a>)</span>
<p>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L. Heuer,
and Peter M. Rice. The Sanger FASTQ file format for sequences with quality
scores, and the Solexa/Illumina FASTQ variants. Nucl. Acids Res. (2010) 38
(6): 1767-1771. first published online December 16, 2009.
doi:10.1093/nar/gkp1137
http://nar.oxfordjournals.org/content/38/6/1767</p>
</div>
<div class="citation" id="r951e5a4cbfda-2" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://en.wikipedia.org/wiki/FASTQ_format">http://en.wikipedia.org/wiki/FASTQ_format</a></p>
</div>
<div class="citation" id="r951e5a4cbfda-3" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">3</a><span class="fn-bracket">]</span></span>
<p><a class="reference external" href="http://www.open-bio.org/">http://www.open-bio.org/</a></p>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="skbio.io.format.fasta.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">FASTA/QUAL format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.fasta</span></code>)</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.io.format.genbank.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">GenBank format (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.io.format.genbank</span></code>)</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-support">Format Support</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-specification">Format Specification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-score-variants">Quality Score Variants</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#format-parameters">Format Parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#examples">Examples</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.4.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>