
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="Storing and working with tree structures, including phylogenetic trees and hierarchies." name="description" />

    <title>Trees and Phylogenetics (skbio.tree) &#8212; scikit-bio 0.7.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=fe7df9b0"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tree';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = '0.7.0';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../tree.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.tree.TreeNode" href="generated/skbio.tree.TreeNode.html" />
    <link rel="prev" title="skbio.alignment.local_pairwise_align" href="generated/skbio.alignment.local_pairwise_align.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.7.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.7.0 documentation"/>
    <img src="_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.7.0 documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.TreeNode.html">skbio.tree.TreeNode</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.upgma.html">skbio.tree.upgma</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.nj.html">skbio.tree.nj</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.gme.html">skbio.tree.gme</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.bme.html">skbio.tree.bme</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.nni.html">skbio.tree.nni</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.rf_dists.html">skbio.tree.rf_dists</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.wrf_dists.html">skbio.tree.wrf_dists</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.path_dists.html">skbio.tree.path_dists</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.tree.majority_rule.html">skbio.tree.majority_rule</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Trees and Phylogenetics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code>)</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="trees-and-phylogenetics-skbio-tree">
<span id="module-skbio.tree"></span><h1>Trees and Phylogenetics (<a class="reference internal" href="#module-skbio.tree" title="skbio.tree"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.tree</span></code></a>)<a class="headerlink" href="#trees-and-phylogenetics-skbio-tree" title="Link to this heading">#</a></h1>
<p>This module provides functionality for working with trees, including phylogenetic trees
and various types of hierarchies with relevance in biology or beyond. It supports trees
that are multifurcating and nodes that have single descendants. Functionality is
provided for constructing trees, for traversing in multiple ways, comparisons, fetching
subtrees, and more.</p>
<p>See the <a class="reference internal" href="#tree-tutorial"><strong>Tutorial</strong></a> section for working with trees using scikit-bio.</p>
<section id="tree-structure-and-operations">
<h2>Tree structure and operations<a class="headerlink" href="#tree-structure-and-operations" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> is the data structure for tree representation. A tree consists of an
arbitrary number of interconnected <code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> objects representing individual nodes.
Each node has pointers to its parent and children. The class provides a large number of
methods to achieve various tasks for tree analysis and manipulation.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.TreeNode.html#skbio.tree.TreeNode" title="skbio.tree.TreeNode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeNode</span></code></a>([name, length, support, parent, ...])</p></td>
<td><p>Represent a node within a tree.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tree-construction">
<h2>Tree Construction<a class="headerlink" href="#tree-construction" title="Link to this heading">#</a></h2>
<p>Algorithms for phylogenetic reconstruction based on distance matrices.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.upgma.html#skbio.tree.upgma" title="skbio.tree.upgma"><code class="xref py py-obj docutils literal notranslate"><span class="pre">upgma</span></code></a>(dm[, weighted])</p></td>
<td><p>Perform unweighted pair group method with arithmetic mean (UPGMA) or its weighted variant (WPGMA) for phylogenetic reconstruction.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.tree.nj.html#skbio.tree.nj" title="skbio.tree.nj"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nj</span></code></a>(dm[, neg_as_zero, result_constructor, ...])</p></td>
<td><p>Perform neighbor joining (NJ) for phylogenetic reconstruction.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.gme.html#skbio.tree.gme" title="skbio.tree.gme"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gme</span></code></a>(dm[, neg_as_zero])</p></td>
<td><p>Perform greedy minimum evolution (GME) for phylogenetic reconstruction.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.tree.bme.html#skbio.tree.bme" title="skbio.tree.bme"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bme</span></code></a>(dm[, neg_as_zero])</p></td>
<td><p>Perform balanced minimum evolution (BME) for phylogenetic reconstruction.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tree-rearrangement">
<h2>Tree Rearrangement<a class="headerlink" href="#tree-rearrangement" title="Link to this heading">#</a></h2>
<p>Algorithms for improving an existing phylogenetic tree via rearrangement.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.nni.html#skbio.tree.nni" title="skbio.tree.nni"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nni</span></code></a>(tree, dm[, balanced, neg_as_zero])</p></td>
<td><p>Perform nearest neighbor interchange (NNI) to improve a phylogenetic tree.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tree-comparison">
<h2>Tree Comparison<a class="headerlink" href="#tree-comparison" title="Link to this heading">#</a></h2>
<p>Metrics for assessing the topological or length dissimilarity among trees.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.rf_dists.html#skbio.tree.rf_dists" title="skbio.tree.rf_dists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rf_dists</span></code></a>(trees[, ids, shared_by_all, ...])</p></td>
<td><p>Calculate Robinson-Foulds (RF) distances among trees.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.tree.wrf_dists.html#skbio.tree.wrf_dists" title="skbio.tree.wrf_dists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">wrf_dists</span></code></a>(trees[, ids, shared_by_all, ...])</p></td>
<td><p>Calculate weighted Robinson-Foulds (wRF) distances or variants among trees.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.path_dists.html#skbio.tree.path_dists" title="skbio.tree.path_dists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">path_dists</span></code></a>(trees[, ids, shared_by_all, ...])</p></td>
<td><p>Calculate path-length distances or variants among trees.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tree-utilities">
<h2>Tree utilities<a class="headerlink" href="#tree-utilities" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.tree.majority_rule.html#skbio.tree.majority_rule" title="skbio.tree.majority_rule"><code class="xref py py-obj docutils literal notranslate"><span class="pre">majority_rule</span></code></a>(trees[, weights, cutoff, ...])</p></td>
<td><p>Determine consensus trees from a list of rooted trees.</p></td>
</tr>
</tbody>
</table>
</div>
<section id="exceptions">
<h3>Exceptions<a class="headerlink" href="#exceptions" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeError</span></code></p></td>
<td><p>General tree error.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">NoLengthError</span></code></p></td>
<td><p>Missing length when expected.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">DuplicateNodeError</span></code></p></td>
<td><p>Duplicate nodes with identical names.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">MissingNodeError</span></code></p></td>
<td><p>Expecting a node.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">NoParentError</span></code></p></td>
<td><p>Missing a parent.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="tutorial">
<span id="tree-tutorial"></span><h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<p>In scikit-bio, the <a class="reference internal" href="generated/skbio.tree.TreeNode.html#skbio.tree.TreeNode" title="skbio.tree.TreeNode"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeNode</span></code></a> class is the main and only data structure for tree
represention. It is simple, flexible and powerful.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">TreeNode</span>
</pre></div>
</div>
<section id="loading-a-tree">
<h3>Loading a tree<a class="headerlink" href="#loading-a-tree" title="Link to this heading">#</a></h3>
<p>A tree can be constructed from a Newick string. <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick</a> is a
common file format used to represent tree structure based on nesting with parentheses.
For instance, the following string describes a 3-taxon tree, with one internal node:</p>
<blockquote>
<div><p>((A,B)C,D)root;</p>
</div></blockquote>
<p>Where A, B, and D are tips of the tree, and C is an internal node that unites tips A
and B. C and D are in turn united by the root of the tree.</p>
<p>If this string is stored in a file <code class="docutils literal notranslate"><span class="pre">input.nwk</span></code>, we can read it into scikit-bio with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;input.nwk&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, we can directly convert a Newick string into a tree with:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((A,B)C,D)root;&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span>
<span class="go">&lt;TreeNode, name: root, internal node count: 1, tips count: 3&gt;</span>
</pre></div>
</div>
<p>Let’s display an ASCII representation of the tree:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                    /-A</span>
<span class="go">          /C-------|</span>
<span class="go">-root----|          \-B</span>
<span class="go">         |</span>
<span class="go">          \-D</span>
</pre></div>
</div>
<p>Optionally, we can attach branch lengths to the nodes in a Newick string. In
phylogenetics, the branch length usually represents the amount of evolution
accumulated from one node to another. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((A:2,B:3)C:1,D:4)root;&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>To save a tree to a Newick file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;output.nwk&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="navigating-in-a-tree">
<h3>Navigating in a tree<a class="headerlink" href="#navigating-in-a-tree" title="Link to this heading">#</a></h3>
<p>The variable <code class="docutils literal notranslate"><span class="pre">tree</span></code> is a <a class="reference internal" href="generated/skbio.tree.TreeNode.html#skbio.tree.TreeNode" title="skbio.tree.TreeNode"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeNode</span></code></a> object. It is merely the <strong>root</strong> node of
the tree, but one can refer to the entire tree with just the root.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;root&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">is_root</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Starting from the root node one can navigate to its child nodes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>The first child of root has the name “C”. It is an <strong>internal node</strong> of the tree.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">is_tip</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
<p>The branch length can be accessed via attribute <code class="docutils literal notranslate"><span class="pre">length</span></code>. Note, that it is the length
of the branch connecting the current node and its parent.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">length</span>
<span class="go">1.0</span>
</pre></div>
</div>
<p>scikit-bio’s <code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> is highly flexible that we can refer to a subtree (a.k.a., a
clade) within a larger tree using its root node (here “C”). There is no need to detach
the subtree and we can just work with it in-place. Most tree analyses implemented in
scikit-bio also apply to subtrees.</p>
<p>The second child “D” is a <strong>tip</strong> (a.k.a., leaf). In a phylogenetic tree, a tip often
represents a <strong>taxon</strong>, i.e. an organism or a group of organisms being studied.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;D&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">is_tip</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Each node has exactly one parent node, with the exception of the root node, whose
parent is None.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;root&#39;</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">children</span></code> attributes connect multiple nodes into a tree
structure. Additionally, one can use methods like <a class="reference internal" href="generated/skbio.tree.TreeNode.siblings.html#skbio.tree.TreeNode.siblings" title="skbio.tree.TreeNode.siblings"><code class="xref py py-meth docutils literal notranslate"><span class="pre">siblings</span></code></a>,
<a class="reference internal" href="generated/skbio.tree.TreeNode.neighbors.html#skbio.tree.TreeNode.neighbors" title="skbio.tree.TreeNode.neighbors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">neighbors</span></code></a>, <a class="reference internal" href="generated/skbio.tree.TreeNode.ancestors.html#skbio.tree.TreeNode.ancestors" title="skbio.tree.TreeNode.ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ancestors</span></code></a> and <a class="reference internal" href="generated/skbio.tree.TreeNode.root.html#skbio.tree.TreeNode.root" title="skbio.tree.TreeNode.root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">root</span></code></a>
to navigate to node with certain relationships to the current node. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">siblings</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
<p>One can quickly find a node by name using <a class="reference internal" href="generated/skbio.tree.TreeNode.find.html#skbio.tree.TreeNode.find" title="skbio.tree.TreeNode.find"><code class="xref py py-meth docutils literal notranslate"><span class="pre">find</span></code></a>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">node</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;A&#39;</span>
</pre></div>
</div>
<p>If the tree is giant, the first call of <code class="docutils literal notranslate"><span class="pre">find</span></code> may take a while, but all subsequent
calls will complete in a flash. That’s because <code class="docutils literal notranslate"><span class="pre">find</span></code> will cache and reuse a lookup
table under the hood.</p>
</section>
<section id="traversing-a-tree">
<h3>Traversing a tree<a class="headerlink" href="#traversing-a-tree" title="Link to this heading">#</a></h3>
<p>The process of visiting each and every node within a tree is called <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Tree_traversal">tree
traversal</a>. There are a few common ways to traverse a tree, and
depending on your use, some methods are more appropriate than others. The wiki page
linked above will go into further depth on this topic. Here, we’re only going to cover
two of the commonly used traversal methods: <a class="reference internal" href="generated/skbio.tree.TreeNode.preorder.html#skbio.tree.TreeNode.preorder" title="skbio.tree.TreeNode.preorder"><code class="xref py py-meth docutils literal notranslate"><span class="pre">preorder</span></code></a> and
<a class="reference internal" href="generated/skbio.tree.TreeNode.postorder.html#skbio.tree.TreeNode.postorder" title="skbio.tree.TreeNode.postorder"><code class="xref py py-meth docutils literal notranslate"><span class="pre">postorder</span></code></a>.</p>
<p>A preorder traversal moves from root to tips, looking at the left most child first.
For instance:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">preorder</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="go">root</span>
<span class="go">C</span>
<span class="go">A</span>
<span class="go">B</span>
<span class="go">D</span>
</pre></div>
</div>
<p>A postorder traversal moves from tips to root,  accessing the left subtree tips first
before walking back up the tree:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">postorder</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="go">A</span>
<span class="go">B</span>
<span class="go">C</span>
<span class="go">D</span>
<span class="go">root</span>
</pre></div>
</div>
<p>If you are not certain (or don’t mind) what method should be used to traverse the tree,
simply use <a class="reference internal" href="generated/skbio.tree.TreeNode.traverse.html#skbio.tree.TreeNode.traverse" title="skbio.tree.TreeNode.traverse"><code class="xref py py-meth docutils literal notranslate"><span class="pre">traverse</span></code></a>.</p>
<p>Note, by default, those traversal methods will include self, i.e., the root node from
which traversal starts. One can pass the <code class="docutils literal notranslate"><span class="pre">include_self=False</span></code> flag to the method call
if you wish to exclude it.</p>
<p><code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> provides multiple methods for iterating over certain nodes. For examples,
<a class="reference internal" href="generated/skbio.tree.TreeNode.tips.html#skbio.tree.TreeNode.tips" title="skbio.tree.TreeNode.tips"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tips</span></code></a> and <a class="reference internal" href="generated/skbio.tree.TreeNode.non_tips.html#skbio.tree.TreeNode.non_tips" title="skbio.tree.TreeNode.non_tips"><code class="xref py py-meth docutils literal notranslate"><span class="pre">non_tips</span></code></a> iterate over tips and internal
nodes, respectively:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">tips</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="go">A</span>
<span class="go">B</span>
<span class="go">D</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">non_tips</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="go">C</span>
</pre></div>
</div>
<p>If we just want to know the total number of nodes (or just tips) in a tree, we can do:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
<span class="go">5</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">tips</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
<p>Instead, if we want to obtain a set of taxa (tip names) from a tree, we can use the
<a class="reference internal" href="generated/skbio.tree.TreeNode.subset.html#skbio.tree.TreeNode.subset" title="skbio.tree.TreeNode.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subset</span></code></a> method (it is called the “subset” because one can call it at
any node and get a subset of taxa descending from it).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">taxa</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">subset</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
<span class="go">[&#39;A&#39;, &#39;B&#39;, &#39;D&#39;]</span>
</pre></div>
</div>
</section>
<section id="editing-a-tree">
<h3>Editing a tree<a class="headerlink" href="#editing-a-tree" title="Link to this heading">#</a></h3>
<p>As discussed above, a tree consists of multiple <code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> objects connected to each
other. To manipulate a tree, one just needs to change the connections between nodes.
<code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> provides several standard methods that let you work with nodes like Python
lists.</p>
<p>Create a new tip “E” and attach it to the internal node “C” as a child of it:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="p">(</span><span class="s2">&quot;E&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                    /-A</span>
<span class="go">                   |</span>
<span class="go">          /C-------|--B</span>
<span class="go">         |         |</span>
<span class="go">-root----|          \-E</span>
<span class="go">         |</span>
<span class="go">          \-D</span>
</pre></div>
</div>
<p>Create two tips “F” and “G” and attach both of them to tip “D”. Now “D” becomes an
internal node because it has children:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">D</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;D&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">TreeNode</span><span class="p">(</span><span class="s2">&quot;F&quot;</span><span class="p">),</span> <span class="n">TreeNode</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                    /-A</span>
<span class="go">                   |</span>
<span class="go">          /C-------|--B</span>
<span class="go">         |         |</span>
<span class="go">-root----|          \-E</span>
<span class="go">         |</span>
<span class="go">         |          /-F</span>
<span class="go">          \D-------|</span>
<span class="go">                    \-G</span>
</pre></div>
</div>
<p>Prune clade “D” and graft it to node “B”. Once a node is attached to a new parent, the
connection to its old parent is automatically removed.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">B</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;B&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                    /-A</span>
<span class="go">                   |</span>
<span class="go">                   |                    /-F</span>
<span class="go">-root---- /C-------|-B------- /D-------|</span>
<span class="go">                   |                    \-G</span>
<span class="go">                   |</span>
<span class="go">                    \-E</span>
</pre></div>
</div>
<p>Remove clade “D” from node “B”. This operation detaches the two nodes, but the removed
node still exists in the memory and once can still use it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">B</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                    /-A</span>
<span class="go">                   |</span>
<span class="go">-root---- /C-------|--B</span>
<span class="go">                   |</span>
<span class="go">                    \-E</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">D</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">          /-F</span>
<span class="go">-D-------|</span>
<span class="go">          \-G</span>
</pre></div>
</div>
<p>These methods automatically clear lookup tables and other caches because they may be
obsolete for the modified tree. This safety measure could impact performance when
working with very large trees. If caches are not present or relevant, we can add
<code class="docutils literal notranslate"><span class="pre">uncache=False</span></code> to these methods to skip this check. See <a class="reference internal" href="generated/skbio.tree.TreeNode.has_caches.html#skbio.tree.TreeNode.has_caches" title="skbio.tree.TreeNode.has_caches"><code class="xref py py-meth docutils literal notranslate"><span class="pre">has_caches</span></code></a>
for details.</p>
<p><code class="docutils literal notranslate"><span class="pre">TreeNode</span></code> provides several advanced methods for editing a tree as a batch. For
examples, <a class="reference internal" href="generated/skbio.tree.TreeNode.bifurcate.html#skbio.tree.TreeNode.bifurcate" title="skbio.tree.TreeNode.bifurcate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">bifurcate</span></code></a> converts an arbitrary tree into a bifurcating
tree (binary tree), <a class="reference internal" href="generated/skbio.tree.TreeNode.shear.html#skbio.tree.TreeNode.shear" title="skbio.tree.TreeNode.shear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">shear</span></code></a> refines a tree to a given set of taxa,
<a class="reference internal" href="generated/skbio.tree.TreeNode.unpack_by_func.html#skbio.tree.TreeNode.unpack_by_func" title="skbio.tree.TreeNode.unpack_by_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unpack_by_func</span></code></a> unpacks (contracts) branches that suffice certain
criteria (e.g., bootstrap support below a threshold).</p>
</section>
<section id="building-a-tree">
<h3>Building a tree<a class="headerlink" href="#building-a-tree" title="Link to this heading">#</a></h3>
<p>scikit-bio provides several common algorithms for building a tree structure based on
a distance matrix (<a class="reference internal" href="generated/skbio.stats.distance.DistanceMatrix.html#skbio.stats.distance.DistanceMatrix" title="skbio.stats.distance.DistanceMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">DistanceMatrix</span></code></a>) that stores pairwise
distances between taxa.</p>
<p>In the field of <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Phylogenetics">phylogenetics</a>, the evolutionary relationships
among organisms are inferred using genetic data.
<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Distance_matrices_in_phylogeny">Distance-based methods</a> are a common category
of methods for such purpose. The distances are usually computed from sequence alignment
(see <a class="reference internal" href="alignment.html#module-skbio.alignment" title="skbio.alignment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code></a>), but they can also be inferred using alternative methods
such as <em>k</em>-mer frequency (see <a class="reference internal" href="generated/skbio.sequence.Sequence.kmer_frequencies.html#skbio.sequence.Sequence.kmer_frequencies" title="skbio.sequence.Sequence.kmer_frequencies"><code class="xref py py-func docutils literal notranslate"><span class="pre">kmer_frequencies</span></code></a>) and
community diversity (see <a class="reference internal" href="generated/skbio.diversity.beta.html#module-skbio.diversity.beta" title="skbio.diversity.beta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity.beta</span></code></a>), and the applications of
distance-based tree building are not limited to evolutionary biology, but can be
extended to other fields (e.g., <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Hierarchical_clustering">hierarchical clustering</a> in data science).</p>
<p>To begin with, let’s create a distance matrix of seven taxa.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">DistanceMatrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dm</span> <span class="o">=</span> <span class="n">DistanceMatrix</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.005</span><span class="p">,</span> <span class="mf">0.033</span><span class="p">,</span> <span class="mf">0.387</span><span class="p">,</span> <span class="mf">0.626</span><span class="p">,</span> <span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.665</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.005</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.028</span><span class="p">,</span> <span class="mf">0.387</span><span class="p">,</span> <span class="mf">0.626</span><span class="p">,</span> <span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.665</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.033</span><span class="p">,</span> <span class="mf">0.028</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.396</span><span class="p">,</span> <span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.64</span> <span class="p">,</span> <span class="mf">0.679</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.387</span><span class="p">,</span> <span class="mf">0.387</span><span class="p">,</span> <span class="mf">0.396</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.611</span><span class="p">,</span> <span class="mf">0.611</span><span class="p">,</span> <span class="mf">0.66</span> <span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.626</span><span class="p">,</span> <span class="mf">0.626</span><span class="p">,</span> <span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.611</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.147</span><span class="p">,</span> <span class="mf">0.758</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.635</span><span class="p">,</span> <span class="mf">0.64</span> <span class="p">,</span> <span class="mf">0.611</span><span class="p">,</span> <span class="mf">0.147</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">,</span> <span class="mf">0.754</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">0.665</span><span class="p">,</span> <span class="mf">0.665</span><span class="p">,</span> <span class="mf">0.679</span><span class="p">,</span> <span class="mf">0.66</span> <span class="p">,</span> <span class="mf">0.758</span><span class="p">,</span> <span class="mf">0.754</span><span class="p">,</span> <span class="mf">0.</span>   <span class="p">],</span>
<span class="gp">... </span><span class="p">],</span> <span class="n">ids</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>    <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;chimp&#39;</span><span class="p">,</span> <span class="s1">&#39;monkey&#39;</span><span class="p">,</span> <span class="s1">&#39;pig&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span><span class="p">,</span> <span class="s1">&#39;rat&#39;</span><span class="p">,</span> <span class="s1">&#39;chicken&#39;</span>
<span class="gp">... </span><span class="p">])</span>
</pre></div>
</div>
<p>The following code applies <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Neighbor_joining">neighbor joining</a> (<a class="reference internal" href="generated/skbio.tree.nj.html#skbio.tree.nj" title="skbio.tree.nj"><code class="xref py py-func docutils literal notranslate"><span class="pre">nj</span></code></a>), a
classic distance-based tree-building algorithm:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.tree</span><span class="w"> </span><span class="kn">import</span> <span class="n">nj</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
</pre></div>
</div>
<p>Display the reconstructed phylogenetic tree:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">          /-human</span>
<span class="go">         |</span>
<span class="go">         |--chimp</span>
<span class="go">---------|</span>
<span class="go">         |          /-monkey</span>
<span class="go">         |         |</span>
<span class="go">          \--------|          /-pig</span>
<span class="go">                   |         |</span>
<span class="go">                    \--------|                    /-mouse</span>
<span class="go">                             |          /--------|</span>
<span class="go">                              \--------|          \-rat</span>
<span class="go">                                       |</span>
<span class="go">                                        \-chicken</span>
</pre></div>
</div>
<p>Other methods include <a class="reference internal" href="generated/skbio.tree.upgma.html#skbio.tree.upgma" title="skbio.tree.upgma"><code class="xref py py-func docutils literal notranslate"><span class="pre">upgma</span></code></a> (simplest), <a class="reference internal" href="generated/skbio.tree.gme.html#skbio.tree.gme" title="skbio.tree.gme"><code class="xref py py-func docutils literal notranslate"><span class="pre">gme</span></code></a> (highly scalable) and
<a class="reference internal" href="generated/skbio.tree.bme.html#skbio.tree.bme" title="skbio.tree.bme"><code class="xref py py-func docutils literal notranslate"><span class="pre">bme</span></code></a>. All of them can be applied in the same way.</p>
<p>These four methods create a tree from scratch through a fixed set of procedures.
Additionally, starting from an existing tree, <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Tree_rearrangement">tree rearrangement</a> methods will explore the tree space further in order to
improve it. Nearest neighbor interchange (<a class="reference internal" href="generated/skbio.tree.nni.html#skbio.tree.nni" title="skbio.tree.nni"><code class="xref py py-func docutils literal notranslate"><span class="pre">nni</span></code></a>) is one such method.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.tree</span><span class="w"> </span><span class="kn">import</span> <span class="n">nni</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">nni</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">dm</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="rooting-a-tree">
<h3>Rooting a tree<a class="headerlink" href="#rooting-a-tree" title="Link to this heading">#</a></h3>
<p>The phylogenetic tree created above seems to properly reflect the relatedness among the
seven organisms. However, the “root” of the tree doesn’t appear to be aligned with our
knowledge of evolution. This is because the tree is in fact <strong>unrooted</strong>. That is, the
placement of branches does not inform the direction of evolution.</p>
<p>In scikit-bio, the distinguishment between rooted and unrooted trees is implicit: Every
tree has a root node. A tree is considered as “rooted” if its root node has exactly two
children. In contrast, an “unrooted” tree may have three (the most common case), one,
or more than three children attached to its root node.</p>
<p>There are several approaches to root an unrooted tree to reflect the correct direction
of evolution. Midpoint rooting (<a class="reference internal" href="generated/skbio.tree.TreeNode.root_at_midpoint.html#skbio.tree.TreeNode.root_at_midpoint" title="skbio.tree.TreeNode.root_at_midpoint"><code class="xref py py-meth docutils literal notranslate"><span class="pre">root_at_midpoint</span></code></a>) is a simple method
that places the root at the midpoint between the two most distant tips in the tree.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rooted</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root_at_midpoint</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rooted</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">          /-chicken</span>
<span class="go">         |</span>
<span class="go">---------|                    /-mouse</span>
<span class="go">         |          /--------|</span>
<span class="go">         |         |          \-rat</span>
<span class="go">          \--------|</span>
<span class="go">                   |          /-pig</span>
<span class="go">                    \--------|</span>
<span class="go">                             |          /-monkey</span>
<span class="go">                              \--------|</span>
<span class="go">                                       |          /-chimp</span>
<span class="go">                                        \--------|</span>
<span class="go">                                                  \-human</span>
</pre></div>
</div>
<p>If we already know which taxon or taxa (called an “outgroup”) were separated from all
other taxa in evolution, we can perform outgroup rooting
(<a class="reference internal" href="generated/skbio.tree.TreeNode.root_by_outgroup.html#skbio.tree.TreeNode.root_by_outgroup" title="skbio.tree.TreeNode.root_by_outgroup"><code class="xref py py-meth docutils literal notranslate"><span class="pre">root_by_outgroup</span></code></a>):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rooted</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root_by_outgroup</span><span class="p">([</span><span class="s2">&quot;chicken&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Or we can directly root a tree at a given taxon, node or branch:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rooted</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root_at</span><span class="p">(</span><span class="s2">&quot;chicken&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To convert a rooted tree into unrooted, we can apply the <a class="reference internal" href="generated/skbio.tree.TreeNode.unroot.html#skbio.tree.TreeNode.unroot" title="skbio.tree.TreeNode.unroot"><code class="xref py py-meth docutils literal notranslate"><span class="pre">unroot</span></code></a>
method.</p>
</section>
<section id="analyzing-topology">
<h3>Analyzing topology<a class="headerlink" href="#analyzing-topology" title="Link to this heading">#</a></h3>
<p>The topology of a tree describes the grouping pattern among taxa. For example, in the
rooted tree displayed above, human, chimp and monkey form a clade (a monophyletic
group), while mouse and rat form another clade. A clade can be referred to by its root
node, which is the lowest common ancestor (<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Lowest_common_ancestor">LCA</a>) of all
descending taxa:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">primate</span> <span class="o">=</span> <span class="n">rooted</span><span class="o">.</span><span class="n">lca</span><span class="p">([</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="s2">&quot;chimp&quot;</span><span class="p">,</span> <span class="s2">&quot;monkey&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rodent</span> <span class="o">=</span> <span class="n">rooted</span><span class="o">.</span><span class="n">lca</span><span class="p">([</span><span class="s2">&quot;mouse&quot;</span><span class="p">,</span> <span class="s2">&quot;rat&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Note that one doesn’t need to list all descending taxa in order to find the LCA. Two
or more taxa that join at the LCA node on their paths to the root are sufficient for
locating the LCA.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">primate</span> <span class="o">=</span> <span class="n">rooted</span><span class="o">.</span><span class="n">lca</span><span class="p">([</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="s2">&quot;monkey&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>A clade regardless of its internal structure can be simply defined by the
<a class="reference internal" href="generated/skbio.tree.TreeNode.subset.html#skbio.tree.TreeNode.subset" title="skbio.tree.TreeNode.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subset</span></code></a> of taxa descending from it (see above). All
<a class="reference internal" href="generated/skbio.tree.TreeNode.subsets.html#skbio.tree.TreeNode.subsets" title="skbio.tree.TreeNode.subsets"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subsets</span></code></a> within a rooted tree describes all possible taxon groups.</p>
<p>Assume we have another rooted tree <code class="docutils literal notranslate"><span class="pre">tree2</span></code> of the same taxa. To assess whether
that tree also support human, chimp and monkey grouped together, we can simply do:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">primate</span><span class="o">.</span><span class="n">subset</span><span class="p">()</span> <span class="ow">in</span> <span class="n">tree2</span><span class="o">.</span><span class="n">subsets</span><span class="p">()</span>
</pre></div>
</div>
<p>When working with unrooted trees, we should replace the notion of subset with
bipartition (<a class="reference internal" href="generated/skbio.tree.TreeNode.bipart.html#skbio.tree.TreeNode.bipart" title="skbio.tree.TreeNode.bipart"><code class="xref py py-meth docutils literal notranslate"><span class="pre">bipart</span></code></a>), which defines the separation of all taxa into
two parts by a given branch, regardless of the direction of evolution. For example,
both the rooted and unrooted trees support the separation of primates from other taxa:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">primate</span><span class="o">.</span><span class="n">bipart</span><span class="p">()</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">biparts</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The topology of two trees with shared taxa can be compared using the Robinson-Foulds
(RF) distance, which is the number of bipartitions (unrooted trees) or subsets
(rooted trees) that are distinct between the two trees. For example, let’s define
three trees, the first two of which have the same topology despite different orders
of clades.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((A,B),(C,D),(E,F));&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree2</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((F,E),(C,D),(B,A));&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree3</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((C,B),(A,D),(E,F));&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Calculate RF distances using (<a class="reference internal" href="generated/skbio.tree.TreeNode.compare_rfd.html#skbio.tree.TreeNode.compare_rfd" title="skbio.tree.TreeNode.compare_rfd"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compare_rfd</span></code></a>):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span><span class="o">.</span><span class="n">compare_rfd</span><span class="p">(</span><span class="n">tree1</span><span class="p">)</span>  <span class="c1"># identity case</span>
<span class="go">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span><span class="o">.</span><span class="n">compare_rfd</span><span class="p">(</span><span class="n">tree2</span><span class="p">)</span>  <span class="c1"># same tree but different clade order</span>
<span class="go">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span><span class="o">.</span><span class="n">compare_rfd</span><span class="p">(</span><span class="n">tree3</span><span class="p">)</span>  <span class="c1"># only 1 of 3 common clade</span>
<span class="go">4.0</span>
</pre></div>
</div>
<p>We can calculate pairwise RF distances between any number of trees using a single call
of <a class="reference internal" href="generated/skbio.tree.rf_dists.html#skbio.tree.rf_dists" title="skbio.tree.rf_dists"><code class="xref py py-func docutils literal notranslate"><span class="pre">rf_dists</span></code></a>, which returns a distance matrix:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.tree</span><span class="w"> </span><span class="kn">import</span> <span class="n">rf_dists</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dm</span> <span class="o">=</span> <span class="n">rf_dists</span><span class="p">([</span><span class="n">tree1</span><span class="p">,</span> <span class="n">tree2</span><span class="p">,</span> <span class="n">tree3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
<span class="go">3x3 distance matrix</span>
<span class="go">IDs:</span>
<span class="go">&#39;0&#39;, &#39;1&#39;, &#39;2&#39;</span>
<span class="go">Data:</span>
<span class="go">[[ 0.  0.  4.]</span>
<span class="go"> [ 0.  0.  4.]</span>
<span class="go"> [ 4.  4.  0.]]</span>
</pre></div>
</div>
</section>
<section id="analyzing-distances">
<h3>Analyzing distances<a class="headerlink" href="#analyzing-distances" title="Link to this heading">#</a></h3>
<p>With branch lengths annotated, a phylogenetic tree describes the evolutionary distances
among taxa. This important property is the basis for multiple downstream analyses, such
as community diversity (see <a class="reference internal" href="diversity.html#module-skbio.diversity" title="skbio.diversity"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code></a>).</p>
<p>The <a class="reference internal" href="generated/skbio.tree.TreeNode.distance.html#skbio.tree.TreeNode.distance" title="skbio.tree.TreeNode.distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance</span></code></a> between two nodes in a tree (a.k.a., patristic distance)
is the total length of branches constituting the path connecting them. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">human</span><span class="p">,</span> <span class="n">mouse</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;human&quot;</span><span class="p">),</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;mouse&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">human</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">mouse</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">0.628</span>
</pre></div>
</div>
<p>The maximum distance between any pair of taxa (<a class="reference internal" href="generated/skbio.tree.TreeNode.maxdist.html#skbio.tree.TreeNode.maxdist" title="skbio.tree.TreeNode.maxdist"><code class="xref py py-meth docutils literal notranslate"><span class="pre">maxdist</span></code></a>) and the total
branch length connecting all taxa (<a class="reference internal" href="generated/skbio.tree.TreeNode.total_length.html#skbio.tree.TreeNode.total_length" title="skbio.tree.TreeNode.total_length"><code class="xref py py-meth docutils literal notranslate"><span class="pre">total_length</span></code></a>) can be used to
measure the biodiversity of groups of organisms.</p>
<p>The distances between all pairs of tips (taxa) in a tree can be efficiently calculated
using <a class="reference internal" href="generated/skbio.tree.TreeNode.cophenet.html#skbio.tree.TreeNode.cophenet" title="skbio.tree.TreeNode.cophenet"><code class="xref py py-meth docutils literal notranslate"><span class="pre">cophenet</span></code></a>, which returns a distance matrix.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dm</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">cophenet</span><span class="p">()</span>
</pre></div>
</div>
<p>The evolutionary distances reflected by two trees can be compared. First, let’s create
two trees with the same taxa but different branch lengths.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((A:0.1,B:0.3):0.3,C:0.4,D:0.5);&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree2</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="s2">&quot;((A:0.4,B:0.7):0.3,C:0.1,D:0.5);&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Calculate the distance between the cophenetic distance matrices using
(<a class="reference internal" href="generated/skbio.tree.TreeNode.compare_cophenet.html#skbio.tree.TreeNode.compare_cophenet" title="skbio.tree.TreeNode.compare_cophenet"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compare_cophenet</span></code></a>):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree1</span><span class="o">.</span><span class="n">compare_cophenet</span><span class="p">(</span><span class="n">tree2</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">0.376</span>
</pre></div>
</div>
<p>Likewise, we can calculate this metric between multiple trees using <a class="reference internal" href="generated/skbio.tree.path_dists.html#skbio.tree.path_dists" title="skbio.tree.path_dists"><code class="xref py py-func docutils literal notranslate"><span class="pre">path_dists</span></code></a>.</p>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.alignment.local_pairwise_align.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.alignment.local_pairwise_align</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.tree.TreeNode.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.tree.TreeNode</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-structure-and-operations">Tree structure and operations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-construction">Tree Construction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-rearrangement">Tree Rearrangement</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-comparison">Tree Comparison</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-utilities">Tree utilities</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exceptions">Exceptions</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-a-tree">Loading a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#navigating-in-a-tree">Navigating in a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#traversing-a-tree">Traversing a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#editing-a-tree">Editing a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#building-a-tree">Building a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rooting-a-tree">Rooting a tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analyzing-topology">Analyzing topology</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analyzing-distances">Analyzing distances</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>