
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="description" content="Store generic sequence data and optional associated metadata." />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>skbio.sequence.Sequence &#8212; scikit-bio 0.7.2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=6dfdf7c3"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/skbio.sequence.Sequence';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = '0.7.2';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../../generated/skbio.sequence.Sequence.html" />
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="skbio.sequence.Sequence.concat" href="skbio.sequence.Sequence.concat.html" />
    <link rel="prev" title="Biological Sequences (skbio.sequence)" href="../sequence.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.7.2" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.7.2 documentation"/>
    <img src="../_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.7.2 documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">skbio.sequence.Sequence</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.concat.html">skbio.sequence.Sequence.concat</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.count.html">skbio.sequence.Sequence.count</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.distance.html">skbio.sequence.Sequence.distance</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.find_with_regex.html">skbio.sequence.Sequence.find_with_regex</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.frequencies.html">skbio.sequence.Sequence.frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.index.html">skbio.sequence.Sequence.index</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.iter_contiguous.html">skbio.sequence.Sequence.iter_contiguous</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.iter_kmers.html">skbio.sequence.Sequence.iter_kmers</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.kmer_frequencies.html">skbio.sequence.Sequence.kmer_frequencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.lowercase.html">skbio.sequence.Sequence.lowercase</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.match_frequency.html">skbio.sequence.Sequence.match_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.matches.html">skbio.sequence.Sequence.matches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.mismatch_frequency.html">skbio.sequence.Sequence.mismatch_frequency</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.mismatches.html">skbio.sequence.Sequence.mismatches</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.read.html">skbio.sequence.Sequence.read</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.replace.html">skbio.sequence.Sequence.replace</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.to_indices.html">skbio.sequence.Sequence.to_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="skbio.sequence.Sequence.write.html">skbio.sequence.Sequence.write</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GrammaredSequence.html">skbio.sequence.GrammaredSequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.DNA.html">skbio.sequence.DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.RNA.html">skbio.sequence.RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.Protein.html">skbio.sequence.Protein</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.GeneticCode.html">skbio.sequence.GeneticCode</a></li>
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.SubstitutionMatrix.html">skbio.sequence.SubstitutionMatrix</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.distance.html">Sequence distance metrics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.distance</span></code>)</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="skbio.sequence.NucleotideMixin.html">skbio.sequence.NucleotideMixin</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../sequence.html" class="nav-link">Biological Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">skbio.sequence.Sequence</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="skbio-sequence-sequence">
<h1>skbio.sequence.Sequence<a class="headerlink" href="#skbio-sequence-sequence" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="skbio.sequence.Sequence">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">skbio.sequence.</span></span><span class="sig-name descname"><span class="pre">Sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positional_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L35"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence" title="Link to this definition">#</a></dt>
<dd><p>Store generic sequence data and optional associated metadata.</p>
<p><code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects do not enforce an alphabet or grammar and are thus the
most generic objects for storing sequence data. <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects do not
necessarily represent biological sequences. For example, <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> can
be used to represent a position in a multiple sequence alignment.
Subclasses <code class="docutils literal notranslate"><span class="pre">DNA</span></code>, <code class="docutils literal notranslate"><span class="pre">RNA</span></code>, and <code class="docutils literal notranslate"><span class="pre">Protein</span></code> enforce the IUPAC character
set <a class="reference internal" href="#rde5f2cb97c52-1" id="id1">[1]</a> for, and provide operations specific to, each respective molecule
type.</p>
<p><code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects consist of the underlying sequence data, as well
as optional metadata and positional metadata. The underlying sequence
is immutable, while the metdata and positional metadata are mutable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, Sequence, or 1D np.ndarray (np.uint8 or ‘|S1’)</span></dt><dd><p>Characters representing the sequence itself.</p>
</dd>
<dt><strong>metadata</strong><span class="classifier">dict, optional</span></dt><dd><p>Arbitrary metadata which applies to the entire sequence. A shallow copy
of the <code class="docutils literal notranslate"><span class="pre">dict</span></code> will be made (see Examples section below for details).</p>
</dd>
<dt><strong>positional_metadata</strong><span class="classifier">pd.DataFrame consumable, optional</span></dt><dd><p>Arbitrary per-character metadata (e.g., sequence read quality
scores). Must be able to be passed directly to <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>
constructor. Each column of metadata must be the same length as
<cite>sequence</cite>. A shallow copy of the positional metadata will be made if
necessary (see Examples section below for details).</p>
</dd>
<dt><strong>interval_metadata</strong><span class="classifier">IntervalMetadata</span></dt><dd><p>Arbitrary metadata which applies to intervals within a sequence to
store interval features (such as genes, ncRNA on the sequence).</p>
</dd>
<dt><strong>lowercase</strong><span class="classifier">bool or str, optional</span></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, lowercase sequence characters will be converted to
uppercase characters. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, no characters will be converted.
If a str, it will be treated as a key into the positional metadata of
the object. All lowercase characters will be converted to uppercase,
and a <code class="docutils literal notranslate"><span class="pre">True</span></code> value will be stored in a boolean array in the
positional metadata under the key.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DNA</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA</span></code></a></dt><dd></dd>
<dt><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Protein</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">References</p>
<div role="list" class="citation-list">
<div class="citation" id="rde5f2cb97c52-1" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Cornish-Bowden, A. (1985). Nomenclature for incompletely specified bases in
nucleic acid sequences: recommendations 1984. Nucleic Acids Res, 13(9), 3021.</p>
</div>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.metadata</span><span class="w"> </span><span class="kn">import</span> <span class="n">IntervalMetadata</span>
</pre></div>
</div>
<p><strong>Creating sequences:</strong></p>
<p>Create a sequence without any metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUCGUGAAGGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Sequence</span>
<span class="go">---------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 12</span>
<span class="go">---------------</span>
<span class="go">0 GGUCGUGAAG GA</span>
</pre></div>
</div>
<p>Create a sequence with metadata, positional metadata and interval metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;authors&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Alice&#39;</span><span class="p">],</span> <span class="s1">&#39;desc&#39;</span><span class="p">:</span><span class="s1">&#39;seq desc&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;seq-id&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">positional_metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;exons&#39;</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">],</span>
<span class="gp">... </span>                       <span class="s1">&#39;quality&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval_metadata</span> <span class="o">=</span> <span class="n">IntervalMetadata</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;sagA&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span>
<span class="gp">... </span>               <span class="n">positional_metadata</span><span class="o">=</span><span class="n">positional_metadata</span><span class="p">,</span>
<span class="gp">... </span>               <span class="n">interval_metadata</span><span class="o">=</span><span class="n">interval_metadata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Sequence</span>
<span class="go">-----------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;authors&#39;: &lt;class &#39;list&#39;&gt;</span>
<span class="go">    &#39;desc&#39;: &#39;seq desc&#39;</span>
<span class="go">    &#39;id&#39;: &#39;seq-id&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;exons&#39;: &lt;dtype: bool&gt;</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: int64&gt;</span>
<span class="go">Interval metadata:</span>
<span class="go">    1 interval feature</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">-----------------------------</span>
<span class="go">0 ACGT</span>
</pre></div>
</div>
<p><strong>Retrieving underlying sequence data:</strong></p>
<p>Retrieve underlying sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">values</span>
<span class="go">array([b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;, b&#39;T&#39;],</span>
<span class="go">      dtype=&#39;|S1&#39;)</span>
</pre></div>
</div>
<p>Underlying sequence immutable:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="sa">b</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="sa">b</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="sa">b</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="sa">b</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;|S1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">values</span> <span class="o">=</span> <span class="n">values</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">AttributeError</span>: <span class="n">can&#39;t set attribute</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="sa">b</span><span class="s1">&#39;T&#39;</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">assignment destination is read-only</span>
</pre></div>
</div>
<p><strong>Retrieving sequence metadata:</strong></p>
<p>Retrieve metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;authors&#39;: [&#39;Alice&#39;], &#39;desc&#39;: &#39;seq desc&#39;, &#39;id&#39;: &#39;seq-id&#39;}</span>
</pre></div>
</div>
<p>Retrieve positional metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">   exons  quality</span>
<span class="go">0   True        3</span>
<span class="go">1   True        3</span>
<span class="go">2  False        4</span>
<span class="go">3   True       10</span>
</pre></div>
</div>
<p>Retrieve interval metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span>
<span class="go">1 interval feature</span>
<span class="go">------------------</span>
<span class="go">Interval(interval_metadata=&lt;...&gt;, bounds=[(1, 3)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;sagA&#39;})</span>
</pre></div>
</div>
<p><strong>Updating sequence metadata:</strong></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Be aware that a shallow copy of <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and
<code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code> is made for performance. Since a deep copy is
not made, changes made to mutable Python objects stored as metadata may
affect the metadata of other <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects or anything else that
shares a reference to the object. The following examples illustrate this
behavior.</p>
</div>
<p>First, let’s create a sequence and update its metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;seq-id&#39;</span><span class="p">,</span> <span class="s1">&#39;desc&#39;</span><span class="p">:</span> <span class="s1">&#39;seq desc&#39;</span><span class="p">,</span> <span class="s1">&#39;authors&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Alice&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;new-id&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;pubmed&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">12345</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;id&#39;: &#39;new-id&#39;, &#39;desc&#39;: &#39;seq desc&#39;, &#39;authors&#39;: [&#39;Alice&#39;], &#39;pubmed&#39;: 12345}</span>
</pre></div>
</div>
<p>Note that the original metadata dictionary (stored in variable
<code class="docutils literal notranslate"><span class="pre">metadata</span></code>) hasn’t changed because a shallow copy was made:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span>
<span class="go">{&#39;id&#39;: &#39;seq-id&#39;, &#39;desc&#39;: &#39;seq desc&#39;, &#39;authors&#39;: [&#39;Alice&#39;]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span> <span class="o">==</span> <span class="n">metadata</span>
<span class="go">False</span>
</pre></div>
</div>
<p>Note however that since only a <em>shallow</em> copy was made, updates to mutable
objects will also change the original metadata dictionary:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Bob&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span>
<span class="go">[&#39;Alice&#39;, &#39;Bob&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span>
<span class="go">[&#39;Alice&#39;, &#39;Bob&#39;]</span>
</pre></div>
</div>
<p>This behavior can also occur when manipulating a sequence that has been
derived from another sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span>
<span class="go">Sequence</span>
<span class="go">-----------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;authors&#39;: &lt;class &#39;list&#39;&gt;</span>
<span class="go">    &#39;desc&#39;: &#39;seq desc&#39;</span>
<span class="go">    &#39;id&#39;: &#39;new-id&#39;</span>
<span class="go">    &#39;pubmed&#39;: 12345</span>
<span class="go">Stats:</span>
<span class="go">    length: 2</span>
<span class="go">-----------------------------</span>
<span class="go">0 CG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;id&#39;: &#39;new-id&#39;, &#39;desc&#39;: &#39;seq desc&#39;, &#39;authors&#39;: [&#39;Alice&#39;, &#39;Bob&#39;], &#39;pubmed&#39;: 12345}</span>
</pre></div>
</div>
<p>The subsequence has inherited the metadata of its parent sequence. If we
update the subsequence’s author list, we see the changes propagated in the
parent sequence and original metadata dictionary:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Carol&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span>
<span class="go">[&#39;Alice&#39;, &#39;Bob&#39;, &#39;Carol&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span>
<span class="go">[&#39;Alice&#39;, &#39;Bob&#39;, &#39;Carol&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;authors&#39;</span><span class="p">]</span>
<span class="go">[&#39;Alice&#39;, &#39;Bob&#39;, &#39;Carol&#39;]</span>
</pre></div>
</div>
<p>The behavior for updating positional metadata is similar. Let’s create a
new sequence with positional metadata that is already stored in a
<code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">positional_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">{</span><span class="s1">&#39;list&#39;</span><span class="p">:</span> <span class="p">[[],</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]],</span> <span class="s1">&#39;quality&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">positional_metadata</span><span class="o">=</span><span class="n">positional_metadata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Sequence</span>
<span class="go">-----------------------------</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;list&#39;: &lt;dtype: object&gt;</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: int64&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">-----------------------------</span>
<span class="go">0 ACGT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">  list  quality</span>
<span class="go">0   []        3</span>
<span class="go">1   []        3</span>
<span class="go">2   []        4</span>
<span class="go">3   []       10</span>
</pre></div>
</div>
<p>Now let’s update the sequence’s positional metadata by adding a new column
and changing a value in another column:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="p">[</span><span class="s1">&#39;gaps&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;quality&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">999</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">  list  quality   gaps</span>
<span class="go">0   []      999  False</span>
<span class="go">1   []        3  False</span>
<span class="go">2   []        4  False</span>
<span class="go">3   []       10  False</span>
</pre></div>
</div>
<p>Note that the original positional metadata (stored in variable
<code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code>) hasn’t changed because a shallow copy was made:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">positional_metadata</span>
<span class="go">  list  quality</span>
<span class="go">0   []        3</span>
<span class="go">1   []        3</span>
<span class="go">2   []        4</span>
<span class="go">3   []       10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">positional_metadata</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
<p>Next let’s create a sequence that has been derived from another sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span>
<span class="go">Sequence</span>
<span class="go">-----------------------------</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;list&#39;: &lt;dtype: object&gt;</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: int64&gt;</span>
<span class="go">    &#39;gaps&#39;: &lt;dtype: bool&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 2</span>
<span class="go">-----------------------------</span>
<span class="go">0 CG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">  list  quality   gaps</span>
<span class="go">0   []        3  False</span>
<span class="go">1   []        4  False</span>
</pre></div>
</div>
<p>As described above for metadata, since only a <em>shallow</em> copy was made of
the positional metadata, updates to mutable objects will also change the
parent sequence’s positional metadata and the original positional metadata
<code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;list&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;item&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subseq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">     list  quality   gaps</span>
<span class="go">0  [item]        3  False</span>
<span class="go">1      []        4  False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span>
<span class="go">     list  quality   gaps</span>
<span class="go">0      []      999  False</span>
<span class="go">1  [item]        3  False</span>
<span class="go">2      []        4  False</span>
<span class="go">3      []       10  False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">positional_metadata</span>
<span class="go">     list  quality</span>
<span class="go">0      []        3</span>
<span class="go">1  [item]        3</span>
<span class="go">2      []        4</span>
<span class="go">3      []       10</span>
</pre></div>
</div>
<p>You can also update the interval metadata. Let’s re-create a
<code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object with interval metadata at first:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;foo&#39;</span><span class="p">})</span>
</pre></div>
</div>
<p>You can update directly on the <code class="docutils literal notranslate"><span class="pre">Interval</span></code> object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span>
<span class="go">Interval(interval_metadata=&lt;...&gt;, bounds=[(1, 3)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;foo&#39;})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span><span class="o">.</span><span class="n">bounds</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">interval</span>
<span class="go">Interval(interval_metadata=&lt;...&gt;, bounds=[(0, 2)], fuzzy=[(False, False)], metadata={&#39;gene&#39;: &#39;foo&#39;})</span>
</pre></div>
</div>
<p>You can also query and obtain the interval features you are
interested and then modify them:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;foo&#39;</span><span class="p">}))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intervals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fuzzy</span> <span class="o">=</span> <span class="p">[(</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">intervals</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">Interval(interval_metadata=&lt;...&gt;, bounds=[(0, 2)], fuzzy=[(True, False)], metadata={&#39;gene&#39;: &#39;foo&#39;})</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.default_write_format" title="skbio.sequence.Sequence.default_write_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">default_write_format</span></code></a></p></td>
<td><p>Default write format for this object: <code class="docutils literal notranslate"><span class="pre">fasta</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.observed_chars" title="skbio.sequence.Sequence.observed_chars"><code class="xref py py-obj docutils literal notranslate"><span class="pre">observed_chars</span></code></a></p></td>
<td><p>Set of observed characters in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.values" title="skbio.sequence.Sequence.values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code></a></p></td>
<td><p>Array containing underlying sequence characters.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Attributes (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.IntervalMetadataMixin.html#skbio.metadata.IntervalMetadataMixin.interval_metadata" title="skbio.metadata.IntervalMetadataMixin.interval_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">interval_metadata</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">IntervalMetadata</span></code> object containing info about interval features.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.metadata.MetadataMixin.html#skbio.metadata.MetadataMixin.metadata" title="skbio.metadata.MetadataMixin.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metadata</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">dict</span></code> containing metadata which applies to the entire object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.PositionalMetadataMixin.html#skbio.metadata.PositionalMetadataMixin.positional_metadata" title="skbio.metadata.PositionalMetadataMixin.positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_metadata</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> containing metadata along an axis.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.concat.html#skbio.sequence.Sequence.concat" title="skbio.sequence.Sequence.concat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code></a></p></td>
<td><p>Concatenate an iterable of <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.count.html#skbio.sequence.Sequence.count" title="skbio.sequence.Sequence.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a></p></td>
<td><p>Count occurrences of a subsequence in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.distance.html#skbio.sequence.Sequence.distance" title="skbio.sequence.Sequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a></p></td>
<td><p>Compute the distance to another sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.find_with_regex.html#skbio.sequence.Sequence.find_with_regex" title="skbio.sequence.Sequence.find_with_regex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_with_regex</span></code></a></p></td>
<td><p>Generate slices for patterns matched by a regular expression.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.frequencies.html#skbio.sequence.Sequence.frequencies" title="skbio.sequence.Sequence.frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frequencies</span></code></a></p></td>
<td><p>Compute frequencies of characters in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.index.html#skbio.sequence.Sequence.index" title="skbio.sequence.Sequence.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code></a></p></td>
<td><p>Find position where subsequence first occurs in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.iter_contiguous.html#skbio.sequence.Sequence.iter_contiguous" title="skbio.sequence.Sequence.iter_contiguous"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_contiguous</span></code></a></p></td>
<td><p>Yield contiguous subsequences based on <cite>included</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.iter_kmers.html#skbio.sequence.Sequence.iter_kmers" title="skbio.sequence.Sequence.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a></p></td>
<td><p>Generate k-mers of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.kmer_frequencies.html#skbio.sequence.Sequence.kmer_frequencies" title="skbio.sequence.Sequence.kmer_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kmer_frequencies</span></code></a></p></td>
<td><p>Return counts of words of length <cite>k</cite> from this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.lowercase.html#skbio.sequence.Sequence.lowercase" title="skbio.sequence.Sequence.lowercase"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lowercase</span></code></a></p></td>
<td><p>Return a case-sensitive string representation of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.match_frequency.html#skbio.sequence.Sequence.match_frequency" title="skbio.sequence.Sequence.match_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">match_frequency</span></code></a></p></td>
<td><p>Return count of positions that are the same between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.matches.html#skbio.sequence.Sequence.matches" title="skbio.sequence.Sequence.matches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matches</span></code></a></p></td>
<td><p>Find positions that match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.mismatch_frequency.html#skbio.sequence.Sequence.mismatch_frequency" title="skbio.sequence.Sequence.mismatch_frequency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatch_frequency</span></code></a></p></td>
<td><p>Return count of positions that differ between two sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.mismatches.html#skbio.sequence.Sequence.mismatches" title="skbio.sequence.Sequence.mismatches"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mismatches</span></code></a></p></td>
<td><p>Find positions that do not match with another sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.read.html#skbio.sequence.Sequence.read" title="skbio.sequence.Sequence.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></a></p></td>
<td><p>Create a new <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> instance from a file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.replace.html#skbio.sequence.Sequence.replace" title="skbio.sequence.Sequence.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a></p></td>
<td><p>Replace values in this sequence with a different character.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.sequence.Sequence.to_indices.html#skbio.sequence.Sequence.to_indices" title="skbio.sequence.Sequence.to_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_indices</span></code></a></p></td>
<td><p>Convert the sequence into indices of characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.sequence.Sequence.write.html#skbio.sequence.Sequence.write" title="skbio.sequence.Sequence.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a></p></td>
<td><p>Write an instance of <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> to a file.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.IntervalMetadataMixin.has_interval_metadata.html#skbio.metadata.IntervalMetadataMixin.has_interval_metadata" title="skbio.metadata.IntervalMetadataMixin.has_interval_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_interval_metadata</span></code></a></p></td>
<td><p>Determine if the object has interval metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="skbio.metadata.MetadataMixin.has_metadata.html#skbio.metadata.MetadataMixin.has_metadata" title="skbio.metadata.MetadataMixin.has_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_metadata</span></code></a></p></td>
<td><p>Determine if the object has metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="skbio.metadata.PositionalMetadataMixin.has_positional_metadata.html#skbio.metadata.PositionalMetadataMixin.has_positional_metadata" title="skbio.metadata.PositionalMetadataMixin.has_positional_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_positional_metadata</span></code></a></p></td>
<td><p>Determine if the object has positional metadata.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__bool__" title="skbio.sequence.Sequence.__bool__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__bool__</span></code></a></p></td>
<td><p>Return truth value (truthiness) of sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__contains__" title="skbio.sequence.Sequence.__contains__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__contains__</span></code></a></p></td>
<td><p>Determine if a subsequence is contained in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__copy__" title="skbio.sequence.Sequence.__copy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__copy__</span></code></a></p></td>
<td><p>Return a shallow copy of this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__deepcopy__" title="skbio.sequence.Sequence.__deepcopy__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__deepcopy__</span></code></a></p></td>
<td><p>Return a deep copy of this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__eq__" title="skbio.sequence.Sequence.__eq__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__eq__</span></code></a></p></td>
<td><p>Determine if this sequence is equal to another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__getitem__" title="skbio.sequence.Sequence.__getitem__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getitem__</span></code></a></p></td>
<td><p>Slice this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__iter__" title="skbio.sequence.Sequence.__iter__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__iter__</span></code></a></p></td>
<td><p>Iterate over positions in this sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__len__" title="skbio.sequence.Sequence.__len__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__len__</span></code></a></p></td>
<td><p>Return the number of characters in this sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__ne__" title="skbio.sequence.Sequence.__ne__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ne__</span></code></a></p></td>
<td><p>Determine if this sequence is not equal to another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__reversed__" title="skbio.sequence.Sequence.__reversed__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__reversed__</span></code></a></p></td>
<td><p>Iterate over positions in this sequence in reverse order.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#skbio.sequence.Sequence.__str__" title="skbio.sequence.Sequence.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__str__</span></code></a></p></td>
<td><p>Return sequence characters as a string.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Special methods (inherited)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__ge__</span></code></p></td>
<td><p>Return self&gt;=value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__getstate__</span></code></p></td>
<td><p>Helper for pickle.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__gt__</span></code></p></td>
<td><p>Return self&gt;value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__le__</span></code></p></td>
<td><p>Return self&lt;=value.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">__lt__</span></code></p></td>
<td><p>Return self&lt;value.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Details</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.default_write_format">
<span class="sig-name descname"><span class="pre">default_write_format</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'fasta'</span></em><a class="headerlink" href="#skbio.sequence.Sequence.default_write_format" title="Link to this definition">#</a></dt>
<dd><p>Default write format for this object: <code class="docutils literal notranslate"><span class="pre">fasta</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.observed_chars">
<span class="sig-name descname"><span class="pre">observed_chars</span></span><a class="headerlink" href="#skbio.sequence.Sequence.observed_chars" title="Link to this definition">#</a></dt>
<dd><p>Set of observed characters in the sequence.</p>
<p class="rubric">Notes</p>
<p>This property is not writeable.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;AACGAC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">observed_chars</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">}</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.values">
<span class="sig-name descname"><span class="pre">values</span></span><a class="headerlink" href="#skbio.sequence.Sequence.values" title="Link to this definition">#</a></dt>
<dd><p>Array containing underlying sequence characters.</p>
<p class="rubric">Notes</p>
<p>This property is not writeable.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;AACGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">values</span>
<span class="go">array([b&#39;A&#39;, b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;, b&#39;A&#39;],</span>
<span class="go">      dtype=&#39;|S1&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__bool__">
<span class="sig-name descname"><span class="pre">__bool__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1003"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__bool__" title="Link to this definition">#</a></dt>
<dd><p>Return truth value (truthiness) of sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>bool</dt><dd><p>True if length of sequence is greater than 0, else False.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">))</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bool</span><span class="p">(</span><span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__contains__">
<span class="sig-name descname"><span class="pre">__contains__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subsequence</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L696"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__contains__" title="Link to this definition">#</a></dt>
<dd><p>Determine if a subsequence is contained in this sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>subsequence</strong><span class="classifier">str, Sequence, or 1D np.ndarray (np.uint8 or ‘|S1’)</span></dt><dd><p>The putative subsequence.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether <cite>subsequence</cite> is contained in this sequence.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>TypeError</dt><dd><p>If <cite>subsequence</cite> is a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object with a different type
than this sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUCGUGAAGGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;GGU&#39;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;CCC&#39;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__copy__">
<span class="sig-name descname"><span class="pre">__copy__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1214"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__copy__" title="Link to this definition">#</a></dt>
<dd><p>Return a shallow copy of this sequence.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference external" href="https://docs.python.org/3/library/copy.html#module-copy" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>This method is equivalent to <code class="docutils literal notranslate"><span class="pre">seq.copy(deep=False)</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1228"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__deepcopy__" title="Link to this definition">#</a></dt>
<dd><p>Return a deep copy of this sequence.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference external" href="https://docs.python.org/3/library/copy.html#module-copy" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>This method is equivalent to <code class="docutils literal notranslate"><span class="pre">seq.copy(deep=True)</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L727"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__eq__" title="Link to this definition">#</a></dt>
<dd><p>Determine if this sequence is equal to another.</p>
<p>Sequences are equal if they are <em>exactly</em> the same type and their
sequence characters, metadata, and positional metadata are the same.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong><span class="classifier">Sequence</span></dt><dd><p>Sequence to test for equality against.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether this sequence is equal to <cite>other</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Define two <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> objects that have the same underlying sequence
of characters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The two sequences are considered equal because they are the same type,
their underlying sequence of characters are the same, and their
optional metadata attributes (<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code>)
were not provided:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="n">t</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">==</span> <span class="n">s</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Define another sequence object with a different sequence of characters
than the previous two sequence objects:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">==</span> <span class="n">t</span>
<span class="go">False</span>
</pre></div>
</div>
<p>Define a sequence with the same sequence of characters as <code class="docutils literal notranslate"><span class="pre">u</span></code> but
with different metadata, positional metadata, and interval metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">v</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;abc&#39;</span><span class="p">},</span>
<span class="gp">... </span>             <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;quality&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">([(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)])</span>
</pre></div>
</div>
<p>The two sequences are not considered equal because their metadata,
positional metadata, and interval metadata do not match:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">==</span> <span class="n">v</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indexable</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L834"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>Slice this sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>indexable</strong><span class="classifier">int, slice, iterable (int and slice), 1D array_like (bool)</span></dt><dd><p>The position(s) to return from this sequence. If <cite>indexable</cite> is an
iterable of integers, these are assumed to be indices in the
sequence to keep. If <cite>indexable</cite> is a 1D <code class="docutils literal notranslate"><span class="pre">array_like</span></code> of
booleans, these are assumed to be the positions in the sequence to
keep.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Sequence</dt><dd><p>New sequence containing the position(s) specified by <cite>indexable</cite> in
this sequence. Positional metadata will be sliced in the same
manner and included in the returned sequence. <cite>metadata</cite> is
included in the returned sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This drops the <code class="docutils literal notranslate"><span class="pre">self.interval_metadata</span></code> from the returned
new <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUCGUGAAGGA&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Obtain a single character from the sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">-------------</span>
<span class="go">0 G</span>
</pre></div>
</div>
<p>Obtain a slice:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">7</span><span class="p">:]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 5</span>
<span class="go">-------------</span>
<span class="go">0 AAGGA</span>
</pre></div>
</div>
<p>Obtain characters at the following indices:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">]]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 5</span>
<span class="go">-------------</span>
<span class="go">0 CGAGC</span>
</pre></div>
</div>
<p>Obtain characters at positions evaluating to <cite>True</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span> <span class="o">==</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">-------------</span>
<span class="go">0 GUC</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1022"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__iter__" title="Link to this definition">#</a></dt>
<dd><p>Iterate over positions in this sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Yields<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Sequence</dt><dd><p>Single character subsequence, one for each position in the
sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">s</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="go">&#39;G&#39;</span>
<span class="go">&#39;G&#39;</span>
<span class="go">&#39;U&#39;</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L985"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__len__" title="Link to this definition">#</a></dt>
<dd><p>Return the number of characters in this sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>The length of this sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="go">4</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__ne__">
<span class="sig-name descname"><span class="pre">__ne__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L801"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__ne__" title="Link to this definition">#</a></dt>
<dd><p>Determine if this sequence is not equal to another.</p>
<p>Sequences are not equal if they are not <em>exactly</em> the same type, or
their sequence characters, metadata, or positional metadata differ.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong><span class="classifier">Sequence</span></dt><dd><p>Sequence to test for inequality against.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>Indicates whether this sequence is not equal to <cite>other</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">!=</span> <span class="n">t</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">!=</span> <span class="n">t</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;ACGA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;v&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">u</span> <span class="o">!=</span> <span class="n">v</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__reversed__">
<span class="sig-name descname"><span class="pre">__reversed__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1046"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__reversed__" title="Link to this definition">#</a></dt>
<dd><p>Iterate over positions in this sequence in reverse order.</p>
<dl class="field-list simple">
<dt class="field-odd">Yields<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Sequence</dt><dd><p>Single character subsequence, one for each position in the
sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="go">&#39;C&#39;</span>
<span class="go">&#39;U&#39;</span>
<span class="go">&#39;G&#39;</span>
<span class="go">&#39;G&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="skbio.sequence.Sequence.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/scikit-bio/scikit-bio/blob/0.7.2/skbio/sequence/_sequence.py#L1069"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#skbio.sequence.Sequence.__str__" title="Link to this definition">#</a></dt>
<dd><p>Return sequence characters as a string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>Sequence characters as a string. No metadata or positional
metadata will be included.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sequence</span><span class="p">(</span><span class="s1">&#39;GGUCGUAAAGGA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;hello&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="go">&#39;GGUCGUAAAGGA&#39;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../sequence.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Biological Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</p>
      </div>
    </a>
    <a class="right-next"
       href="skbio.sequence.Sequence.concat.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.sequence.Sequence.concat</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.default_write_format"><code class="docutils literal notranslate"><span class="pre">Sequence.default_write_format</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.observed_chars"><code class="docutils literal notranslate"><span class="pre">Sequence.observed_chars</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.values"><code class="docutils literal notranslate"><span class="pre">Sequence.values</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__bool__"><code class="docutils literal notranslate"><span class="pre">Sequence.__bool__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__contains__"><code class="docutils literal notranslate"><span class="pre">Sequence.__contains__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__copy__"><code class="docutils literal notranslate"><span class="pre">Sequence.__copy__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__deepcopy__"><code class="docutils literal notranslate"><span class="pre">Sequence.__deepcopy__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__eq__"><code class="docutils literal notranslate"><span class="pre">Sequence.__eq__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__getitem__"><code class="docutils literal notranslate"><span class="pre">Sequence.__getitem__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__iter__"><code class="docutils literal notranslate"><span class="pre">Sequence.__iter__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__len__"><code class="docutils literal notranslate"><span class="pre">Sequence.__len__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__ne__"><code class="docutils literal notranslate"><span class="pre">Sequence.__ne__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__reversed__"><code class="docutils literal notranslate"><span class="pre">Sequence.__reversed__</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#skbio.sequence.Sequence.__str__"><code class="docutils literal notranslate"><span class="pre">Sequence.__str__</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2026, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>