<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.core.sequence.ProteinSequence &mdash; scikit-bio 0.1.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.1.1 documentation" href="../index.html" />
    <link rel="up" title="Biological sequences (skbio.core.sequence)" href="../core.sequence.html" />
    <link rel="next" title="skbio.core.sequence.ProteinSequence.description" href="skbio.core.sequence.ProteinSequence.description.html" />
    <link rel="prev" title="skbio.core.sequence.RNASequence.upper" href="skbio.core.sequence.RNASequence.upper.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.1</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../core.alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.core.alignment</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../core.sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.distance.html">Dissimilarity and distance matrices (<tt class="docutils literal"><span class="pre">skbio.core.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.exception.html">Exceptions (<tt class="docutils literal"><span class="pre">skbio.core.exception</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.warning.html">Warnings (<tt class="docutils literal"><span class="pre">skbio.core.warning</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.genetic_code.html">Genetic Code (<tt class="docutils literal"><span class="pre">skbio.core.genetic_code</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.core.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.core.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../core.ssw.html">Performing Striped Smith Waterman Alignments (<tt class="docutils literal"><span class="pre">skbio.core.ssw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.distributions.html">Distribution visualizations (<tt class="docutils literal"><span class="pre">skbio.draw.distributions</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.diversity.alpha.html">Alpha diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.alpha</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.distance.html">Distance-based statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.spatial.html">Spatial Statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.spatial</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.stats.ordination.html">Ordination methods (<tt class="docutils literal"><span class="pre">skbio.math.stats.ordination</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.gradient.html">Gradient analyses (<tt class="docutils literal"><span class="pre">skbio.math.gradient</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../math.subsample.html">Subsampling (<tt class="docutils literal"><span class="pre">skbio.math.subsample</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.io.html">I/O utils (<tt class="docutils literal"><span class="pre">skbio.util.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.misc.html">Miscellaneous Utilities (<tt class="docutils literal"><span class="pre">skbio.util.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.testing.html">Testing utilities (<tt class="docutils literal"><span class="pre">skbio.util.testing</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.core.sequence.ProteinSequence</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-core-sequence-proteinsequence">
<h1>skbio.core.sequence.ProteinSequence<a class="headerlink" href="#skbio-core-sequence-proteinsequence" title="Permalink to this headline">Â¶</a></h1>
<dl class="class">
<dt id="skbio.core.sequence.ProteinSequence">
<em class="property">class </em><tt class="descclassname">skbio.core.sequence.</tt><tt class="descname">ProteinSequence</tt><big>(</big><em>sequence</em>, <em>id=''</em>, <em>description=''</em>, <em>validate=False</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.1.1/skbio/core/sequence.py#L1561"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.core.sequence.ProteinSequence" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Base class for protein sequences.</p>
<p>A <cite>ProteinSequence</cite> is a <cite>BiologicalSequence</cite> containing only characters
used in the IUPAC protein lexicon.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><tt class="xref py py-obj docutils literal"><span class="pre">BiologialSequence</span></tt></p>
</div>
<p class="rubric">Notes</p>
<p>All uppercase and lowercase IUPAC protein characters are supported.</p>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.description.html#skbio.core.sequence.ProteinSequence.description" title="skbio.core.sequence.ProteinSequence.description"><tt class="xref py py-obj docutils literal"><span class="pre">description</span></tt></a></td>
<td>Return the description of the <cite>BiologicalSequence</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.id.html#skbio.core.sequence.ProteinSequence.id" title="skbio.core.sequence.ProteinSequence.id"><tt class="xref py py-obj docutils literal"><span class="pre">id</span></tt></a></td>
<td>Return the id of the <cite>BiologicalSequence</cite></td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__contains__.html#skbio.core.sequence.ProteinSequence.__contains__" title="skbio.core.sequence.ProteinSequence.__contains__"><tt class="xref py py-obj docutils literal"><span class="pre">__contains__</span></tt></a>(other)</td>
<td>The in operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__eq__.html#skbio.core.sequence.ProteinSequence.__eq__" title="skbio.core.sequence.ProteinSequence.__eq__"><tt class="xref py py-obj docutils literal"><span class="pre">__eq__</span></tt></a>(other)</td>
<td>The equality operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__getitem__.html#skbio.core.sequence.ProteinSequence.__getitem__" title="skbio.core.sequence.ProteinSequence.__getitem__"><tt class="xref py py-obj docutils literal"><span class="pre">__getitem__</span></tt></a>(i)</td>
<td>The indexing operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__hash__.html#skbio.core.sequence.ProteinSequence.__hash__" title="skbio.core.sequence.ProteinSequence.__hash__"><tt class="xref py py-obj docutils literal"><span class="pre">__hash__</span></tt></a>()</td>
<td>The hash operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__iter__.html#skbio.core.sequence.ProteinSequence.__iter__" title="skbio.core.sequence.ProteinSequence.__iter__"><tt class="xref py py-obj docutils literal"><span class="pre">__iter__</span></tt></a>()</td>
<td>The iter operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__len__.html#skbio.core.sequence.ProteinSequence.__len__" title="skbio.core.sequence.ProteinSequence.__len__"><tt class="xref py py-obj docutils literal"><span class="pre">__len__</span></tt></a>()</td>
<td>The len operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__ne__.html#skbio.core.sequence.ProteinSequence.__ne__" title="skbio.core.sequence.ProteinSequence.__ne__"><tt class="xref py py-obj docutils literal"><span class="pre">__ne__</span></tt></a>(other)</td>
<td>The inequality operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__repr__.html#skbio.core.sequence.ProteinSequence.__repr__" title="skbio.core.sequence.ProteinSequence.__repr__"><tt class="xref py py-obj docutils literal"><span class="pre">__repr__</span></tt></a>()</td>
<td>The repr method.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__reversed__.html#skbio.core.sequence.ProteinSequence.__reversed__" title="skbio.core.sequence.ProteinSequence.__reversed__"><tt class="xref py py-obj docutils literal"><span class="pre">__reversed__</span></tt></a>()</td>
<td>The reversed operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.__str__.html#skbio.core.sequence.ProteinSequence.__str__" title="skbio.core.sequence.ProteinSequence.__str__"><tt class="xref py py-obj docutils literal"><span class="pre">__str__</span></tt></a>()</td>
<td>The str operator</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.alphabet.html#skbio.core.sequence.ProteinSequence.alphabet" title="skbio.core.sequence.ProteinSequence.alphabet"><tt class="xref py py-obj docutils literal"><span class="pre">alphabet</span></tt></a>()</td>
<td>Return the set of characters allowed in a <cite>BiologicalSequence</cite>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.count.html#skbio.core.sequence.ProteinSequence.count" title="skbio.core.sequence.ProteinSequence.count"><tt class="xref py py-obj docutils literal"><span class="pre">count</span></tt></a>(subsequence)</td>
<td>Returns the number of occurences of subsequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.degap.html#skbio.core.sequence.ProteinSequence.degap" title="skbio.core.sequence.ProteinSequence.degap"><tt class="xref py py-obj docutils literal"><span class="pre">degap</span></tt></a>()</td>
<td>Returns a new <cite>BiologicalSequence</cite> with gaps characters removed.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.distance.html#skbio.core.sequence.ProteinSequence.distance" title="skbio.core.sequence.ProteinSequence.distance"><tt class="xref py py-obj docutils literal"><span class="pre">distance</span></tt></a>(other[,&nbsp;distance_fn])</td>
<td>Returns the distance to other</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.fraction_diff.html#skbio.core.sequence.ProteinSequence.fraction_diff" title="skbio.core.sequence.ProteinSequence.fraction_diff"><tt class="xref py py-obj docutils literal"><span class="pre">fraction_diff</span></tt></a>(other)</td>
<td>Return fraction of positions that differ relative to <cite>other</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.fraction_same.html#skbio.core.sequence.ProteinSequence.fraction_same" title="skbio.core.sequence.ProteinSequence.fraction_same"><tt class="xref py py-obj docutils literal"><span class="pre">fraction_same</span></tt></a>(other)</td>
<td>Return fraction of positions that are the same relative to <cite>other</cite></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.gap_alphabet.html#skbio.core.sequence.ProteinSequence.gap_alphabet" title="skbio.core.sequence.ProteinSequence.gap_alphabet"><tt class="xref py py-obj docutils literal"><span class="pre">gap_alphabet</span></tt></a>()</td>
<td>Return the set of characters defined as gaps.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.gap_maps.html#skbio.core.sequence.ProteinSequence.gap_maps" title="skbio.core.sequence.ProteinSequence.gap_maps"><tt class="xref py py-obj docutils literal"><span class="pre">gap_maps</span></tt></a>()</td>
<td>Return tuples mapping b/w gapped and ungapped positions</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.gap_vector.html#skbio.core.sequence.ProteinSequence.gap_vector" title="skbio.core.sequence.ProteinSequence.gap_vector"><tt class="xref py py-obj docutils literal"><span class="pre">gap_vector</span></tt></a>()</td>
<td>Return list indicating positions containing gaps</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.has_unsupported_characters.html#skbio.core.sequence.ProteinSequence.has_unsupported_characters" title="skbio.core.sequence.ProteinSequence.has_unsupported_characters"><tt class="xref py py-obj docutils literal"><span class="pre">has_unsupported_characters</span></tt></a>()</td>
<td>Return bool indicating presence/absence of unsupported characters</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.index.html#skbio.core.sequence.ProteinSequence.index" title="skbio.core.sequence.ProteinSequence.index"><tt class="xref py py-obj docutils literal"><span class="pre">index</span></tt></a>(subsequence)</td>
<td>Return the position where subsequence first occurs</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.is_gap.html#skbio.core.sequence.ProteinSequence.is_gap" title="skbio.core.sequence.ProteinSequence.is_gap"><tt class="xref py py-obj docutils literal"><span class="pre">is_gap</span></tt></a>(char)</td>
<td>Return True if <cite>char</cite> is in the <cite>gap_alphabet</cite> set</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.is_gapped.html#skbio.core.sequence.ProteinSequence.is_gapped" title="skbio.core.sequence.ProteinSequence.is_gapped"><tt class="xref py py-obj docutils literal"><span class="pre">is_gapped</span></tt></a>()</td>
<td>Return True if char(s) in <cite>gap_alphabet</cite> are present</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.is_valid.html#skbio.core.sequence.ProteinSequence.is_valid" title="skbio.core.sequence.ProteinSequence.is_valid"><tt class="xref py py-obj docutils literal"><span class="pre">is_valid</span></tt></a>()</td>
<td>Return True if the sequence is valid</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.iupac_characters.html#skbio.core.sequence.ProteinSequence.iupac_characters" title="skbio.core.sequence.ProteinSequence.iupac_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_characters</span></tt></a>()</td>
<td>Return the non-degenerate and degenerate characters.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.iupac_degeneracies.html#skbio.core.sequence.ProteinSequence.iupac_degeneracies" title="skbio.core.sequence.ProteinSequence.iupac_degeneracies"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_degeneracies</span></tt></a>()</td>
<td>Return the mapping of degenerate to non-degenerate characters.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.iupac_degenerate_characters.html#skbio.core.sequence.ProteinSequence.iupac_degenerate_characters" title="skbio.core.sequence.ProteinSequence.iupac_degenerate_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_degenerate_characters</span></tt></a>()</td>
<td>Return the degenerate IUPAC characters.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.iupac_standard_characters.html#skbio.core.sequence.ProteinSequence.iupac_standard_characters" title="skbio.core.sequence.ProteinSequence.iupac_standard_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_standard_characters</span></tt></a>()</td>
<td>Return the non-degenerate IUPAC protein characters.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.k_word_counts.html#skbio.core.sequence.ProteinSequence.k_word_counts" title="skbio.core.sequence.ProteinSequence.k_word_counts"><tt class="xref py py-obj docutils literal"><span class="pre">k_word_counts</span></tt></a>(k[,&nbsp;overlapping,&nbsp;constructor])</td>
<td>Get the counts of words of length k</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.k_word_frequencies.html#skbio.core.sequence.ProteinSequence.k_word_frequencies" title="skbio.core.sequence.ProteinSequence.k_word_frequencies"><tt class="xref py py-obj docutils literal"><span class="pre">k_word_frequencies</span></tt></a>(k[,&nbsp;overlapping,&nbsp;constructor])</td>
<td>Get the frequencies of words of length k</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.k_words.html#skbio.core.sequence.ProteinSequence.k_words" title="skbio.core.sequence.ProteinSequence.k_words"><tt class="xref py py-obj docutils literal"><span class="pre">k_words</span></tt></a>(k[,&nbsp;overlapping,&nbsp;constructor])</td>
<td>Get the list of words of length k</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.lower.html#skbio.core.sequence.ProteinSequence.lower" title="skbio.core.sequence.ProteinSequence.lower"><tt class="xref py py-obj docutils literal"><span class="pre">lower</span></tt></a>()</td>
<td>Convert the BiologicalSequence to lowercase</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.nondegenerates.html#skbio.core.sequence.ProteinSequence.nondegenerates" title="skbio.core.sequence.ProteinSequence.nondegenerates"><tt class="xref py py-obj docutils literal"><span class="pre">nondegenerates</span></tt></a>()</td>
<td>Yield all nondegenerate versions of the sequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.to_fasta.html#skbio.core.sequence.ProteinSequence.to_fasta" title="skbio.core.sequence.ProteinSequence.to_fasta"><tt class="xref py py-obj docutils literal"><span class="pre">to_fasta</span></tt></a>([field_delimiter,&nbsp;terminal_character])</td>
<td>Return the sequence as a fasta-formatted string</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.unsupported_characters.html#skbio.core.sequence.ProteinSequence.unsupported_characters" title="skbio.core.sequence.ProteinSequence.unsupported_characters"><tt class="xref py py-obj docutils literal"><span class="pre">unsupported_characters</span></tt></a>()</td>
<td>Return the set of unsupported characters in the <cite>BiologicalSequence</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.core.sequence.ProteinSequence.upper.html#skbio.core.sequence.ProteinSequence.upper" title="skbio.core.sequence.ProteinSequence.upper"><tt class="xref py py-obj docutils literal"><span class="pre">upper</span></tt></a>()</td>
<td>Convert the BiologicalSequence to uppercase</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>