<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QSeq format (skbio.io.format.qseq) &mdash; scikit-bio 0.4.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.4/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.4.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.4.0 documentation" href="../index.html" />
    <link rel="up" title="File I/O (skbio.io)" href="../io.html" />
    <link rel="next" title="skbio.io.registry.write" href="skbio.io.registry.write.html" />
    <link rel="prev" title="PHYLIP multiple sequence alignment format (skbio.io.format.phylip)" href="skbio.io.format.phylip.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.4.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sequence.html">Sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Alignments and Sequence collections (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../user/api_stability.html">API Stability</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">QSeq format (<tt class="docutils literal"><span class="pre">skbio.io.format.qseq</span></tt>)</a><ul>
<li><a class="reference internal" href="#format-support">Format Support</a></li>
<li><a class="reference internal" href="#format-specification">Format Specification</a></li>
<li><a class="reference internal" href="#format-parameters">Format Parameters</a><ul>
<li><a class="reference internal" href="#sequencecollection-and-generators-only">SequenceCollection and Generators Only</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <span class="target" id="module-skbio.io.format.qseq"></span><div class="section" id="qseq-format-skbio-io-format-qseq">
<h1>QSeq format (<a class="reference internal" href="#module-skbio.io.format.qseq" title="skbio.io.format.qseq"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.io.format.qseq</span></tt></a>)<a class="headerlink" href="#qseq-format-skbio-io-format-qseq" title="Permalink to this headline">¶</a></h1>
<p>The QSeq format (<cite>qseq</cite>) is a record-based, plain text output format produced
by some DNA sequencers for storing biological sequence data, quality scores,
per-sequence filtering information, and run-specific metadata.</p>
<div class="section" id="format-support">
<h2>Format Support<a class="headerlink" href="#format-support" title="Permalink to this headline">¶</a></h2>
<p><strong>Has Sniffer: Yes</strong></p>
<table border="1" class="docutils">
<colgroup>
<col width="8%" />
<col width="8%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Reader</th>
<th class="head">Writer</th>
<th class="head">Object Class</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Yes</td>
<td>No</td>
<td>generator of <a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.Sequence</span></tt></a> objects</td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>No</td>
<td><a class="reference internal" href="skbio.alignment.SequenceCollection.html#skbio.alignment.SequenceCollection" title="skbio.alignment.SequenceCollection"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.alignment.SequenceCollection</span></tt></a></td>
</tr>
<tr class="row-even"><td>Yes</td>
<td>No</td>
<td><a class="reference internal" href="skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.Sequence</span></tt></a></td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>No</td>
<td><a class="reference internal" href="skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.DNA</span></tt></a></td>
</tr>
<tr class="row-even"><td>Yes</td>
<td>No</td>
<td><a class="reference internal" href="skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.RNA</span></tt></a></td>
</tr>
<tr class="row-odd"><td>Yes</td>
<td>No</td>
<td><a class="reference internal" href="skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.sequence.Protein</span></tt></a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="format-specification">
<h2>Format Specification<a class="headerlink" href="#format-specification" title="Permalink to this headline">¶</a></h2>
<p>A QSeq file is composed of single-line records, delimited by tabs. There are
11 fields in a record:</p>
<ul class="simple">
<li>Machine name</li>
<li>Run number</li>
<li>Lane number (positive int)</li>
<li>Tile number (positive int)</li>
<li>X coordinate (integer)</li>
<li>Y coordinate (integer)</li>
<li>Index</li>
<li>Read number (1-3)</li>
<li>Sequence data (typically IUPAC characters)</li>
<li>Quality scores (quality scores encoded as printable ASCII)</li>
<li>Filter boolean (1 if sequence has passed CASAVA&#8217;s filter, 0 otherwise)</li>
</ul>
<p>For more details please refer to the CASAVA documentation <a class="reference internal" href="#r162" id="id1">[R162]</a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When a QSeq file is read into a scikit-bio object, the object&#8217;s
<cite>metadata</cite> attribute is automatically populated with data corresponding
to the names above.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><cite>lowercase</cite> functionality is supported when reading QSeq files,
depending on the object type the file is being read into. Refer to
specific object constructor documentation for details.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">scikit-bio allows for the filter field to be ommitted, but it is not
clear if this is part of the original format specification.</p>
</div>
</div>
<div class="section" id="format-parameters">
<h2>Format Parameters<a class="headerlink" href="#format-parameters" title="Permalink to this headline">¶</a></h2>
<p>The following parameters are the same as in FASTQ format
(<a class="reference internal" href="skbio.io.format.fastq.html#module-skbio.io.format.fastq" title="skbio.io.format.fastq"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.io.format.fastq</span></tt></a>):</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">variant</span></tt>: see <tt class="docutils literal"><span class="pre">variant</span></tt> parameter in FASTQ format</li>
<li><tt class="docutils literal"><span class="pre">phred_offset</span></tt>: see <tt class="docutils literal"><span class="pre">phred_offset</span></tt> parameter in FASTQ format</li>
</ul>
<p>The following additional parameters are the same as in FASTA format
(<a class="reference internal" href="skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.io.format.fasta</span></tt></a>):</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">constructor</span></tt>: see <tt class="docutils literal"><span class="pre">constructor</span></tt> parameter in FASTA format</li>
<li><tt class="docutils literal"><span class="pre">seq_num</span></tt>: see <tt class="docutils literal"><span class="pre">seq_num</span></tt> parameter in FASTA format</li>
</ul>
<div class="section" id="sequencecollection-and-generators-only">
<h3>SequenceCollection and Generators Only<a class="headerlink" href="#sequencecollection-and-generators-only" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">filter</span></tt>: If <cite>True</cite>, excludes sequences that did not pass filtering
(i.e., filter field is 0). Default is <cite>True</cite>.</li>
</ul>
</div>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Suppose we have the following QSeq file:</p>
<div class="highlight-python"><div class="highlight"><pre>illumina    1       3       34      -30     30      0       1       ACG....ACGTAC   ruBBBBrBCEFGH   1
illumina    1       3       34      30      -30     0       1       CGGGCATTGCA     CGGGCasdGCA     0
illumina    1       3       35      -30     30      0       2       ACGTA.AATAAAC   geTaAafhwqAAf   1
illumina    1       3       35      30      -30     0       3       CATTTAGGA.TGCA  tjflkAFnkKghvM  0
</pre></div>
</div>
<p>Let&#8217;s define this file in-memory as a <tt class="docutils literal"><span class="pre">StringIO</span></tt>, though this could be a real
file path, file handle, or anything that&#8217;s supported by scikit-bio&#8217;s I/O
registry in practice:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fs</span> <span class="o">=</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s">&#39;illumina</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">3</span><span class="se">\t</span><span class="s">34</span><span class="se">\t</span><span class="s">-30</span><span class="se">\t</span><span class="s">30</span><span class="se">\t</span><span class="s">0</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">ACG....ACGTAC</span><span class="se">\t</span><span class="s">ruBBBBrBCEFGH</span><span class="se">\t</span><span class="s">1&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s">&#39;illumina</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">3</span><span class="se">\t</span><span class="s">34</span><span class="se">\t</span><span class="s">30</span><span class="se">\t</span><span class="s">-30</span><span class="se">\t</span><span class="s">0</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">CGGGCATTGCA</span><span class="se">\t</span><span class="s">CGGGCasdGCA</span><span class="se">\t</span><span class="s">0&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s">&#39;illumina</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">3</span><span class="se">\t</span><span class="s">35</span><span class="se">\t</span><span class="s">-30</span><span class="se">\t</span><span class="s">30</span><span class="se">\t</span><span class="s">0</span><span class="se">\t</span><span class="s">2</span><span class="se">\t</span><span class="s">ACGTA.AATAAAC</span><span class="se">\t</span><span class="s">geTaAafhwqAAf</span><span class="se">\t</span><span class="s">1&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s">&#39;illumina</span><span class="se">\t</span><span class="s">1</span><span class="se">\t</span><span class="s">3</span><span class="se">\t</span><span class="s">35</span><span class="se">\t</span><span class="s">30</span><span class="se">\t</span><span class="s">-30</span><span class="se">\t</span><span class="s">0</span><span class="se">\t</span><span class="s">3</span><span class="se">\t</span><span class="s">CATTTAGGA.TGCA</span><span class="se">\t</span><span class="s">tjflkAFnkKghvM</span><span class="se">\t</span><span class="s">0&#39;</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fh</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">fs</span><span class="p">)</span>
</pre></div>
</div>
<p>To load the sequences into a <tt class="docutils literal"><span class="pre">SequenceCollection</span></tt>, we run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">SequenceCollection</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">SequenceCollection</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">variant</span><span class="o">=</span><span class="s">&#39;illumina1.3&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span>
<span class="go">&lt;SequenceCollection: n=2; mean +/- std length=13.00 +/- 0.00&gt;</span>
</pre></div>
</div>
<p>Note that only two sequences were loaded because the QSeq reader filters out
sequences whose filter field is 0 (unless <tt class="docutils literal"><span class="pre">filter=False</span></tt> is supplied).</p>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<table class="docutils citation" frame="void" id="r162" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[R162]</a></td><td><a class="reference external" href="http://biowulf.nih.gov/apps/CASAVA_UG_15011196B.pdf">http://biowulf.nih.gov/apps/CASAVA_UG_15011196B.pdf</a></td></tr>
</tbody>
</table>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>