
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="Computing and manipulating biological sequence alignments." name="description" />

    <title>Sequence Alignments (skbio.alignment) &#8212; scikit-bio 0.6.3-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=4317e44f"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'alignment';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.4';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <link rel="canonical" href="../../alignment.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.alignment.TabularMSA" href="generated/skbio.alignment.TabularMSA.html" />
    <link rel="prev" title="skbio.sequence.NucleotideMixin.reverse_complement" href="generated/skbio.sequence.NucleotideMixin.reverse_complement.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.3-dev documentation"/>
    <script>document.write(`<img src="_static/logo_inv.svg" class="logo__image only-dark" alt="scikit-bio 0.6.3-dev documentation"/>`);</script>
  
  
</a></div>
    
      <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.TabularMSA.html">skbio.alignment.TabularMSA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.AlignPath.html">skbio.alignment.AlignPath</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.PairAlignPath.html">skbio.alignment.PairAlignPath</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.StripedSmithWaterman.html">skbio.alignment.StripedSmithWaterman</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.AlignmentStructure.html">skbio.alignment.AlignmentStructure</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_ssw.html">skbio.alignment.local_pairwise_align_ssw</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_nucleotide.html">skbio.alignment.global_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_protein.html">skbio.alignment.global_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align.html">skbio.alignment.global_pairwise_align</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_nucleotide.html">skbio.alignment.local_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_protein.html">skbio.alignment.local_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align.html">skbio.alignment.local_pairwise_align</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.make_identity_substitution_matrix.html">skbio.alignment.make_identity_substitution_matrix</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">Sequence...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="sequence-alignments-skbio-alignment">
<span id="module-skbio.alignment"></span><h1>Sequence Alignments (<a class="reference internal" href="#module-skbio.alignment" title="skbio.alignment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code></a>)<a class="headerlink" href="#sequence-alignments-skbio-alignment" title="Link to this heading">#</a></h1>
<p>This module provides functionality for computing and manipulating sequence
alignments. DNA, RNA, and protein sequences can be aligned, as well as
sequences with custom alphabets.</p>
<section id="alignment-structure">
<h2>Alignment structure<a class="headerlink" href="#alignment-structure" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TabularMSA</span></code></a>(sequences[, metadata, ...])</p></td>
<td><p>Store a multiple sequence alignment in tabular (row/column) form.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.AlignPath.html#skbio.alignment.AlignPath" title="skbio.alignment.AlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignPath</span></code></a>(lengths, states, starts)</p></td>
<td><p>Create an alignment path from segment lengths and states.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a>(lengths, states, starts)</p></td>
<td><p>Create a pairwise alignment path from segment lengths and states.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="alignment-algorithms">
<h2>Alignment algorithms<a class="headerlink" href="#alignment-algorithms" title="Link to this heading">#</a></h2>
<p class="rubric">Optimized (i.e., production-ready) algorithms</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.StripedSmithWaterman.html#skbio.alignment.StripedSmithWaterman" title="skbio.alignment.StripedSmithWaterman"><code class="xref py py-obj docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code></a></p></td>
<td><p>Performs a striped (banded) Smith Waterman Alignment.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.AlignmentStructure.html#skbio.alignment.AlignmentStructure" title="skbio.alignment.AlignmentStructure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignmentStructure</span></code></a></p></td>
<td><p>Wraps the result of an alignment c struct so it is accessible to Python</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_ssw.html#skbio.alignment.local_pairwise_align_ssw" title="skbio.alignment.local_pairwise_align_ssw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_ssw</span></code></a>(sequence1, ...)</p></td>
<td><p>Align query and target sequences with Striped Smith-Waterman.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Slow (i.e., educational-purposes only) algorithms</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_nucleotide.html#skbio.alignment.global_pairwise_align_nucleotide" title="skbio.alignment.global_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Globally align nucleotide seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_protein.html#skbio.alignment.global_pairwise_align_protein" title="skbio.alignment.global_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_protein</span></code></a>(seq1, seq2[, ...])</p></td>
<td><p>Globally align pair of protein seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align.html#skbio.alignment.global_pairwise_align" title="skbio.alignment.global_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align</span></code></a>(seq1, seq2, ...[, ...])</p></td>
<td><p>Globally align a pair of seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_nucleotide.html#skbio.alignment.local_pairwise_align_nucleotide" title="skbio.alignment.local_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Locally align exactly two nucleotide seqs with Smith-Waterman.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_protein.html#skbio.alignment.local_pairwise_align_protein" title="skbio.alignment.local_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_protein</span></code></a>(seq1, seq2[, ...])</p></td>
<td><p>Locally align exactly two protein seqs with Smith-Waterman.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align.html#skbio.alignment.local_pairwise_align" title="skbio.alignment.local_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align</span></code></a>(seq1, seq2, ...)</p></td>
<td><p>Locally align exactly two seqs with Smith-Waterman.</p></td>
</tr>
</tbody>
</table>
</div>
<section id="deprecated-functionality">
<h3>Deprecated functionality<a class="headerlink" href="#deprecated-functionality" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.make_identity_substitution_matrix.html#skbio.alignment.make_identity_substitution_matrix" title="skbio.alignment.make_identity_substitution_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_identity_substitution_matrix</span></code></a>(...[, ...])</p></td>
<td><p>Generate substitution matrix where all matches are scored equally.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<section id="alignment-data-structure">
<h3>Alignment data structure<a class="headerlink" href="#alignment-data-structure" title="Link to this heading">#</a></h3>
<p>Load two DNA sequences that have been previously aligned into a <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code>
object, using sequence IDs as the MSA’s index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TabularMSA</span><span class="p">,</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACC--G-GGTA..&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s2">&quot;seq1&quot;</span><span class="p">}),</span>
<span class="gp">... </span>        <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;TCC--G-GGCA..&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s2">&quot;seq2&quot;</span><span class="p">})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">minter</span><span class="o">=</span><span class="s1">&#39;id&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">----------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 13</span>
<span class="go">----------------------</span>
<span class="go">ACC--G-GGTA..</span>
<span class="go">TCC--G-GGCA..</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">Index([&#39;seq1&#39;, &#39;seq2&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
</section>
<section id="using-the-optimized-alignment-algorithm">
<h3>Using the optimized alignment algorithm<a class="headerlink" href="#using-the-optimized-alignment-algorithm" title="Link to this heading">#</a></h3>
<p>Using the convenient <code class="docutils literal notranslate"><span class="pre">local_pairwise_align_ssw</span></code> function:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">local_pairwise_align_ssw</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">local_pairwise_align_ssw</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCCTAACCCCCTTTTCTCAGA&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 30</span>
<span class="go">------------------------------</span>
<span class="go">ACTAAGGCTCTCT-ACCCC----TCTCAGA</span>
<span class="go">ACTAAGGCTC-CTAACCCCCTTTTCTCAGA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">27</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_end_positions</span>
<span class="go">[(0, 24), (0, 28)]</span>
</pre></div>
</div>
<p>Using the <code class="docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code> object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">StripedSmithWaterman</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="n">StripedSmithWaterman</span><span class="p">(</span><span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">query</span><span class="p">(</span><span class="s2">&quot;AAAAAACTCTCTAAACTCACTAAGGCTCTCTACCCCTCTTCAGAGAAGTCGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">Score: 49</span>
<span class="go">Length: 28</span>
</pre></div>
</div>
<p>Using the <code class="docutils literal notranslate"><span class="pre">StripedSmithWaterman</span></code> object for multiple targets in an efficient
way and finding the aligned sequence representations:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">StripedSmithWaterman</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target_sequences</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="s2">&quot;GCTAACTAGGCTCCCTTCTACCCCTCTCAGAGA&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;GCCCAGTAGCTTCCCAATATGAGAGCATCAATTGTAGATCGGGCC&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;TCTATAAGATTCCGCATGCGTTACTTATAAGATGTCTCAACGG&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;TAGAGATTAATTGCCACTGCCAAAATTCTG&quot;</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_sequence</span> <span class="o">=</span> <span class="s2">&quot;ACTAAGGCTCTCTACCCCTCTCAGAGA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="n">StripedSmithWaterman</span><span class="p">(</span><span class="n">query_sequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">target_sequence</span> <span class="ow">in</span> <span class="n">target_sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">alignment</span> <span class="o">=</span> <span class="n">query</span><span class="p">(</span><span class="n">target_sequence</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">ACTAAGGCTC...</span>
<span class="go">ACT-AGGCTC...</span>
<span class="go">Score: 38</span>
<span class="go">Length: 30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned_query_sequence</span><span class="p">)</span>
<span class="go">ACTAAGGCTC---TCTACCCCTCTCAGAGA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned_target_sequence</span><span class="p">)</span>
<span class="go">ACT-AGGCTCCCTTCTACCCCTCTCAGAGA</span>
</pre></div>
</div>
</section>
<section id="using-the-slow-alignment-algorithm">
<h3>Using the slow alignment algorithm<a class="headerlink" href="#using-the-slow-alignment-algorithm" title="Link to this heading">#</a></h3>
<p>scikit-bio also provides pure-Python implementations of Smith-Waterman and
Needleman-Wunsch alignment. These are much slower than the methods described
above, but serve as useful educational examples as they’re simpler to
experiment with. Functions are provided for local and global alignment of
protein and nucleotide sequences. The <code class="docutils literal notranslate"><span class="pre">global*</span></code> and <code class="docutils literal notranslate"><span class="pre">local*</span></code> functions
differ in the underlying algorithm that is applied (<code class="docutils literal notranslate"><span class="pre">global*</span></code> uses Needleman-
Wunsch while <code class="docutils literal notranslate"><span class="pre">local*</span></code> uses Smith-Waterman), and <code class="docutils literal notranslate"><span class="pre">*protein</span></code> and
<code class="docutils literal notranslate"><span class="pre">*nucleotide</span></code> differ in their default scoring of matches, mismatches, and
gaps.</p>
<p>Here we locally align a pair of protein sequences using gap open penalty
of 11 and a gap extend penalty of 1 (in other words, it is much more
costly to open a new gap than extend an existing one).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">local_pairwise_align_protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;HEAGAWGHEE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;PAWHEAE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">local_pairwise_align_protein</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>This returns an <code class="docutils literal notranslate"><span class="pre">skbio.TabularMSA</span></code> object, the alignment score, and start/end
positions of each aligned sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[Protein]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 5</span>
<span class="go">---------------------</span>
<span class="go">AWGHE</span>
<span class="go">AW-HE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">25.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_end_positions</span>
<span class="go">[(4, 8), (1, 4)]</span>
</pre></div>
</div>
<p>Similarly, we can perform global alignment of nucleotide sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.alignment</span> <span class="kn">import</span> <span class="n">global_pairwise_align_nucleotide</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;GCGTGCCTAAGGTATGCAAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s2">&quot;ACGTGCCTAGGTACGCAAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">start_end_positions</span> <span class="o">=</span> <span class="n">global_pairwise_align_nucleotide</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">----------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 20</span>
<span class="go">----------------------</span>
<span class="go">GCGTGCCTAAGGTATGCAAG</span>
<span class="go">ACGTGCCTA-GGTACGCAAG</span>
</pre></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.sequence.NucleotideMixin.reverse_complement.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.sequence.NucleotideMixin.reverse_complement</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.alignment.TabularMSA.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.alignment.TabularMSA</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-structure">Alignment structure</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-algorithms">Alignment algorithms</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deprecated-functionality">Deprecated functionality</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-data-structure">Alignment data structure</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-the-optimized-alignment-algorithm">Using the optimized alignment algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-the-slow-alignment-algorithm">Using the slow alignment algorithm</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.0.2.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>