
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="Computing and manipulating biological sequence alignments." name="description" />

    <title>Sequence Alignments (skbio.alignment) &#8212; scikit-bio 0.7.2-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=ae22e3c7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'alignment';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../alignment.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.alignment.TabularMSA" href="generated/skbio.alignment.TabularMSA.html" />
    <link rel="prev" title="skbio.sequence.NucleotideMixin.reverse_complement" href="generated/skbio.sequence.NucleotideMixin.reverse_complement.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.7.2-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.7.2-dev documentation"/>
    <img src="_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.7.2-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.TabularMSA.html">skbio.alignment.TabularMSA</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.AlignPath.html">skbio.alignment.AlignPath</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.PairAlignPath.html">skbio.alignment.PairAlignPath</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.pair_align.html">skbio.alignment.pair_align</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.pair_align_nucl.html">skbio.alignment.pair_align_nucl</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.pair_align_prot.html">skbio.alignment.pair_align_prot</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.align_score.html">skbio.alignment.align_score</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_nucleotide.html">skbio.alignment.global_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_protein.html">skbio.alignment.global_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align.html">skbio.alignment.global_pairwise_align</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_nucleotide.html">skbio.alignment.local_pairwise_align_nucleotide</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_protein.html">skbio.alignment.local_pairwise_align_protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align.html">skbio.alignment.local_pairwise_align</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Sequence Alignments (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code>)</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="sequence-alignments-skbio-alignment">
<span id="module-skbio.alignment"></span><h1>Sequence Alignments (<a class="reference internal" href="#module-skbio.alignment" title="skbio.alignment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code></a>)<a class="headerlink" href="#sequence-alignments-skbio-alignment" title="Link to this heading">#</a></h1>
<p>This module provides functionality for computing and manipulating sequence
alignments. DNA, RNA, and protein sequences can be aligned, as well as
sequences with or without custom alphabets.</p>
<p>See the <a class="reference internal" href="#alignment-tutorial"><strong>Tutorial</strong></a> section for performing and working with sequence
alignments using scikit-bio.</p>
<section id="alignment-structures">
<h2>Alignment structures<a class="headerlink" href="#alignment-structures" title="Link to this heading">#</a></h2>
<p>Tabular format of aligned sequences.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TabularMSA</span></code></a>(sequences[, metadata, ...])</p></td>
<td><p>Store a multiple sequence alignment in tabular (row/column) form.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Compact format of alignment paths.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.AlignPath.html#skbio.alignment.AlignPath" title="skbio.alignment.AlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignPath</span></code></a>(lengths, states, *[, ranges, ...])</p></td>
<td><p>Store an alignment path between sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a>(lengths, states, *[, ranges, ...])</p></td>
<td><p>Store a pairwise alignment path between two sequences.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="pairwise-alignment">
<h2>Pairwise alignment<a class="headerlink" href="#pairwise-alignment" title="Link to this heading">#</a></h2>
<p>A versatile, efficient and generalizable pairwise alignment algorithm.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.pair_align.html#skbio.alignment.pair_align" title="skbio.alignment.pair_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_align</span></code></a>(seq1, seq2, /[, mode, sub_score, ...])</p></td>
<td><p>Perform pairwise alignment of two sequences.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Convenience wrappers with preset scoring schemes.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.pair_align_nucl.html#skbio.alignment.pair_align_nucl" title="skbio.alignment.pair_align_nucl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_align_nucl</span></code></a>(seq1, seq2, /, **kwargs)</p></td>
<td><p>Align two nucleotide sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.pair_align_prot.html#skbio.alignment.pair_align_prot" title="skbio.alignment.pair_align_prot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_align_prot</span></code></a>(seq1, seq2, /, **kwargs)</p></td>
<td><p>Align two protein sequences.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="alignment-statistics">
<h2>Alignment statistics<a class="headerlink" href="#alignment-statistics" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.align_score.html#skbio.alignment.align_score" title="skbio.alignment.align_score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">align_score</span></code></a>(alignment[, sub_score, ...])</p></td>
<td><p>Calculate the alignment score of two or more aligned sequences.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="deprecated-functionality">
<h2>Deprecated functionality<a class="headerlink" href="#deprecated-functionality" title="Link to this heading">#</a></h2>
<p>Pure Python algorithms (slow; educational-purposes only)</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_nucleotide.html#skbio.alignment.global_pairwise_align_nucleotide" title="skbio.alignment.global_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Globally align nucleotide seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align_protein.html#skbio.alignment.global_pairwise_align_protein" title="skbio.alignment.global_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align_protein</span></code></a>(seq1, seq2[, ...])</p></td>
<td><p>Globally align pair of protein seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.global_pairwise_align.html#skbio.alignment.global_pairwise_align" title="skbio.alignment.global_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_pairwise_align</span></code></a>(seq1, seq2, ...[, ...])</p></td>
<td><p>Globally align a pair of seqs or alignments with Needleman-Wunsch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_nucleotide.html#skbio.alignment.local_pairwise_align_nucleotide" title="skbio.alignment.local_pairwise_align_nucleotide"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_nucleotide</span></code></a>(seq1, seq2)</p></td>
<td><p>Locally align exactly two nucleotide seqs with Smith-Waterman.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align_protein.html#skbio.alignment.local_pairwise_align_protein" title="skbio.alignment.local_pairwise_align_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align_protein</span></code></a>(seq1, seq2[, ...])</p></td>
<td><p>Locally align exactly two protein seqs with Smith-Waterman.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.alignment.local_pairwise_align.html#skbio.alignment.local_pairwise_align" title="skbio.alignment.local_pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">local_pairwise_align</span></code></a>(seq1, seq2, ...)</p></td>
<td><p>Locally align exactly two seqs with Smith-Waterman.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tutorial">
<span id="alignment-tutorial"></span><h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<p>Sequence alignment is the process of arranging biological sequences such that they
resemble each other when positioned together. The alignment reveals the degree of
similarity between sequences, which may indicate evolutionary relationships (i.e.,
“homology”) and/or structural and functional relevance. The alignment process permits
only one operation: shifting characters by inserting gaps (<cite>-</cite>) between them. Shuffling
characters is not permitted.</p>
<section id="quick-start">
<h3>Quick start<a class="headerlink" href="#quick-start" title="Link to this heading">#</a></h3>
<p>To demonstrate this process, let’s create two DNA sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACTACCAGATTACTTACGGATCAGGTACTTGCCAACAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGAAACTACTAGATTACGGATCTTACTTTCCAGCAAGG&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Just by eyeballing, we can observe some local similarities between the sequences,
particularly in the middle sections. However, the beginning and ending portions aren’t
so similar between the two.</p>
<p>Apply <a class="reference internal" href="generated/skbio.alignment.pair_align_nucl.html#skbio.alignment.pair_align_nucl" title="skbio.alignment.pair_align_nucl"><code class="xref py py-func docutils literal notranslate"><span class="pre">pair_align_nucl</span></code></a> to perform pairwise alignment of the two nucleotide
sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">pair_align_nucl</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">pair_align_nucl</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">PairAlignResult(score=22.0, paths=[&lt;PairAlignPath, positions: 44, segments: 7, ...</span>
</pre></div>
</div>
<p>The result contains <code class="docutils literal notranslate"><span class="pre">score</span></code> and <code class="docutils literal notranslate"><span class="pre">paths</span></code>. The latter is a list of optimal alignment
paths calculated by the algorithm. By default, only one path will be returned, which is
all we need in most applications.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&lt;PairAlignPath, positions: 44, segments: 7, CIGAR: &#39;4I13M4D6M2D13M2I&#39;&gt;</span>
</pre></div>
</div>
<p>To obtain aligned sequences, we can call the <a class="reference internal" href="generated/skbio.alignment.AlignPath.to_aligned.html#skbio.alignment.AlignPath.to_aligned" title="skbio.alignment.AlignPath.to_aligned"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_aligned</span></code></a> method of
the path.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_aligned</span><span class="p">((</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="go">[&#39;----ACTACCAGATTACTTACGGATCAGGTACTTGCCAACAA--&#39;,</span>
<span class="go"> &#39;CGAAACTACTAGATTAC----GGATCT--TACTTTCCAGCAAGG&#39;]</span>
</pre></div>
</div>
<p>As we can see, by inserting four consecutive gaps in the appropriate positions, the two
sequences appear more similar, or “aligned”, with one another.</p>
<p>The score measures the goodness of the alignment. The value is dependent on the
sequences and the alignment parameters (discussed below).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">score</span>
<span class="go">22.0</span>
</pre></div>
</div>
<p>Likewise, we can align protein sequences with <a class="reference internal" href="generated/skbio.alignment.pair_align_prot.html#skbio.alignment.pair_align_prot" title="skbio.alignment.pair_align_prot"><code class="xref py py-func docutils literal notranslate"><span class="pre">pair_align_prot</span></code></a>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">pair_align_prot</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pair_align_prot</span><span class="p">(</span><span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;HEAGAWGHEE&quot;</span><span class="p">),</span> <span class="n">Protein</span><span class="p">(</span><span class="s2">&quot;PAWHEAE&quot;</span><span class="p">))</span>
<span class="go">PairAlignResult(score=15.0, paths=[&lt;PairAlignPath, positions: 13, segments: 3, ...</span>
</pre></div>
</div>
</section>
<section id="alignment-algorithm">
<h3>Alignment algorithm<a class="headerlink" href="#alignment-algorithm" title="Link to this heading">#</a></h3>
<p>The alignment algorithm aims to find the optimal alignment path(s) that yield the
highest possible alignment score for two sequences. For decades,
<a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Dynamic_programming">dynamic programming</a> (DP) has been the gold standard for
pairwise sequence alignment, and is widely covered in many bioinformatics textbooks.
scikit-bio also implements this algorithm. <code class="docutils literal notranslate"><span class="pre">pair_align_nucl</span></code> and <code class="docutils literal notranslate"><span class="pre">pair_align_prot</span></code>
are but convenience wrappers for the function <a class="reference internal" href="generated/skbio.alignment.pair_align.html#skbio.alignment.pair_align" title="skbio.alignment.pair_align"><code class="xref py py-func docutils literal notranslate"><span class="pre">pair_align</span></code></a>, which offers multiple
customizable parameters and comprehensive documentation explaining the algorithm and
everything you need to know for using it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">pair_align</span>
</pre></div>
</div>
<p>By referring to the documentation of <code class="docutils literal notranslate"><span class="pre">pair_align</span></code>, we can customize the algorithm’s
behavior to best suit our research objectives. The default mode is <em>global</em> alignment,
as in the Needleman-Wunsch algorithm (which you may recall from textbooks), with both
ends free of gap penalty (a.k.a., “semi-global alignment”, or “overlap alignment”, in
some literature).</p>
<p>Next, we will switch to <em>local</em> alignment, as in the Smith-Waterman algorithm.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;local&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">PairAlignResult(score=13.0, paths=[&lt;PairAlignPath, positions: 25, segments: 3, ...</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_aligned</span><span class="p">((</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="go">[&#39;TTACGGATCAGGTACTTGCCAACAA&#39;, &#39;TTACGGATCT--TACTTTCCAGCAA&#39;]</span>
</pre></div>
</div>
<p>As shown, the alignment is shorter than that of the global alignment, and truncated:
only the highly similar regions in the middle of the two sequences are preserved, while
the less similar regions at both ends are discarded.</p>
<p>We can further tweak the alignment parameters. The wrapper function <code class="docutils literal notranslate"><span class="pre">pair_align_nucl</span></code>
uses parameters that are consistent with NCBI BLASTN’s defaults. To replicate this
behavior using <code class="docutils literal notranslate"><span class="pre">pair_align</span></code>, we can proceed as follows:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">,</span> <span class="n">sub_score</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">),</span> <span class="n">gap_cost</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">PairAlignResult(score=22.0, paths=[&lt;PairAlignPath, positions: 44, segments: 7, ...</span>
</pre></div>
</div>
<p>Here, 2 and -3 in <code class="docutils literal notranslate"><span class="pre">sub_score</span></code> represent match and mismatch scores (added to the
alignment score). 5 and 2 in <code class="docutils literal notranslate"><span class="pre">gap_cost</span></code> represent gap opening and gap extension
penalties (subtracted from the alignment score). The use of two gap penalty values is
known to as <em>affine</em> gap penalty.</p>
<p>Let’s try an alternative setting. The follow parameters uses a nucleotide substitution
matrix “NUC.4.4” to replace match/mismatch scores. This provides more flexibility in
defining the score between pairs of characters. For example, when there are degenerate
characters in the DNA sequences, such as “R” and “Y”, the substitution matrix usually
models the substitution scores more realistically. In addition, this setting uses a
simple <em>linear</em> gap penalty scheme to replace the more sophisticated affine gap penalty
scheme.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">,</span> <span class="n">sub_score</span><span class="o">=</span><span class="s2">&quot;NUC.4.4&quot;</span><span class="p">,</span> <span class="n">gap_cost</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span>
<span class="go">PairAlignResult(score=106.0, paths=[&lt;PairAlignPath, positions: 44, segments: 7, ...</span>
</pre></div>
</div>
<p>Scikit-bio also provides an <a class="reference internal" href="generated/skbio.alignment.align_score.html#skbio.alignment.align_score" title="skbio.alignment.align_score"><code class="xref py py-func docutils literal notranslate"><span class="pre">align_score</span></code></a> function to calculate the alignment
score of an existing alignment. You will need to supply the same parameters to get the
same outcome.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">align_score</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_score</span><span class="p">((</span><span class="n">aln</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)),</span> <span class="n">sub_score</span><span class="o">=</span><span class="s2">&quot;NUC.4.4&quot;</span><span class="p">,</span> <span class="n">gap_cost</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">106.0</span>
</pre></div>
</div>
<p>Pairwise alignment is often taught in bioinformatics courses. By adding
<code class="docutils literal notranslate"><span class="pre">keep_matrices=True</span></code> to the function, the alignment matrix calculated during the DP
process will be retained. This can be particularly useful for educational purposes.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">score</span><span class="p">,</span> <span class="p">(</span><span class="n">path</span><span class="p">,),</span> <span class="p">(</span><span class="n">matrix</span><span class="p">,)</span> <span class="o">=</span> <span class="n">pair_align</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s2">&quot;GATCT&quot;</span><span class="p">,</span> <span class="s2">&quot;GTAC&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;global&quot;</span><span class="p">,</span> <span class="n">free_ends</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">keep_matrices</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span>
<span class="go">array([[  0.,  -2.,  -4.,  -6.,  -8.],</span>
<span class="go">       [ -2.,   1.,  -1.,  -3.,  -5.],</span>
<span class="go">       [ -4.,  -1.,   0.,   0.,  -2.],</span>
<span class="go">       [ -6.,  -3.,   0.,  -1.,  -1.],</span>
<span class="go">       [ -8.,  -5.,  -2.,  -1.,   0.],</span>
<span class="go">       [-10.,  -7.,  -4.,  -3.,  -2.]]...</span>
</pre></div>
</div>
</section>
<section id="alignment-path">
<h3>Alignment path<a class="headerlink" href="#alignment-path" title="Link to this heading">#</a></h3>
<p>A pairwise alignment path is an instance of the <a class="reference internal" href="generated/skbio.alignment.PairAlignPath.html#skbio.alignment.PairAlignPath" title="skbio.alignment.PairAlignPath"><code class="xref py py-class docutils literal notranslate"><span class="pre">PairAlignPath</span></code></a> class, which in
turn is a subclass of the <a class="reference internal" href="generated/skbio.alignment.AlignPath.html#skbio.alignment.AlignPath" title="skbio.alignment.AlignPath"><code class="xref py py-class docutils literal notranslate"><span class="pre">AlignPath</span></code></a> class. This class implements an efficient
data structure to store the “path”, i.e., the events of placing gaps inside sequences.
It uses run-length encoding (RLE) under the hood, and does not store the sequence data,
such that its memory cost is minimum. Refer to the class documentation for details of
this data structure.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">lengths</span>
<span class="go">array([ 4, 13,  4,  6,  2, 13,  2])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">states</span>
<span class="go">array([[1, 0, 2, 0, 2, 0, 1]], dtype=uint8)</span>
</pre></div>
</div>
<p>The path can be represented by a CIGAR string, which is a common format used by many
high-throughput pairwise alignment tools.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">to_cigar</span><span class="p">()</span>
<span class="go">&#39;4I13M4D6M2D13M2I&#39;</span>
</pre></div>
</div>
<p>The letters in the CIGAR string represent the gap-placing events: <code class="docutils literal notranslate"><span class="pre">I</span></code> (insertion):
gap in sequence 1, <code class="docutils literal notranslate"><span class="pre">D</span></code> (deletion): gap in sequence 2, <code class="docutils literal notranslate"><span class="pre">M</span></code> (mis/match): gap in
neither sequence. The numbers represent the length of each segment. As you may have
noticed, the <code class="docutils literal notranslate"><span class="pre">PairAlignPath</span></code> data structure resembles the CIGAR encoding. In this
example, the alignment path contains four consecutive gaps flanking four chunks of
(mis)matched characters.</p>
<p>If we pass the sequences to the <a class="reference internal" href="generated/skbio.alignment.PairAlignPath.to_cigar.html#skbio.alignment.PairAlignPath.to_cigar" title="skbio.alignment.PairAlignPath.to_cigar"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_cigar</span></code></a> method, we will get
<code class="docutils literal notranslate"><span class="pre">=</span></code> (match, i.e., same character) and <code class="docutils literal notranslate"><span class="pre">X</span></code> (mismatch, i.e., different characters)
instead of <code class="docutils literal notranslate"><span class="pre">M</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">to_cigar</span><span class="p">((</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="go">&#39;4I5=1X7=4D5=1X2D5=1X3=1X3=2I&#39;</span>
</pre></div>
</div>
<p>One can construct a <code class="docutils literal notranslate"><span class="pre">PairAlignPath</span></code> from a CIGAR string using the
<a class="reference internal" href="generated/skbio.alignment.PairAlignPath.from_cigar.html#skbio.alignment.PairAlignPath.from_cigar" title="skbio.alignment.PairAlignPath.from_cigar"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_cigar</span></code></a> method. Likewise, this process doesn’t rely on the
sequence data. Therefore, it is suitable for handling many CIGAR strings from large
data files, such as those in SAM/BAM formatted files.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">PairAlignPath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">PairAlignPath</span><span class="o">.</span><span class="n">from_cigar</span><span class="p">(</span><span class="s1">&#39;1I8M2D5M2I&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span>
<span class="go">&lt;PairAlignPath, positions: 18, segments: 5, CIGAR: &#39;1I8M2D5M2I&#39;&gt;</span>
</pre></div>
</div>
<p>This method can also be used to parse CIGAR-formatted alignment results of external
sequence alignment tools, such as <strong>Parasail</strong>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">parasail</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">parasail</span><span class="o">.</span><span class="n">nw_trace</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">PairAlignPath</span><span class="o">.</span><span class="n">from_cigar</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">cigar</span><span class="o">.</span><span class="n">decode</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> is a more generalized data structure that supports an arbitrary number
of sequences. For example, we can directly construct a path of three aligned sequences
using the <a class="reference internal" href="generated/skbio.alignment.AlignPath.from_aligned.html#skbio.alignment.AlignPath.from_aligned" title="skbio.alignment.AlignPath.from_aligned"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_aligned</span></code></a> method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">AlignPath</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">AlignPath</span><span class="o">.</span><span class="n">from_aligned</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s1">&#39;CGTCGTGC&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;CA--GT-C&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;CGTCGT-T&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span>
<span class="go">&lt;AlignPath, sequences: 3, positions: 8, segments: 5&gt;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">lengths</span>
<span class="go">array([2, 2, 2, 1, 1])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">.</span><span class="n">states</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">array([0, 2, 0, 6, 0], dtype=uint8)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">AlignPath</span></code> (as well as <code class="docutils literal notranslate"><span class="pre">PairAlignPath</span></code>) class serves as a central hub for
interacting with different alignment formats and tools. For example, the
<a class="reference internal" href="generated/skbio.alignment.AlignPath.from_coordinates.html#skbio.alignment.AlignPath.from_coordinates" title="skbio.alignment.AlignPath.from_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_coordinates</span></code></a> method can parse <strong>BioPython</strong>’s alignment results
(and vice versa with <a class="reference internal" href="generated/skbio.alignment.AlignPath.to_coordinates.html#skbio.alignment.AlignPath.to_coordinates" title="skbio.alignment.AlignPath.to_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_coordinates</span></code></a>).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Align</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">AlignPath</span><span class="o">.</span><span class="n">from_coordinates</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="generated/skbio.alignment.AlignPath.from_indices.html#skbio.alignment.AlignPath.from_indices" title="skbio.alignment.AlignPath.from_indices"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_indices</span></code></a> method can parse <strong>Biotite</strong>’s alignment results
(and vice versa with <a class="reference internal" href="generated/skbio.alignment.AlignPath.to_indices.html#skbio.alignment.AlignPath.to_indices" title="skbio.alignment.AlignPath.to_indices"><code class="xref py py-meth docutils literal notranslate"><span class="pre">to_indices</span></code></a>).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">biotite.sequence.align</span><span class="w"> </span><span class="kn">import</span> <span class="n">align_optimal</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">align_optimal</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">AlignPath</span><span class="o">.</span><span class="n">from_indices</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">trace</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="tabular-alignment">
<h3>Tabular alignment<a class="headerlink" href="#tabular-alignment" title="Link to this heading">#</a></h3>
<p>In addition to alignment paths, scikit-bio provides <a class="reference internal" href="generated/skbio.alignment.TabularMSA.html#skbio.alignment.TabularMSA" title="skbio.alignment.TabularMSA"><code class="xref py py-class docutils literal notranslate"><span class="pre">TabularMSA</span></code></a>, a tabular data
structure that hosts aligned sequences (“MSA” stands for multiple sequence alignment).</p>
<p>One can convert an alignment path and the original (unaligned) sequences into a
<code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object using the <a class="reference internal" href="generated/skbio.alignment.TabularMSA.from_path_seqs.html#skbio.alignment.TabularMSA.from_path_seqs" title="skbio.alignment.TabularMSA.from_path_seqs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">from_path_seqs</span></code></a> method (and vice
versa with <a class="reference internal" href="generated/skbio.alignment.AlignPath.from_tabular.html#skbio.alignment.AlignPath.from_tabular" title="skbio.alignment.AlignPath.from_tabular"><code class="xref py py-meth docutils literal notranslate"><span class="pre">AlignPath.from_tabular</span></code></a>).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.alignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">TabularMSA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span><span class="p">,</span> <span class="p">(</span><span class="n">path</span><span class="p">,),</span> <span class="n">_</span> <span class="o">=</span> <span class="n">pair_align_nucl</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">from_path_seqs</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">--------------------------------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 2</span>
<span class="go">    position count: 44</span>
<span class="go">--------------------------------------------</span>
<span class="go">----ACTACCAGATTACTTACGGATCAGGTACTTGCCAACAA--</span>
<span class="go">CGAAACTACTAGATTAC----GGATCT--TACTTTCCAGCAAGG</span>
</pre></div>
</div>
<p>Alternatively, one can read from a multi-FASTA file of aligned sequences:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;input.fasta&#39;</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">)</span>
</pre></div>
</div>
<p>Or directly from aligned sequence objects:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">TabularMSA</span><span class="p">([</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGTCGTGC&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CA--GT-C&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;CGTCGT-T&#39;</span><span class="p">),</span>
<span class="gp">... </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span>
<span class="go">TabularMSA[DNA]</span>
<span class="go">---------------------</span>
<span class="go">Stats:</span>
<span class="go">    sequence count: 3</span>
<span class="go">    position count: 8</span>
<span class="go">---------------------</span>
<span class="go">CGTCGTGC</span>
<span class="go">CA--GT-C</span>
<span class="go">CGTCGT-T</span>
</pre></div>
</div>
<p>A <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> object is iterally like a table. If you are also familiar with pandas
DataFrame, you can apply similar indexing methods to it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">index</span>
<span class="go">RangeIndex(start=0, stop=3, step=1)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 75.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 CGTCGTGC</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">Sequence</span>
<span class="go">-------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">-------------</span>
<span class="go">0 GAG</span>
</pre></div>
</div>
<p>Like <a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <code class="docutils literal notranslate"><span class="pre">TabularMSA</span></code> can be annotated with metadata.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;test alignment&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">positional_metadata</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;identity&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">80</span><span class="p">,</span> <span class="mi">85</span><span class="p">,</span> <span class="mi">88</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">95</span><span class="p">,</span> <span class="mi">92</span><span class="p">,</span> <span class="mi">86</span><span class="p">]}</span>
</pre></div>
</div>
<p>The class provides some utilities for downstream analysis. For example, to obtain a
consensus sequence of the alignment:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">consensus</span><span class="p">()</span>
<span class="go">DNA</span>
<span class="go">------------------------------</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;identity&#39;: &lt;dtype: int64&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: True</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 71.43%</span>
<span class="go">------------------------------</span>
<span class="go">0 CGTCGT-C</span>
</pre></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.sequence.NucleotideMixin.reverse_complement.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.sequence.NucleotideMixin.reverse_complement</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.alignment.TabularMSA.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.alignment.TabularMSA</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-structures">Alignment structures</a><ul class="nav section-nav flex-column">
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pairwise-alignment">Pairwise alignment</a><ul class="nav section-nav flex-column">
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-statistics">Alignment statistics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#deprecated-functionality">Deprecated functionality</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-start">Quick start</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-algorithm">Alignment algorithm</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-path">Alignment path</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tabular-alignment">Tabular alignment</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2025, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>