
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="Analyzing biodiversity of communities -- groups of organisms living in the same area." name="description" />

    <title>Community Diversity (skbio.diversity) &#8212; scikit-bio 0.6.3-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=4317e44f"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'diversity';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../diversity.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Alpha diversity measures (skbio.diversity.alpha)" href="generated/skbio.diversity.alpha.html" />
    <link rel="prev" title="skbio.tree.majority_rule" href="generated/skbio.tree.majority_rule.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.3-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.3-dev documentation"/>
    <img src="_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.6.3-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="sequence.html">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.alpha.html">Alpha diversity measures (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity.alpha</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.get_alpha_diversity_metrics.html">skbio.diversity.get_alpha_diversity_metrics</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.alpha_diversity.html">skbio.diversity.alpha_diversity</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.beta.html">Beta diversity measures (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity.beta</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.get_beta_diversity_metrics.html">skbio.diversity.get_beta_diversity_metrics</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.beta_diversity.html">skbio.diversity.beta_diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.partial_beta_diversity.html">skbio.diversity.partial_beta_diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.diversity.block_beta_diversity.html">skbio.diversity.block_beta_diversity</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Community Diversity (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code>)</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="community-diversity-skbio-diversity">
<span id="module-skbio.diversity"></span><h1>Community Diversity (<a class="reference internal" href="#module-skbio.diversity" title="skbio.diversity"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity</span></code></a>)<a class="headerlink" href="#community-diversity-skbio-diversity" title="Link to this heading">#</a></h1>
<p>This module provides functionality for analyzing biodiversity of communities
– groups of organisms living in the same area. It implements various metrics
of alpha (within-community) and beta (between-community) diversity, and
provides “driver functions” for computing alpha and beta diversity for an
entire data table. Additional utilities are provided to support discovery of
available diversity metrics. While diversity metrics were originally designed
to study biological communities, they can be generalized to the analysis of
various biological data types.</p>
<section id="alpha-diversity">
<h2>Alpha diversity<a class="headerlink" href="#alpha-diversity" title="Link to this heading">#</a></h2>
<p class="rubric">Alpha diversity metrics</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.diversity.alpha.html#module-skbio.diversity.alpha" title="skbio.diversity.alpha"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alpha</span></code></a></p></td>
<td><p>Alpha diversity measures (skbio.diversity.alpha)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.diversity.get_alpha_diversity_metrics.html#skbio.diversity.get_alpha_diversity_metrics" title="skbio.diversity.get_alpha_diversity_metrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_alpha_diversity_metrics</span></code></a>()</p></td>
<td><p>List scikit-bio's alpha diversity metrics.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Driver function</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.diversity.alpha_diversity.html#skbio.diversity.alpha_diversity" title="skbio.diversity.alpha_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alpha_diversity</span></code></a>(metric, counts[, ids, validate])</p></td>
<td><p>Compute alpha diversity for one or more samples.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="beta-diversity">
<h2>Beta diversity<a class="headerlink" href="#beta-diversity" title="Link to this heading">#</a></h2>
<p class="rubric">Beta diversity metrics</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.diversity.beta.html#module-skbio.diversity.beta" title="skbio.diversity.beta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">beta</span></code></a></p></td>
<td><p>Beta diversity measures (skbio.diversity.beta)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.diversity.get_beta_diversity_metrics.html#skbio.diversity.get_beta_diversity_metrics" title="skbio.diversity.get_beta_diversity_metrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_beta_diversity_metrics</span></code></a>()</p></td>
<td><p>List scikit-bio's beta diversity metrics.</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Driver functions</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.diversity.beta_diversity.html#skbio.diversity.beta_diversity" title="skbio.diversity.beta_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">beta_diversity</span></code></a>(metric, counts[, ids, ...])</p></td>
<td><p>Compute distances between all pairs of samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.diversity.partial_beta_diversity.html#skbio.diversity.partial_beta_diversity" title="skbio.diversity.partial_beta_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">partial_beta_diversity</span></code></a>(metric, counts, ids, ...)</p></td>
<td><p>Compute distances only between specified ID pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.diversity.block_beta_diversity.html#skbio.diversity.block_beta_diversity" title="skbio.diversity.block_beta_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">block_beta_diversity</span></code></a>(metric, counts, ids[, ...])</p></td>
<td><p>Perform a block-decomposition beta diversity calculation.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>A community (i.e., sample) is represented by a vector of frequencies of taxa
within the sample. The term “taxon” (plural: “taxa”) describes a group of
biologically related organisms that constitute a unit in the community. Taxa
are usually defined at a uniform taxonomic rank, such as species, genus or
family. In community ecology, taxon is usually referred to as “species”
(singular = plural), but its definition is not limited to species as a
taxonomic rank. The term “taxonomic group” is a synonym of taxon in many
situations.</p>
<p>In scikit-bio, the term “taxon/taxa” is used very loosely, as these can in
practice represent diverse feature types including organisms, genes, and
metabolites. The term “sample” is also loosely defined for these purposes.
These are intended to represent a single unit of sampling, and as such what a
single sample represents can vary widely. For example, in a microbiome survey,
these could represent all 16S rRNA gene sequences from a single oral swab. In
a comparative genomics study on the other hand, a sample could represent an
individual organism’s genome.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Previous versions of scikit-bio referred to taxon as operational taxonomic
unit (OTU), a historically important term in microbiome research. However,
as the field advances and the research targets diverge (e.g., amplicon
sequence variant, or ASV), a more generic term such as “taxon” becomes more
appropriate. Therefore, the term OTU was replaced by taxon in scikit-bio
0.6.0.</p>
</div>
<p>Each frequency in a given vector represents the number of individuals observed
for a particular taxon. We will refer to the frequencies associated with a
single sample as a <em>counts vector</em> or <code class="docutils literal notranslate"><span class="pre">counts</span></code> throughout the documentation.
Counts vectors are <cite>array_like</cite>: anything that can be converted into a 1-D
numpy array is acceptable input. For example, you can provide a numpy array or
a native Python list and the results will be identical.</p>
<p>The driver functions <a class="reference internal" href="generated/skbio.diversity.alpha_diversity.html#skbio.diversity.alpha_diversity" title="skbio.diversity.alpha_diversity"><code class="xref py py-func docutils literal notranslate"><span class="pre">alpha_diversity()</span></code></a> and <a class="reference internal" href="generated/skbio.diversity.beta_diversity.html#skbio.diversity.beta_diversity" title="skbio.diversity.beta_diversity"><code class="xref py py-func docutils literal notranslate"><span class="pre">beta_diversity()</span></code></a> are
designed to compute alpha diversity for one or more samples, or beta diversity
for one or more pairs of samples. The driver functions accept a matrix
containing vectors of frequencies of taxa within each sample. Each row in the
matrix represents a single sample’s count vector, so that rows represent
samples and columns represent taxa.</p>
<p>Some diversity metrics incorporate relationships between the taxa in their
computation through reference to a phylogenetic tree. These metrics
additionally take a <code class="xref py py-class docutils literal notranslate"><span class="pre">skbio.TreeNode</span></code> object and a list of taxa mapping
the values in the counts vector to tips in the tree.</p>
<p>The driver functions are optimized so that computing a diversity metric more
than one time (i.e., for more than one sample for alpha diversity metrics, or
more than one pair of samples for beta diversity metrics) is often much faster
than repeated calls to the metric. For this reason, the driver functions take
matrices of counts vectors rather than a single counts vector for alpha
diversity metrics or two counts vectors for beta diversity metrics. The
<code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code> driver function will thus compute alpha diversity for all
counts vectors in the matrix, and the <code class="docutils literal notranslate"><span class="pre">beta_diversity</span></code> driver function will
compute beta diversity for all pairs of counts vectors in the matrix.</p>
<section id="input-validation">
<h3>Input validation<a class="headerlink" href="#input-validation" title="Link to this heading">#</a></h3>
<p>The driver functions perform validation of input by default. Validation can be
slow so it is possible to disable this step by passing <code class="docutils literal notranslate"><span class="pre">validate=False</span></code>. This
can be dangerous however. If invalid input is encountered when validation is
disabled it can result in difficult-to-interpret error messages or incorrect
results. We therefore recommend that users are careful to ensure that their
input data is valid before disabling validation.</p>
<p>The conditions that the driver functions validate follow. If disabling
validation, users should be confident that these conditions are met.</p>
<ul class="simple">
<li><p>the data in the counts vectors can be safely cast to integers</p></li>
<li><p>there are no negative values in the counts vectors</p></li>
<li><p>each counts vector is one dimensional</p></li>
<li><p>the counts matrix is two dimensional</p></li>
<li><p>all counts vectors are of equal length</p></li>
</ul>
<p>Additionally, if a phylogenetic diversity metric is being computed, the
following conditions are also confirmed:</p>
<ul class="simple">
<li><p>the provided taxa are all unique</p></li>
<li><p>the length of each counts vector is equal to the number of taxa</p></li>
<li><p>the provided tree is rooted</p></li>
<li><p>the tree has more than one node</p></li>
<li><p>all nodes in the provided tree except for the root node have branch lengths</p></li>
<li><p>all tip names in the provided tree are unique</p></li>
<li><p>all provided taxa correspond to tip names in the provided tree</p></li>
</ul>
</section>
<section id="count-vectors">
<h3>Count vectors<a class="headerlink" href="#count-vectors" title="Link to this heading">#</a></h3>
<p>There are different ways that count vectors are represented in the ecological
literature and in related software. The diversity measures provided here
<em>always</em> assume that the input contains abundance data: each count represents
the number of individuals observed for a particular taxon in the sample. For
example, if you have two taxa, where three individuals were observed from the
first taxon and only a single individual was observed from the second taxon,
you could represent this data in the following forms (among others).</p>
<p>As a vector of counts. This is the expected type of input for the diversity
measures in this module. There are 3 individuals from the taxon at index 0,
and 1 individual from the taxon at index 1:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>As a vector of indices. The taxon at index 0 is observed 3 times, while the
taxon at index 1 is observed 1 time:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">indices</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>As a vector of frequencies. We have 1 taxon that is a singleton and 1 taxon
that is a tripleton. We do not have any 0-tons or doubletons:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">frequencies</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>Always use the first representation (a counts vector) with this module.</p>
</section>
<section id="specifying-a-diversity-metric">
<h3>Specifying a diversity metric<a class="headerlink" href="#specifying-a-diversity-metric" title="Link to this heading">#</a></h3>
<p>The driver functions take a parameter, <code class="docutils literal notranslate"><span class="pre">metric</span></code>, that specifies which
diversity metric should be applied. The value that you provide for <code class="docutils literal notranslate"><span class="pre">metric</span></code>
can be either a string (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;faith_pd&quot;</span></code>) or a function (e.g.,
<code class="docutils literal notranslate"><span class="pre">skbio.diversity.alpha.faith_pd</span></code>). The metric should generally be passed as a
string, as this often uses an optimized version of the metric. For example,
passing  <code class="docutils literal notranslate"><span class="pre">metric=&quot;faith_pd&quot;</span></code> (a string) to <code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code> will be tens
of times faster than passing <code class="docutils literal notranslate"><span class="pre">metric=skbio.diversity.alpha.faith_pd</span></code> (a
function) when computing Faith’s PD on about 100 samples.  Similarly, passing
<code class="docutils literal notranslate"><span class="pre">metric=&quot;unweighted_unifrac&quot;</span></code> (a string) will be hundreds of times
faster than passing <code class="docutils literal notranslate"><span class="pre">metric=skbio.diversity.beta.unweighted_unifrac</span></code> (a
function) when computing unweighted UniFrac on about 100 samples. The latter
may be faster if computing only one alpha or beta diversity value, but in
these cases the run times will likely be so small that the difference will be
negligible. <strong>We therefore recommend that you always pass the metric as a
string when possible.</strong></p>
<p>Passing a metric as a string will not be possible if the metric you’d like to
run is not one that scikit-bio knows about. This might be the case, for
example, if you’re applying a custom metric that you’ve developed. To discover
the metric names that scikit-bio knows about as strings that can be passed as
<code class="docutils literal notranslate"><span class="pre">metric</span></code> to <code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code> or <code class="docutils literal notranslate"><span class="pre">beta_diversity</span></code>, you can call
<code class="docutils literal notranslate"><span class="pre">get_alpha_diversity_metrics</span></code> or <code class="docutils literal notranslate"><span class="pre">get_beta_diversity_metrics</span></code>,
respectively. These functions return lists of alpha and beta diversity metrics
that are implemented in scikit-bio. There may be additional metrics that can be
passed as strings which won’t be listed here, such as those implemented in
<code class="docutils literal notranslate"><span class="pre">scipy.spatial.distance.pdist</span></code>.</p>
</section>
</section>
<section id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<p>Create a matrix containing 6 samples (rows) and 7 taxa (columns):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">23</span><span class="p">,</span> <span class="mi">64</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">42</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">44</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">19</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s1">&#39;ABCDEF&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>First, we’ll compute <span class="math notranslate nohighlight">\(S_{obs}\)</span>, an alpha diversity metric, for each
sample using the <code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code> driver function:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.diversity</span> <span class="kn">import</span> <span class="n">alpha_diversity</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adiv_sobs</span> <span class="o">=</span> <span class="n">alpha_diversity</span><span class="p">(</span><span class="s1">&#39;sobs&#39;</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">ids</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adiv_sobs</span> 
<span class="go">A    5</span>
<span class="go">B    5</span>
<span class="go">C    4</span>
<span class="go">D    4</span>
<span class="go">E    5</span>
<span class="go">F    3</span>
<span class="go">dtype: ...</span>
</pre></div>
</div>
<p>Next we’ll compute Faith’s PD on the same samples. Since this is a
phylogenetic diversity metric, we’ll first create a tree and an ordered
list of taxa.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">TreeNode</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span>
<span class="gp">... </span>                     <span class="s1">&#39;(((((U1:0.5,U2:0.5):0.5,U3:1.0):1.0):0.0,&#39;</span>
<span class="gp">... </span>                     <span class="s1">&#39;(U4:0.75,(U5:0.5,(U6:0.5,U7:0.5):0.5):&#39;</span>
<span class="gp">... </span>                     <span class="s1">&#39;0.5):1.25):0.0)root;&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">taxa</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;U1&#39;</span><span class="p">,</span> <span class="s1">&#39;U2&#39;</span><span class="p">,</span> <span class="s1">&#39;U3&#39;</span><span class="p">,</span> <span class="s1">&#39;U4&#39;</span><span class="p">,</span> <span class="s1">&#39;U5&#39;</span><span class="p">,</span> <span class="s1">&#39;U6&#39;</span><span class="p">,</span> <span class="s1">&#39;U7&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adiv_faith_pd</span> <span class="o">=</span> <span class="n">alpha_diversity</span><span class="p">(</span><span class="s1">&#39;faith_pd&#39;</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">ids</span><span class="o">=</span><span class="n">ids</span><span class="p">,</span>
<span class="gp">... </span>                                <span class="n">taxa</span><span class="o">=</span><span class="n">taxa</span><span class="p">,</span> <span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adiv_faith_pd</span>
<span class="go">A    6.75</span>
<span class="go">B    7.00</span>
<span class="go">C    6.25</span>
<span class="go">D    5.75</span>
<span class="go">E    6.75</span>
<span class="go">F    5.50</span>
<span class="go">dtype: float64</span>
</pre></div>
</div>
<p>Now we’ll compute Bray-Curtis distances, a beta diversity metric, between
all pairs of samples. Notice that the <code class="docutils literal notranslate"><span class="pre">data</span></code> and <code class="docutils literal notranslate"><span class="pre">ids</span></code> parameters
provided to <code class="docutils literal notranslate"><span class="pre">beta_diversity</span></code> are the same as those provided to
<code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.diversity</span> <span class="kn">import</span> <span class="n">beta_diversity</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bc_dm</span> <span class="o">=</span> <span class="n">beta_diversity</span><span class="p">(</span><span class="s2">&quot;braycurtis&quot;</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">ids</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bc_dm</span><span class="p">)</span>
<span class="go">6x6 distance matrix</span>
<span class="go">IDs:</span>
<span class="go">&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;</span>
<span class="go">Data:</span>
<span class="go">[[ 0.          0.78787879  0.86666667  0.30927835  0.85714286  0.81521739]</span>
<span class="go"> [ 0.78787879  0.          0.78142077  0.86813187  0.75        0.1627907 ]</span>
<span class="go"> [ 0.86666667  0.78142077  0.          0.87709497  0.09392265  0.71597633]</span>
<span class="go"> [ 0.30927835  0.86813187  0.87709497  0.          0.87777778  0.89285714]</span>
<span class="go"> [ 0.85714286  0.75        0.09392265  0.87777778  0.          0.68235294]</span>
<span class="go"> [ 0.81521739  0.1627907   0.71597633  0.89285714  0.68235294  0.        ]]</span>
</pre></div>
</div>
<p>Next, we’ll compute weighted UniFrac distances between all pairs of samples.
Because weighted UniFrac is a phylogenetic beta diversity metric, we’ll need
to pass the <code class="docutils literal notranslate"><span class="pre">skbio.TreeNode</span></code> and list of taxa that we created above.
Again, these are the same values that were provided to <code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">wu_dm</span> <span class="o">=</span> <span class="n">beta_diversity</span><span class="p">(</span><span class="s2">&quot;weighted_unifrac&quot;</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">ids</span><span class="p">,</span> <span class="n">tree</span><span class="o">=</span><span class="n">tree</span><span class="p">,</span>
<span class="gp">... </span>                       <span class="n">taxa</span><span class="o">=</span><span class="n">taxa</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">wu_dm</span><span class="p">)</span>
<span class="go">6x6 distance matrix</span>
<span class="go">IDs:</span>
<span class="go">&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;</span>
<span class="go">Data:</span>
<span class="go">[[ 0.          2.77549923  3.82857143  0.42512039  3.8547619   3.10937312]</span>
<span class="go"> [ 2.77549923  0.          2.26433692  2.98435423  2.24270353  0.46774194]</span>
<span class="go"> [ 3.82857143  2.26433692  0.          3.95224719  0.16025641  1.86111111]</span>
<span class="go"> [ 0.42512039  2.98435423  3.95224719  0.          3.98796148  3.30870431]</span>
<span class="go"> [ 3.8547619   2.24270353  0.16025641  3.98796148  0.          1.82967033]</span>
<span class="go"> [ 3.10937312  0.46774194  1.86111111  3.30870431  1.82967033  0.        ]]</span>
</pre></div>
</div>
<p>Next we’ll do some work with these beta diversity distance matrices. First,
we’ll determine if the UniFrac and Bray-Curtis distance matrices are
significantly correlated by computing the Mantel correlation between them.
Then we’ll determine if the p-value is significant based on an alpha of
0.05.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.stats.distance</span> <span class="kn">import</span> <span class="n">mantel</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span><span class="p">,</span> <span class="n">p_value</span><span class="p">,</span> <span class="n">n</span> <span class="o">=</span> <span class="n">mantel</span><span class="p">(</span><span class="n">wu_dm</span><span class="p">,</span> <span class="n">bc_dm</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
<span class="go">0.922404392093</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p_value</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Next, we’ll perform principal coordinates analysis (PCoA) on our weighted
UniFrac distance matrix.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.stats.ordination</span> <span class="kn">import</span> <span class="n">pcoa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wu_pc</span> <span class="o">=</span> <span class="n">pcoa</span><span class="p">(</span><span class="n">wu_dm</span><span class="p">)</span>
</pre></div>
</div>
<p>PCoA plots are only really interesting in the context of sample metadata, so
let’s define some before we visualize these results.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;gut&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;skin&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;tongue&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;gut&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;tongue&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">[</span><span class="s1">&#39;skin&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">]],</span>
<span class="gp">... </span>   <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;F&#39;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;body_site&#39;</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span>
<span class="go">  body_site subject</span>
<span class="go">A       gut      s1</span>
<span class="go">B      skin      s1</span>
<span class="go">C    tongue      s1</span>
<span class="go">D       gut      s2</span>
<span class="go">E    tongue      s2</span>
<span class="go">F      skin      s2</span>
</pre></div>
</div>
<p>Now let’s plot our PCoA results, coloring each sample by the subject it
was taken from:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">wu_pc</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">sample_md</span><span class="p">,</span> <span class="s1">&#39;subject&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">axis_labels</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;PC 1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC 2&#39;</span><span class="p">,</span> <span class="s1">&#39;PC 3&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Samples colored by subject&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span>
<span class="gp">... </span><span class="p">)</span>  
</pre></div>
</div>
<figure class="align-default">
<img alt="_images/diversity-1.png" class="plot-directive" src="_images/diversity-1.png" />
</figure>
<p>We don’t see any clustering/grouping of samples. If we were to instead color
the samples by the body site they were taken from, we see that the samples
from the same body site (those that are colored the same) appear to be
closer to one another in the 3-D space then they are to samples from
other body sites.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">wu_pc</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">sample_md</span><span class="p">,</span> <span class="s1">&#39;body_site&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">axis_labels</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;PC 1&#39;</span><span class="p">,</span> <span class="s1">&#39;PC 2&#39;</span><span class="p">,</span> <span class="s1">&#39;PC 3&#39;</span><span class="p">),</span>
<span class="gp">... </span>    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Samples colored by body site&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">50</span>
<span class="gp">... </span><span class="p">)</span>  
</pre></div>
</div>
<figure class="align-default">
<img alt="_images/diversity-2_00.png" class="plot-directive" src="_images/diversity-2_00.png" />
</figure>
<figure class="align-default">
<img alt="_images/diversity-2_01.png" class="plot-directive" src="_images/diversity-2_01.png" />
</figure>
<p>Ordination techniques, such as PCoA, are useful for exploratory analysis.
The next step is to quantify the strength of the grouping/clustering that we
see in ordination plots. There are many statistical methods available to
accomplish this; many operate on distance matrices. Let’s use ANOSIM to
quantify the strength of the clustering we see in the ordination plots
above, using our weighted UniFrac distance matrix and sample metadata.</p>
<p>First test the grouping of samples by subject:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.stats.distance</span> <span class="kn">import</span> <span class="n">anosim</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">anosim</span><span class="p">(</span><span class="n">wu_dm</span><span class="p">,</span> <span class="n">sample_md</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;subject&#39;</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">999</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;test statistic&#39;</span><span class="p">]</span>
<span class="go">-0.33333333333333331</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span>
<span class="go">False</span>
</pre></div>
</div>
<p>The negative value of ANOSIM’s R statistic indicates anti-clustering and the
p-value is insignificant at an alpha of 0.1.</p>
<p>Now let’s test the grouping of samples by body site:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">anosim</span><span class="p">(</span><span class="n">wu_dm</span><span class="p">,</span> <span class="n">sample_md</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;body_site&#39;</span><span class="p">,</span> <span class="n">permutations</span><span class="o">=</span><span class="mi">999</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;test statistic&#39;</span><span class="p">]</span>
<span class="go">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The R statistic indicates strong separation of samples based on body site.
The p-value is significant at an alpha of 0.1.</p>
<p>We can also explore the alpha diversity in the context of sample metadata.
To do this, let’s add the observed richness and Faith’s PD metrics to our
sample metadata. This is straight-forward because <code class="docutils literal notranslate"><span class="pre">alpha_diversity</span></code>
returns a Pandas <code class="docutils literal notranslate"><span class="pre">Series</span></code> object, and we’re representing our sample
metadata in a Pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span><span class="p">[</span><span class="s1">&#39;Obs. richness&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adiv_sobs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span><span class="p">[</span><span class="s1">&#39;Faith PD&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adiv_faith_pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span>
<span class="go">  body_site subject  Obs. richness  Faith PD</span>
<span class="go">A       gut      s1              5      6.75</span>
<span class="go">B      skin      s1              5      7.00</span>
<span class="go">C    tongue      s1              4      6.25</span>
<span class="go">D       gut      s2              4      5.75</span>
<span class="go">E    tongue      s2              5      6.75</span>
<span class="go">F      skin      s2              3      5.50</span>
</pre></div>
</div>
<p>We can investigate these alpha diversity data in the context of our metadata
categories. For example, we can generate boxplots showing Faith PD by body
site.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">sample_md</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">column</span><span class="o">=</span><span class="s1">&#39;Faith PD&#39;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;body_site&#39;</span><span class="p">)</span>  
</pre></div>
</div>
<figure class="align-default">
<img alt="_images/diversity-3_00.png" class="plot-directive" src="_images/diversity-3_00.png" />
</figure>
<figure class="align-default">
<img alt="_images/diversity-3_01.png" class="plot-directive" src="_images/diversity-3_01.png" />
</figure>
<figure class="align-default">
<img alt="_images/diversity-3_02.png" class="plot-directive" src="_images/diversity-3_02.png" />
</figure>
<p>We can also compute Spearman correlations between all pairs of columns in
this <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code>. Since our alpha diversity metrics are the only two
numeric columns (and thus the only columns for which Spearman correlation
is relevant), this will give us a symmetric 2x2 correlation matrix.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample_md</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;spearman&quot;</span><span class="p">,</span> <span class="n">numeric_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">               Obs. richness  Faith PD</span>
<span class="go">Obs. richness       1.000000  0.939336</span>
<span class="go">Faith PD            0.939336  1.000000</span>
</pre></div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.tree.majority_rule.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.tree.majority_rule</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.diversity.alpha.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Alpha diversity measures (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.diversity.alpha</span></code>)</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-diversity">Alpha diversity</a><ul class="nav section-nav flex-column">
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beta-diversity">Beta diversity</a><ul class="nav section-nav flex-column">
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">Introduction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#input-validation">Input validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#count-vectors">Count vectors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-a-diversity-metric">Specifying a diversity metric</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>