
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Biological Sequences (skbio.sequence) &#8212; scikit-bio 0.6.2-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=87e54e7c" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=be67db09"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'sequence';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.15.4';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner = false;
        </script>
    <link rel="canonical" href="../../sequence.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.sequence.Sequence" href="generated/skbio.sequence.Sequence.html" />
    <link rel="prev" title="skbio.io.registry.create_format" href="generated/skbio.io.registry.create_format.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.6.2-dev documentation"/>
    <script>document.write(`<img src="_static/logo_inv.svg" class="logo__image only-dark" alt="scikit-bio 0.6.2-dev documentation"/>`);</script>
  
  
</a></div>
    
      <div class="navbar-item">
<script>
document.write(`
  <div class="version-switcher__container dropdown">
    <button id="pst-version-switcher-button-2"
      type="button"
      class="version-switcher__button btn btn-sm dropdown-toggle"
      data-bs-toggle="dropdown"
      aria-haspopup="listbox"
      aria-controls="pst-version-switcher-list-2"
      aria-label="Version switcher list"
    >
      Choose version  <!-- this text may get changed later by javascript -->
      <span class="caret"></span>
    </button>
    <div id="pst-version-switcher-list-2"
      class="version-switcher__menu dropdown-menu list-group-flush py-0"
      role="listbox" aria-labelledby="pst-version-switcher-button-2">
      <!-- dropdown will be populated by javascript on page load -->
    </div>
  </div>
`);
</script></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.Sequence.html">skbio.sequence.Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html">skbio.sequence.GrammaredSequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.DNA.html">skbio.sequence.DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.RNA.html">skbio.sequence.RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.Protein.html">skbio.sequence.Protein</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.GeneticCode.html">skbio.sequence.GeneticCode</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html">skbio.sequence.SubstitutionMatrix</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.distance.html">Sequence distance metrics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.distance</span></code>)</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page">Biological...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="biological-sequences-skbio-sequence">
<span id="module-skbio.sequence"></span><h1>Biological Sequences (<a class="reference internal" href="#module-skbio.sequence" title="skbio.sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code></a>)<a class="headerlink" href="#biological-sequences-skbio-sequence" title="Link to this heading">#</a></h1>
<p>This module provides functionality for storing and working with sequences, including
molecular sequences based on IUPAC-defined alphabets (<a class="reference internal" href="generated/skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-class docutils literal notranslate"><span class="pre">DNA</span></code></a>, <a class="reference internal" href="generated/skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA</span></code></a>,
<a class="reference internal" href="generated/skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-class docutils literal notranslate"><span class="pre">Protein</span></code></a>), sequences based on custom alphabets (<a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a>),
and generic/non-biological sequences with no alphabet restrictions (<a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>).</p>
<p>Additionally, this module defines the <a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticCode</span></code></a> class, which represents an
immutable object that translates DNA or RNA sequences into protein sequences, and
the <a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix" title="skbio.sequence.SubstitutionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a> class, which stores scores of substitutions between
sequence characters.</p>
<section id="sequence-types">
<h2>Sequence types<a class="headerlink" href="#sequence-types" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Sequence</span></code></a>(sequence[, metadata, ...])</p></td>
<td><p>Store generic sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a>(sequence[, metadata, ...])</p></td>
<td><p>Store sequence data conforming to a character set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DNA</span></code></a>(sequence[, metadata, ...])</p></td>
<td><p>Store DNA sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA</span></code></a>(sequence[, metadata, ...])</p></td>
<td><p>Store RNA sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Protein</span></code></a>(sequence[, metadata, ...])</p></td>
<td><p>Store protein sequence data and optional associated metadata.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sequence-utilities">
<h2>Sequence utilities<a class="headerlink" href="#sequence-utilities" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GeneticCode</span></code></a>(amino_acids, starts[, name])</p></td>
<td><p>Genetic code for translating codons to amino acids.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix" title="skbio.sequence.SubstitutionMatrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a>(alphabet, scores, **kwargs)</p></td>
<td><p>Scoring matrix between characters in biological sequences.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="distance-calculation">
<h2>Distance calculation<a class="headerlink" href="#distance-calculation" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.distance.html#module-skbio.sequence.distance" title="skbio.sequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a></p></td>
<td><p>Sequence distance metrics (skbio.sequence.distance)</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<p>The primary information stored for each different type of sequence object is the
underlying sequence data itself. This is stored as an immutable NumPy array.
Additionally, each type of sequence may include optional metadata and positional
metadata. Note that metadata and positional metadata are mutable.</p>
<p>Common operations are defined as methods, for example computing the reverse complement
of a DNA sequence, or searching for N-glycosylation motifs in protein sequences. Class
attributes provide valid character sets, complement maps for different sequence types,
and degenerate character definitions.</p>
<p>New sequences are created with optional metadata and positional metadata.
Metadata is stored as a Python <code class="docutils literal notranslate"><span class="pre">dict</span></code>, while positional metadata is stored as
a pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio</span> <span class="kn">import</span> <span class="n">DNA</span><span class="p">,</span> <span class="n">RNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ACCGGGTA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s2">&quot;my-sequence&quot;</span><span class="p">,</span> <span class="s1">&#39;description&#39;</span><span class="p">:</span><span class="s2">&quot;GFP&quot;</span><span class="p">},</span>
<span class="gp">... </span>         <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;quality&#39;</span><span class="p">:[</span><span class="mi">22</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">22</span><span class="p">,</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">23</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">25</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">DNA</span>
<span class="go">-----------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;GFP&#39;</span>
<span class="go">    &#39;id&#39;: &#39;my-sequence&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: int64&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 62.50%</span>
<span class="go">-----------------------------</span>
<span class="go">0 ACCGGGTA</span>
</pre></div>
</div>
<p>New sequences can also be created from existing sequences, for example as their
reverse complement or degapped (i.e., unaligned) version.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;.ACC--GGG-TA...&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span><span class="s1">&#39;my-sequence&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="n">d1</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;my-sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 62.50%</span>
<span class="go">--------------------------</span>
<span class="go">0 ACCGGGTA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;id&#39;: &#39;my-sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 8</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 62.50%</span>
<span class="go">--------------------------</span>
<span class="go">0 TACCCGGT</span>
</pre></div>
</div>
<p>It’s also straightforward to compute distances between sequences (optionally
using user-defined distance metrics, the default is Hamming distance which
requires that the sequences being compared are the same length) for use in
sequence clustering, phylogenetic reconstruction, etc.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r1</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;GACCCGCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r2</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;GCCCCCCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r1</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>
<span class="go">0.2</span>
</pre></div>
</div>
<p>Similarly, you can calculate the percent (dis)similarity between a pair of
aligned sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;ACCGUUAGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r4</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;ACGGGU--UC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span><span class="o">.</span><span class="n">match_frequency</span><span class="p">(</span><span class="n">r4</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">0.6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r3</span><span class="o">.</span><span class="n">mismatch_frequency</span><span class="p">(</span><span class="n">r4</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">0.4</span>
</pre></div>
</div>
<p>Sequences can be searched for known motif types. This returns the slices that
describe the matches.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r5</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;AGG-GGACUGAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s1">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">motif</span>
<span class="go">slice(0, 3, None)</span>
<span class="go">slice(4, 7, None)</span>
<span class="go">slice(9, 12, None)</span>
</pre></div>
</div>
<p>Those slices can be used to extract the relevant subsequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s1">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">--------------------------</span>
<span class="go">0 AGG</span>

<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">--------------------------</span>
<span class="go">0 GGA</span>

<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>And gaps or other features can be ignored while searching, as these may disrupt
otherwise meaningful motifs.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s1">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignore</span><span class="o">=</span><span class="n">r5</span><span class="o">.</span><span class="n">gaps</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 7</span>
<span class="go">    has gaps: True</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">--------------------------</span>
<span class="go">0 AGG-GGA</span>

<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>In the above example, removing gaps from the resulting motif matches is easily
achieved, as the sliced matches themselves are sequences of the same type as
the input.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s1">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ignore</span><span class="o">=</span><span class="n">r5</span><span class="o">.</span><span class="n">gaps</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="n">r5</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 6</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">--------------------------</span>
<span class="go">0 AGGGGA</span>

<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 GAA</span>
</pre></div>
</div>
<p>Sequences can similarly be searched for arbitrary patterns using regular
expressions.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">r5</span><span class="o">.</span><span class="n">find_with_regex</span><span class="p">(</span><span class="s1">&#39;(G+AC[UT])&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">match</span>
<span class="go">slice(4, 9, None)</span>
</pre></div>
</div>
<p>DNA can be transcribed to RNA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ATGTGTATTTGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span>
<span class="go">RNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 12</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 25.00%</span>
<span class="go">--------------------------</span>
<span class="go">0 AUGUGUAUUU GA</span>
</pre></div>
</div>
<p>Both DNA and RNA can be translated into a protein sequence. For example, let’s
translate our DNA and RNA sequences using NCBI’s standard genetic code (table
ID 1, the default genetic code in scikit-bio):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span>
<span class="go">Protein</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: True</span>
<span class="go">--------------------------</span>
<span class="go">0 MCI*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_rna</span> <span class="o">=</span> <span class="n">rna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_rna</span>
<span class="go">Protein</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 4</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: True</span>
<span class="go">--------------------------</span>
<span class="go">0 MCI*</span>
</pre></div>
</div>
<p>The two translations are equivalent:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">protein_from_dna</span> <span class="o">==</span> <span class="n">protein_from_rna</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Class-level methods contain information about the molecule types.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">DNA</span><span class="o">.</span><span class="n">degenerate_map</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">])</span>
<span class="go">[&#39;C&#39;, &#39;G&#39;, &#39;T&#39;]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">RNA</span><span class="o">.</span><span class="n">degenerate_map</span><span class="p">[</span><span class="s1">&#39;B&#39;</span><span class="p">])</span>
<span class="go">[&#39;C&#39;, &#39;G&#39;, &#39;U&#39;]</span>
</pre></div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.io.registry.create_format.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.io.registry.create_format</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.sequence.Sequence.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.sequence.Sequence</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-types">Sequence types</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-utilities">Sequence utilities</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-calculation">Distance calculation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2024, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>