
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="Storing and working with sequences: nucleotides, amino acids, custom alphabets and unrestricted ones." name="description" />

    <title>Biological Sequences (skbio.sequence) &#8212; scikit-bio 0.7.2-dev documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/css/style.css?v=240ceacb" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=ae22e3c7"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script async="async" src="https://www.googletagmanager.com/gtag/js?id=UA-6636235-9"></script>
    <script>
                window.dataLayer = window.dataLayer || [];
                function gtag(){ dataLayer.push(arguments); }
                gtag('js', new Date());
                gtag('config', 'UA-6636235-9');
            </script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'sequence';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://scikit.bio/versions.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="canonical" href="../../sequence.html" />
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="skbio.sequence.Sequence" href="generated/skbio.sequence.Sequence.html" />
    <link rel="prev" title="skbio.io.registry.create_format" href="generated/skbio.io.registry.create_format.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.7.2-dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
     
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="scikit-bio 0.7.2-dev documentation"/>
    <img src="_static/logo_inv.svg" class="logo__image only-dark pst-js-only" alt="scikit-bio 0.7.2-dev documentation"/>
  
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-external" href="../../install.html">
    Install
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../learn.html">
    Learn
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="index.html">
    Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../contribute.html">
    Contribute
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/discussions">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md">
    Releases
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="../../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="io.html">
    Input and Output (skbio.io)
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Biological Sequences (skbio.sequence)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="alignment.html">
    Sequence Alignments (skbio.alignment)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tree.html">
    Trees and Phylogenetics (skbio.tree)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="diversity.html">
    Community Diversity (skbio.diversity)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="stats.html">
    Multivariate Statistics (skbio.stats)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="embedding.html">
    Biological Embeddings (skbio.embedding)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="table.html">
    Data Table (skbio.table)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="metadata.html">
    Metadata (skbio.metadata)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="workflow.html">
    Workflow Construction (skbio.workflow)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="util.html">
    Utilities for Developers (skbio.util)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="config.html">
    Configuration Options
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/scikit-bio/scikit-bio" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://twitter.com/scikitbio" title="Twitter" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-twitter fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Twitter</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.Sequence.html">skbio.sequence.Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html">skbio.sequence.GrammaredSequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.DNA.html">skbio.sequence.DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.RNA.html">skbio.sequence.RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.Protein.html">skbio.sequence.Protein</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.GeneticCode.html">skbio.sequence.GeneticCode</a></li>
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html">skbio.sequence.SubstitutionMatrix</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.distance.html">Sequence distance metrics (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence.distance</span></code>)</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="generated/skbio.sequence.NucleotideMixin.html">skbio.sequence.NucleotideMixin</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Biological Sequences (<code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code>)</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="biological-sequences-skbio-sequence">
<span id="module-skbio.sequence"></span><h1>Biological Sequences (<a class="reference internal" href="#module-skbio.sequence" title="skbio.sequence"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.sequence</span></code></a>)<a class="headerlink" href="#biological-sequences-skbio-sequence" title="Link to this heading">#</a></h1>
<p>This module provides functionality for storing and working with sequences, including
molecular sequences based on IUPAC-defined alphabets (<a class="reference internal" href="generated/skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-class docutils literal notranslate"><span class="pre">DNA</span></code></a>, <a class="reference internal" href="generated/skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-class docutils literal notranslate"><span class="pre">RNA</span></code></a>,
<a class="reference internal" href="generated/skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-class docutils literal notranslate"><span class="pre">Protein</span></code></a>), sequences based on custom alphabets (<a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a>),
and generic/non-biological sequences with no alphabet restrictions (<a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>).</p>
<p>Additionally, this module defines the <a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticCode</span></code></a> class, which represents an
immutable object that translates DNA or RNA sequences into protein sequences, and
the <a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix" title="skbio.sequence.SubstitutionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a> class, which stores scores of substitutions between
sequence characters.</p>
<p>See the <a class="reference internal" href="#sequence-tutorial"><strong>Tutorial</strong></a> section for working with biological sequences using
scikit-bio.</p>
<section id="sequence-types">
<h2>Sequence types<a class="headerlink" href="#sequence-types" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence" title="skbio.sequence.Sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Sequence</span></code></a></p></td>
<td><p>Store generic sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a></p></td>
<td><p>Store sequence data conforming to a character set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.DNA.html#skbio.sequence.DNA" title="skbio.sequence.DNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DNA</span></code></a></p></td>
<td><p>Store DNA sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.RNA.html#skbio.sequence.RNA" title="skbio.sequence.RNA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RNA</span></code></a></p></td>
<td><p>Store RNA sequence data and optional associated metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.Protein.html#skbio.sequence.Protein" title="skbio.sequence.Protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Protein</span></code></a></p></td>
<td><p>Store protein sequence data and optional associated metadata.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sequence-utilities">
<h2>Sequence utilities<a class="headerlink" href="#sequence-utilities" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GeneticCode</span></code></a></p></td>
<td><p>Genetic code for translating codons to amino acids.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/skbio.sequence.SubstitutionMatrix.html#skbio.sequence.SubstitutionMatrix" title="skbio.sequence.SubstitutionMatrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SubstitutionMatrix</span></code></a></p></td>
<td><p>Scoring matrix between characters in biological sequences.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="distance-calculation">
<h2>Distance calculation<a class="headerlink" href="#distance-calculation" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.distance.html#module-skbio.sequence.distance" title="skbio.sequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a></p></td>
<td><p>Sequence distance metrics (skbio.sequence.distance)</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="abstract-classes">
<h2>Abstract classes<a class="headerlink" href="#abstract-classes" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/skbio.sequence.NucleotideMixin.html#skbio.sequence.NucleotideMixin" title="skbio.sequence.NucleotideMixin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NucleotideMixin</span></code></a></p></td>
<td><p>Mixin for adding functionality for working with sequences of nucleotides.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tutorial">
<span id="sequence-tutorial"></span><h2>Tutorial<a class="headerlink" href="#tutorial" title="Link to this heading">#</a></h2>
<section id="creating-sequences">
<h3>Creating sequences<a class="headerlink" href="#creating-sequences" title="Link to this heading">#</a></h3>
<p>Create a DNA sequence from a string of nucleotide characters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">--------------------------</span>
<span class="go">Stats:</span>
<span class="go">    length: 6</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">--------------------------</span>
<span class="go">0 GAATTC</span>
</pre></div>
</div>
<p>Optionally, one may attach metadata to a sequence object. Details of metadata are
introduced in <a class="reference internal" href="#annotate-sequences"><span class="std std-ref">Annotating sequences</span></a> below.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;S1&#39;</span><span class="p">,</span> <span class="s1">&#39;description&#39;</span><span class="p">:</span> <span class="s1">&#39;EcoRI recognition site&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">-------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;EcoRI recognition site&#39;</span>
<span class="go">    &#39;id&#39;: &#39;S1&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 6</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 33.33%</span>
<span class="go">-------------------------------------------</span>
<span class="go">0 GAATTC</span>
</pre></div>
</div>
<p>Sequences can be read from <a class="reference internal" href="io.html#module-skbio.io" title="skbio.io"><code class="xref py py-mod docutils literal notranslate"><span class="pre">file</span> <span class="pre">formats</span></code></a> recognizable by scikit-bio.
For example, to read a DNA sequence from a <a class="reference internal" href="generated/skbio.io.format.fasta.html#module-skbio.io.format.fasta" title="skbio.io.format.fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">FASTA</span></code></a> file:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;input.fa&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The filename “input.fa” can also be a file handle or other readable objects.</p>
<p>To read all sequences from a multi-FASTA file into a list of sequence objects, you will
need:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">read</span> <span class="k">as</span> <span class="n">sk_read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sk_read</span><span class="p">(</span><span class="s1">&#39;input.fa&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span> <span class="n">constructor</span><span class="o">=</span><span class="n">DNA</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="sequence-grammar">
<h3>Sequence grammar<a class="headerlink" href="#sequence-grammar" title="Link to this heading">#</a></h3>
<p>The three common biological sequence types: <code class="docutils literal notranslate"><span class="pre">DNA</span></code>, <code class="docutils literal notranslate"><span class="pre">RNA</span></code> and <code class="docutils literal notranslate"><span class="pre">Protein</span></code>, are
<a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">grammared</span> <span class="pre">sequences</span></code></a>. That is, each of them has a defined
alphabet (character set), which typically consists of:</p>
<ul class="simple">
<li><p><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence.definite_chars" title="skbio.sequence.GrammaredSequence.definite_chars"><code class="xref py py-attr docutils literal notranslate"><span class="pre">definite</span> <span class="pre">characters</span></code></a>, such as <code class="docutils literal notranslate"><span class="pre">ACGT</span></code>,
which represent the four canonical nucleotides in a DNA sequence,</p></li>
<li><p><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence.degenerate_chars" title="skbio.sequence.GrammaredSequence.degenerate_chars"><code class="xref py py-attr docutils literal notranslate"><span class="pre">degenerate</span> <span class="pre">characters</span></code></a>, such as <code class="docutils literal notranslate"><span class="pre">R</span></code>,
which represents <code class="docutils literal notranslate"><span class="pre">A</span></code> or <code class="docutils literal notranslate"><span class="pre">G</span></code>,</p></li>
<li><p><a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence.gap_chars" title="skbio.sequence.GrammaredSequence.gap_chars"><code class="xref py py-attr docutils literal notranslate"><span class="pre">gap</span> <span class="pre">character(s)</span></code></a>, such as <code class="docutils literal notranslate"><span class="pre">-</span></code>, and</p></li>
<li><p>a <a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence.wildcard_char" title="skbio.sequence.GrammaredSequence.wildcard_char"><code class="xref py py-attr docutils literal notranslate"><span class="pre">wildcard</span> <span class="pre">character</span></code></a>, such as <code class="docutils literal notranslate"><span class="pre">N</span></code>, which
is also a degenerate character.</p></li>
</ul>
<p>Use of any of these characters in the sequence data is valid. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GCCRCCATGG&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Kozak consensus sequence&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">--------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;name&#39;: &#39;Kozak consensus sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 10</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: True</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 70.00%</span>
<span class="go">--------------------------------------</span>
<span class="go">0 GCCRCCATGG</span>
</pre></div>
</div>
<p>One can customize grammared sequence types, for example to include modified amino
acids, methylation sites, or non-biological molecules. Refer to
<a class="reference internal" href="generated/skbio.sequence.GrammaredSequence.html#skbio.sequence.GrammaredSequence" title="skbio.sequence.GrammaredSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">GrammaredSequence</span></code></a> for instructions.</p>
<p>Each class contains a degenerate map for lookup:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">DNA</span><span class="o">.</span><span class="n">degenerate_map</span><span class="p">[</span><span class="s1">&#39;R&#39;</span><span class="p">])</span>
<span class="go">[&#39;A&#39;, &#39;G&#39;]</span>
</pre></div>
</div>
<p>Check if a sequence contains degenerate characters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">has_degenerates</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Identify the positions of degenerate characters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">degenerates</span><span class="p">()</span>
<span class="go">array([False, False, False,  True, False, False, False, False, False,</span>
<span class="go">       False], dtype=bool)</span>
</pre></div>
</div>
<p>Likewise, one can create a sequence with gaps, and identify the gap positions.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gapped</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;ATG--CAC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gapped</span><span class="o">.</span><span class="n">gaps</span><span class="p">()</span>
<span class="go">array([False, False, False,  True,  True, False, False, False], dtype=bool)</span>
</pre></div>
</div>
<p>One can efficiently remove all gaps from a sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ungapped</span> <span class="o">=</span> <span class="n">gapped</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ungapped</span><span class="p">)</span>
<span class="go">ATGCAC</span>
</pre></div>
</div>
</section>
<section id="exploring-sequences">
<h3>Exploring sequences<a class="headerlink" href="#exploring-sequences" title="Link to this heading">#</a></h3>
<p>The primary information stored in a sequence object is the sequence data itself. It can
be retrieved from the string representation of the sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">&#39;GCCRCCATGG&#39;</span>
</pre></div>
</div>
<p>Or:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GCCRCCATGG</span>
</pre></div>
</div>
<p>Under the hood, the sequence data is stored as a NumPy array of bytes, which permit
compact storage and efficient, vectorized operations.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">values</span>
<span class="go">array([b&#39;G&#39;, b&#39;C&#39;, b&#39;C&#39;, b&#39;R&#39;, b&#39;C&#39;, b&#39;C&#39;, b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;, b&#39;G&#39;],</span>
<span class="go">      dtype=&#39;|S1&#39;)</span>
</pre></div>
</div>
<p>This array can also be viewed as ASCII code points:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="s1">&#39;uint8&#39;</span><span class="p">)</span>
<span class="go">array([71, 67, 67, 82, 67, 67, 65, 84, 71, 71], dtype=uint8)</span>
</pre></div>
</div>
<p>A sequence object shares typical behaviors of a Python string. This means that you can
apply Python string <strong>indexing</strong> and <strong>slicing</strong> approaches on a sequence. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;name&#39;: &#39;Kozak consensus sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 1</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 100.00%</span>
<span class="go">--------------------------------------</span>
<span class="go">0 G</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">DNA</span>
<span class="go">--------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;name&#39;: &#39;Kozak consensus sequence&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 3</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: True</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 66.67%</span>
<span class="go">--------------------------------------</span>
<span class="go">0 CRC</span>
</pre></div>
</div>
<p>The output of indexing or slicing is also a sequence object. This process doesn’t
copy the original sequence data. Instead, it returns a <strong>view</strong> of the same memory
space, which is efficient. Refer to NumPy’s <a class="reference external" href="https://numpy.org/doc/stable/user/basics.copies.html">copies and views</a> for more information.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">base</span> <span class="ow">is</span> <span class="n">seq</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">base</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Some Python string methods, such as <code class="docutils literal notranslate"><span class="pre">count</span></code> and <code class="docutils literal notranslate"><span class="pre">index</span></code>, also work on scikit-bio
sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
<span class="go">4</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span>
<span class="go">6</span>
</pre></div>
</div>
<p>To concatenate multiple sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GGATCC&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;AAGCTT&#39;</span><span class="p">),</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;GAATTC&#39;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GGATCCAAGCTTGAATTC</span>
</pre></div>
</div>
</section>
<section id="sequence-conversion">
<h3>Sequence conversion<a class="headerlink" href="#sequence-conversion" title="Link to this heading">#</a></h3>
<p>scikit-bio supports convenient conversion between DNA (two strands), RNA and protein
sequences. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;TGCTACATCCAGAACTGCCCCCTGGGA&#39;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;oxytocin&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dna</span>
<span class="go">DNA</span>
<span class="go">-------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;name&#39;: &#39;oxytocin&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 27</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 59.26%</span>
<span class="go">-------------------------------</span>
<span class="go">0 TGCTACATCC AGAACTGCCC CCTGGGA</span>
</pre></div>
</div>
<p>Reverse complement a DNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
<span class="go">TCCCAGGGGGCAGTTCTGGATGTAGCA</span>
</pre></div>
</div>
<p>Transcribe a DNA sequence into mRNA:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rna</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rna</span><span class="p">)</span>
<span class="go">UGCUACAUCCAGAACUGCCCCCUGGGA</span>
</pre></div>
</div>
<p>Translate an mRNA sequence into protein:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prot</span> <span class="o">=</span> <span class="n">rna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">prot</span><span class="p">)</span>
<span class="go">CYIQNCPLG</span>
</pre></div>
</div>
<p>By default, the standard genetic code table (1) is used. To use a specific table (see
the <a class="reference internal" href="generated/skbio.sequence.GeneticCode.html#skbio.sequence.GeneticCode" title="skbio.sequence.GeneticCode"><code class="xref py py-class docutils literal notranslate"><span class="pre">GeneticCode</span></code></a> class for details), simply specify its index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prot</span> <span class="o">=</span> <span class="n">rna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="mi">11</span><span class="p">)</span>  <span class="c1"># Bacteria</span>
</pre></div>
</div>
<p>A DNA sequence can be directly translated into protein. The output is the same as
transcription followed by translation.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prot_</span> <span class="o">=</span> <span class="n">dna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prot_</span> <span class="o">==</span> <span class="n">prot</span>
<span class="go">True</span>
</pre></div>
</div>
</section>
<section id="searching-for-patterns">
<h3>Searching for patterns<a class="headerlink" href="#searching-for-patterns" title="Link to this heading">#</a></h3>
<p>Sequences can be searched for patterns, a.k.a., <strong>motifs</strong> in molecular biology. Except
for searching for exact substrings using the <code class="docutils literal notranslate"><span class="pre">index</span></code> method (see above), one can also
search for flexibly defined patterns using <a class="extlink-wiki reference external" href="https://en.wikipedia.org/wiki/Regular_expression">regular expressions</a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">RNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;AGG-GGACUGAA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">find_with_regex</span><span class="p">(</span><span class="s1">&#39;(G+AC[UT])&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">match</span>
<span class="go">slice(4, 9, None)</span>
</pre></div>
</div>
<p>The returned values are slices, with which one can extract the relevant subsequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">match</span><span class="p">])</span>
<span class="go">GGACU</span>
</pre></div>
</div>
<p>scikit-bio has a convenience shortcut <code class="docutils literal notranslate"><span class="pre">find_motifs</span></code>, which searches for contiguous
runs of purines or pyrimidines.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">find_motifs</span><span class="p">(</span><span class="s1">&#39;purine-run&#39;</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">motif</span><span class="p">])</span>
<span class="go">AGG</span>
<span class="go">GGA</span>
<span class="go">GAA</span>
</pre></div>
</div>
</section>
<section id="comparing-sequences">
<h3>Comparing sequences<a class="headerlink" href="#comparing-sequences" title="Link to this heading">#</a></h3>
<p>It is straightforward to compute the distance between two sequences. By default, the
Hamming distance (<a class="reference internal" href="generated/skbio.sequence.distance.hamming.html#skbio.sequence.distance.hamming" title="skbio.sequence.distance.hamming"><code class="xref py py-func docutils literal notranslate"><span class="pre">hamming</span></code></a>, i.e., proportion of different sites) is
calculated, This metric is useful in analyses such as sequence clustering and
phylogenetic reconstruction.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;GACCCGCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;GCCCCCCUUU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="go">0.2</span>
</pre></div>
</div>
<p>Hamming distance requires that the two sequences have the same length. This often
requires that you <strong>align</strong> the sequences before comparison. See <a class="reference internal" href="alignment.html#module-skbio.alignment" title="skbio.alignment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">skbio.alignment</span></code></a>
for how to align sequences.</p>
<p>Similarly, you can calculate the percent (dis)similarity between a pair of
aligned sequences.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;ACCGUUAGUC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">RNA</span><span class="p">(</span><span class="s1">&#39;ACGGGU--UC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="o">.</span><span class="n">match_frequency</span><span class="p">(</span><span class="n">seq2</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">0.6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="o">.</span><span class="n">mismatch_frequency</span><span class="p">(</span><span class="n">seq2</span><span class="p">,</span> <span class="n">relative</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">0.4</span>
</pre></div>
</div>
</section>
<section id="annotating-sequences">
<span id="annotate-sequences"></span><h3>Annotating sequences<a class="headerlink" href="#annotating-sequences" title="Link to this heading">#</a></h3>
<p>Metadata describe the properties of a sequence or some of its regions/sites. scikit-bio
supports three types of sequence metadata: <strong>metadata</strong> (for the entire sequence),
<strong>positional metadata</strong> (for sites), and <strong>interval metadata</strong> (for regions).</p>
<p><code class="docutils literal notranslate"><span class="pre">metadata</span></code> stores the properties of the entire sequence, such as ID, description,
source, function, etc., as a Python dictionary. The following example creates a DNA
sequence with “gene” and “product” properties.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="p">(</span><span class="s1">&#39;TGGTATATAGTTTAAACAAAACGAATGATTTCGACTCATTAAATTATGATAATCATATTTACCAA&#39;</span><span class="p">,</span>
<span class="gp">... </span>          <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;TRNR&#39;</span><span class="p">,</span> <span class="s1">&#39;product&#39;</span><span class="p">:</span> <span class="s1">&#39;tRNA-Arg&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">--------------------------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;gene&#39;: &#39;TRNR&#39;</span>
<span class="go">    &#39;product&#39;: &#39;tRNA-Arg&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 65</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 23.08%</span>
<span class="go">--------------------------------------------------------------------</span>
<span class="go">0  TGGTATATAG TTTAAACAAA ACGAATGATT TCGACTCATT AAATTATGAT AATCATATTT</span>
<span class="go">60 ACCAA</span>
</pre></div>
</div>
<p>Access the metadata by key:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;product&#39;</span><span class="p">]</span>
<span class="go">&#39;tRNA-Arg&#39;</span>
</pre></div>
</div>
<p>Modify the metadata of an existing sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;organism&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Homo sapiens&#39;</span>
</pre></div>
</div>
<p>Many of scikit-bio’s file format parsers can automatically populate metadata fields.
For example, the FASTA parser reads the sequence ID and description into the <code class="docutils literal notranslate"><span class="pre">id</span></code>
and <code class="docutils literal notranslate"><span class="pre">description</span></code> fields:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.sequence</span><span class="w"> </span><span class="kn">import</span> <span class="n">Protein</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;&gt;NP_000591.1 interleukin-6 isoform 1 precursor [Homo sapiens]</span>
<span class="gp">... </span><span class="s1">MNSFSTSAFGPVAFSLGLLLVLPAAFPAPVPPGEDSKDVAAPHRQPLTSSERIDKQIRYILDGISALRKE</span>
<span class="gp">... </span><span class="s1">TCNKSNMCESSKEALAENNLNLPKMAEKDGCFQSGFNEETCLVKIITGLLEFEVYLEYLQNRFESSEEQA</span>
<span class="gp">... </span><span class="s1">RAVQMSTKVLIQFLQKKAKNLDAITTPDPTTNASLLTKLQAQNQWLQDMTTHLILRSFKEFLQSSLRALR</span>
<span class="gp">... </span><span class="s1">QM&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Protein</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="n">fasta</span><span class="p">],</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Protein</span>
<span class="go">---------------------------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;interleukin-6 isoform 1 precursor [Homo sapiens]&#39;</span>
<span class="go">    &#39;id&#39;: &#39;NP_000591.1&#39;</span>
<span class="go">Stats:</span>
<span class="go">    length: 212</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    has stops: False</span>
<span class="go">---------------------------------------------------------------------</span>
<span class="go">0   MNSFSTSAFG PVAFSLGLLL VLPAAFPAPV PPGEDSKDVA APHRQPLTSS ERIDKQIRYI</span>
<span class="go">60  LDGISALRKE TCNKSNMCES SKEALAENNL NLPKMAEKDG CFQSGFNEET CLVKIITGLL</span>
<span class="go">120 EFEVYLEYLQ NRFESSEEQA RAVQMSTKVL IQFLQKKAKN LDAITTPDPT TNASLLTKLQ</span>
<span class="go">180 AQNQWLQDMT THLILRSFKE FLQSSLRALR QM</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">positional_metadata</span></code> stores per-character properties. Each property has the same
number of values as the sequence length. The following example annotates three types
of sites on the human TP53 protein sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="s1">&#39;MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAA&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;PPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKT&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;CPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRN&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;TFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGR&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;DRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALEL&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;KDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD&#39;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sites</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">acetylation</span> <span class="o">=</span> <span class="n">sites</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">acetylation</span><span class="p">[[</span><span class="mi">119</span><span class="p">,</span> <span class="mi">304</span><span class="p">,</span> <span class="mi">320</span><span class="p">,</span> <span class="mi">372</span><span class="p">,</span> <span class="mi">380</span><span class="p">,</span> <span class="mi">381</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">methylation</span> <span class="o">=</span> <span class="n">sites</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">methylation</span><span class="p">[[</span><span class="mi">332</span><span class="p">,</span> <span class="mi">334</span><span class="p">,</span> <span class="mi">336</span><span class="p">,</span> <span class="mi">369</span><span class="p">,</span> <span class="mi">371</span><span class="p">,</span> <span class="mi">372</span><span class="p">,</span> <span class="mi">381</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phosphorylation</span> <span class="o">=</span> <span class="n">sites</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phosphorylation</span><span class="p">[[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="mi">19</span><span class="p">,</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">36</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">54</span><span class="p">,</span> <span class="mi">182</span><span class="p">,</span> <span class="mi">268</span><span class="p">,</span> <span class="mi">314</span><span class="p">,</span> <span class="mi">391</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Protein</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="s1">&#39;NP_000537.3&#39;</span><span class="p">,</span> <span class="s1">&#39;description&#39;</span><span class="p">:</span> <span class="s1">&#39;TP53&#39;</span><span class="p">},</span>
<span class="gp">... </span>              <span class="n">positional_metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;acetylation&#39;</span><span class="p">:</span> <span class="n">acetylation</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="s1">&#39;methylation&#39;</span><span class="p">:</span> <span class="n">methylation</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="s1">&#39;phosphorylation&#39;</span><span class="p">:</span> <span class="n">phosphorylation</span><span class="p">})</span>
</pre></div>
</div>
<p>Once created, positional metadata are stored as a dataframe.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="p">)</span>
<span class="go">     acetylation  methylation  phosphorylation</span>
<span class="go">0          False        False            False</span>
<span class="go">1          False        False            False</span>
<span class="go">2          False        False            False</span>
<span class="go">3          False        False            False</span>
<span class="go">4          False        False            False</span>
<span class="go">..           ...          ...              ...</span>
<span class="go">388        False        False            False</span>
<span class="go">389        False        False            False</span>
<span class="go">390        False        False            False</span>
<span class="go">391        False        False             True</span>
<span class="go">392        False        False            False</span>

<span class="go">[393 rows x 3 columns]</span>
</pre></div>
</div>
<p>Likewise, scikit-bio’s <a class="reference internal" href="generated/skbio.io.format.fastq.html#module-skbio.io.format.fastq" title="skbio.io.format.fastq"><code class="xref py py-mod docutils literal notranslate"><span class="pre">FASTQ</span></code></a> parser can automatically
populate the “quality” column of the positional metadata:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fastq</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;@S00001/1</span>
<span class="gp">... </span><span class="s1">CGTGCTCCACGCTCTCCTGCACGCCCGAGCCGTTGCGGTCGGCGAGAGCCATGTCGACGGCGGGCTTGAG</span>
<span class="gp">... </span><span class="s1">+</span>
<span class="gp">... </span><span class="s1">DDAEGGGGIIIGIJKJKKKHIKJAKKKHCGIKKJAJGICKKKAJEJA:G1EJJKJHJFKJIDIEII?=DC</span>
<span class="gp">... </span><span class="s1">&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="n">fastq</span><span class="p">],</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fastq&#39;</span><span class="p">,</span> <span class="n">phred_offset</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">DNA</span>
<span class="go">--------------------------------------------------------------------</span>
<span class="go">Metadata:</span>
<span class="go">    &#39;description&#39;: &#39;&#39;</span>
<span class="go">    &#39;id&#39;: &#39;S00001/1&#39;</span>
<span class="go">Positional metadata:</span>
<span class="go">    &#39;quality&#39;: &lt;dtype: uint8&gt;</span>
<span class="go">Stats:</span>
<span class="go">    length: 70</span>
<span class="go">    has gaps: False</span>
<span class="go">    has degenerates: False</span>
<span class="go">    has definites: True</span>
<span class="go">    GC-content: 71.43%</span>
<span class="go">--------------------------------------------------------------------</span>
<span class="go">0  CGTGCTCCAC GCTCTCCTGC ACGCCCGAGC CGTTGCGGTC GGCGAGAGCC ATGTCGACGG</span>
<span class="go">60 CGGGCTTGAG</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">positional_metadata</span><span class="p">[</span><span class="s1">&#39;quality&#39;</span><span class="p">])</span>
<span class="go">0     35</span>
<span class="go">1     35</span>
<span class="go">2     32</span>
<span class="go">3     36</span>
<span class="go">4     38</span>
<span class="go">      ..</span>
<span class="go">65    40</span>
<span class="go">66    30</span>
<span class="go">67    28</span>
<span class="go">68    35</span>
<span class="go">69    34</span>
<span class="go">Name: quality, Length: 70, dtype: uint8</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">interval_metadata</span></code> stores properties that span a range in the sequence, such as
genes and regulatory elements. See the <a class="reference internal" href="generated/skbio.metadata.IntervalMetadata.html#skbio.metadata.IntervalMetadata" title="skbio.metadata.IntervalMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntervalMetadata</span></code></a> class
for details. The following example annotates three regions of the human H4C6 gene.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fasta</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;&gt;NM_003540.4 Homo sapiens H4 clustered histone 6 (H4C6), mRNA</span>
<span class="gp">... </span><span class="s1">GCAAAAGTTAAGAGTTGTTGTTTGTCTTCGATCATGTCTGGTAGAGGCAAAGGTGGTAAAGGTTTAGGAA</span>
<span class="gp">... </span><span class="s1">AGGGAGGCGCCAAGCGCCATCGCAAAGTGCTGCGTGACAACATACAGGGCATCACGAAGCCCGCCATCCG</span>
<span class="gp">... </span><span class="s1">TCGCTTGGCCCGACGCGGCGGCGTGAAACGCATTTCGGGCCTCATTTATGAGGAGACCCGCGGTGTTCTT</span>
<span class="gp">... </span><span class="s1">AAGGTGTTCCTGGAGAATGTGATACGGGACGCCGTAACCTACACGGAGCACGCCAAGCGTAAGACAGTCA</span>
<span class="gp">... </span><span class="s1">CTGCAATGGATGTTGTCTACGCGCTCAAGCGCCAGGGACGCACTCTGTACGGCTTTGGTGGCTGAGCCTC</span>
<span class="gp">... </span><span class="s1">ACCCCGGCTTTTTATTTAACAGCTCACCCATAAAAGGCCCTTTTCAGGGCC&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">([</span><span class="n">fasta</span><span class="p">],</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">401</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;exon&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">33</span><span class="p">,</span> <span class="mi">345</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;cds&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">bounds</span><span class="o">=</span><span class="p">[(</span><span class="mi">385</span><span class="p">,</span> <span class="mi">401</span><span class="p">)],</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;stem_loop&#39;</span><span class="p">})</span>
</pre></div>
</div>
<p>scikit-bio’s <a class="reference internal" href="generated/skbio.io.format.gff3.html#module-skbio.io.format.gff3" title="skbio.io.format.gff3"><code class="xref py py-mod docutils literal notranslate"><span class="pre">GFF3</span></code></a> parser can read sequence annotation
into interval metadata. The following code reads a genome’s sequence and annotation
from separate files and combine.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">skbio.metadata</span><span class="w"> </span><span class="kn">import</span> <span class="n">IntervalMetadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;genomic.fna&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">interval_metadata</span> <span class="o">=</span> <span class="n">IntervalMetadata</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s1">&#39;genomic.gff&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;gff3&#39;</span><span class="p">,</span> <span class="n">seq_id</span><span class="o">=</span><span class="n">seq</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>If the GFF3 contains the sequence, one can read the sequence and annotation all at
once:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;genomic.gff&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;gff3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/skbio.io.registry.create_format.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">skbio.io.registry.create_format</p>
      </div>
    </a>
    <a class="right-next"
       href="generated/skbio.sequence.Sequence.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">skbio.sequence.Sequence</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-types">Sequence types</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-utilities">Sequence utilities</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#distance-calculation">Distance calculation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#abstract-classes">Abstract classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tutorial">Tutorial</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-sequences">Creating sequences</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-grammar">Sequence grammar</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exploring-sequences">Exploring sequences</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-conversion">Sequence conversion</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#searching-for-patterns">Searching for patterns</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparing-sequences">Comparing sequences</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#annotating-sequences">Annotating sequences</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2014-2026, scikit-bio development team.
      <br/>
    
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__center">
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>