<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.sequence.BiologicalSequence &mdash; scikit-bio 0.2.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.2.3 documentation" href="../index.html" />
    <link rel="up" title="Biological sequences (skbio.sequence)" href="../sequence.html" />
    <link rel="next" title="skbio.sequence.BiologicalSequence.sequence" href="skbio.sequence.BiologicalSequence.sequence.html" />
    <link rel="prev" title="skbio.sequence.InvalidCodonError" href="skbio.sequence.InvalidCodonError.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.sequence.BiologicalSequence</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-sequence-biologicalsequence">
<h1>skbio.sequence.BiologicalSequence<a class="headerlink" href="#skbio-sequence-biologicalsequence" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="skbio.sequence.BiologicalSequence">
<em class="property">class </em><tt class="descclassname">skbio.sequence.</tt><tt class="descname">BiologicalSequence</tt><big>(</big><em>sequence</em>, <em>id=''</em>, <em>description=''</em>, <em>quality=None</em>, <em>validate=False</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.2.3/skbio/sequence/_sequence.py#L26"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.sequence.BiologicalSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for biological sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sequence</strong> : python Sequence (e.g., str, list or tuple)</p>
<blockquote>
<div><p>The biological sequence.</p>
</div></blockquote>
<p><strong>id</strong> : str, optional</p>
<blockquote>
<div><p>The sequence id (e.g., an accession number).</p>
</div></blockquote>
<p><strong>description</strong> : str, optional</p>
<blockquote>
<div><p>A description or comment about the sequence (e.g., &#8220;green
fluorescent protein&#8221;).</p>
</div></blockquote>
<p><strong>quality</strong> : 1-D array_like, int, optional</p>
<blockquote>
<div><p>Phred quality scores stored as nonnegative integers, one per sequence
character. If provided, must be the same length as the biological
sequence. Can be a 1-D <tt class="docutils literal"><span class="pre">numpy.ndarray</span></tt> of integers, or a structure
that can be converted to this representation using <tt class="docutils literal"><span class="pre">numpy.asarray</span></tt>. A
copy will <em>not</em> be made if <cite>quality</cite> is already a 1-D <tt class="docutils literal"><span class="pre">numpy.ndarray</span></tt>
with an <tt class="docutils literal"><span class="pre">int</span></tt> <tt class="docutils literal"><span class="pre">dtype</span></tt>. The array will be made read-only (i.e., its
<tt class="docutils literal"><span class="pre">WRITEABLE</span></tt> flag will be set to <tt class="docutils literal"><span class="pre">False</span></tt>).</p>
</div></blockquote>
<p><strong>validate</strong> : bool, optional</p>
<blockquote>
<div><p>If True, runs the <cite>is_valid</cite> method after construction and raises
BiologicalSequenceError if <tt class="docutils literal"><span class="pre">is_valid</span> <span class="pre">==</span> <span class="pre">False</span></tt>.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>skbio.sequence.BiologicalSequenceError</strong></p>
<blockquote class="last">
<div><p>If <tt class="docutils literal"><span class="pre">validate</span> <span class="pre">==</span> <span class="pre">True</span></tt> and <tt class="docutils literal"><span class="pre">is_valid</span> <span class="pre">==</span> <span class="pre">False</span></tt>, or if <cite>quality</cite> is
not the correct shape.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="skbio.sequence.NucleotideSequence.html#skbio.sequence.NucleotideSequence" title="skbio.sequence.NucleotideSequence"><tt class="xref py py-obj docutils literal"><span class="pre">NucleotideSequence</span></tt></a>, <a class="reference internal" href="skbio.sequence.DNASequence.html#skbio.sequence.DNASequence" title="skbio.sequence.DNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">DNASequence</span></tt></a>, <a class="reference internal" href="skbio.sequence.RNASequence.html#skbio.sequence.RNASequence" title="skbio.sequence.RNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">RNASequence</span></tt></a></p>
</div>
<p class="rubric">Notes</p>
<p><cite>BiologicalSequence</cite> objects are immutable. Where applicable, methods
return a new object of the same class.
Subclasses are typically defined by methods relevant to only a specific
type of biological sequence, and by containing characters only contained in
the IUPAC standard character set <a class="reference internal" href="#r171" id="id1">[R171]</a> for that molecule type.</p>
<p class="rubric">References</p>
<table class="docutils citation" frame="void" id="r171" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[R171]</td><td><em>(<a class="fn-backref" href="#id1">1</a>, <a class="fn-backref" href="#id2">2</a>)</em> Nomenclature for incompletely specified bases in nucleic acid
sequences: recommendations 1984.
Nucleic Acids Res. May 10, 1985; 13(9): 3021-3030.
A Cornish-Bowden</td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.sequence</span> <span class="kn">import</span> <span class="n">BiologicalSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">BiologicalSequence</span><span class="p">(</span><span class="s">&#39;GGUCGUGAAGGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">BiologicalSequence</span><span class="p">(</span><span class="s">&#39;GGUCCUGAAGGU&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.sequence.html#skbio.sequence.BiologicalSequence.sequence" title="skbio.sequence.BiologicalSequence.sequence"><tt class="xref py py-obj docutils literal"><span class="pre">sequence</span></tt></a></td>
<td>String containing underlying biological sequence characters.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.id.html#skbio.sequence.BiologicalSequence.id" title="skbio.sequence.BiologicalSequence.id"><tt class="xref py py-obj docutils literal"><span class="pre">id</span></tt></a></td>
<td>ID of the biological sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.description.html#skbio.sequence.BiologicalSequence.description" title="skbio.sequence.BiologicalSequence.description"><tt class="xref py py-obj docutils literal"><span class="pre">description</span></tt></a></td>
<td>Description of the biological sequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.quality.html#skbio.sequence.BiologicalSequence.quality" title="skbio.sequence.BiologicalSequence.quality"><tt class="xref py py-obj docutils literal"><span class="pre">quality</span></tt></a></td>
<td>Quality scores of the characters in the biological sequence.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__contains__.html#skbio.sequence.BiologicalSequence.__contains__" title="skbio.sequence.BiologicalSequence.__contains__"><tt class="xref py py-obj docutils literal"><span class="pre">__contains__</span></tt></a>(other)</td>
<td>The in operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__eq__.html#skbio.sequence.BiologicalSequence.__eq__" title="skbio.sequence.BiologicalSequence.__eq__"><tt class="xref py py-obj docutils literal"><span class="pre">__eq__</span></tt></a>(other)</td>
<td>The equality operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__getitem__.html#skbio.sequence.BiologicalSequence.__getitem__" title="skbio.sequence.BiologicalSequence.__getitem__"><tt class="xref py py-obj docutils literal"><span class="pre">__getitem__</span></tt></a>(i)</td>
<td>The indexing operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__hash__.html#skbio.sequence.BiologicalSequence.__hash__" title="skbio.sequence.BiologicalSequence.__hash__"><tt class="xref py py-obj docutils literal"><span class="pre">__hash__</span></tt></a>()</td>
<td>The hash operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__iter__.html#skbio.sequence.BiologicalSequence.__iter__" title="skbio.sequence.BiologicalSequence.__iter__"><tt class="xref py py-obj docutils literal"><span class="pre">__iter__</span></tt></a>()</td>
<td>The iter operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__len__.html#skbio.sequence.BiologicalSequence.__len__" title="skbio.sequence.BiologicalSequence.__len__"><tt class="xref py py-obj docutils literal"><span class="pre">__len__</span></tt></a>()</td>
<td>The len operator.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__ne__.html#skbio.sequence.BiologicalSequence.__ne__" title="skbio.sequence.BiologicalSequence.__ne__"><tt class="xref py py-obj docutils literal"><span class="pre">__ne__</span></tt></a>(other)</td>
<td>The inequality operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__repr__.html#skbio.sequence.BiologicalSequence.__repr__" title="skbio.sequence.BiologicalSequence.__repr__"><tt class="xref py py-obj docutils literal"><span class="pre">__repr__</span></tt></a>()</td>
<td>The repr method.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__reversed__.html#skbio.sequence.BiologicalSequence.__reversed__" title="skbio.sequence.BiologicalSequence.__reversed__"><tt class="xref py py-obj docutils literal"><span class="pre">__reversed__</span></tt></a>()</td>
<td>The reversed operator.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.__str__.html#skbio.sequence.BiologicalSequence.__str__" title="skbio.sequence.BiologicalSequence.__str__"><tt class="xref py py-obj docutils literal"><span class="pre">__str__</span></tt></a>()</td>
<td>The str operator</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.alphabet.html#skbio.sequence.BiologicalSequence.alphabet" title="skbio.sequence.BiologicalSequence.alphabet"><tt class="xref py py-obj docutils literal"><span class="pre">alphabet</span></tt></a>()</td>
<td>Return the set of characters allowed in a <cite>BiologicalSequence</cite>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.copy.html#skbio.sequence.BiologicalSequence.copy" title="skbio.sequence.BiologicalSequence.copy"><tt class="xref py py-obj docutils literal"><span class="pre">copy</span></tt></a>(**kwargs)</td>
<td>Return a copy of the current biological sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.count.html#skbio.sequence.BiologicalSequence.count" title="skbio.sequence.BiologicalSequence.count"><tt class="xref py py-obj docutils literal"><span class="pre">count</span></tt></a>(subsequence)</td>
<td>Returns the number of occurences of subsequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.degap.html#skbio.sequence.BiologicalSequence.degap" title="skbio.sequence.BiologicalSequence.degap"><tt class="xref py py-obj docutils literal"><span class="pre">degap</span></tt></a>()</td>
<td>Returns a new <cite>BiologicalSequence</cite> with gap characters removed.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.distance.html#skbio.sequence.BiologicalSequence.distance" title="skbio.sequence.BiologicalSequence.distance"><tt class="xref py py-obj docutils literal"><span class="pre">distance</span></tt></a>(other[,&nbsp;distance_fn])</td>
<td>Returns the distance to other</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.equals.html#skbio.sequence.BiologicalSequence.equals" title="skbio.sequence.BiologicalSequence.equals"><tt class="xref py py-obj docutils literal"><span class="pre">equals</span></tt></a>(other[,&nbsp;ignore])</td>
<td>Compare two biological sequences for equality.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.fraction_diff.html#skbio.sequence.BiologicalSequence.fraction_diff" title="skbio.sequence.BiologicalSequence.fraction_diff"><tt class="xref py py-obj docutils literal"><span class="pre">fraction_diff</span></tt></a>(other)</td>
<td>Return fraction of positions that differ relative to <cite>other</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.fraction_same.html#skbio.sequence.BiologicalSequence.fraction_same" title="skbio.sequence.BiologicalSequence.fraction_same"><tt class="xref py py-obj docutils literal"><span class="pre">fraction_same</span></tt></a>(other)</td>
<td>Return fraction of positions that are the same relative to <cite>other</cite></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.gap_alphabet.html#skbio.sequence.BiologicalSequence.gap_alphabet" title="skbio.sequence.BiologicalSequence.gap_alphabet"><tt class="xref py py-obj docutils literal"><span class="pre">gap_alphabet</span></tt></a>()</td>
<td>Return the set of characters defined as gaps.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.gap_maps.html#skbio.sequence.BiologicalSequence.gap_maps" title="skbio.sequence.BiologicalSequence.gap_maps"><tt class="xref py py-obj docutils literal"><span class="pre">gap_maps</span></tt></a>()</td>
<td>Return tuples mapping b/w gapped and ungapped positions</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.gap_vector.html#skbio.sequence.BiologicalSequence.gap_vector" title="skbio.sequence.BiologicalSequence.gap_vector"><tt class="xref py py-obj docutils literal"><span class="pre">gap_vector</span></tt></a>()</td>
<td>Return list indicating positions containing gaps</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.has_quality.html#skbio.sequence.BiologicalSequence.has_quality" title="skbio.sequence.BiologicalSequence.has_quality"><tt class="xref py py-obj docutils literal"><span class="pre">has_quality</span></tt></a>()</td>
<td>Return bool indicating presence of quality scores in the sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.has_unsupported_characters.html#skbio.sequence.BiologicalSequence.has_unsupported_characters" title="skbio.sequence.BiologicalSequence.has_unsupported_characters"><tt class="xref py py-obj docutils literal"><span class="pre">has_unsupported_characters</span></tt></a>()</td>
<td>Return bool indicating presence/absence of unsupported characters</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.index.html#skbio.sequence.BiologicalSequence.index" title="skbio.sequence.BiologicalSequence.index"><tt class="xref py py-obj docutils literal"><span class="pre">index</span></tt></a>(subsequence)</td>
<td>Return the position where subsequence first occurs</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.is_gap.html#skbio.sequence.BiologicalSequence.is_gap" title="skbio.sequence.BiologicalSequence.is_gap"><tt class="xref py py-obj docutils literal"><span class="pre">is_gap</span></tt></a>(char)</td>
<td>Return True if <cite>char</cite> is in the <cite>gap_alphabet</cite> set</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.is_gapped.html#skbio.sequence.BiologicalSequence.is_gapped" title="skbio.sequence.BiologicalSequence.is_gapped"><tt class="xref py py-obj docutils literal"><span class="pre">is_gapped</span></tt></a>()</td>
<td>Return True if char(s) in <cite>gap_alphabet</cite> are present</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.is_valid.html#skbio.sequence.BiologicalSequence.is_valid" title="skbio.sequence.BiologicalSequence.is_valid"><tt class="xref py py-obj docutils literal"><span class="pre">is_valid</span></tt></a>()</td>
<td>Return True if the sequence is valid</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.iupac_characters.html#skbio.sequence.BiologicalSequence.iupac_characters" title="skbio.sequence.BiologicalSequence.iupac_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_characters</span></tt></a>()</td>
<td>Return the non-degenerate and degenerate characters.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.iupac_degeneracies.html#skbio.sequence.BiologicalSequence.iupac_degeneracies" title="skbio.sequence.BiologicalSequence.iupac_degeneracies"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_degeneracies</span></tt></a>()</td>
<td>Return the mapping of degenerate to non-degenerate characters.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.iupac_degenerate_characters.html#skbio.sequence.BiologicalSequence.iupac_degenerate_characters" title="skbio.sequence.BiologicalSequence.iupac_degenerate_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_degenerate_characters</span></tt></a>()</td>
<td>Return the degenerate IUPAC characters.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.iupac_standard_characters.html#skbio.sequence.BiologicalSequence.iupac_standard_characters" title="skbio.sequence.BiologicalSequence.iupac_standard_characters"><tt class="xref py py-obj docutils literal"><span class="pre">iupac_standard_characters</span></tt></a>()</td>
<td>Return the non-degenerate IUPAC characters.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.k_word_counts.html#skbio.sequence.BiologicalSequence.k_word_counts" title="skbio.sequence.BiologicalSequence.k_word_counts"><tt class="xref py py-obj docutils literal"><span class="pre">k_word_counts</span></tt></a>(k[,&nbsp;overlapping])</td>
<td>Get the counts of words of length k</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.k_word_frequencies.html#skbio.sequence.BiologicalSequence.k_word_frequencies" title="skbio.sequence.BiologicalSequence.k_word_frequencies"><tt class="xref py py-obj docutils literal"><span class="pre">k_word_frequencies</span></tt></a>(k[,&nbsp;overlapping])</td>
<td>Get the frequencies of words of length <cite>k</cite></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.k_words.html#skbio.sequence.BiologicalSequence.k_words" title="skbio.sequence.BiologicalSequence.k_words"><tt class="xref py py-obj docutils literal"><span class="pre">k_words</span></tt></a>(k[,&nbsp;overlapping])</td>
<td>Get the list of words of length k</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.lower.html#skbio.sequence.BiologicalSequence.lower" title="skbio.sequence.BiologicalSequence.lower"><tt class="xref py py-obj docutils literal"><span class="pre">lower</span></tt></a>()</td>
<td>Convert the BiologicalSequence to lowercase</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.nondegenerates.html#skbio.sequence.BiologicalSequence.nondegenerates" title="skbio.sequence.BiologicalSequence.nondegenerates"><tt class="xref py py-obj docutils literal"><span class="pre">nondegenerates</span></tt></a>()</td>
<td>Yield all nondegenerate versions of the sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.read.html#skbio.sequence.BiologicalSequence.read" title="skbio.sequence.BiologicalSequence.read"><tt class="xref py py-obj docutils literal"><span class="pre">read</span></tt></a>(fp[,&nbsp;format])</td>
<td>Create a new <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> instance from a file.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.regex_iter.html#skbio.sequence.BiologicalSequence.regex_iter" title="skbio.sequence.BiologicalSequence.regex_iter"><tt class="xref py py-obj docutils literal"><span class="pre">regex_iter</span></tt></a>(regex[,&nbsp;retrieve_group_0])</td>
<td>Find patterns specified by regular expression</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.to_fasta.html#skbio.sequence.BiologicalSequence.to_fasta" title="skbio.sequence.BiologicalSequence.to_fasta"><tt class="xref py py-obj docutils literal"><span class="pre">to_fasta</span></tt></a>([field_delimiter,&nbsp;terminal_character])</td>
<td>Return the sequence as a fasta-formatted string</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.unsupported_characters.html#skbio.sequence.BiologicalSequence.unsupported_characters" title="skbio.sequence.BiologicalSequence.unsupported_characters"><tt class="xref py py-obj docutils literal"><span class="pre">unsupported_characters</span></tt></a>()</td>
<td>Return the set of unsupported characters in the <cite>BiologicalSequence</cite></td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.upper.html#skbio.sequence.BiologicalSequence.upper" title="skbio.sequence.BiologicalSequence.upper"><tt class="xref py py-obj docutils literal"><span class="pre">upper</span></tt></a>()</td>
<td>Convert the BiologicalSequence to uppercase</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="skbio.sequence.BiologicalSequence.write.html#skbio.sequence.BiologicalSequence.write" title="skbio.sequence.BiologicalSequence.write"><tt class="xref py py-obj docutils literal"><span class="pre">write</span></tt></a>(fp[,&nbsp;format])</td>
<td>Write an instance of <tt class="docutils literal"><span class="pre">BiologicalSequence</span></tt> to a file.</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>