<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Biological sequences (skbio.core.sequence) &mdash; scikit-bio 0.1.3 documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.1.3 documentation" href="index.html" />
    <link rel="next" title="skbio.core.sequence.BiologicalSequence" href="generated/skbio.core.sequence.BiologicalSequence.html" />
    <link rel="prev" title="skbio.core.alignment.align_striped_smith_waterman" href="generated/skbio.core.alignment.align_striped_smith_waterman.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="core.alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.core.alignment</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.distance.html">Dissimilarity and distance matrices (<tt class="docutils literal"><span class="pre">skbio.core.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.exception.html">Exceptions (<tt class="docutils literal"><span class="pre">skbio.core.exception</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.warning.html">Warnings (<tt class="docutils literal"><span class="pre">skbio.core.warning</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.genetic_code.html">Genetic Code (<tt class="docutils literal"><span class="pre">skbio.core.genetic_code</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.core.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="core.workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.core.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="draw.distributions.html">Distribution visualizations (<tt class="docutils literal"><span class="pre">skbio.draw.distributions</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.diversity.alpha.html">Alpha diversity measures (<tt class="docutils literal"><span class="pre">skbio.math.diversity.alpha</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.stats.distance.html">Distance-based statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.distance</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.stats.misc.html">Miscellaneous statistics utilities (<tt class="docutils literal"><span class="pre">skbio.math.stats.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.stats.ordination.html">Ordination methods (<tt class="docutils literal"><span class="pre">skbio.math.stats.ordination</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.stats.spatial.html">Spatial Statistics (<tt class="docutils literal"><span class="pre">skbio.math.stats.spatial</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.gradient.html">Gradient analyses (<tt class="docutils literal"><span class="pre">skbio.math.gradient</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="math.subsample.html">Subsampling (<tt class="docutils literal"><span class="pre">skbio.math.subsample</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="util.io.html">I/O utils (<tt class="docutils literal"><span class="pre">skbio.util.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="util.misc.html">Miscellaneous Utilities (<tt class="docutils literal"><span class="pre">skbio.util.misc</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="util.testing.html">Testing utilities (<tt class="docutils literal"><span class="pre">skbio.util.testing</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.core.sequence</span></tt>)</a><ul>
<li><a class="reference internal" href="#classes">Classes</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <span class="target" id="module-skbio.core.sequence"></span><div class="section" id="biological-sequences-skbio-core-sequence">
<h1>Biological sequences (<a class="reference internal" href="#module-skbio.core.sequence" title="skbio.core.sequence"><tt class="xref py py-mod docutils literal"><span class="pre">skbio.core.sequence</span></tt></a>)<a class="headerlink" href="#biological-sequences-skbio-core-sequence" title="Permalink to this headline">Â¶</a></h1>
<p>This module provides functionality for working with biological sequences,
including generic sequences, nucelotide sequences, DNA sequences, and RNA
sequences. Class methods and attributes are also available to obtain valid
character sets, complement maps for different sequence types, and for
obtaining degenerate character definitions.</p>
<div class="section" id="classes">
<h2>Classes<a class="headerlink" href="#classes" title="Permalink to this headline">Â¶</a></h2>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.core.sequence.BiologicalSequence.html#skbio.core.sequence.BiologicalSequence" title="skbio.core.sequence.BiologicalSequence"><tt class="xref py py-obj docutils literal"><span class="pre">BiologicalSequence</span></tt></a>(sequence[,&nbsp;id,&nbsp;...])</td>
<td>Base class for biological sequences.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/skbio.core.sequence.NucleotideSequence.html#skbio.core.sequence.NucleotideSequence" title="skbio.core.sequence.NucleotideSequence"><tt class="xref py py-obj docutils literal"><span class="pre">NucleotideSequence</span></tt></a>(sequence[,&nbsp;id,&nbsp;...])</td>
<td>Base class for nucleotide sequences.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.core.sequence.DNASequence.html#skbio.core.sequence.DNASequence" title="skbio.core.sequence.DNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">DNASequence</span></tt></a>(sequence[,&nbsp;id,&nbsp;description,&nbsp;...])</td>
<td>Base class for DNA sequences.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/skbio.core.sequence.RNASequence.html#skbio.core.sequence.RNASequence" title="skbio.core.sequence.RNASequence"><tt class="xref py py-obj docutils literal"><span class="pre">RNASequence</span></tt></a>(sequence[,&nbsp;id,&nbsp;description,&nbsp;...])</td>
<td>Base class for RNA sequences.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/skbio.core.sequence.ProteinSequence.html#skbio.core.sequence.ProteinSequence" title="skbio.core.sequence.ProteinSequence"><tt class="xref py py-obj docutils literal"><span class="pre">ProteinSequence</span></tt></a>(sequence[,&nbsp;id,&nbsp;description,&nbsp;...])</td>
<td>Base class for protein sequences.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">Â¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.core.sequence</span> <span class="kn">import</span> <span class="n">DNASequence</span><span class="p">,</span> <span class="n">RNASequence</span>
</pre></div>
</div>
<p>New sequences are created with optional id and description fields.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;ACC--G-GGTA..&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;ACC--G-GGTA..&#39;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="s">&quot;seq1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;ACC--G-GGTA..&#39;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="s">&quot;seq1&quot;</span><span class="p">,</span><span class="n">description</span><span class="o">=</span><span class="s">&quot;GFP&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>New sequences can also be created from existing sequences, for example as their
reverse complement or degapped (i.e., unaligned) version.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="n">d1</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span>
<span class="go">&lt;DNASequence: ACC--G-GGT... (length: 13)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span>
<span class="go">&lt;DNASequence: ACCGGGTA (length: 8)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span>
<span class="go">&lt;DNASequence: TACCCGGT (length: 8)&gt;</span>
</pre></div>
</div>
<p>It&#8217;s also straight-forward to compute distances between sequences (optionally
using user-defined distance metrics, default is Hamming distance) for use in
sequence clustering, phylogenetic reconstruction, etc.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d4</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;GACCCGCT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d5</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;GACCCCCT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">d4</span><span class="p">)</span>
<span class="go">0.25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d3</span><span class="o">.</span><span class="n">distance</span><span class="p">(</span><span class="n">d5</span><span class="p">)</span>
<span class="go">0.375</span>
</pre></div>
</div>
<p>Class-level methods contain information about the molecule types.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">DNASequence</span><span class="o">.</span><span class="n">iupac_degeneracies</span><span class="p">()[</span><span class="s">&#39;B&#39;</span><span class="p">]</span>
<span class="go">set([&#39;C&#39;, &#39;T&#39;, &#39;G&#39;])</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">RNASequence</span><span class="o">.</span><span class="n">iupac_degeneracies</span><span class="p">()[</span><span class="s">&#39;B&#39;</span><span class="p">]</span>
<span class="go">set([&#39;C&#39;, &#39;U&#39;, &#39;G&#39;])</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">DNASequence</span><span class="o">.</span><span class="n">is_gap</span><span class="p">(</span><span class="s">&#39;-&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>NucleotideSequences can be translated using a GeneticCode object.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">d6</span> <span class="o">=</span> <span class="n">DNASequence</span><span class="p">(</span><span class="s">&#39;ATGTCTAAATGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.core.genetic_code</span> <span class="kn">import</span> <span class="n">GeneticCodes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">GeneticCodes</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">d6</span><span class="p">)</span>
<span class="go">&lt;ProteinSequence: MSK* (length: 4)&gt;</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>