<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>skbio.stats.power.subsample_power &mdash; scikit-bio 0.2.3 documentation</title>
    
    <link rel="stylesheet" href="../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/style.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootswatch-3.1.0/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.2.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/copybutton.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-3.1.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="scikit-bio 0.2.3 documentation" href="../../index.html" />
    <link rel="up" title="Empirical Power Estimation (skbio.stats.power)" href="../skbio.stats.power.html" />
    <link rel="next" title="skbio.stats.power.subsample_paired_power" href="skbio.stats.power.subsample_paired_power.html" />
    <link rel="prev" title="skbio.stats.power.paired_subsamples" href="skbio.stats.power.paired_subsamples.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>
  
  <a href="https://github.com/biocore/scikit-bio"
     class="visible-desktop hidden-xs"><img
    id="gh-banner"
    style="position: absolute; top: 50px; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"
    alt="Fork me on GitHub"></a>
  <script>
    // Adjust banner height.
    $(function () {
      var navHeight = $(".navbar .container").css("height");
      $("#gh-banner").css("top", navHeight);
    });
  </script>


  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
          scikit-bio docs</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            <li class="divider-vertical"></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../alignment.html">Sequence collections and alignments (<tt class="docutils literal"><span class="pre">skbio.alignment</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../sequence.html">Biological sequences (<tt class="docutils literal"><span class="pre">skbio.sequence</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tree.html">Tree representations (<tt class="docutils literal"><span class="pre">skbio.tree</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../workflow.html">Constructing workflows (<tt class="docutils literal"><span class="pre">skbio.workflow</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../draw.html">Visualizations (<tt class="docutils literal"><span class="pre">skbio.draw</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../format.sequences.html">Format biological sequences (<tt class="docutils literal"><span class="pre">skbio.format.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../diversity.html">Diversity calculations (<tt class="docutils literal"><span class="pre">skbio.diversity</span></tt>)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../stats.html">Statistics (<tt class="docutils literal"><span class="pre">skbio.stats</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../parse.sequences.html">Parse biological sequences (<tt class="docutils literal"><span class="pre">skbio.parse.sequences</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../io.html">File I/O (<tt class="docutils literal"><span class="pre">skbio.io</span></tt>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../util.html">Utility functionality (<tt class="docutils literal"><span class="pre">skbio.util</span></tt>)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../development/py3.html">Supporting Python 2 and Python 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development/coding_guidelines.html">Coding guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development/new_module.html">Adding a new module to skbio</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">skbio.stats.power.subsample_power</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="skbio-stats-power-subsample-power">
<h1>skbio.stats.power.subsample_power<a class="headerlink" href="#skbio-stats-power-subsample-power" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="skbio.stats.power.subsample_power">
<tt class="descclassname">skbio.stats.power.</tt><tt class="descname">subsample_power</tt><big>(</big><em>test</em>, <em>samples</em>, <em>draw_mode='ind'</em>, <em>alpha_pwr=0.05</em>, <em>min_observations=20</em>, <em>max_counts=50</em>, <em>counts_interval=10</em>, <em>min_counts=None</em>, <em>num_iter=500</em>, <em>num_runs=10</em><big>)</big><a class="reference external" href="http://github.com/biocore/scikit-bio/blob/0.2.3/skbio/stats/power.py#L152"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#skbio.stats.power.subsample_power" title="Permalink to this definition">¶</a></dt>
<dd><p>Subsamples data to iteratively calculate power</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>test</strong> : function</p>
<blockquote>
<div><p>The statistical test which accepts a list of arrays of values
(sample ids or numeric values) and returns a p value.</p>
</div></blockquote>
<p><strong>samples</strong> : array_like</p>
<blockquote>
<div><p><cite>samples</cite> can be a list of lists or a list of arrays where each
sublist or row in the array corresponds to a sampled group.</p>
</div></blockquote>
<p><strong>draw_mode</strong> : {&#8220;ind&#8221;, &#8220;matched&#8221;}, optional</p>
<blockquote>
<div><p>&#8220;matched&#8221; samples should be used when observations in
samples have corresponding observations in other groups. For instance,
this may be useful when working with regression data where
<span class="math">\(x_{1}, x_{2}, ..., x_{n}\)</span> maps to
<span class="math">\(y_{1}, y_{2}, ..., y_{n}\)</span>. Sample vectors must be the same
length in &#8220;matched&#8221; mode.
If there is no reciprocal relationship between samples, then
&#8220;ind&#8221; mode should be used.</p>
</div></blockquote>
<p><strong>alpha_pwr</strong> : float, optional</p>
<blockquote>
<div><p>The critical value used to calculate the power.</p>
</div></blockquote>
<p><strong>min_observations</strong> : unsigned int, optional</p>
<blockquote>
<div><p>The minimum number of observations in any sample to
perform power analysis. Note that this is not the same as the minimum
number of samples drawn per group.</p>
</div></blockquote>
<p><strong>max_counts</strong> : unsigned int, optional</p>
<blockquote>
<div><p>The maximum number of samples per group to draw for
effect size calculation.</p>
</div></blockquote>
<p><strong>counts_interval</strong> : unsigned int, optional</p>
<blockquote>
<div><p>The difference between each subsampling count.</p>
</div></blockquote>
<p><strong>min_counts</strong> : unsigned int, optional</p>
<blockquote>
<div><p>How many samples should be drawn for the smallest
subsample. If this is None, the <cite>counts_interval</cite> will be used.</p>
</div></blockquote>
<p><strong>num_iter</strong> : unsigned int, optional</p>
<blockquote>
<div><p>The number of p-values to generate for each point
on the curve.</p>
</div></blockquote>
<p><strong>num_runs</strong> : unsigned int, optional</p>
<blockquote>
<div><p>The number of times to calculate each curve.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>power</strong> : array</p>
<blockquote>
<div><p>The power calculated for each subsample at each count.</p>
</div></blockquote>
<p><strong>sample_counts</strong> : array</p>
<blockquote>
<div><p>The number of samples drawn at each power calculation.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>ValueError</strong></p>
<blockquote>
<div><p>If the <cite>mode</cite> is &#8220;matched&#8221;, an error will occur if the arrays in
<cite>samples</cite> are not the same length.</p>
</div></blockquote>
<p><strong>ValueError</strong></p>
<blockquote>
<div><p>There is a ValueError if there are fewer samples than the minimum
count.</p>
</div></blockquote>
<p><strong>ValueError</strong></p>
<blockquote class="last">
<div><p>If the <cite>counts_interval</cite> is greater than the difference between the
sample start and the max value, the function raises a ValueError.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Let&#8217;s say we wanted to look at the relationship between the presence of a
specific bacteria and the probability of a pre or post menopausal woman
experiencing a health outcome. Healthy women were enrolled in the study
either before or after menopause, and followed for five years. They
submitted fecal samples at regular intervals during that period, and were
assessed for a particular irreversible health outcome over that period.</p>
<p>16S sequencing and available literature suggest a set of candidate taxa
may be associated with the health outcome. Assume there are 100 samples
(50 premenopausal samples and 50 postmenopausal samples) where the taxa
of interest was identified by 16S sequencing and the taxonomic abundance
was confirmed in a certain fraction of samples at a minimum level.</p>
<p>We can simulate the probability that a woman positive for this taxa
experiences the health outcome using a binomial distribution.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">25</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pre_rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">50</span><span class="p">,))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pre_rate</span>
<span class="go">array([0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,</span>
<span class="go">       0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0,</span>
<span class="go">       1, 1, 1, 1])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">50</span><span class="p">,))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_rate</span>
<span class="go">array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,</span>
<span class="go">       0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,</span>
<span class="go">       0, 1, 0, 0])</span>
</pre></div>
</div>
<p>Let&#8217;s set up a test function, so we can test the probability of
finding a difference in frequency between the two groups. We&#8217;ll use
<cite>scipy.stats.chisquare</cite> to look for the difference in frequency between
groups.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chisquare</span><span class="p">,</span> <span class="n">nanmean</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">test</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">chisquare</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>
<span class="gp">... </span>    <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))]))[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>Let&#8217;s make sure that our two distributions are different.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">round</span><span class="p">(</span><span class="n">test</span><span class="p">([</span><span class="n">pre_rate</span><span class="p">,</span> <span class="n">pos_rate</span><span class="p">]),</span> <span class="mi">5</span><span class="p">)</span>
<span class="go">9e-05</span>
</pre></div>
</div>
<p>Since there are an even number of samples, and we don&#8217;t have enough
information to try controlling the data, we&#8217;ll use
<cite>skbio.stats.power.subsample_power</cite> to compare the two groups. If we had
metadata about other risk factors, like a family history, BMI, tobacco use,
we might want to use <cite>skbio.stats.power.subsample_paired_power</cite>.
We&#8217;ll also use &#8220;ind&#8221; <cite>draw_mode</cite>, since there is no linkage between the
two groups of samples.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">skbio.stats.power</span> <span class="kn">import</span> <span class="n">subsample_power</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pwr_est</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">subsample_power</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
<span class="gp">... </span>                                  <span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="n">pre_rate</span><span class="p">,</span> <span class="n">pos_rate</span><span class="p">],</span>
<span class="gp">... </span>                                  <span class="n">counts_interval</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span>
<span class="go">array([ 5, 10, 15, 20, 25, 30, 35, 40, 45])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nanmean</span><span class="p">(</span><span class="n">pwr_est</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">array([ 0.1776,  0.3392,  0.658 ,  0.8856,  0.9804,  0.9982,  1.    ,</span>
<span class="go">        1.    ,  1.    ])</span>
</pre></div>
</div>
<p>So, we can estimate that we will see a significant difference between
the two groups (<span class="math">\(\alpha \leq 0.05\)</span>) at least 80% of the time if we
use 20 observations per group.</p>
<p>If we wanted to test the relationship of a second candidate taxa which is
more rare in the population, but may have a similar effect, based on
available literature, we might also start by trying to identify 20
samples per group where the second candidate taxa is present.</p>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
  
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014--, scikit-bio development team.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.2.<br/>
    </p>
  </div>
</footer>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-6636235-9', 'scikit-bio.org');
    ga('send', 'pageview');
  </script>

  </body>
</html>